// @frc-web-components/fwc/dist/components/line-chart.es.js
import { E as fe, b as tt, r as Pt, n as D, i as un, x as St } from "../property-C4BPH-wS.js";
import { r as cn } from "../state-Cn3aen3G.js";
import { e as zt } from "../query-BrpQj_Qi.js";
import { c as ln, i as qt, a as he, b as ge, d as fn, s as xt } from "../transform-DR-Bw75L.js";
import { f as hn } from "../directive-helpers-CDfQTKPA.js";
import { i as gn, t as mn, e as dn } from "../directive-B76A7YXI.js";
import { o as Dt } from "../style-map-CXHve4fi.js";
function $t(t, e) {
  return t == null || e == null ? NaN : t < e ? -1 : t > e ? 1 : t >= e ? 0 : NaN;
}
function pn(t, e) {
  return t == null || e == null ? NaN : e < t ? -1 : e > t ? 1 : e >= t ? 0 : NaN;
}
function te(t) {
  let e, n, r;
  t.length !== 2 ? (e = $t, n = (u, l) => $t(t(u), l), r = (u, l) => t(u) - l) : (e = t === $t || t === pn ? t : yn, n = t, r = t);
  function i(u, l, s = 0, g = u.length) {
    if (s < g) {
      if (e(l, l) !== 0) return g;
      do {
        const h = s + g >>> 1;
        n(u[h], l) < 0 ? s = h + 1 : g = h;
      } while (s < g);
    }
    return s;
  }
  function a(u, l, s = 0, g = u.length) {
    if (s < g) {
      if (e(l, l) !== 0) return g;
      do {
        const h = s + g >>> 1;
        n(u[h], l) <= 0 ? s = h + 1 : g = h;
      } while (s < g);
    }
    return s;
  }
  function o(u, l, s = 0, g = u.length) {
    const h = i(u, l, s, g - 1);
    return h > s && r(u[h - 1], l) > -r(u[h], l) ? h - 1 : h;
  }
  return { left: i, center: o, right: a };
}
function yn() {
  return 0;
}
function Mn(t) {
  return t === null ? NaN : +t;
}
const xn = te($t), vn = xn.right;
te(Mn).center;
class me extends Map {
  constructor(e, n = Tn) {
    if (super(), Object.defineProperties(this, { _intern: { value: /* @__PURE__ */ new Map() }, _key: { value: n } }), e != null) for (const [r, i] of e) this.set(r, i);
  }
  get(e) {
    return super.get(de(this, e));
  }
  has(e) {
    return super.has(de(this, e));
  }
  set(e, n) {
    return super.set(wn(this, e), n);
  }
  delete(e) {
    return super.delete(Cn(this, e));
  }
}
function de({ _intern: t, _key: e }, n) {
  const r = e(n);
  return t.has(r) ? t.get(r) : n;
}
function wn({ _intern: t, _key: e }, n) {
  const r = e(n);
  return t.has(r) ? t.get(r) : (t.set(r, n), n);
}
function Cn({ _intern: t, _key: e }, n) {
  const r = e(n);
  return t.has(r) && (n = t.get(r), t.delete(r)), n;
}
function Tn(t) {
  return t !== null && typeof t == "object" ? t.valueOf() : t;
}
const kn = Math.sqrt(50), bn = Math.sqrt(10), Sn = Math.sqrt(2);
function Nt(t, e, n) {
  const r = (e - t) / Math.max(0, n), i = Math.floor(Math.log10(r)), a = r / Math.pow(10, i), o = a >= kn ? 10 : a >= bn ? 5 : a >= Sn ? 2 : 1;
  let u, l, s;
  return i < 0 ? (s = Math.pow(10, -i) / o, u = Math.round(t * s), l = Math.round(e * s), u / s < t && ++u, l / s > e && --l, s = -s) : (s = Math.pow(10, i) * o, u = Math.round(t / s), l = Math.round(e / s), u * s < t && ++u, l * s > e && --l), l < u && 0.5 <= n && n < 2 ? Nt(t, e, n * 2) : [u, l, s];
}
function Dn(t, e, n) {
  if (e = +e, t = +t, n = +n, !(n > 0)) return [];
  if (t === e) return [t];
  const r = e < t, [i, a, o] = r ? Nt(e, t, n) : Nt(t, e, n);
  if (!(a >= i)) return [];
  const u = a - i + 1, l = new Array(u);
  if (r)
    if (o < 0) for (let s = 0; s < u; ++s) l[s] = (a - s) / -o;
    else for (let s = 0; s < u; ++s) l[s] = (a - s) * o;
  else if (o < 0) for (let s = 0; s < u; ++s) l[s] = (i + s) / -o;
  else for (let s = 0; s < u; ++s) l[s] = (i + s) * o;
  return l;
}
function Xt(t, e, n) {
  return e = +e, t = +t, n = +n, Nt(t, e, n)[2];
}
function Qt(t, e, n) {
  e = +e, t = +t, n = +n;
  const r = e < t, i = r ? Xt(e, t, n) : Xt(t, e, n);
  return (r ? -1 : 1) * (i < 0 ? 1 / -i : i);
}
function Un(t) {
  return t;
}
var Et = 1, At = 2, Jt = 3, Mt = 4, pe = 1e-6;
function $n(t) {
  return "translate(" + t + ",0)";
}
function An(t) {
  return "translate(0," + t + ")";
}
function Nn(t) {
  return (e) => +t(e);
}
function Fn(t, e) {
  return e = Math.max(0, t.bandwidth() - e * 2) / 2, t.round() && (e = Math.round(e)), (n) => +t(n) + e;
}
function Yn() {
  return !this.__axis;
}
function ee(t, e) {
  var n = [], r = null, i = null, a = 6, o = 6, u = 3, l = typeof window < "u" && window.devicePixelRatio > 1 ? 0 : 0.5, s = t === Et || t === Mt ? -1 : 1, g = t === Mt || t === At ? "x" : "y", h = t === Et || t === Jt ? $n : An;
  function f(d) {
    var b = r ?? (e.ticks ? e.ticks.apply(e, n) : e.domain()), L = i ?? (e.tickFormat ? e.tickFormat.apply(e, n) : Un), F = Math.max(a, 0) + u, _ = e.range(), P = +_[0] + l, T = +_[_.length - 1] + l, N = (e.bandwidth ? Fn : Nn)(e.copy(), l), k = d.selection ? d.selection() : d, p = k.selectAll(".domain").data([null]), w = k.selectAll(".tick").data(b, e).order(), O = w.exit(), q = w.enter().append("g").attr("class", "tick"), E = w.select("line"), y = w.select("text");
    p = p.merge(p.enter().insert("path", ".tick").attr("class", "domain").attr("stroke", "currentColor")), w = w.merge(q), E = E.merge(q.append("line").attr("stroke", "currentColor").attr(g + "2", s * a)), y = y.merge(q.append("text").attr("fill", "currentColor").attr(g, s * F).attr("dy", t === Et ? "0em" : t === Jt ? "0.71em" : "0.32em")), d !== k && (p = p.transition(d), w = w.transition(d), E = E.transition(d), y = y.transition(d), O = O.transition(d).attr("opacity", pe).attr("transform", function($) {
      return isFinite($ = N($)) ? h($ + l) : this.getAttribute("transform");
    }), q.attr("opacity", pe).attr("transform", function($) {
      var S = this.parentNode.__axis;
      return h((S && isFinite(S = S($)) ? S : N($)) + l);
    })), O.remove(), p.attr("d", t === Mt || t === At ? o ? "M" + s * o + "," + P + "H" + l + "V" + T + "H" + s * o : "M" + l + "," + P + "V" + T : o ? "M" + P + "," + s * o + "V" + l + "H" + T + "V" + s * o : "M" + P + "," + l + "H" + T), w.attr("opacity", 1).attr("transform", function($) {
      return h(N($) + l);
    }), E.attr(g + "2", s * a), y.attr(g, s * F).text(L), k.filter(Yn).attr("fill", "none").attr("font-size", 10).attr("font-family", "sans-serif").attr("text-anchor", t === At ? "start" : t === Mt ? "end" : "middle"), k.each(function() {
      this.__axis = N;
    });
  }
  return f.scale = function(d) {
    return arguments.length ? (e = d, f) : e;
  }, f.ticks = function() {
    return n = Array.from(arguments), f;
  }, f.tickArguments = function(d) {
    return arguments.length ? (n = d == null ? [] : Array.from(d), f) : n.slice();
  }, f.tickValues = function(d) {
    return arguments.length ? (r = d == null ? null : Array.from(d), f) : r && r.slice();
  }, f.tickFormat = function(d) {
    return arguments.length ? (i = d, f) : i;
  }, f.tickSize = function(d) {
    return arguments.length ? (a = o = +d, f) : a;
  }, f.tickSizeInner = function(d) {
    return arguments.length ? (a = +d, f) : a;
  }, f.tickSizeOuter = function(d) {
    return arguments.length ? (o = +d, f) : o;
  }, f.tickPadding = function(d) {
    return arguments.length ? (u = +d, f) : u;
  }, f.offset = function(d) {
    return arguments.length ? (l = +d, f) : l;
  }, f;
}
function Ln(t) {
  return ee(At, t);
}
function He(t) {
  return ee(Jt, t);
}
function ye(t) {
  return ee(Mt, t);
}
function _n(t, e) {
  e || (e = []);
  var n = t ? Math.min(e.length, t.length) : 0, r = e.slice(), i;
  return function(a) {
    for (i = 0; i < n; ++i) r[i] = t[i] * (1 - a) + e[i] * a;
    return r;
  };
}
function Hn(t) {
  return ArrayBuffer.isView(t) && !(t instanceof DataView);
}
function Wn(t, e) {
  var n = e ? e.length : 0, r = t ? Math.min(n, t.length) : 0, i = new Array(r), a = new Array(n), o;
  for (o = 0; o < r; ++o) i[o] = ne(t[o], e[o]);
  for (; o < n; ++o) a[o] = e[o];
  return function(u) {
    for (o = 0; o < r; ++o) a[o] = i[o](u);
    return a;
  };
}
function Pn(t, e) {
  var n = /* @__PURE__ */ new Date();
  return t = +t, e = +e, function(r) {
    return n.setTime(t * (1 - r) + e * r), n;
  };
}
function zn(t, e) {
  var n = {}, r = {}, i;
  (t === null || typeof t != "object") && (t = {}), (e === null || typeof e != "object") && (e = {});
  for (i in e)
    i in t ? n[i] = ne(t[i], e[i]) : r[i] = e[i];
  return function(a) {
    for (i in n) r[i] = n[i](a);
    return r;
  };
}
function ne(t, e) {
  var n = typeof e, r;
  return e == null || n === "boolean" ? ln(e) : (n === "number" ? qt : n === "string" ? (r = he(e)) ? (e = r, ge) : fn : e instanceof he ? ge : e instanceof Date ? Pn : Hn(e) ? _n : Array.isArray(e) ? Wn : typeof e.valueOf != "function" && typeof e.toString != "function" || isNaN(e) ? zn : qt)(t, e);
}
function On(t, e) {
  return t = +t, e = +e, function(n) {
    return Math.round(t * (1 - n) + e * n);
  };
}
function In(t) {
  return Math.abs(t = Math.round(t)) >= 1e21 ? t.toLocaleString("en").replace(/,/g, "") : t.toString(10);
}
function Ft(t, e) {
  if ((n = (t = e ? t.toExponential(e - 1) : t.toExponential()).indexOf("e")) < 0) return null;
  var n, r = t.slice(0, n);
  return [
    r.length > 1 ? r[0] + r.slice(2) : r,
    +t.slice(n + 1)
  ];
}
function ft(t) {
  return t = Ft(Math.abs(t)), t ? t[1] : NaN;
}
function En(t, e) {
  return function(n, r) {
    for (var i = n.length, a = [], o = 0, u = t[0], l = 0; i > 0 && u > 0 && (l + u + 1 > r && (u = Math.max(1, r - l)), a.push(n.substring(i -= u, i + u)), !((l += u + 1) > r)); )
      u = t[o = (o + 1) % t.length];
    return a.reverse().join(e);
  };
}
function Vn(t) {
  return function(e) {
    return e.replace(/[0-9]/g, function(n) {
      return t[+n];
    });
  };
}
var Bn = /^(?:(.)?([<>=^]))?([+\-( ])?([$#])?(0)?(\d+)?(,)?(\.\d+)?(~)?([a-z%])?$/i;
function Yt(t) {
  if (!(e = Bn.exec(t))) throw new Error("invalid format: " + t);
  var e;
  return new re({
    fill: e[1],
    align: e[2],
    sign: e[3],
    symbol: e[4],
    zero: e[5],
    width: e[6],
    comma: e[7],
    precision: e[8] && e[8].slice(1),
    trim: e[9],
    type: e[10]
  });
}
Yt.prototype = re.prototype;
function re(t) {
  this.fill = t.fill === void 0 ? " " : t.fill + "", this.align = t.align === void 0 ? ">" : t.align + "", this.sign = t.sign === void 0 ? "-" : t.sign + "", this.symbol = t.symbol === void 0 ? "" : t.symbol + "", this.zero = !!t.zero, this.width = t.width === void 0 ? void 0 : +t.width, this.comma = !!t.comma, this.precision = t.precision === void 0 ? void 0 : +t.precision, this.trim = !!t.trim, this.type = t.type === void 0 ? "" : t.type + "";
}
re.prototype.toString = function() {
  return this.fill + this.align + this.sign + this.symbol + (this.zero ? "0" : "") + (this.width === void 0 ? "" : Math.max(1, this.width | 0)) + (this.comma ? "," : "") + (this.precision === void 0 ? "" : "." + Math.max(0, this.precision | 0)) + (this.trim ? "~" : "") + this.type;
};
function Rn(t) {
  t: for (var e = t.length, n = 1, r = -1, i; n < e; ++n)
    switch (t[n]) {
      case ".":
        r = i = n;
        break;
      case "0":
        r === 0 && (r = n), i = n;
        break;
      default:
        if (!+t[n]) break t;
        r > 0 && (r = 0);
        break;
    }
  return r > 0 ? t.slice(0, r) + t.slice(i + 1) : t;
}
var We;
function jn(t, e) {
  var n = Ft(t, e);
  if (!n) return t + "";
  var r = n[0], i = n[1], a = i - (We = Math.max(-8, Math.min(8, Math.floor(i / 3))) * 3) + 1, o = r.length;
  return a === o ? r : a > o ? r + new Array(a - o + 1).join("0") : a > 0 ? r.slice(0, a) + "." + r.slice(a) : "0." + new Array(1 - a).join("0") + Ft(t, Math.max(0, e + a - 1))[0];
}
function Me(t, e) {
  var n = Ft(t, e);
  if (!n) return t + "";
  var r = n[0], i = n[1];
  return i < 0 ? "0." + new Array(-i).join("0") + r : r.length > i + 1 ? r.slice(0, i + 1) + "." + r.slice(i + 1) : r + new Array(i - r.length + 2).join("0");
}
const xe = {
  "%": (t, e) => (t * 100).toFixed(e),
  b: (t) => Math.round(t).toString(2),
  c: (t) => t + "",
  d: In,
  e: (t, e) => t.toExponential(e),
  f: (t, e) => t.toFixed(e),
  g: (t, e) => t.toPrecision(e),
  o: (t) => Math.round(t).toString(8),
  p: (t, e) => Me(t * 100, e),
  r: Me,
  s: jn,
  X: (t) => Math.round(t).toString(16).toUpperCase(),
  x: (t) => Math.round(t).toString(16)
};
function ve(t) {
  return t;
}
var we = Array.prototype.map, Ce = ["y", "z", "a", "f", "p", "n", "µ", "m", "", "k", "M", "G", "T", "P", "E", "Z", "Y"];
function Gn(t) {
  var e = t.grouping === void 0 || t.thousands === void 0 ? ve : En(we.call(t.grouping, Number), t.thousands + ""), n = t.currency === void 0 ? "" : t.currency[0] + "", r = t.currency === void 0 ? "" : t.currency[1] + "", i = t.decimal === void 0 ? "." : t.decimal + "", a = t.numerals === void 0 ? ve : Vn(we.call(t.numerals, String)), o = t.percent === void 0 ? "%" : t.percent + "", u = t.minus === void 0 ? "−" : t.minus + "", l = t.nan === void 0 ? "NaN" : t.nan + "";
  function s(h) {
    h = Yt(h);
    var f = h.fill, d = h.align, b = h.sign, L = h.symbol, F = h.zero, _ = h.width, P = h.comma, T = h.precision, N = h.trim, k = h.type;
    k === "n" ? (P = !0, k = "g") : xe[k] || (T === void 0 && (T = 12), N = !0, k = "g"), (F || f === "0" && d === "=") && (F = !0, f = "0", d = "=");
    var p = L === "$" ? n : L === "#" && /[boxX]/.test(k) ? "0" + k.toLowerCase() : "", w = L === "$" ? r : /[%p]/.test(k) ? o : "", O = xe[k], q = /[defgprs%]/.test(k);
    T = T === void 0 ? 6 : /[gprs]/.test(k) ? Math.max(1, Math.min(21, T)) : Math.max(0, Math.min(20, T));
    function E(y) {
      var $ = p, S = w, X, bt, at;
      if (k === "c")
        S = O(y) + S, y = "";
      else {
        y = +y;
        var ot = y < 0 || 1 / y < 0;
        if (y = isNaN(y) ? l : O(Math.abs(y), T), N && (y = Rn(y)), ot && +y == 0 && b !== "+" && (ot = !1), $ = (ot ? b === "(" ? b : u : b === "-" || b === "(" ? "" : b) + $, S = (k === "s" ? Ce[8 + We / 3] : "") + S + (ot && b === "(" ? ")" : ""), q) {
          for (X = -1, bt = y.length; ++X < bt; )
            if (at = y.charCodeAt(X), 48 > at || at > 57) {
              S = (at === 46 ? i + y.slice(X + 1) : y.slice(X)) + S, y = y.slice(0, X);
              break;
            }
        }
      }
      P && !F && (y = e(y, 1 / 0));
      var st = $.length + y.length + S.length, I = st < _ ? new Array(_ - st + 1).join(f) : "";
      switch (P && F && (y = e(I + y, I.length ? _ - S.length : 1 / 0), I = ""), d) {
        case "<":
          y = $ + y + S + I;
          break;
        case "=":
          y = $ + I + y + S;
          break;
        case "^":
          y = I.slice(0, st = I.length >> 1) + $ + y + S + I.slice(st);
          break;
        default:
          y = I + $ + y + S;
          break;
      }
      return a(y);
    }
    return E.toString = function() {
      return h + "";
    }, E;
  }
  function g(h, f) {
    var d = s((h = Yt(h), h.type = "f", h)), b = Math.max(-8, Math.min(8, Math.floor(ft(f) / 3))) * 3, L = Math.pow(10, -b), F = Ce[8 + b / 3];
    return function(_) {
      return d(L * _) + F;
    };
  }
  return {
    format: s,
    formatPrefix: g
  };
}
var Ut, Pe, ze;
Zn({
  thousands: ",",
  grouping: [3],
  currency: ["$", ""]
});
function Zn(t) {
  return Ut = Gn(t), Pe = Ut.format, ze = Ut.formatPrefix, Ut;
}
function qn(t) {
  return Math.max(0, -ft(Math.abs(t)));
}
function Xn(t, e) {
  return Math.max(0, Math.max(-8, Math.min(8, Math.floor(ft(e) / 3))) * 3 - ft(Math.abs(t)));
}
function Qn(t, e) {
  return t = Math.abs(t), e = Math.abs(e) - t, Math.max(0, ft(e) - ft(t)) + 1;
}
function ie(t, e) {
  switch (arguments.length) {
    case 0:
      break;
    case 1:
      this.range(t);
      break;
    default:
      this.range(e).domain(t);
      break;
  }
  return this;
}
const Te = Symbol("implicit");
function Oe() {
  var t = new me(), e = [], n = [], r = Te;
  function i(a) {
    let o = t.get(a);
    if (o === void 0) {
      if (r !== Te) return r;
      t.set(a, o = e.push(a) - 1);
    }
    return n[o % n.length];
  }
  return i.domain = function(a) {
    if (!arguments.length) return e.slice();
    e = [], t = new me();
    for (const o of a)
      t.has(o) || t.set(o, e.push(o) - 1);
    return i;
  }, i.range = function(a) {
    return arguments.length ? (n = Array.from(a), i) : n.slice();
  }, i.unknown = function(a) {
    return arguments.length ? (r = a, i) : r;
  }, i.copy = function() {
    return Oe(e, n).unknown(r);
  }, ie.apply(i, arguments), i;
}
function Jn(t) {
  return function() {
    return t;
  };
}
function Kn(t) {
  return +t;
}
var ke = [0, 1];
function ct(t) {
  return t;
}
function Kt(t, e) {
  return (e -= t = +t) ? function(n) {
    return (n - t) / e;
  } : Jn(isNaN(e) ? NaN : 0.5);
}
function tr(t, e) {
  var n;
  return t > e && (n = t, t = e, e = n), function(r) {
    return Math.max(t, Math.min(e, r));
  };
}
function er(t, e, n) {
  var r = t[0], i = t[1], a = e[0], o = e[1];
  return i < r ? (r = Kt(i, r), a = n(o, a)) : (r = Kt(r, i), a = n(a, o)), function(u) {
    return a(r(u));
  };
}
function nr(t, e, n) {
  var r = Math.min(t.length, e.length) - 1, i = new Array(r), a = new Array(r), o = -1;
  for (t[r] < t[0] && (t = t.slice().reverse(), e = e.slice().reverse()); ++o < r; )
    i[o] = Kt(t[o], t[o + 1]), a[o] = n(e[o], e[o + 1]);
  return function(u) {
    var l = vn(t, u, 1, r) - 1;
    return a[l](i[l](u));
  };
}
function Ie(t, e) {
  return e.domain(t.domain()).range(t.range()).interpolate(t.interpolate()).clamp(t.clamp()).unknown(t.unknown());
}
function rr() {
  var t = ke, e = ke, n = ne, r, i, a, o = ct, u, l, s;
  function g() {
    var f = Math.min(t.length, e.length);
    return o !== ct && (o = tr(t[0], t[f - 1])), u = f > 2 ? nr : er, l = s = null, h;
  }
  function h(f) {
    return f == null || isNaN(f = +f) ? a : (l || (l = u(t.map(r), e, n)))(r(o(f)));
  }
  return h.invert = function(f) {
    return o(i((s || (s = u(e, t.map(r), qt)))(f)));
  }, h.domain = function(f) {
    return arguments.length ? (t = Array.from(f, Kn), g()) : t.slice();
  }, h.range = function(f) {
    return arguments.length ? (e = Array.from(f), g()) : e.slice();
  }, h.rangeRound = function(f) {
    return e = Array.from(f), n = On, g();
  }, h.clamp = function(f) {
    return arguments.length ? (o = f ? !0 : ct, g()) : o !== ct;
  }, h.interpolate = function(f) {
    return arguments.length ? (n = f, g()) : n;
  }, h.unknown = function(f) {
    return arguments.length ? (a = f, h) : a;
  }, function(f, d) {
    return r = f, i = d, g();
  };
}
function Ee() {
  return rr()(ct, ct);
}
function ir(t, e, n, r) {
  var i = Qt(t, e, n), a;
  switch (r = Yt(r ?? ",f"), r.type) {
    case "s": {
      var o = Math.max(Math.abs(t), Math.abs(e));
      return r.precision == null && !isNaN(a = Xn(i, o)) && (r.precision = a), ze(r, o);
    }
    case "":
    case "e":
    case "g":
    case "p":
    case "r": {
      r.precision == null && !isNaN(a = Qn(i, Math.max(Math.abs(t), Math.abs(e)))) && (r.precision = a - (r.type === "e"));
      break;
    }
    case "f":
    case "%": {
      r.precision == null && !isNaN(a = qn(i)) && (r.precision = a - (r.type === "%") * 2);
      break;
    }
  }
  return Pe(r);
}
function ar(t) {
  var e = t.domain;
  return t.ticks = function(n) {
    var r = e();
    return Dn(r[0], r[r.length - 1], n ?? 10);
  }, t.tickFormat = function(n, r) {
    var i = e();
    return ir(i[0], i[i.length - 1], n ?? 10, r);
  }, t.nice = function(n) {
    n == null && (n = 10);
    var r = e(), i = 0, a = r.length - 1, o = r[i], u = r[a], l, s, g = 10;
    for (u < o && (s = o, o = u, u = s, s = i, i = a, a = s); g-- > 0; ) {
      if (s = Xt(o, u, n), s === l)
        return r[i] = o, r[a] = u, e(r);
      if (s > 0)
        o = Math.floor(o / s) * s, u = Math.ceil(u / s) * s;
      else if (s < 0)
        o = Math.ceil(o * s) / s, u = Math.floor(u * s) / s;
      else
        break;
      l = s;
    }
    return t;
  }, t;
}
function Ve() {
  var t = Ee();
  return t.copy = function() {
    return Ie(t, Ve());
  }, ie.apply(t, arguments), ar(t);
}
function or(t, e) {
  t = t.slice();
  var n = 0, r = t.length - 1, i = t[n], a = t[r], o;
  return a < i && (o = n, n = r, r = o, o = i, i = a, a = o), t[n] = e.floor(i), t[r] = e.ceil(a), t;
}
const Vt = /* @__PURE__ */ new Date(), Bt = /* @__PURE__ */ new Date();
function U(t, e, n, r) {
  function i(a) {
    return t(a = arguments.length === 0 ? /* @__PURE__ */ new Date() : /* @__PURE__ */ new Date(+a)), a;
  }
  return i.floor = (a) => (t(a = /* @__PURE__ */ new Date(+a)), a), i.ceil = (a) => (t(a = new Date(a - 1)), e(a, 1), t(a), a), i.round = (a) => {
    const o = i(a), u = i.ceil(a);
    return a - o < u - a ? o : u;
  }, i.offset = (a, o) => (e(a = /* @__PURE__ */ new Date(+a), o == null ? 1 : Math.floor(o)), a), i.range = (a, o, u) => {
    const l = [];
    if (a = i.ceil(a), u = u == null ? 1 : Math.floor(u), !(a < o) || !(u > 0)) return l;
    let s;
    do
      l.push(s = /* @__PURE__ */ new Date(+a)), e(a, u), t(a);
    while (s < a && a < o);
    return l;
  }, i.filter = (a) => U((o) => {
    if (o >= o) for (; t(o), !a(o); ) o.setTime(o - 1);
  }, (o, u) => {
    if (o >= o)
      if (u < 0) for (; ++u <= 0; )
        for (; e(o, -1), !a(o); )
          ;
      else for (; --u >= 0; )
        for (; e(o, 1), !a(o); )
          ;
  }), n && (i.count = (a, o) => (Vt.setTime(+a), Bt.setTime(+o), t(Vt), t(Bt), Math.floor(n(Vt, Bt))), i.every = (a) => (a = Math.floor(a), !isFinite(a) || !(a > 0) ? null : a > 1 ? i.filter(r ? (o) => r(o) % a === 0 : (o) => i.count(0, o) % a === 0) : i)), i;
}
const Lt = U(() => {
}, (t, e) => {
  t.setTime(+t + e);
}, (t, e) => e - t);
Lt.every = (t) => (t = Math.floor(t), !isFinite(t) || !(t > 0) ? null : t > 1 ? U((e) => {
  e.setTime(Math.floor(e / t) * t);
}, (e, n) => {
  e.setTime(+e + n * t);
}, (e, n) => (n - e) / t) : Lt);
Lt.range;
const V = 1e3, z = V * 60, B = z * 60, j = B * 24, ae = j * 7, be = j * 30, Rt = j * 365, lt = U((t) => {
  t.setTime(t - t.getMilliseconds());
}, (t, e) => {
  t.setTime(+t + e * V);
}, (t, e) => (e - t) / V, (t) => t.getUTCSeconds());
lt.range;
const oe = U((t) => {
  t.setTime(t - t.getMilliseconds() - t.getSeconds() * V);
}, (t, e) => {
  t.setTime(+t + e * z);
}, (t, e) => (e - t) / z, (t) => t.getMinutes());
oe.range;
const sr = U((t) => {
  t.setUTCSeconds(0, 0);
}, (t, e) => {
  t.setTime(+t + e * z);
}, (t, e) => (e - t) / z, (t) => t.getUTCMinutes());
sr.range;
const se = U((t) => {
  t.setTime(t - t.getMilliseconds() - t.getSeconds() * V - t.getMinutes() * z);
}, (t, e) => {
  t.setTime(+t + e * B);
}, (t, e) => (e - t) / B, (t) => t.getHours());
se.range;
const ur = U((t) => {
  t.setUTCMinutes(0, 0, 0);
}, (t, e) => {
  t.setTime(+t + e * B);
}, (t, e) => (e - t) / B, (t) => t.getUTCHours());
ur.range;
const Ct = U(
  (t) => t.setHours(0, 0, 0, 0),
  (t, e) => t.setDate(t.getDate() + e),
  (t, e) => (e - t - (e.getTimezoneOffset() - t.getTimezoneOffset()) * z) / j,
  (t) => t.getDate() - 1
);
Ct.range;
const ue = U((t) => {
  t.setUTCHours(0, 0, 0, 0);
}, (t, e) => {
  t.setUTCDate(t.getUTCDate() + e);
}, (t, e) => (e - t) / j, (t) => t.getUTCDate() - 1);
ue.range;
const cr = U((t) => {
  t.setUTCHours(0, 0, 0, 0);
}, (t, e) => {
  t.setUTCDate(t.getUTCDate() + e);
}, (t, e) => (e - t) / j, (t) => Math.floor(t / j));
cr.range;
function nt(t) {
  return U((e) => {
    e.setDate(e.getDate() - (e.getDay() + 7 - t) % 7), e.setHours(0, 0, 0, 0);
  }, (e, n) => {
    e.setDate(e.getDate() + n * 7);
  }, (e, n) => (n - e - (n.getTimezoneOffset() - e.getTimezoneOffset()) * z) / ae);
}
const Ot = nt(0), _t = nt(1), lr = nt(2), fr = nt(3), ht = nt(4), hr = nt(5), gr = nt(6);
Ot.range;
_t.range;
lr.range;
fr.range;
ht.range;
hr.range;
gr.range;
function rt(t) {
  return U((e) => {
    e.setUTCDate(e.getUTCDate() - (e.getUTCDay() + 7 - t) % 7), e.setUTCHours(0, 0, 0, 0);
  }, (e, n) => {
    e.setUTCDate(e.getUTCDate() + n * 7);
  }, (e, n) => (n - e) / ae);
}
const Be = rt(0), Ht = rt(1), mr = rt(2), dr = rt(3), gt = rt(4), pr = rt(5), yr = rt(6);
Be.range;
Ht.range;
mr.range;
dr.range;
gt.range;
pr.range;
yr.range;
const ce = U((t) => {
  t.setDate(1), t.setHours(0, 0, 0, 0);
}, (t, e) => {
  t.setMonth(t.getMonth() + e);
}, (t, e) => e.getMonth() - t.getMonth() + (e.getFullYear() - t.getFullYear()) * 12, (t) => t.getMonth());
ce.range;
const Mr = U((t) => {
  t.setUTCDate(1), t.setUTCHours(0, 0, 0, 0);
}, (t, e) => {
  t.setUTCMonth(t.getUTCMonth() + e);
}, (t, e) => e.getUTCMonth() - t.getUTCMonth() + (e.getUTCFullYear() - t.getUTCFullYear()) * 12, (t) => t.getUTCMonth());
Mr.range;
const G = U((t) => {
  t.setMonth(0, 1), t.setHours(0, 0, 0, 0);
}, (t, e) => {
  t.setFullYear(t.getFullYear() + e);
}, (t, e) => e.getFullYear() - t.getFullYear(), (t) => t.getFullYear());
G.every = (t) => !isFinite(t = Math.floor(t)) || !(t > 0) ? null : U((e) => {
  e.setFullYear(Math.floor(e.getFullYear() / t) * t), e.setMonth(0, 1), e.setHours(0, 0, 0, 0);
}, (e, n) => {
  e.setFullYear(e.getFullYear() + n * t);
});
G.range;
const et = U((t) => {
  t.setUTCMonth(0, 1), t.setUTCHours(0, 0, 0, 0);
}, (t, e) => {
  t.setUTCFullYear(t.getUTCFullYear() + e);
}, (t, e) => e.getUTCFullYear() - t.getUTCFullYear(), (t) => t.getUTCFullYear());
et.every = (t) => !isFinite(t = Math.floor(t)) || !(t > 0) ? null : U((e) => {
  e.setUTCFullYear(Math.floor(e.getUTCFullYear() / t) * t), e.setUTCMonth(0, 1), e.setUTCHours(0, 0, 0, 0);
}, (e, n) => {
  e.setUTCFullYear(e.getUTCFullYear() + n * t);
});
et.range;
function xr(t, e, n, r, i, a) {
  const o = [
    [lt, 1, V],
    [lt, 5, 5 * V],
    [lt, 15, 15 * V],
    [lt, 30, 30 * V],
    [a, 1, z],
    [a, 5, 5 * z],
    [a, 15, 15 * z],
    [a, 30, 30 * z],
    [i, 1, B],
    [i, 3, 3 * B],
    [i, 6, 6 * B],
    [i, 12, 12 * B],
    [r, 1, j],
    [r, 2, 2 * j],
    [n, 1, ae],
    [e, 1, be],
    [e, 3, 3 * be],
    [t, 1, Rt]
  ];
  function u(s, g, h) {
    const f = g < s;
    f && ([s, g] = [g, s]);
    const d = h && typeof h.range == "function" ? h : l(s, g, h), b = d ? d.range(s, +g + 1) : [];
    return f ? b.reverse() : b;
  }
  function l(s, g, h) {
    const f = Math.abs(g - s) / h, d = te(([, , F]) => F).right(o, f);
    if (d === o.length) return t.every(Qt(s / Rt, g / Rt, h));
    if (d === 0) return Lt.every(Math.max(Qt(s, g, h), 1));
    const [b, L] = o[f / o[d - 1][2] < o[d][2] / f ? d - 1 : d];
    return b.every(L);
  }
  return [u, l];
}
const [vr, wr] = xr(G, ce, Ot, Ct, se, oe);
function jt(t) {
  if (0 <= t.y && t.y < 100) {
    var e = new Date(-1, t.m, t.d, t.H, t.M, t.S, t.L);
    return e.setFullYear(t.y), e;
  }
  return new Date(t.y, t.m, t.d, t.H, t.M, t.S, t.L);
}
function Gt(t) {
  if (0 <= t.y && t.y < 100) {
    var e = new Date(Date.UTC(-1, t.m, t.d, t.H, t.M, t.S, t.L));
    return e.setUTCFullYear(t.y), e;
  }
  return new Date(Date.UTC(t.y, t.m, t.d, t.H, t.M, t.S, t.L));
}
function dt(t, e, n) {
  return { y: t, m: e, d: n, H: 0, M: 0, S: 0, L: 0 };
}
function Cr(t) {
  var e = t.dateTime, n = t.date, r = t.time, i = t.periods, a = t.days, o = t.shortDays, u = t.months, l = t.shortMonths, s = pt(i), g = yt(i), h = pt(a), f = yt(a), d = pt(o), b = yt(o), L = pt(u), F = yt(u), _ = pt(l), P = yt(l), T = {
    a: ot,
    A: st,
    b: I,
    B: Je,
    c: null,
    d: Ne,
    e: Ne,
    f: Rr,
    g: ei,
    G: ri,
    H: Er,
    I: Vr,
    j: Br,
    L: Re,
    m: jr,
    M: Gr,
    p: Ke,
    q: tn,
    Q: Le,
    s: _e,
    S: Zr,
    u: qr,
    U: Xr,
    V: Qr,
    w: Jr,
    W: Kr,
    x: null,
    X: null,
    y: ti,
    Y: ni,
    Z: ii,
    "%": Ye
  }, N = {
    a: en,
    A: nn,
    b: rn,
    B: an,
    c: null,
    d: Fe,
    e: Fe,
    f: ui,
    g: Mi,
    G: vi,
    H: ai,
    I: oi,
    j: si,
    L: Ge,
    m: ci,
    M: li,
    p: on,
    q: sn,
    Q: Le,
    s: _e,
    S: fi,
    u: hi,
    U: gi,
    V: mi,
    w: di,
    W: pi,
    x: null,
    X: null,
    y: yi,
    Y: xi,
    Z: wi,
    "%": Ye
  }, k = {
    a: E,
    A: y,
    b: $,
    B: S,
    c: X,
    d: $e,
    e: $e,
    f: Pr,
    g: Ue,
    G: De,
    H: Ae,
    I: Ae,
    j: Lr,
    L: Wr,
    m: Yr,
    M: _r,
    p: q,
    q: Fr,
    Q: Or,
    s: Ir,
    S: Hr,
    u: Dr,
    U: Ur,
    V: $r,
    w: Sr,
    W: Ar,
    x: bt,
    X: at,
    y: Ue,
    Y: De,
    Z: Nr,
    "%": zr
  };
  T.x = p(n, T), T.X = p(r, T), T.c = p(e, T), N.x = p(n, N), N.X = p(r, N), N.c = p(e, N);
  function p(m, M) {
    return function(x) {
      var c = [], Y = -1, C = 0, H = m.length, W, K, le;
      for (x instanceof Date || (x = /* @__PURE__ */ new Date(+x)); ++Y < H; )
        m.charCodeAt(Y) === 37 && (c.push(m.slice(C, Y)), (K = Se[W = m.charAt(++Y)]) != null ? W = m.charAt(++Y) : K = W === "e" ? " " : "0", (le = M[W]) && (W = le(x, K)), c.push(W), C = Y + 1);
      return c.push(m.slice(C, Y)), c.join("");
    };
  }
  function w(m, M) {
    return function(x) {
      var c = dt(1900, void 0, 1), Y = O(c, m, x += "", 0), C, H;
      if (Y != x.length) return null;
      if ("Q" in c) return new Date(c.Q);
      if ("s" in c) return new Date(c.s * 1e3 + ("L" in c ? c.L : 0));
      if (M && !("Z" in c) && (c.Z = 0), "p" in c && (c.H = c.H % 12 + c.p * 12), c.m === void 0 && (c.m = "q" in c ? c.q : 0), "V" in c) {
        if (c.V < 1 || c.V > 53) return null;
        "w" in c || (c.w = 1), "Z" in c ? (C = Gt(dt(c.y, 0, 1)), H = C.getUTCDay(), C = H > 4 || H === 0 ? Ht.ceil(C) : Ht(C), C = ue.offset(C, (c.V - 1) * 7), c.y = C.getUTCFullYear(), c.m = C.getUTCMonth(), c.d = C.getUTCDate() + (c.w + 6) % 7) : (C = jt(dt(c.y, 0, 1)), H = C.getDay(), C = H > 4 || H === 0 ? _t.ceil(C) : _t(C), C = Ct.offset(C, (c.V - 1) * 7), c.y = C.getFullYear(), c.m = C.getMonth(), c.d = C.getDate() + (c.w + 6) % 7);
      } else ("W" in c || "U" in c) && ("w" in c || (c.w = "u" in c ? c.u % 7 : "W" in c ? 1 : 0), H = "Z" in c ? Gt(dt(c.y, 0, 1)).getUTCDay() : jt(dt(c.y, 0, 1)).getDay(), c.m = 0, c.d = "W" in c ? (c.w + 6) % 7 + c.W * 7 - (H + 5) % 7 : c.w + c.U * 7 - (H + 6) % 7);
      return "Z" in c ? (c.H += c.Z / 100 | 0, c.M += c.Z % 100, Gt(c)) : jt(c);
    };
  }
  function O(m, M, x, c) {
    for (var Y = 0, C = M.length, H = x.length, W, K; Y < C; ) {
      if (c >= H) return -1;
      if (W = M.charCodeAt(Y++), W === 37) {
        if (W = M.charAt(Y++), K = k[W in Se ? M.charAt(Y++) : W], !K || (c = K(m, x, c)) < 0) return -1;
      } else if (W != x.charCodeAt(c++))
        return -1;
    }
    return c;
  }
  function q(m, M, x) {
    var c = s.exec(M.slice(x));
    return c ? (m.p = g.get(c[0].toLowerCase()), x + c[0].length) : -1;
  }
  function E(m, M, x) {
    var c = d.exec(M.slice(x));
    return c ? (m.w = b.get(c[0].toLowerCase()), x + c[0].length) : -1;
  }
  function y(m, M, x) {
    var c = h.exec(M.slice(x));
    return c ? (m.w = f.get(c[0].toLowerCase()), x + c[0].length) : -1;
  }
  function $(m, M, x) {
    var c = _.exec(M.slice(x));
    return c ? (m.m = P.get(c[0].toLowerCase()), x + c[0].length) : -1;
  }
  function S(m, M, x) {
    var c = L.exec(M.slice(x));
    return c ? (m.m = F.get(c[0].toLowerCase()), x + c[0].length) : -1;
  }
  function X(m, M, x) {
    return O(m, e, M, x);
  }
  function bt(m, M, x) {
    return O(m, n, M, x);
  }
  function at(m, M, x) {
    return O(m, r, M, x);
  }
  function ot(m) {
    return o[m.getDay()];
  }
  function st(m) {
    return a[m.getDay()];
  }
  function I(m) {
    return l[m.getMonth()];
  }
  function Je(m) {
    return u[m.getMonth()];
  }
  function Ke(m) {
    return i[+(m.getHours() >= 12)];
  }
  function tn(m) {
    return 1 + ~~(m.getMonth() / 3);
  }
  function en(m) {
    return o[m.getUTCDay()];
  }
  function nn(m) {
    return a[m.getUTCDay()];
  }
  function rn(m) {
    return l[m.getUTCMonth()];
  }
  function an(m) {
    return u[m.getUTCMonth()];
  }
  function on(m) {
    return i[+(m.getUTCHours() >= 12)];
  }
  function sn(m) {
    return 1 + ~~(m.getUTCMonth() / 3);
  }
  return {
    format: function(m) {
      var M = p(m += "", T);
      return M.toString = function() {
        return m;
      }, M;
    },
    parse: function(m) {
      var M = w(m += "", !1);
      return M.toString = function() {
        return m;
      }, M;
    },
    utcFormat: function(m) {
      var M = p(m += "", N);
      return M.toString = function() {
        return m;
      }, M;
    },
    utcParse: function(m) {
      var M = w(m += "", !0);
      return M.toString = function() {
        return m;
      }, M;
    }
  };
}
var Se = { "-": "", _: " ", 0: "0" }, A = /^\s*\d+/, Tr = /^%/, kr = /[\\^$*+?|[\]().{}]/g;
function v(t, e, n) {
  var r = t < 0 ? "-" : "", i = (r ? -t : t) + "", a = i.length;
  return r + (a < n ? new Array(n - a + 1).join(e) + i : i);
}
function br(t) {
  return t.replace(kr, "\\$&");
}
function pt(t) {
  return new RegExp("^(?:" + t.map(br).join("|") + ")", "i");
}
function yt(t) {
  return new Map(t.map((e, n) => [e.toLowerCase(), n]));
}
function Sr(t, e, n) {
  var r = A.exec(e.slice(n, n + 1));
  return r ? (t.w = +r[0], n + r[0].length) : -1;
}
function Dr(t, e, n) {
  var r = A.exec(e.slice(n, n + 1));
  return r ? (t.u = +r[0], n + r[0].length) : -1;
}
function Ur(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.U = +r[0], n + r[0].length) : -1;
}
function $r(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.V = +r[0], n + r[0].length) : -1;
}
function Ar(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.W = +r[0], n + r[0].length) : -1;
}
function De(t, e, n) {
  var r = A.exec(e.slice(n, n + 4));
  return r ? (t.y = +r[0], n + r[0].length) : -1;
}
function Ue(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.y = +r[0] + (+r[0] > 68 ? 1900 : 2e3), n + r[0].length) : -1;
}
function Nr(t, e, n) {
  var r = /^(Z)|([+-]\d\d)(?::?(\d\d))?/.exec(e.slice(n, n + 6));
  return r ? (t.Z = r[1] ? 0 : -(r[2] + (r[3] || "00")), n + r[0].length) : -1;
}
function Fr(t, e, n) {
  var r = A.exec(e.slice(n, n + 1));
  return r ? (t.q = r[0] * 3 - 3, n + r[0].length) : -1;
}
function Yr(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.m = r[0] - 1, n + r[0].length) : -1;
}
function $e(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.d = +r[0], n + r[0].length) : -1;
}
function Lr(t, e, n) {
  var r = A.exec(e.slice(n, n + 3));
  return r ? (t.m = 0, t.d = +r[0], n + r[0].length) : -1;
}
function Ae(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.H = +r[0], n + r[0].length) : -1;
}
function _r(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.M = +r[0], n + r[0].length) : -1;
}
function Hr(t, e, n) {
  var r = A.exec(e.slice(n, n + 2));
  return r ? (t.S = +r[0], n + r[0].length) : -1;
}
function Wr(t, e, n) {
  var r = A.exec(e.slice(n, n + 3));
  return r ? (t.L = +r[0], n + r[0].length) : -1;
}
function Pr(t, e, n) {
  var r = A.exec(e.slice(n, n + 6));
  return r ? (t.L = Math.floor(r[0] / 1e3), n + r[0].length) : -1;
}
function zr(t, e, n) {
  var r = Tr.exec(e.slice(n, n + 1));
  return r ? n + r[0].length : -1;
}
function Or(t, e, n) {
  var r = A.exec(e.slice(n));
  return r ? (t.Q = +r[0], n + r[0].length) : -1;
}
function Ir(t, e, n) {
  var r = A.exec(e.slice(n));
  return r ? (t.s = +r[0], n + r[0].length) : -1;
}
function Ne(t, e) {
  return v(t.getDate(), e, 2);
}
function Er(t, e) {
  return v(t.getHours(), e, 2);
}
function Vr(t, e) {
  return v(t.getHours() % 12 || 12, e, 2);
}
function Br(t, e) {
  return v(1 + Ct.count(G(t), t), e, 3);
}
function Re(t, e) {
  return v(t.getMilliseconds(), e, 3);
}
function Rr(t, e) {
  return Re(t, e) + "000";
}
function jr(t, e) {
  return v(t.getMonth() + 1, e, 2);
}
function Gr(t, e) {
  return v(t.getMinutes(), e, 2);
}
function Zr(t, e) {
  return v(t.getSeconds(), e, 2);
}
function qr(t) {
  var e = t.getDay();
  return e === 0 ? 7 : e;
}
function Xr(t, e) {
  return v(Ot.count(G(t) - 1, t), e, 2);
}
function je(t) {
  var e = t.getDay();
  return e >= 4 || e === 0 ? ht(t) : ht.ceil(t);
}
function Qr(t, e) {
  return t = je(t), v(ht.count(G(t), t) + (G(t).getDay() === 4), e, 2);
}
function Jr(t) {
  return t.getDay();
}
function Kr(t, e) {
  return v(_t.count(G(t) - 1, t), e, 2);
}
function ti(t, e) {
  return v(t.getFullYear() % 100, e, 2);
}
function ei(t, e) {
  return t = je(t), v(t.getFullYear() % 100, e, 2);
}
function ni(t, e) {
  return v(t.getFullYear() % 1e4, e, 4);
}
function ri(t, e) {
  var n = t.getDay();
  return t = n >= 4 || n === 0 ? ht(t) : ht.ceil(t), v(t.getFullYear() % 1e4, e, 4);
}
function ii(t) {
  var e = t.getTimezoneOffset();
  return (e > 0 ? "-" : (e *= -1, "+")) + v(e / 60 | 0, "0", 2) + v(e % 60, "0", 2);
}
function Fe(t, e) {
  return v(t.getUTCDate(), e, 2);
}
function ai(t, e) {
  return v(t.getUTCHours(), e, 2);
}
function oi(t, e) {
  return v(t.getUTCHours() % 12 || 12, e, 2);
}
function si(t, e) {
  return v(1 + ue.count(et(t), t), e, 3);
}
function Ge(t, e) {
  return v(t.getUTCMilliseconds(), e, 3);
}
function ui(t, e) {
  return Ge(t, e) + "000";
}
function ci(t, e) {
  return v(t.getUTCMonth() + 1, e, 2);
}
function li(t, e) {
  return v(t.getUTCMinutes(), e, 2);
}
function fi(t, e) {
  return v(t.getUTCSeconds(), e, 2);
}
function hi(t) {
  var e = t.getUTCDay();
  return e === 0 ? 7 : e;
}
function gi(t, e) {
  return v(Be.count(et(t) - 1, t), e, 2);
}
function Ze(t) {
  var e = t.getUTCDay();
  return e >= 4 || e === 0 ? gt(t) : gt.ceil(t);
}
function mi(t, e) {
  return t = Ze(t), v(gt.count(et(t), t) + (et(t).getUTCDay() === 4), e, 2);
}
function di(t) {
  return t.getUTCDay();
}
function pi(t, e) {
  return v(Ht.count(et(t) - 1, t), e, 2);
}
function yi(t, e) {
  return v(t.getUTCFullYear() % 100, e, 2);
}
function Mi(t, e) {
  return t = Ze(t), v(t.getUTCFullYear() % 100, e, 2);
}
function xi(t, e) {
  return v(t.getUTCFullYear() % 1e4, e, 4);
}
function vi(t, e) {
  var n = t.getUTCDay();
  return t = n >= 4 || n === 0 ? gt(t) : gt.ceil(t), v(t.getUTCFullYear() % 1e4, e, 4);
}
function wi() {
  return "+0000";
}
function Ye() {
  return "%";
}
function Le(t) {
  return +t;
}
function _e(t) {
  return Math.floor(+t / 1e3);
}
var ut, qe;
Ci({
  dateTime: "%x, %X",
  date: "%-m/%-d/%Y",
  time: "%-I:%M:%S %p",
  periods: ["AM", "PM"],
  days: ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"],
  shortDays: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  months: ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
  shortMonths: ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
});
function Ci(t) {
  return ut = Cr(t), qe = ut.format, ut.parse, ut.utcFormat, ut.utcParse, ut;
}
function Ti(t) {
  return new Date(t);
}
function ki(t) {
  return t instanceof Date ? +t : +/* @__PURE__ */ new Date(+t);
}
function Xe(t, e, n, r, i, a, o, u, l, s) {
  var g = Ee(), h = g.invert, f = g.domain, d = s(".%L"), b = s(":%S"), L = s("%I:%M"), F = s("%I %p"), _ = s("%a %d"), P = s("%b %d"), T = s("%B"), N = s("%Y");
  function k(p) {
    return (l(p) < p ? d : u(p) < p ? b : o(p) < p ? L : a(p) < p ? F : r(p) < p ? i(p) < p ? _ : P : n(p) < p ? T : N)(p);
  }
  return g.invert = function(p) {
    return new Date(h(p));
  }, g.domain = function(p) {
    return arguments.length ? f(Array.from(p, ki)) : f().map(Ti);
  }, g.ticks = function(p) {
    var w = f();
    return t(w[0], w[w.length - 1], p ?? 10);
  }, g.tickFormat = function(p, w) {
    return w == null ? k : s(w);
  }, g.nice = function(p) {
    var w = f();
    return (!p || typeof p.range != "function") && (p = e(w[0], w[w.length - 1], p ?? 10)), p ? f(or(w, p)) : g;
  }, g.copy = function() {
    return Ie(g, Xe(t, e, n, r, i, a, o, u, l, s));
  }, g;
}
function bi() {
  return ie.apply(Xe(vr, wr, G, ce, Ot, Ct, se, oe, lt, qe).domain([new Date(2e3, 0, 1), new Date(2e3, 0, 2)]), arguments);
}
function Si(t) {
  for (var e = t.length / 6 | 0, n = new Array(e), r = 0; r < e; ) n[r] = "#" + t.slice(r * 6, ++r * 6);
  return n;
}
const Di = Si("66c2a5fc8d628da0cbe78ac3a6d854ffd92fe5c494b3b3b3");
/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const vt = (t, e) => {
  var r;
  const n = t._$AN;
  if (n === void 0) return !1;
  for (const i of n) (r = i._$AO) == null || r.call(i, e, !1), vt(i, e);
  return !0;
}, Wt = (t) => {
  let e, n;
  do {
    if ((e = t._$AM) === void 0) break;
    n = e._$AN, n.delete(t), t = e;
  } while ((n == null ? void 0 : n.size) === 0);
}, Qe = (t) => {
  for (let e; e = t._$AM; t = e) {
    let n = e._$AN;
    if (n === void 0) e._$AN = n = /* @__PURE__ */ new Set();
    else if (n.has(t)) break;
    n.add(t), Ai(e);
  }
};
function Ui(t) {
  this._$AN !== void 0 ? (Wt(this), this._$AM = t, Qe(this)) : this._$AM = t;
}
function $i(t, e = !1, n = 0) {
  const r = this._$AH, i = this._$AN;
  if (i !== void 0 && i.size !== 0) if (e) if (Array.isArray(r)) for (let a = n; a < r.length; a++) vt(r[a], !1), Wt(r[a]);
  else r != null && (vt(r, !1), Wt(r));
  else vt(this, t);
}
const Ai = (t) => {
  t.type == mn.CHILD && (t._$AP ?? (t._$AP = $i), t._$AQ ?? (t._$AQ = Ui));
};
class Ni extends gn {
  constructor() {
    super(...arguments), this._$AN = void 0;
  }
  _$AT(e, n, r) {
    super._$AT(e, n, r), Qe(this), this.isConnected = e._$AU;
  }
  _$AO(e, n = !0) {
    var r, i;
    e !== this.isConnected && (this.isConnected = e, e ? (r = this.reconnected) == null || r.call(this) : (i = this.disconnected) == null || i.call(this)), n && (vt(this, e), Wt(this));
  }
  setValue(e) {
    if (hn(this._$Ct)) this._$Ct._$AI(e, this);
    else {
      const n = [...this._$Ct._$AH];
      n[this._$Ci] = e, this._$Ct._$AI(n, this, 0);
    }
  }
  disconnected() {
  }
  reconnected() {
  }
}
const Zt = /* @__PURE__ */ new WeakMap(), wt = dn(class extends Ni {
  render(t) {
    return fe;
  }
  update(t, [e]) {
    var r;
    const n = e !== this.Y;
    return n && this.Y !== void 0 && this.rt(void 0), (n || this.lt !== this.ct) && (this.Y = e, this.ht = (r = t.options) == null ? void 0 : r.host, this.rt(this.ct = t.element)), fe;
  }
  rt(t) {
    if (this.isConnected || (t = void 0), typeof this.Y == "function") {
      const e = this.ht ?? globalThis;
      let n = Zt.get(e);
      n === void 0 && (n = /* @__PURE__ */ new WeakMap(), Zt.set(e, n)), n.get(this.Y) !== void 0 && this.Y.call(this.ht, void 0), n.set(this.Y, t), t !== void 0 && this.Y.call(this.ht, t);
    } else this.Y.value = t;
  }
  get lt() {
    var t, e;
    return typeof this.Y == "function" ? (t = Zt.get(this.ht ?? globalThis)) == null ? void 0 : t.get(this.Y) : (e = this.Y) == null ? void 0 : e.value;
  }
  disconnected() {
    this.lt === this.ct && this.rt(void 0);
  }
  reconnected() {
    this.rt(this.ct);
  }
});
function Fi(t, e, n) {
  return bi().domain([n - e, n]).range([0, t]);
}
function Yi(t) {
  const e = He(t).tickFormat((n) => `${n.getTime() / 1e3}`);
  return tt`
    <g 
      class="axis axis--x" 
      ${wt((n) => {
    e(xt(n));
  })}>
    </g> 
  `;
}
function Li(t, e) {
  const n = He(t).tickSize(-e).tickFormat("");
  return tt`
    <g 
      class="axis-x-grid" 
      ${wt((r) => {
    n(xt(r));
  })}>
    </g> 
  `;
}
var _i = Object.defineProperty, Tt = (t, e, n, r) => {
  for (var i = void 0, a = t.length - 1, o; a >= 0; a--)
    (o = t[a]) && (i = o(e, n, i) || i);
  return i && _i(e, n, i), i;
};
class mt extends Pt {
  constructor() {
    super(...arguments), this.value = 0, this.color = "", this.hide = !1, this.yAxis = 0, this.displayName = "";
  }
}
Tt([
  D({ type: Number })
], mt.prototype, "value");
Tt([
  D({ type: String })
], mt.prototype, "color");
Tt([
  D({ type: Boolean })
], mt.prototype, "hide");
Tt([
  D({ type: Number, attribute: "y-axis" })
], mt.prototype, "yAxis");
Tt([
  D({ type: String, attribute: "display-name", reflect: !0 })
], mt.prototype, "displayName");
customElements.get("frc-line-chart-data") || customElements.define("frc-line-chart-data", mt);
var Hi = Object.defineProperty, it = (t, e, n, r) => {
  for (var i = void 0, a = t.length - 1, o; a >= 0; a--)
    (o = t[a]) && (i = o(e, n, i) || i);
  return i && Hi(e, n, i), i;
};
const Wi = Oe().domain(
  Array(8).fill(0).map((t, e) => e)
).range(Di);
var R;
const Q = (R = class extends Pt {
  constructor() {
    super(...arguments), this.viewTime = 10, this.chartTitle = "", this.data = [], this.startTime = 0, this.elapsedTimeMs = 0;
  }
  firstUpdated() {
    this.startTime = Date.now(), requestAnimationFrame(() => {
      this.updateChart();
    });
  }
  getFilteredData(e) {
    const n = e.findIndex(
      ({ timeMs: r }) => this.elapsedTimeMs - r < Math.round(this.viewTime * 1e3)
    ) - 1;
    return n < 0 ? e : e.slice(n);
  }
  updateChart() {
    this.elapsedTimeMs = Date.now() - this.startTime;
    const e = [...this.children].filter(
      (n) => n.tagName.toLowerCase() === "frc-line-chart-data"
    );
    this.data = e.map((n, r) => {
      var a;
      const i = ((a = this.data[r]) == null ? void 0 : a.data) ?? [];
      return i.push({
        value: n.value ?? 0,
        timeMs: this.elapsedTimeMs
      }), {
        data: this.getFilteredData(i),
        color: n.color || Wi(r % 8),
        hide: n.hide ?? !1,
        yAxis: n.yAxis ?? 0,
        displayName: n.displayName || `Data ${r}`
      };
    }), requestAnimationFrame(() => {
      this.updateChart();
    });
  }
  static getYScale(e, n, r) {
    const { invert: i, lockMin: a, lockMax: o } = r;
    let { min: u, max: l } = r;
    (!a || !o) && n.forEach((g) => {
      g.data.forEach(({ value: h }) => {
        a || (u = Math.min(u, h)), o || (l = Math.max(l, h));
      });
    });
    const s = i ? [0, e] : [e, 0];
    return Ve().domain([u, l]).range(s);
  }
  getDimensions() {
    var u;
    const e = { top: 0, right: 40, bottom: 20, left: 40 }, n = ((u = this.chartContainer) == null ? void 0 : u.getBoundingClientRect()) ?? {
      width: 960,
      height: 500
    }, r = n.width, i = n.height, a = r - e.left - e.right, o = i - e.top - e.bottom;
    return {
      margin: e,
      svgWidth: r,
      svgHeight: i,
      width: a,
      height: o
    };
  }
  getYScales() {
    const { height: e } = this.getDimensions(), n = [...this.children].filter(
      (r) => r.tagName.toLowerCase() === "frc-line-chart-axis"
    );
    if (n.length === 0) {
      const r = {
        min: -1,
        max: 1,
        lockMin: !1,
        lockMax: !1,
        invert: !1,
        side: "left",
        hideGridLines: !1,
        hide: !1
      };
      return [
        {
          chartAxis: r,
          scale: R.getYScale(
            e,
            this.data.filter(({ yAxis: i }) => i === 0),
            r
          )
        }
      ];
    }
    return n.map((r, i) => {
      const a = r, o = {
        min: a.min ?? -1,
        max: a.max ?? 1,
        lockMin: a.lockMin ?? !1,
        lockMax: a.lockMax ?? !1,
        invert: a.invert ?? !1,
        side: a.side ?? "left",
        hideGridLines: a.hideGridLines ?? !1,
        hide: a.hide ?? !1
      };
      return {
        chartAxis: o,
        scale: R.getYScale(
          e,
          this.data.filter(({ yAxis: u }) => u === i),
          o
        )
      };
    });
  }
  static getPath(e, n, r) {
    return `M${e.map(
      (a) => [n(a).toFixed(3), r(a).toFixed(3)].join(",")
    ).join("L")}`;
  }
  getLegend() {
    const e = [...this.children].find(
      (i) => i.tagName.toLowerCase() === "frc-line-chart-legend"
    );
    if (!e)
      return {
        direction: "horizontal",
        position: "n",
        inside: !1,
        hide: !1
      };
    const n = e;
    return {
      direction: n.direction ?? "horizontal",
      position: n.position ?? "n",
      inside: n.inside ?? !1,
      hide: n.hide ?? !1
    };
  }
  render() {
    const { inside: e, position: n, direction: r, hide: i } = this.getLegend(), a = n === "w" || n === "e" || ["ne", "nw", "se", "sw"].includes(n) && r === "vertical", o = a && ["ne", "e", "se"].includes(n) || !a && ["sw", "s", "se"].includes(n), u = Dt({
      display: e ? "block" : "flex",
      flexDirection: `${a ? "row" : "column"}${o ? "-reverse" : ""}`
    }), l = e ? {
      width: "100%",
      height: "100%"
    } : { flex: "1" };
    return St`
      <div class="chart-and-header">
        ${this.chartTitle ? St`<header>${this.chartTitle}</header>` : ""}
        <div class="chart-and-legend" style=${u}>
          ${i ? "" : this.renderLegend()}
          <div class="chart-container" style=${Dt(l)}>
            ${this.renderChart()}
          </div>
        </div>
      </div>
    `;
  }
  renderLegend() {
    const { margin: e } = this.getDimensions(), { inside: n, direction: r, position: i } = this.getLegend(), a = ["n", "s", "e", "w"].includes(i) ? "center" : i === "nw" ? "flex-start" : i === "se" ? "flex-end" : i === "ne" ? r === "horizontal" ? "flex-end" : "flex-start" : r === "vertical" ? "flex-end" : "flex-start", o = (() => {
      if (!n)
        return {};
      const s = {
        alignItems: "center"
      };
      return i.includes("n") ? s.top = `${e.top}px` : i.includes("s") && (s.bottom = `${e.bottom}px`), i.includes("w") ? s.left = `${e.left}px` : i.includes("e") && (s.right = `${e.right}px`), ["n", "s", "e", "w"].includes(i) && (s.justifyContent = "center"), ["n", "s"].includes(i) && (s.left = "0", s.right = "0"), ["e", "w"].includes(i) && (s.top = "0", s.bottom = "0"), s;
    })(), u = Dt({
      alignSelf: a,
      ...o
    });
    return St`
      <div
        class="legend ${n ? "inside" : "outside"} ${r === "horizontal" ? "horizontal" : "vertical"}"
        style=${u}
      >
        ${this.data.map(
      ({ color: s, displayName: g }) => St`
            <div class="legend-item">
              <div
                class="legend-item-box"
                style=${Dt({
        background: s
      })}
              ></div>
              <span class="legend-item-label">${g}</span>
            </div>
          `
    )}
      </div>
    `;
  }
  renderChart() {
    const { margin: e, svgWidth: n, svgHeight: r, width: i, height: a } = this.getDimensions(), o = Fi(
      i,
      Math.round(this.viewTime * 1e3),
      this.elapsedTimeMs
    ), u = this.getYScales().filter(
      ({ chartAxis: s }) => !s.hide
    ), l = u.map((s) => (h) => R.getPath(
      h,
      (f) => o(new Date(f.timeMs)),
      (f) => s.scale(f.value)
    ));
    return tt`
      <svg width=${n} height=${r} style="position: absolute">
        <g transform="translate(${e.left},${e.top})">
          <defs>
            <clipPath id="clip">
              <rect width=${i} height=${a}></rect>
            </clipPath>
          </defs>
          <g transform="translate(0,${a})">
            ${Yi(o)}
            ${Li(o, a)}
          </g>
          ${this.renderScales(u, "left")}
          ${this.renderScales(u, "right")}
          <rect class="chart-border" width=${i} height=${a}></rect>
          ${this.data.filter(
      ({ hide: s, yAxis: g }) => !s && g < u.length && g >= 0
    ).map(
      (s) => tt`
            <g clip-path="url(#clip)">
              <path 
                class="data-path"
                ${wt((g) => {
        const h = xt(g), f = l[s.yAxis](s.data);
        h.attr("class", "line").attr("stroke", s.color).attr("d", f);
      })}>
              ></path>
            </g>
          `
    )}
        </g>
      </svg>
    `;
  }
  renderScales(e, n) {
    const { width: r } = this.getDimensions();
    return tt`
     ${e.filter((i) => i.chartAxis.side === n).map((i, a) => {
      const o = i.chartAxis.side === "left" ? -a * 25 : r;
      return tt`
              <g 
                class="axis axis--y" 
                transform="translate(${o}, 0)"
                ${wt((u) => {
        (i.chartAxis.side === "left" ? ye : Ln)(i.scale)(xt(u));
      })}>
              </g>
              ${i.chartAxis.hideGridLines ? "" : tt`
                    <g 
                      class="axis-y-grid" 
                      ${wt((u) => {
        ye(i.scale).tickSize(-r).tickFormat("")(xt(u));
      })}>
                    </g> 
                  `}
            `;
    })}
    `;
  }
}, R.styles = un`
    :host {
      display: inline-block;
      width: 700px;
      height: 400px;
      color: var(--frc-line-chart-text-color, black);
    }

    .chart-and-header {
      width: 100%;
      height: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    .chart-and-header header {
      font-size: 18px;
      font-family: sans-serif;
      margin: 3px 0;
    }

    .chart-and-legend {
      flex: 1;
      width: 100%;
      position: relative;
    }

    .legend-container {
      position: relative;
    }

    .legend {
      display: flex;
      z-index: 100;
    }

    .legend.outside {
      position: relative;
    }

    .legend.inside {
      position: absolute;
    }

    .legend-item {
      display: flex;
      flex-wrap: no-wrap;
      gap: 5px;
      align-items: center;
      padding: 5px;
    }

    .legend.horizontal {
      flex-direction: row;
    }

    .legend.vertical {
      flex-direction: column;
    }

    .legend-item-box {
      width: 15px;
      height: 15px;
    }

    .legend-item-label {
      font-family: sans-serif;
      font-size: 15px;
    }

    .line {
      fill: none;
      stroke-width: 1.5px;
    }

    .chart-border {
      fill: none;
      stroke-width: 1;
      stroke: var(--frc-line-chart-border-color, black);
    }

    .axis-x-grid line,
    .axis-y-grid line {
      stroke: var(--frc-line-chart-grid-color, #eee);
    }

    .axis--y {
      font-family: monospace;
    }

    svg {
      overflow: visible;
    }
  `, R);
it([
  D({ type: Number, attribute: "view-time" })
], Q.prototype, "viewTime");
it([
  D({ type: String, attribute: "chart-title", reflect: !0 })
], Q.prototype, "chartTitle");
it([
  cn()
], Q.prototype, "data");
it([
  zt(".data-path")
], Q.prototype, "path");
it([
  zt("svg")
], Q.prototype, "svg");
it([
  zt(".axis--x")
], Q.prototype, "xAxis");
it([
  zt(".chart-container")
], Q.prototype, "chartContainer");
let Pi = Q;
customElements.get("frc-line-chart") || customElements.define("frc-line-chart", Pi);
var zi = Object.defineProperty, J = (t, e, n, r) => {
  for (var i = void 0, a = t.length - 1, o; a >= 0; a--)
    (o = t[a]) && (i = o(e, n, i) || i);
  return i && zi(e, n, i), i;
};
class Z extends Pt {
  constructor() {
    super(...arguments), this.min = -1, this.max = 1, this.lockMin = !1, this.lockMax = !1, this.invert = !1, this.side = "left", this.hideGridLines = !1, this.hide = !1;
  }
}
J([
  D({ type: Number })
], Z.prototype, "min");
J([
  D({ type: Number })
], Z.prototype, "max");
J([
  D({ type: Boolean, attribute: "lock-min" })
], Z.prototype, "lockMin");
J([
  D({ type: Boolean, attribute: "lock-max" })
], Z.prototype, "lockMax");
J([
  D({ type: Boolean })
], Z.prototype, "invert");
J([
  D({ type: String, reflect: !0 })
], Z.prototype, "side");
J([
  D({ type: Boolean, attribute: "hide-grid-lines" })
], Z.prototype, "hideGridLines");
J([
  D({ type: Boolean })
], Z.prototype, "hide");
customElements.get("frc-line-chart-axis") || customElements.define("frc-line-chart-axis", Z);
var Oi = Object.defineProperty, It = (t, e, n, r) => {
  for (var i = void 0, a = t.length - 1, o; a >= 0; a--)
    (o = t[a]) && (i = o(e, n, i) || i);
  return i && Oi(e, n, i), i;
};
const Ii = [
  "nw",
  "n",
  "ne",
  "w",
  "e",
  "sw",
  "s",
  "se"
];
class kt extends Pt {
  constructor() {
    super(...arguments), this.position = "n", this.direction = "horizontal", this.hide = !1, this.inside = !1;
  }
}
It([
  D({ type: String })
], kt.prototype, "position");
It([
  D({ type: String })
], kt.prototype, "direction");
It([
  D({ type: Boolean })
], kt.prototype, "hide");
It([
  D({ type: Boolean })
], kt.prototype, "inside");
customElements.get("frc-line-chart-legend") || customElements.define("frc-line-chart-legend", kt);
const Ei = {
  dashboard: {
    displayName: (t) => (t == null ? void 0 : t.getAttribute("chart-title")) || "Line Chart",
    defaultHtml: `
      <frc-line-chart chart-title="Line Chart">
        <frc-line-chart-legend></frc-line-chart-legend>
        <frc-line-chart-axis side="left"></frc-line-chart-axis>
        <frc-line-chart-axis side="right"></frc-line-chart-axis>
        <frc-line-chart-data></frc-line-chart-data>
      </frc-line-chart>
    `
  },
  description: "A component used to graph data over time.",
  properties: {
    chartTitle: { type: "String", attribute: "chart-title", reflect: !0 },
    viewTime: { type: "Number", attribute: "view-time", defaultValue: 10 }
  },
  slots: [
    {
      name: "",
      allowedChildren: [
        "frc-line-chart-data",
        "frc-line-chart-axis",
        "frc-line-chart-legend"
      ]
    }
  ]
}, Vi = {
  dashboard: {
    topLevel: !1,
    displayName: (t) => t ? `Chart Axis (${t.getAttribute("side") || "left"})` : "Line Chart Axis"
  },
  properties: {
    min: { type: "Number", defaultValue: -1 },
    max: { type: "Number", defaultValue: 1 },
    lockMin: { type: "Boolean", attribute: "lock-min" },
    lockMax: { type: "Boolean", attribute: "lock-max" },
    invert: { type: "Boolean" },
    side: {
      type: "String",
      defaultValue: "left",
      reflect: !0,
      input: {
        type: "StringDropdown",
        allowCustomValues: !1,
        getOptions() {
          return ["left", "right"];
        }
      }
    },
    hideGridLines: { type: "Boolean", attribute: "hide-grid-lines" },
    hide: { type: "Boolean" }
  }
}, Bi = {
  dashboard: {
    topLevel: !1,
    displayName: (t) => {
      if (!t)
        return "Line Chart Data";
      const r = [...t.parentElement.children].filter(
        (a) => a.tagName.toLowerCase() === "frc-line-chart-data"
      ).indexOf(t);
      return t.getAttribute("display-name") || `Data ${r}`;
    }
  },
  properties: {
    value: { type: "Number", primary: !0 },
    color: { type: "String", input: { type: "ColorPicker" } },
    hide: { type: "Boolean" },
    yAxis: { type: "Number", attribute: "y-axis" },
    displayName: { type: "String", attribute: "display-name", reflect: !0 }
  }
}, Ri = {
  dashboard: {
    topLevel: !1,
    displayName: "Chart Legend"
  },
  properties: {
    position: {
      type: "String",
      defaultValue: "n",
      input: {
        type: "StringDropdown",
        getOptions: () => Ii
      }
    },
    direction: {
      type: "String",
      defaultValue: "horizontal",
      input: {
        type: "StringDropdown",
        getOptions: () => ["horizontal", "vertical"]
      }
    },
    hide: { type: "Boolean" },
    inside: { type: "Boolean" }
  }
}, Ki = {
  "frc-line-chart": Ei,
  "frc-line-chart-axis": Vi,
  "frc-line-chart-data": Bi,
  "frc-line-chart-legend": Ri
};
export {
  Ii as ChartLegendPositions,
  Pi as LineChart,
  Z as LineChartAxis,
  mt as LineChartData,
  kt as LineChartLegend,
  Pi as default,
  Vi as lineChartAxisDashboardConfig,
  Ei as lineChartDashboardConfig,
  Ki as lineChartDashboardConfigs,
  Bi as lineChartDataDashboardConfig,
  Ri as lineChartLegendDashboardConfig
};
