// @frc-web-components/fwc/dist/components/pdp.es.js
import { r as h, i as g, x as a, n as c } from "../property-C4BPH-wS.js";
import "./number-bar.es.js";
import "./voltage-view.es.js";
var d = Object.defineProperty, p = (n, t, r, o) => {
  for (var e = void 0, s = n.length - 1, m; s >= 0; s--)
    (m = n[s]) && (e = m(t, r, e) || e);
  return e && d(t, r, e), e;
};
function i(n, t) {
  const r = [];
  for (let o = n; o < t; o += 1)
    r.push(o);
  return r;
}
const x = "frc-pdp", u = class u extends h {
  constructor() {
    super(...arguments), this.voltage = 0, this.totalCurrent = 0, this.channels = Array(16).fill(0);
  }
  renderChannel(t) {
    return a`
      <frc-number-bar
        class="channel"
        part="channel"
        value=${this.channels[t] ?? 0}
        min="0"
        max="40"
        center="0"
        precision="2"
        ?hide-text="${!1}"
        num-tick-marks="0"
        unit="A"
      ></frc-number-bar>
    `;
  }
  render() {
    return a`
      <div part="channels">
        ${i(0, 8).map(
      (t) => a`
            <label part="channel-label">
              <slot name="${`channel-label${t}`}">Ch. ${t}</slot>
            </label>
          `
    )}
        ${i(0, 8).map((t) => a` ${this.renderChannel(t)} `)}
        ${i(8, 16).map(
      (t) => a`
            <label part="channel-label">
              <slot name="${`channel-label${t}`}">Ch. ${t}</slot>
            </label>
          `
    )}
        ${i(8, 16).map(
      (t) => a` ${this.renderChannel(t)} `
    )}
      </div>
      <div part="voltage-and-total-current">
        <label part="voltage-label">
          <slot name="voltage-label">Voltage</slot>
        </label>
        <label part="total-current-label">
          <slot name="total-current">Total Current</slot>
        </label>
        <frc-voltage-view
          class="voltage"
          part="voltage"
          value="${this.voltage}"
          min="0"
          max="15"
          center="0"
          precision="2"
          ?hide-text="${!1}"
          num-tick-marks="0"
        ></frc-voltage-view>
        <frc-number-bar
          class="total-current"
          part="total-current"
          value="${this.totalCurrent}"
          min="0"
          max="500"
          center="0"
          precision="2"
          ?hide-text="${!1}"
          num-tick-marks="0"
          unit="A"
        ></frc-number-bar>
      </div>
    `;
  }
};
u.styles = g`
    :host {
      display: inline-block;
      width: 350px;
      margin: 5px;
      color: var(--frc-label-text-color, black);
      font-family: sans-serif;
      font-size: 16px;
    }

    [part='channels'] {
      display: grid;
      grid-auto-flow: column;
      grid-template-columns: min-content 1fr min-content 1fr;
      grid-template-rows: auto auto auto auto auto auto auto auto;
      width: 100%;
      height: auto;
      align-items: center;
      gap: 3px;
    }

    .channel,
    .voltage,
    .total-current {
      width: auto;
    }

    [part='channel-label'] {
      padding-left: 5px;
      text-align: right;
      white-space: nowrap;
    }

    [part='voltage-and-total-current'] {
      margin-top: 5px;
      display: grid;
      grid-auto-flow: column;
      grid-template-columns: min-content auto;
      grid-template-rows: auto auto;
      column-gap: 10px;
      width: 100%;
      height: auto;
      align-items: center;
      gap: 3px;
    }

    [part='voltage-and-total-current'] {
      white-space: nowrap;
    }
  `;
let l = u;
p([
  c({ type: Number })
], l.prototype, "voltage");
p([
  c({ type: Number, attribute: "total-current" })
], l.prototype, "totalCurrent");
p([
  c({ type: Array })
], l.prototype, "channels");
customElements.get("frc-pdp") || customElements.define("frc-pdp", l);
export {
  l as Pdp,
  l as default,
  x as elementName
};
