// @frc-web-components/react/dist/components/LineChart.es.js
import je from "react";
import { U as fn, M as $, H as Pe, O as De, T as K, d as gt, o as Oe } from "../property-C4BPH-wS-7fb0a59c.js";
import { n as hn } from "../state-Cn3aen3G-6b8e5a08.js";
import { h as Ee } from "../query-BrpQj_Qi-db00b239.js";
import { E as xe, A as Je, j as gn, P as dt, M as mt, l as dn } from "../transform-DR-Bw75L-3629b551.js";
import { r as mn } from "../directive-helpers-CDfQTKPA-206aa42d.js";
import { T as pn, _ as vn, s as yn } from "../directive-B76A7YXI-abc0af54.js";
import { p as $e } from "../style-map-CXHve4fi-4b8d68f4.js";
function Ue(e, t) {
  return e == null || t == null ? NaN : e < t ? -1 : e > t ? 1 : e >= t ? 0 : NaN;
}
function Mn(e, t) {
  return e == null || t == null ? NaN : t < e ? -1 : t > e ? 1 : t >= e ? 0 : NaN;
}
function nt(e) {
  let t, n, r;
  e.length !== 2 ? (t = Ue, n = (u, l) => Ue(e(u), l), r = (u, l) => e(u) - l) : (t = e === Ue || e === Mn ? e : xn, n = e, r = e);
  function i(u, l, s = 0, h = u.length) {
    if (s < h) {
      if (t(l, l) !== 0)
        return h;
      do {
        const d = s + h >>> 1;
        n(u[d], l) < 0 ? s = d + 1 : h = d;
      } while (s < h);
    }
    return s;
  }
  function o(u, l, s = 0, h = u.length) {
    if (s < h) {
      if (t(l, l) !== 0)
        return h;
      do {
        const d = s + h >>> 1;
        n(u[d], l) <= 0 ? s = d + 1 : h = d;
      } while (s < h);
    }
    return s;
  }
  function a(u, l, s = 0, h = u.length) {
    const d = i(u, l, s, h - 1);
    return d > s && r(u[d - 1], l) > -r(u[d], l) ? d - 1 : d;
  }
  return { left: i, center: a, right: o };
}
function xn() {
  return 0;
}
function wn(e) {
  return e === null ? NaN : +e;
}
const Tn = nt(Ue), Cn = Tn.right;
nt(wn).center;
class pt extends Map {
  constructor(t, n = Dn) {
    if (super(), Object.defineProperties(this, { _intern: { value: /* @__PURE__ */ new Map() }, _key: { value: n } }), t != null)
      for (const [r, i] of t)
        this.set(r, i);
  }
  get(t) {
    return super.get(vt(this, t));
  }
  has(t) {
    return super.has(vt(this, t));
  }
  set(t, n) {
    return super.set(bn(this, t), n);
  }
  delete(t) {
    return super.delete(An(this, t));
  }
}
function vt({ _intern: e, _key: t }, n) {
  const r = t(n);
  return e.has(r) ? e.get(r) : n;
}
function bn({ _intern: e, _key: t }, n) {
  const r = t(n);
  return e.has(r) ? e.get(r) : (e.set(r, n), n);
}
function An({ _intern: e, _key: t }, n) {
  const r = t(n);
  return e.has(r) && (n = e.get(r), e.delete(r)), n;
}
function Dn(e) {
  return e !== null && typeof e == "object" ? e.valueOf() : e;
}
const $n = Math.sqrt(50), kn = Math.sqrt(10), Un = Math.sqrt(2);
function Ne(e, t, n) {
  const r = (t - e) / Math.max(0, n), i = Math.floor(Math.log10(r)), o = r / Math.pow(10, i), a = o >= $n ? 10 : o >= kn ? 5 : o >= Un ? 2 : 1;
  let u, l, s;
  return i < 0 ? (s = Math.pow(10, -i) / a, u = Math.round(e * s), l = Math.round(t * s), u / s < e && ++u, l / s > t && --l, s = -s) : (s = Math.pow(10, i) * a, u = Math.round(e / s), l = Math.round(t / s), u * s < e && ++u, l * s > t && --l), l < u && 0.5 <= n && n < 2 ? Ne(e, t, n * 2) : [u, l, s];
}
function Fn(e, t, n) {
  if (t = +t, e = +e, n = +n, !(n > 0))
    return [];
  if (e === t)
    return [e];
  const r = t < e, [i, o, a] = r ? Ne(t, e, n) : Ne(e, t, n);
  if (!(o >= i))
    return [];
  const u = o - i + 1, l = new Array(u);
  if (r)
    if (a < 0)
      for (let s = 0; s < u; ++s)
        l[s] = (o - s) / -a;
    else
      for (let s = 0; s < u; ++s)
        l[s] = (o - s) * a;
  else if (a < 0)
    for (let s = 0; s < u; ++s)
      l[s] = (i + s) / -a;
  else
    for (let s = 0; s < u; ++s)
      l[s] = (i + s) * a;
  return l;
}
function Re(e, t, n) {
  return t = +t, e = +e, n = +n, Ne(e, t, n)[2];
}
function Ke(e, t, n) {
  t = +t, e = +e, n = +n;
  const r = t < e, i = r ? Re(t, e, n) : Re(e, t, n);
  return (r ? -1 : 1) * (i < 0 ? 1 / -i : i);
}
function Nn(e) {
  return e;
}
var Ie = 1, Fe = 2, et = 3, Me = 4, yt = 1e-6;
function Sn(e) {
  return "translate(" + e + ",0)";
}
function Yn(e) {
  return "translate(0," + e + ")";
}
function _n(e) {
  return (t) => +e(t);
}
function Ln(e, t) {
  return t = Math.max(0, e.bandwidth() - t * 2) / 2, e.round() && (t = Math.round(t)), (n) => +e(n) + t;
}
function Hn() {
  return !this.__axis;
}
function rt(e, t) {
  var n = [], r = null, i = null, o = 6, a = 6, u = 3, l = typeof window < "u" && window.devicePixelRatio > 1 ? 0 : 0.5, s = e === Ie || e === Me ? -1 : 1, h = e === Me || e === Fe ? "x" : "y", d = e === Ie || e === et ? Sn : Yn;
  function f(m) {
    var A = r ?? (t.ticks ? t.ticks.apply(t, n) : t.domain()), _ = i ?? (t.tickFormat ? t.tickFormat.apply(t, n) : Nn), S = Math.max(o, 0) + u, L = t.range(), j = +L[0] + l, C = +L[L.length - 1] + l, N = (t.bandwidth ? Ln : _n)(t.copy(), l), b = m.selection ? m.selection() : m, p = b.selectAll(".domain").data([null]), w = b.selectAll(".tick").data(A, t).order(), O = w.exit(), W = w.enter().append("g").attr("class", "tick"), V = w.select("line"), v = w.select("text");
    p = p.merge(p.enter().insert("path", ".tick").attr("class", "domain").attr("stroke", "currentColor")), w = w.merge(W), V = V.merge(W.append("line").attr("stroke", "currentColor").attr(h + "2", s * o)), v = v.merge(W.append("text").attr("fill", "currentColor").attr(h, s * S).attr("dy", e === Ie ? "0em" : e === et ? "0.71em" : "0.32em")), m !== b && (p = p.transition(m), w = w.transition(m), V = V.transition(m), v = v.transition(m), O = O.transition(m).attr("opacity", yt).attr("transform", function(U) {
      return isFinite(U = N(U)) ? d(U + l) : this.getAttribute("transform");
    }), W.attr("opacity", yt).attr("transform", function(U) {
      var D = this.parentNode.__axis;
      return d((D && isFinite(D = D(U)) ? D : N(U)) + l);
    })), O.remove(), p.attr("d", e === Me || e === Fe ? a ? "M" + s * a + "," + j + "H" + l + "V" + C + "H" + s * a : "M" + l + "," + j + "V" + C : a ? "M" + j + "," + s * a + "V" + l + "H" + C + "V" + s * a : "M" + j + "," + l + "H" + C), w.attr("opacity", 1).attr("transform", function(U) {
      return d(N(U) + l);
    }), V.attr(h + "2", s * o), v.attr(h, s * S).text(_), b.filter(Hn).attr("fill", "none").attr("font-size", 10).attr("font-family", "sans-serif").attr("text-anchor", e === Fe ? "start" : e === Me ? "end" : "middle"), b.each(function() {
      this.__axis = N;
    });
  }
  return f.scale = function(m) {
    return arguments.length ? (t = m, f) : t;
  }, f.ticks = function() {
    return n = Array.from(arguments), f;
  }, f.tickArguments = function(m) {
    return arguments.length ? (n = m == null ? [] : Array.from(m), f) : n.slice();
  }, f.tickValues = function(m) {
    return arguments.length ? (r = m == null ? null : Array.from(m), f) : r && r.slice();
  }, f.tickFormat = function(m) {
    return arguments.length ? (i = m, f) : i;
  }, f.tickSize = function(m) {
    return arguments.length ? (o = a = +m, f) : o;
  }, f.tickSizeInner = function(m) {
    return arguments.length ? (o = +m, f) : o;
  }, f.tickSizeOuter = function(m) {
    return arguments.length ? (a = +m, f) : a;
  }, f.tickPadding = function(m) {
    return arguments.length ? (u = +m, f) : u;
  }, f.offset = function(m) {
    return arguments.length ? (l = +m, f) : l;
  }, f;
}
function zn(e) {
  return rt(Fe, e);
}
function jt(e) {
  return rt(et, e);
}
function Mt(e) {
  return rt(Me, e);
}
function jn(e, t) {
  t || (t = []);
  var n = e ? Math.min(t.length, e.length) : 0, r = t.slice(), i;
  return function(o) {
    for (i = 0; i < n; ++i)
      r[i] = e[i] * (1 - o) + t[i] * o;
    return r;
  };
}
function Pn(e) {
  return ArrayBuffer.isView(e) && !(e instanceof DataView);
}
function On(e, t) {
  var n = t ? t.length : 0, r = e ? Math.min(n, e.length) : 0, i = new Array(r), o = new Array(n), a;
  for (a = 0; a < r; ++a)
    i[a] = it(e[a], t[a]);
  for (; a < n; ++a)
    o[a] = t[a];
  return function(u) {
    for (a = 0; a < r; ++a)
      o[a] = i[a](u);
    return o;
  };
}
function En(e, t) {
  var n = /* @__PURE__ */ new Date();
  return e = +e, t = +t, function(r) {
    return n.setTime(e * (1 - r) + t * r), n;
  };
}
function Vn(e, t) {
  var n = {}, r = {}, i;
  (e === null || typeof e != "object") && (e = {}), (t === null || typeof t != "object") && (t = {});
  for (i in t)
    i in e ? n[i] = it(e[i], t[i]) : r[i] = t[i];
  return function(o) {
    for (i in n)
      r[i] = n[i](o);
    return r;
  };
}
function it(e, t) {
  var n = typeof t, r;
  return t == null || n === "boolean" ? gn(t) : (n === "number" ? Je : n === "string" ? (r = dt(t)) ? (t = r, mt) : dn : t instanceof dt ? mt : t instanceof Date ? En : Pn(t) ? jn : Array.isArray(t) ? On : typeof t.valueOf != "function" && typeof t.toString != "function" || isNaN(t) ? Vn : Je)(e, t);
}
function Bn(e, t) {
  return e = +e, t = +t, function(n) {
    return Math.round(e * (1 - n) + t * n);
  };
}
function In(e) {
  return Math.abs(e = Math.round(e)) >= 1e21 ? e.toLocaleString("en").replace(/,/g, "") : e.toString(10);
}
function Se(e, t) {
  if ((n = (e = t ? e.toExponential(t - 1) : e.toExponential()).indexOf("e")) < 0)
    return null;
  var n, r = e.slice(0, n);
  return [
    r.length > 1 ? r[0] + r.slice(2) : r,
    +e.slice(n + 1)
  ];
}
function he(e) {
  return e = Se(Math.abs(e)), e ? e[1] : NaN;
}
function Zn(e, t) {
  return function(n, r) {
    for (var i = n.length, o = [], a = 0, u = e[0], l = 0; i > 0 && u > 0 && (l + u + 1 > r && (u = Math.max(1, r - l)), o.push(n.substring(i -= u, i + u)), !((l += u + 1) > r)); )
      u = e[a = (a + 1) % e.length];
    return o.reverse().join(t);
  };
}
function qn(e) {
  return function(t) {
    return t.replace(/[0-9]/g, function(n) {
      return e[+n];
    });
  };
}
var Gn = /^(?:(.)?([<>=^]))?([+\-( ])?([$#])?(0)?(\d+)?(,)?(\.\d+)?(~)?([a-z%])?$/i;
function Ye(e) {
  if (!(t = Gn.exec(e)))
    throw new Error("invalid format: " + e);
  var t;
  return new ot({
    fill: t[1],
    align: t[2],
    sign: t[3],
    symbol: t[4],
    zero: t[5],
    width: t[6],
    comma: t[7],
    precision: t[8] && t[8].slice(1),
    trim: t[9],
    type: t[10]
  });
}
Ye.prototype = ot.prototype;
function ot(e) {
  this.fill = e.fill === void 0 ? " " : e.fill + "", this.align = e.align === void 0 ? ">" : e.align + "", this.sign = e.sign === void 0 ? "-" : e.sign + "", this.symbol = e.symbol === void 0 ? "" : e.symbol + "", this.zero = !!e.zero, this.width = e.width === void 0 ? void 0 : +e.width, this.comma = !!e.comma, this.precision = e.precision === void 0 ? void 0 : +e.precision, this.trim = !!e.trim, this.type = e.type === void 0 ? "" : e.type + "";
}
ot.prototype.toString = function() {
  return this.fill + this.align + this.sign + this.symbol + (this.zero ? "0" : "") + (this.width === void 0 ? "" : Math.max(1, this.width | 0)) + (this.comma ? "," : "") + (this.precision === void 0 ? "" : "." + Math.max(0, this.precision | 0)) + (this.trim ? "~" : "") + this.type;
};
function Wn(e) {
  e:
    for (var t = e.length, n = 1, r = -1, i; n < t; ++n)
      switch (e[n]) {
        case ".":
          r = i = n;
          break;
        case "0":
          r === 0 && (r = n), i = n;
          break;
        default:
          if (!+e[n])
            break e;
          r > 0 && (r = 0);
          break;
      }
  return r > 0 ? e.slice(0, r) + e.slice(i + 1) : e;
}
var Pt;
function Xn(e, t) {
  var n = Se(e, t);
  if (!n)
    return e + "";
  var r = n[0], i = n[1], o = i - (Pt = Math.max(-8, Math.min(8, Math.floor(i / 3))) * 3) + 1, a = r.length;
  return o === a ? r : o > a ? r + new Array(o - a + 1).join("0") : o > 0 ? r.slice(0, o) + "." + r.slice(o) : "0." + new Array(1 - o).join("0") + Se(e, Math.max(0, t + o - 1))[0];
}
function xt(e, t) {
  var n = Se(e, t);
  if (!n)
    return e + "";
  var r = n[0], i = n[1];
  return i < 0 ? "0." + new Array(-i).join("0") + r : r.length > i + 1 ? r.slice(0, i + 1) + "." + r.slice(i + 1) : r + new Array(i - r.length + 2).join("0");
}
const wt = {
  "%": (e, t) => (e * 100).toFixed(t),
  b: (e) => Math.round(e).toString(2),
  c: (e) => e + "",
  d: In,
  e: (e, t) => e.toExponential(t),
  f: (e, t) => e.toFixed(t),
  g: (e, t) => e.toPrecision(t),
  o: (e) => Math.round(e).toString(8),
  p: (e, t) => xt(e * 100, t),
  r: xt,
  s: Xn,
  X: (e) => Math.round(e).toString(16).toUpperCase(),
  x: (e) => Math.round(e).toString(16)
};
function Tt(e) {
  return e;
}
var Ct = Array.prototype.map, bt = ["y", "z", "a", "f", "p", "n", "µ", "m", "", "k", "M", "G", "T", "P", "E", "Z", "Y"];
function Qn(e) {
  var t = e.grouping === void 0 || e.thousands === void 0 ? Tt : Zn(Ct.call(e.grouping, Number), e.thousands + ""), n = e.currency === void 0 ? "" : e.currency[0] + "", r = e.currency === void 0 ? "" : e.currency[1] + "", i = e.decimal === void 0 ? "." : e.decimal + "", o = e.numerals === void 0 ? Tt : qn(Ct.call(e.numerals, String)), a = e.percent === void 0 ? "%" : e.percent + "", u = e.minus === void 0 ? "−" : e.minus + "", l = e.nan === void 0 ? "NaN" : e.nan + "";
  function s(d) {
    d = Ye(d);
    var f = d.fill, m = d.align, A = d.sign, _ = d.symbol, S = d.zero, L = d.width, j = d.comma, C = d.precision, N = d.trim, b = d.type;
    b === "n" ? (j = !0, b = "g") : wt[b] || (C === void 0 && (C = 12), N = !0, b = "g"), (S || f === "0" && m === "=") && (S = !0, f = "0", m = "=");
    var p = _ === "$" ? n : _ === "#" && /[boxX]/.test(b) ? "0" + b.toLowerCase() : "", w = _ === "$" ? r : /[%p]/.test(b) ? a : "", O = wt[b], W = /[defgprs%]/.test(b);
    C = C === void 0 ? 6 : /[gprs]/.test(b) ? Math.max(1, Math.min(21, C)) : Math.max(0, Math.min(20, C));
    function V(v) {
      var U = p, D = w, X, Ae, oe;
      if (b === "c")
        D = O(v) + D, v = "";
      else {
        v = +v;
        var ae = v < 0 || 1 / v < 0;
        if (v = isNaN(v) ? l : O(Math.abs(v), C), N && (v = Wn(v)), ae && +v == 0 && A !== "+" && (ae = !1), U = (ae ? A === "(" ? A : u : A === "-" || A === "(" ? "" : A) + U, D = (b === "s" ? bt[8 + Pt / 3] : "") + D + (ae && A === "(" ? ")" : ""), W) {
          for (X = -1, Ae = v.length; ++X < Ae; )
            if (oe = v.charCodeAt(X), 48 > oe || oe > 57) {
              D = (oe === 46 ? i + v.slice(X + 1) : v.slice(X)) + D, v = v.slice(0, X);
              break;
            }
        }
      }
      j && !S && (v = t(v, 1 / 0));
      var se = U.length + v.length + D.length, E = se < L ? new Array(L - se + 1).join(f) : "";
      switch (j && S && (v = t(E + v, E.length ? L - D.length : 1 / 0), E = ""), m) {
        case "<":
          v = U + v + D + E;
          break;
        case "=":
          v = U + E + v + D;
          break;
        case "^":
          v = E.slice(0, se = E.length >> 1) + U + v + D + E.slice(se);
          break;
        default:
          v = E + U + v + D;
          break;
      }
      return o(v);
    }
    return V.toString = function() {
      return d + "";
    }, V;
  }
  function h(d, f) {
    var m = s((d = Ye(d), d.type = "f", d)), A = Math.max(-8, Math.min(8, Math.floor(he(f) / 3))) * 3, _ = Math.pow(10, -A), S = bt[8 + A / 3];
    return function(L) {
      return m(_ * L) + S;
    };
  }
  return {
    format: s,
    formatPrefix: h
  };
}
var ke, Ot, Et;
Jn({
  thousands: ",",
  grouping: [3],
  currency: ["$", ""]
});
function Jn(e) {
  return ke = Qn(e), Ot = ke.format, Et = ke.formatPrefix, ke;
}
function Rn(e) {
  return Math.max(0, -he(Math.abs(e)));
}
function Kn(e, t) {
  return Math.max(0, Math.max(-8, Math.min(8, Math.floor(he(t) / 3))) * 3 - he(Math.abs(e)));
}
function er(e, t) {
  return e = Math.abs(e), t = Math.abs(t) - e, Math.max(0, he(t) - he(e)) + 1;
}
function at(e, t) {
  switch (arguments.length) {
    case 0:
      break;
    case 1:
      this.range(e);
      break;
    default:
      this.range(t).domain(e);
      break;
  }
  return this;
}
const At = Symbol("implicit");
function Vt() {
  var e = new pt(), t = [], n = [], r = At;
  function i(o) {
    let a = e.get(o);
    if (a === void 0) {
      if (r !== At)
        return r;
      e.set(o, a = t.push(o) - 1);
    }
    return n[a % n.length];
  }
  return i.domain = function(o) {
    if (!arguments.length)
      return t.slice();
    t = [], e = new pt();
    for (const a of o)
      e.has(a) || e.set(a, t.push(a) - 1);
    return i;
  }, i.range = function(o) {
    return arguments.length ? (n = Array.from(o), i) : n.slice();
  }, i.unknown = function(o) {
    return arguments.length ? (r = o, i) : r;
  }, i.copy = function() {
    return Vt(t, n).unknown(r);
  }, at.apply(i, arguments), i;
}
function tr(e) {
  return function() {
    return e;
  };
}
function nr(e) {
  return +e;
}
var Dt = [0, 1];
function le(e) {
  return e;
}
function tt(e, t) {
  return (t -= e = +e) ? function(n) {
    return (n - e) / t;
  } : tr(isNaN(t) ? NaN : 0.5);
}
function rr(e, t) {
  var n;
  return e > t && (n = e, e = t, t = n), function(r) {
    return Math.max(e, Math.min(t, r));
  };
}
function ir(e, t, n) {
  var r = e[0], i = e[1], o = t[0], a = t[1];
  return i < r ? (r = tt(i, r), o = n(a, o)) : (r = tt(r, i), o = n(o, a)), function(u) {
    return o(r(u));
  };
}
function or(e, t, n) {
  var r = Math.min(e.length, t.length) - 1, i = new Array(r), o = new Array(r), a = -1;
  for (e[r] < e[0] && (e = e.slice().reverse(), t = t.slice().reverse()); ++a < r; )
    i[a] = tt(e[a], e[a + 1]), o[a] = n(t[a], t[a + 1]);
  return function(u) {
    var l = Cn(e, u, 1, r) - 1;
    return o[l](i[l](u));
  };
}
function Bt(e, t) {
  return t.domain(e.domain()).range(e.range()).interpolate(e.interpolate()).clamp(e.clamp()).unknown(e.unknown());
}
function ar() {
  var e = Dt, t = Dt, n = it, r, i, o, a = le, u, l, s;
  function h() {
    var f = Math.min(e.length, t.length);
    return a !== le && (a = rr(e[0], e[f - 1])), u = f > 2 ? or : ir, l = s = null, d;
  }
  function d(f) {
    return f == null || isNaN(f = +f) ? o : (l || (l = u(e.map(r), t, n)))(r(a(f)));
  }
  return d.invert = function(f) {
    return a(i((s || (s = u(t, e.map(r), Je)))(f)));
  }, d.domain = function(f) {
    return arguments.length ? (e = Array.from(f, nr), h()) : e.slice();
  }, d.range = function(f) {
    return arguments.length ? (t = Array.from(f), h()) : t.slice();
  }, d.rangeRound = function(f) {
    return t = Array.from(f), n = Bn, h();
  }, d.clamp = function(f) {
    return arguments.length ? (a = f ? !0 : le, h()) : a !== le;
  }, d.interpolate = function(f) {
    return arguments.length ? (n = f, h()) : n;
  }, d.unknown = function(f) {
    return arguments.length ? (o = f, d) : o;
  }, function(f, m) {
    return r = f, i = m, h();
  };
}
function It() {
  return ar()(le, le);
}
function sr(e, t, n, r) {
  var i = Ke(e, t, n), o;
  switch (r = Ye(r ?? ",f"), r.type) {
    case "s": {
      var a = Math.max(Math.abs(e), Math.abs(t));
      return r.precision == null && !isNaN(o = Kn(i, a)) && (r.precision = o), Et(r, a);
    }
    case "":
    case "e":
    case "g":
    case "p":
    case "r": {
      r.precision == null && !isNaN(o = er(i, Math.max(Math.abs(e), Math.abs(t)))) && (r.precision = o - (r.type === "e"));
      break;
    }
    case "f":
    case "%": {
      r.precision == null && !isNaN(o = Rn(i)) && (r.precision = o - (r.type === "%") * 2);
      break;
    }
  }
  return Ot(r);
}
function ur(e) {
  var t = e.domain;
  return e.ticks = function(n) {
    var r = t();
    return Fn(r[0], r[r.length - 1], n ?? 10);
  }, e.tickFormat = function(n, r) {
    var i = t();
    return sr(i[0], i[i.length - 1], n ?? 10, r);
  }, e.nice = function(n) {
    n == null && (n = 10);
    var r = t(), i = 0, o = r.length - 1, a = r[i], u = r[o], l, s, h = 10;
    for (u < a && (s = a, a = u, u = s, s = i, i = o, o = s); h-- > 0; ) {
      if (s = Re(a, u, n), s === l)
        return r[i] = a, r[o] = u, t(r);
      if (s > 0)
        a = Math.floor(a / s) * s, u = Math.ceil(u / s) * s;
      else if (s < 0)
        a = Math.ceil(a * s) / s, u = Math.floor(u * s) / s;
      else
        break;
      l = s;
    }
    return e;
  }, e;
}
function Zt() {
  var e = It();
  return e.copy = function() {
    return Bt(e, Zt());
  }, at.apply(e, arguments), ur(e);
}
function cr(e, t) {
  e = e.slice();
  var n = 0, r = e.length - 1, i = e[n], o = e[r], a;
  return o < i && (a = n, n = r, r = a, a = i, i = o, o = a), e[n] = t.floor(i), e[r] = t.ceil(o), e;
}
const Ze = /* @__PURE__ */ new Date(), qe = /* @__PURE__ */ new Date();
function k(e, t, n, r) {
  function i(o) {
    return e(o = arguments.length === 0 ? /* @__PURE__ */ new Date() : /* @__PURE__ */ new Date(+o)), o;
  }
  return i.floor = (o) => (e(o = /* @__PURE__ */ new Date(+o)), o), i.ceil = (o) => (e(o = new Date(o - 1)), t(o, 1), e(o), o), i.round = (o) => {
    const a = i(o), u = i.ceil(o);
    return o - a < u - o ? a : u;
  }, i.offset = (o, a) => (t(o = /* @__PURE__ */ new Date(+o), a == null ? 1 : Math.floor(a)), o), i.range = (o, a, u) => {
    const l = [];
    if (o = i.ceil(o), u = u == null ? 1 : Math.floor(u), !(o < a) || !(u > 0))
      return l;
    let s;
    do
      l.push(s = /* @__PURE__ */ new Date(+o)), t(o, u), e(o);
    while (s < o && o < a);
    return l;
  }, i.filter = (o) => k((a) => {
    if (a >= a)
      for (; e(a), !o(a); )
        a.setTime(a - 1);
  }, (a, u) => {
    if (a >= a)
      if (u < 0)
        for (; ++u <= 0; )
          for (; t(a, -1), !o(a); )
            ;
      else
        for (; --u >= 0; )
          for (; t(a, 1), !o(a); )
            ;
  }), n && (i.count = (o, a) => (Ze.setTime(+o), qe.setTime(+a), e(Ze), e(qe), Math.floor(n(Ze, qe))), i.every = (o) => (o = Math.floor(o), !isFinite(o) || !(o > 0) ? null : o > 1 ? i.filter(r ? (a) => r(a) % o === 0 : (a) => i.count(0, a) % o === 0) : i)), i;
}
const _e = k(() => {
}, (e, t) => {
  e.setTime(+e + t);
}, (e, t) => t - e);
_e.every = (e) => (e = Math.floor(e), !isFinite(e) || !(e > 0) ? null : e > 1 ? k((t) => {
  t.setTime(Math.floor(t / e) * e);
}, (t, n) => {
  t.setTime(+t + n * e);
}, (t, n) => (n - t) / e) : _e);
_e.range;
const I = 1e3, P = I * 60, Z = P * 60, q = Z * 24, st = q * 7, $t = q * 30, Ge = q * 365, fe = k((e) => {
  e.setTime(e - e.getMilliseconds());
}, (e, t) => {
  e.setTime(+e + t * I);
}, (e, t) => (t - e) / I, (e) => e.getUTCSeconds());
fe.range;
const ut = k((e) => {
  e.setTime(e - e.getMilliseconds() - e.getSeconds() * I);
}, (e, t) => {
  e.setTime(+e + t * P);
}, (e, t) => (t - e) / P, (e) => e.getMinutes());
ut.range;
const lr = k((e) => {
  e.setUTCSeconds(0, 0);
}, (e, t) => {
  e.setTime(+e + t * P);
}, (e, t) => (t - e) / P, (e) => e.getUTCMinutes());
lr.range;
const ct = k((e) => {
  e.setTime(e - e.getMilliseconds() - e.getSeconds() * I - e.getMinutes() * P);
}, (e, t) => {
  e.setTime(+e + t * Z);
}, (e, t) => (t - e) / Z, (e) => e.getHours());
ct.range;
const fr = k((e) => {
  e.setUTCMinutes(0, 0, 0);
}, (e, t) => {
  e.setTime(+e + t * Z);
}, (e, t) => (t - e) / Z, (e) => e.getUTCHours());
fr.range;
const Ce = k(
  (e) => e.setHours(0, 0, 0, 0),
  (e, t) => e.setDate(e.getDate() + t),
  (e, t) => (t - e - (t.getTimezoneOffset() - e.getTimezoneOffset()) * P) / q,
  (e) => e.getDate() - 1
);
Ce.range;
const lt = k((e) => {
  e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCDate(e.getUTCDate() + t);
}, (e, t) => (t - e) / q, (e) => e.getUTCDate() - 1);
lt.range;
const hr = k((e) => {
  e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCDate(e.getUTCDate() + t);
}, (e, t) => (t - e) / q, (e) => Math.floor(e / q));
hr.range;
function te(e) {
  return k((t) => {
    t.setDate(t.getDate() - (t.getDay() + 7 - e) % 7), t.setHours(0, 0, 0, 0);
  }, (t, n) => {
    t.setDate(t.getDate() + n * 7);
  }, (t, n) => (n - t - (n.getTimezoneOffset() - t.getTimezoneOffset()) * P) / st);
}
const Ve = te(0), Le = te(1), gr = te(2), dr = te(3), ge = te(4), mr = te(5), pr = te(6);
Ve.range;
Le.range;
gr.range;
dr.range;
ge.range;
mr.range;
pr.range;
function ne(e) {
  return k((t) => {
    t.setUTCDate(t.getUTCDate() - (t.getUTCDay() + 7 - e) % 7), t.setUTCHours(0, 0, 0, 0);
  }, (t, n) => {
    t.setUTCDate(t.getUTCDate() + n * 7);
  }, (t, n) => (n - t) / st);
}
const qt = ne(0), He = ne(1), vr = ne(2), yr = ne(3), de = ne(4), Mr = ne(5), xr = ne(6);
qt.range;
He.range;
vr.range;
yr.range;
de.range;
Mr.range;
xr.range;
const ft = k((e) => {
  e.setDate(1), e.setHours(0, 0, 0, 0);
}, (e, t) => {
  e.setMonth(e.getMonth() + t);
}, (e, t) => t.getMonth() - e.getMonth() + (t.getFullYear() - e.getFullYear()) * 12, (e) => e.getMonth());
ft.range;
const wr = k((e) => {
  e.setUTCDate(1), e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCMonth(e.getUTCMonth() + t);
}, (e, t) => t.getUTCMonth() - e.getUTCMonth() + (t.getUTCFullYear() - e.getUTCFullYear()) * 12, (e) => e.getUTCMonth());
wr.range;
const G = k((e) => {
  e.setMonth(0, 1), e.setHours(0, 0, 0, 0);
}, (e, t) => {
  e.setFullYear(e.getFullYear() + t);
}, (e, t) => t.getFullYear() - e.getFullYear(), (e) => e.getFullYear());
G.every = (e) => !isFinite(e = Math.floor(e)) || !(e > 0) ? null : k((t) => {
  t.setFullYear(Math.floor(t.getFullYear() / e) * e), t.setMonth(0, 1), t.setHours(0, 0, 0, 0);
}, (t, n) => {
  t.setFullYear(t.getFullYear() + n * e);
});
G.range;
const ee = k((e) => {
  e.setUTCMonth(0, 1), e.setUTCHours(0, 0, 0, 0);
}, (e, t) => {
  e.setUTCFullYear(e.getUTCFullYear() + t);
}, (e, t) => t.getUTCFullYear() - e.getUTCFullYear(), (e) => e.getUTCFullYear());
ee.every = (e) => !isFinite(e = Math.floor(e)) || !(e > 0) ? null : k((t) => {
  t.setUTCFullYear(Math.floor(t.getUTCFullYear() / e) * e), t.setUTCMonth(0, 1), t.setUTCHours(0, 0, 0, 0);
}, (t, n) => {
  t.setUTCFullYear(t.getUTCFullYear() + n * e);
});
ee.range;
function Tr(e, t, n, r, i, o) {
  const a = [
    [fe, 1, I],
    [fe, 5, 5 * I],
    [fe, 15, 15 * I],
    [fe, 30, 30 * I],
    [o, 1, P],
    [o, 5, 5 * P],
    [o, 15, 15 * P],
    [o, 30, 30 * P],
    [i, 1, Z],
    [i, 3, 3 * Z],
    [i, 6, 6 * Z],
    [i, 12, 12 * Z],
    [r, 1, q],
    [r, 2, 2 * q],
    [n, 1, st],
    [t, 1, $t],
    [t, 3, 3 * $t],
    [e, 1, Ge]
  ];
  function u(s, h, d) {
    const f = h < s;
    f && ([s, h] = [h, s]);
    const m = d && typeof d.range == "function" ? d : l(s, h, d), A = m ? m.range(s, +h + 1) : [];
    return f ? A.reverse() : A;
  }
  function l(s, h, d) {
    const f = Math.abs(h - s) / d, m = nt(([, , S]) => S).right(a, f);
    if (m === a.length)
      return e.every(Ke(s / Ge, h / Ge, d));
    if (m === 0)
      return _e.every(Math.max(Ke(s, h, d), 1));
    const [A, _] = a[f / a[m - 1][2] < a[m][2] / f ? m - 1 : m];
    return A.every(_);
  }
  return [u, l];
}
const [Cr, br] = Tr(G, ft, Ve, Ce, ct, ut);
function We(e) {
  if (0 <= e.y && e.y < 100) {
    var t = new Date(-1, e.m, e.d, e.H, e.M, e.S, e.L);
    return t.setFullYear(e.y), t;
  }
  return new Date(e.y, e.m, e.d, e.H, e.M, e.S, e.L);
}
function Xe(e) {
  if (0 <= e.y && e.y < 100) {
    var t = new Date(Date.UTC(-1, e.m, e.d, e.H, e.M, e.S, e.L));
    return t.setUTCFullYear(e.y), t;
  }
  return new Date(Date.UTC(e.y, e.m, e.d, e.H, e.M, e.S, e.L));
}
function pe(e, t, n) {
  return { y: e, m: t, d: n, H: 0, M: 0, S: 0, L: 0 };
}
function Ar(e) {
  var t = e.dateTime, n = e.date, r = e.time, i = e.periods, o = e.days, a = e.shortDays, u = e.months, l = e.shortMonths, s = ve(i), h = ye(i), d = ve(o), f = ye(o), m = ve(a), A = ye(a), _ = ve(u), S = ye(u), L = ve(l), j = ye(l), C = {
    a: ae,
    A: se,
    b: E,
    B: tn,
    c: null,
    d: Yt,
    e: Yt,
    f: Wr,
    g: ii,
    G: ai,
    H: Zr,
    I: qr,
    j: Gr,
    L: Gt,
    m: Xr,
    M: Qr,
    p: nn,
    q: rn,
    Q: Ht,
    s: zt,
    S: Jr,
    u: Rr,
    U: Kr,
    V: ei,
    w: ti,
    W: ni,
    x: null,
    X: null,
    y: ri,
    Y: oi,
    Z: si,
    "%": Lt
  }, N = {
    a: on,
    A: an,
    b: sn,
    B: un,
    c: null,
    d: _t,
    e: _t,
    f: fi,
    g: wi,
    G: Ci,
    H: ui,
    I: ci,
    j: li,
    L: Xt,
    m: hi,
    M: gi,
    p: cn,
    q: ln,
    Q: Ht,
    s: zt,
    S: di,
    u: mi,
    U: pi,
    V: vi,
    w: yi,
    W: Mi,
    x: null,
    X: null,
    y: xi,
    Y: Ti,
    Z: bi,
    "%": Lt
  }, b = {
    a: V,
    A: v,
    b: U,
    B: D,
    c: X,
    d: Nt,
    e: Nt,
    f: Er,
    g: Ft,
    G: Ut,
    H: St,
    I: St,
    j: zr,
    L: Or,
    m: Hr,
    M: jr,
    p: W,
    q: Lr,
    Q: Br,
    s: Ir,
    S: Pr,
    u: Fr,
    U: Nr,
    V: Sr,
    w: Ur,
    W: Yr,
    x: Ae,
    X: oe,
    y: Ft,
    Y: Ut,
    Z: _r,
    "%": Vr
  };
  C.x = p(n, C), C.X = p(r, C), C.c = p(t, C), N.x = p(n, N), N.X = p(r, N), N.c = p(t, N);
  function p(g, y) {
    return function(M) {
      var c = [], Y = -1, T = 0, H = g.length, z, R, ht;
      for (M instanceof Date || (M = /* @__PURE__ */ new Date(+M)); ++Y < H; )
        g.charCodeAt(Y) === 37 && (c.push(g.slice(T, Y)), (R = kt[z = g.charAt(++Y)]) != null ? z = g.charAt(++Y) : R = z === "e" ? " " : "0", (ht = y[z]) && (z = ht(M, R)), c.push(z), T = Y + 1);
      return c.push(g.slice(T, Y)), c.join("");
    };
  }
  function w(g, y) {
    return function(M) {
      var c = pe(1900, void 0, 1), Y = O(c, g, M += "", 0), T, H;
      if (Y != M.length)
        return null;
      if ("Q" in c)
        return new Date(c.Q);
      if ("s" in c)
        return new Date(c.s * 1e3 + ("L" in c ? c.L : 0));
      if (y && !("Z" in c) && (c.Z = 0), "p" in c && (c.H = c.H % 12 + c.p * 12), c.m === void 0 && (c.m = "q" in c ? c.q : 0), "V" in c) {
        if (c.V < 1 || c.V > 53)
          return null;
        "w" in c || (c.w = 1), "Z" in c ? (T = Xe(pe(c.y, 0, 1)), H = T.getUTCDay(), T = H > 4 || H === 0 ? He.ceil(T) : He(T), T = lt.offset(T, (c.V - 1) * 7), c.y = T.getUTCFullYear(), c.m = T.getUTCMonth(), c.d = T.getUTCDate() + (c.w + 6) % 7) : (T = We(pe(c.y, 0, 1)), H = T.getDay(), T = H > 4 || H === 0 ? Le.ceil(T) : Le(T), T = Ce.offset(T, (c.V - 1) * 7), c.y = T.getFullYear(), c.m = T.getMonth(), c.d = T.getDate() + (c.w + 6) % 7);
      } else
        ("W" in c || "U" in c) && ("w" in c || (c.w = "u" in c ? c.u % 7 : "W" in c ? 1 : 0), H = "Z" in c ? Xe(pe(c.y, 0, 1)).getUTCDay() : We(pe(c.y, 0, 1)).getDay(), c.m = 0, c.d = "W" in c ? (c.w + 6) % 7 + c.W * 7 - (H + 5) % 7 : c.w + c.U * 7 - (H + 6) % 7);
      return "Z" in c ? (c.H += c.Z / 100 | 0, c.M += c.Z % 100, Xe(c)) : We(c);
    };
  }
  function O(g, y, M, c) {
    for (var Y = 0, T = y.length, H = M.length, z, R; Y < T; ) {
      if (c >= H)
        return -1;
      if (z = y.charCodeAt(Y++), z === 37) {
        if (z = y.charAt(Y++), R = b[z in kt ? y.charAt(Y++) : z], !R || (c = R(g, M, c)) < 0)
          return -1;
      } else if (z != M.charCodeAt(c++))
        return -1;
    }
    return c;
  }
  function W(g, y, M) {
    var c = s.exec(y.slice(M));
    return c ? (g.p = h.get(c[0].toLowerCase()), M + c[0].length) : -1;
  }
  function V(g, y, M) {
    var c = m.exec(y.slice(M));
    return c ? (g.w = A.get(c[0].toLowerCase()), M + c[0].length) : -1;
  }
  function v(g, y, M) {
    var c = d.exec(y.slice(M));
    return c ? (g.w = f.get(c[0].toLowerCase()), M + c[0].length) : -1;
  }
  function U(g, y, M) {
    var c = L.exec(y.slice(M));
    return c ? (g.m = j.get(c[0].toLowerCase()), M + c[0].length) : -1;
  }
  function D(g, y, M) {
    var c = _.exec(y.slice(M));
    return c ? (g.m = S.get(c[0].toLowerCase()), M + c[0].length) : -1;
  }
  function X(g, y, M) {
    return O(g, t, y, M);
  }
  function Ae(g, y, M) {
    return O(g, n, y, M);
  }
  function oe(g, y, M) {
    return O(g, r, y, M);
  }
  function ae(g) {
    return a[g.getDay()];
  }
  function se(g) {
    return o[g.getDay()];
  }
  function E(g) {
    return l[g.getMonth()];
  }
  function tn(g) {
    return u[g.getMonth()];
  }
  function nn(g) {
    return i[+(g.getHours() >= 12)];
  }
  function rn(g) {
    return 1 + ~~(g.getMonth() / 3);
  }
  function on(g) {
    return a[g.getUTCDay()];
  }
  function an(g) {
    return o[g.getUTCDay()];
  }
  function sn(g) {
    return l[g.getUTCMonth()];
  }
  function un(g) {
    return u[g.getUTCMonth()];
  }
  function cn(g) {
    return i[+(g.getUTCHours() >= 12)];
  }
  function ln(g) {
    return 1 + ~~(g.getUTCMonth() / 3);
  }
  return {
    format: function(g) {
      var y = p(g += "", C);
      return y.toString = function() {
        return g;
      }, y;
    },
    parse: function(g) {
      var y = w(g += "", !1);
      return y.toString = function() {
        return g;
      }, y;
    },
    utcFormat: function(g) {
      var y = p(g += "", N);
      return y.toString = function() {
        return g;
      }, y;
    },
    utcParse: function(g) {
      var y = w(g += "", !0);
      return y.toString = function() {
        return g;
      }, y;
    }
  };
}
var kt = { "-": "", _: " ", 0: "0" }, F = /^\s*\d+/, Dr = /^%/, $r = /[\\^$*+?|[\]().{}]/g;
function x(e, t, n) {
  var r = e < 0 ? "-" : "", i = (r ? -e : e) + "", o = i.length;
  return r + (o < n ? new Array(n - o + 1).join(t) + i : i);
}
function kr(e) {
  return e.replace($r, "\\$&");
}
function ve(e) {
  return new RegExp("^(?:" + e.map(kr).join("|") + ")", "i");
}
function ye(e) {
  return new Map(e.map((t, n) => [t.toLowerCase(), n]));
}
function Ur(e, t, n) {
  var r = F.exec(t.slice(n, n + 1));
  return r ? (e.w = +r[0], n + r[0].length) : -1;
}
function Fr(e, t, n) {
  var r = F.exec(t.slice(n, n + 1));
  return r ? (e.u = +r[0], n + r[0].length) : -1;
}
function Nr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.U = +r[0], n + r[0].length) : -1;
}
function Sr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.V = +r[0], n + r[0].length) : -1;
}
function Yr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.W = +r[0], n + r[0].length) : -1;
}
function Ut(e, t, n) {
  var r = F.exec(t.slice(n, n + 4));
  return r ? (e.y = +r[0], n + r[0].length) : -1;
}
function Ft(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.y = +r[0] + (+r[0] > 68 ? 1900 : 2e3), n + r[0].length) : -1;
}
function _r(e, t, n) {
  var r = /^(Z)|([+-]\d\d)(?::?(\d\d))?/.exec(t.slice(n, n + 6));
  return r ? (e.Z = r[1] ? 0 : -(r[2] + (r[3] || "00")), n + r[0].length) : -1;
}
function Lr(e, t, n) {
  var r = F.exec(t.slice(n, n + 1));
  return r ? (e.q = r[0] * 3 - 3, n + r[0].length) : -1;
}
function Hr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.m = r[0] - 1, n + r[0].length) : -1;
}
function Nt(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.d = +r[0], n + r[0].length) : -1;
}
function zr(e, t, n) {
  var r = F.exec(t.slice(n, n + 3));
  return r ? (e.m = 0, e.d = +r[0], n + r[0].length) : -1;
}
function St(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.H = +r[0], n + r[0].length) : -1;
}
function jr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.M = +r[0], n + r[0].length) : -1;
}
function Pr(e, t, n) {
  var r = F.exec(t.slice(n, n + 2));
  return r ? (e.S = +r[0], n + r[0].length) : -1;
}
function Or(e, t, n) {
  var r = F.exec(t.slice(n, n + 3));
  return r ? (e.L = +r[0], n + r[0].length) : -1;
}
function Er(e, t, n) {
  var r = F.exec(t.slice(n, n + 6));
  return r ? (e.L = Math.floor(r[0] / 1e3), n + r[0].length) : -1;
}
function Vr(e, t, n) {
  var r = Dr.exec(t.slice(n, n + 1));
  return r ? n + r[0].length : -1;
}
function Br(e, t, n) {
  var r = F.exec(t.slice(n));
  return r ? (e.Q = +r[0], n + r[0].length) : -1;
}
function Ir(e, t, n) {
  var r = F.exec(t.slice(n));
  return r ? (e.s = +r[0], n + r[0].length) : -1;
}
function Yt(e, t) {
  return x(e.getDate(), t, 2);
}
function Zr(e, t) {
  return x(e.getHours(), t, 2);
}
function qr(e, t) {
  return x(e.getHours() % 12 || 12, t, 2);
}
function Gr(e, t) {
  return x(1 + Ce.count(G(e), e), t, 3);
}
function Gt(e, t) {
  return x(e.getMilliseconds(), t, 3);
}
function Wr(e, t) {
  return Gt(e, t) + "000";
}
function Xr(e, t) {
  return x(e.getMonth() + 1, t, 2);
}
function Qr(e, t) {
  return x(e.getMinutes(), t, 2);
}
function Jr(e, t) {
  return x(e.getSeconds(), t, 2);
}
function Rr(e) {
  var t = e.getDay();
  return t === 0 ? 7 : t;
}
function Kr(e, t) {
  return x(Ve.count(G(e) - 1, e), t, 2);
}
function Wt(e) {
  var t = e.getDay();
  return t >= 4 || t === 0 ? ge(e) : ge.ceil(e);
}
function ei(e, t) {
  return e = Wt(e), x(ge.count(G(e), e) + (G(e).getDay() === 4), t, 2);
}
function ti(e) {
  return e.getDay();
}
function ni(e, t) {
  return x(Le.count(G(e) - 1, e), t, 2);
}
function ri(e, t) {
  return x(e.getFullYear() % 100, t, 2);
}
function ii(e, t) {
  return e = Wt(e), x(e.getFullYear() % 100, t, 2);
}
function oi(e, t) {
  return x(e.getFullYear() % 1e4, t, 4);
}
function ai(e, t) {
  var n = e.getDay();
  return e = n >= 4 || n === 0 ? ge(e) : ge.ceil(e), x(e.getFullYear() % 1e4, t, 4);
}
function si(e) {
  var t = e.getTimezoneOffset();
  return (t > 0 ? "-" : (t *= -1, "+")) + x(t / 60 | 0, "0", 2) + x(t % 60, "0", 2);
}
function _t(e, t) {
  return x(e.getUTCDate(), t, 2);
}
function ui(e, t) {
  return x(e.getUTCHours(), t, 2);
}
function ci(e, t) {
  return x(e.getUTCHours() % 12 || 12, t, 2);
}
function li(e, t) {
  return x(1 + lt.count(ee(e), e), t, 3);
}
function Xt(e, t) {
  return x(e.getUTCMilliseconds(), t, 3);
}
function fi(e, t) {
  return Xt(e, t) + "000";
}
function hi(e, t) {
  return x(e.getUTCMonth() + 1, t, 2);
}
function gi(e, t) {
  return x(e.getUTCMinutes(), t, 2);
}
function di(e, t) {
  return x(e.getUTCSeconds(), t, 2);
}
function mi(e) {
  var t = e.getUTCDay();
  return t === 0 ? 7 : t;
}
function pi(e, t) {
  return x(qt.count(ee(e) - 1, e), t, 2);
}
function Qt(e) {
  var t = e.getUTCDay();
  return t >= 4 || t === 0 ? de(e) : de.ceil(e);
}
function vi(e, t) {
  return e = Qt(e), x(de.count(ee(e), e) + (ee(e).getUTCDay() === 4), t, 2);
}
function yi(e) {
  return e.getUTCDay();
}
function Mi(e, t) {
  return x(He.count(ee(e) - 1, e), t, 2);
}
function xi(e, t) {
  return x(e.getUTCFullYear() % 100, t, 2);
}
function wi(e, t) {
  return e = Qt(e), x(e.getUTCFullYear() % 100, t, 2);
}
function Ti(e, t) {
  return x(e.getUTCFullYear() % 1e4, t, 4);
}
function Ci(e, t) {
  var n = e.getUTCDay();
  return e = n >= 4 || n === 0 ? de(e) : de.ceil(e), x(e.getUTCFullYear() % 1e4, t, 4);
}
function bi() {
  return "+0000";
}
function Lt() {
  return "%";
}
function Ht(e) {
  return +e;
}
function zt(e) {
  return Math.floor(+e / 1e3);
}
var ue, Jt;
Ai({
  dateTime: "%x, %X",
  date: "%-m/%-d/%Y",
  time: "%-I:%M:%S %p",
  periods: ["AM", "PM"],
  days: ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"],
  shortDays: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  months: ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
  shortMonths: ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
});
function Ai(e) {
  return ue = Ar(e), Jt = ue.format, ue.parse, ue.utcFormat, ue.utcParse, ue;
}
function Di(e) {
  return new Date(e);
}
function $i(e) {
  return e instanceof Date ? +e : +/* @__PURE__ */ new Date(+e);
}
function Rt(e, t, n, r, i, o, a, u, l, s) {
  var h = It(), d = h.invert, f = h.domain, m = s(".%L"), A = s(":%S"), _ = s("%I:%M"), S = s("%I %p"), L = s("%a %d"), j = s("%b %d"), C = s("%B"), N = s("%Y");
  function b(p) {
    return (l(p) < p ? m : u(p) < p ? A : a(p) < p ? _ : o(p) < p ? S : r(p) < p ? i(p) < p ? L : j : n(p) < p ? C : N)(p);
  }
  return h.invert = function(p) {
    return new Date(d(p));
  }, h.domain = function(p) {
    return arguments.length ? f(Array.from(p, $i)) : f().map(Di);
  }, h.ticks = function(p) {
    var w = f();
    return e(w[0], w[w.length - 1], p ?? 10);
  }, h.tickFormat = function(p, w) {
    return w == null ? b : s(w);
  }, h.nice = function(p) {
    var w = f();
    return (!p || typeof p.range != "function") && (p = t(w[0], w[w.length - 1], p ?? 10)), p ? f(cr(w, p)) : h;
  }, h.copy = function() {
    return Bt(h, Rt(e, t, n, r, i, o, a, u, l, s));
  }, h;
}
function ki() {
  return at.apply(Rt(Cr, br, G, ft, Ve, Ce, ct, ut, fe, Jt).domain([new Date(2e3, 0, 1), new Date(2e3, 0, 2)]), arguments);
}
function Ui(e) {
  for (var t = e.length / 6 | 0, n = new Array(t), r = 0; r < t; )
    n[r] = "#" + e.slice(r * 6, ++r * 6);
  return n;
}
const Fi = Ui("66c2a5fc8d628da0cbe78ac3a6d854ffd92fe5c494b3b3b3");
/**
 * @license
 * Copyright 2017 Google LLC
 * SPDX-License-Identifier: BSD-3-Clause
 */
const we = (e, t) => {
  var n;
  const r = e._$AN;
  if (r === void 0)
    return !1;
  for (const i of r)
    (n = i._$AO) == null || n.call(i, t, !1), we(i, t);
  return !0;
}, ze = (e) => {
  let t, n;
  do {
    if ((t = e._$AM) === void 0)
      break;
    n = t._$AN, n.delete(e), e = t;
  } while ((n == null ? void 0 : n.size) === 0);
}, Kt = (e) => {
  for (let t; t = e._$AM; e = t) {
    let n = t._$AN;
    if (n === void 0)
      t._$AN = n = /* @__PURE__ */ new Set();
    else if (n.has(e))
      break;
    n.add(e), Yi(t);
  }
};
function Ni(e) {
  this._$AN !== void 0 ? (ze(this), this._$AM = e, Kt(this)) : this._$AM = e;
}
function Si(e, t = !1, n = 0) {
  const r = this._$AH, i = this._$AN;
  if (i !== void 0 && i.size !== 0)
    if (t)
      if (Array.isArray(r))
        for (let o = n; o < r.length; o++)
          we(r[o], !1), ze(r[o]);
      else
        r != null && (we(r, !1), ze(r));
    else
      we(this, e);
}
const Yi = (e) => {
  e.type == yn.CHILD && (e._$AP ?? (e._$AP = Si), e._$AQ ?? (e._$AQ = Ni));
};
class _i extends vn {
  constructor() {
    super(...arguments), this._$AN = void 0;
  }
  _$AT(t, n, r) {
    super._$AT(t, n, r), Kt(this), this.isConnected = t._$AU;
  }
  _$AO(t, n = !0) {
    var r, i;
    t !== this.isConnected && (this.isConnected = t, t ? (r = this.reconnected) == null || r.call(this) : (i = this.disconnected) == null || i.call(this)), n && (we(this, t), ze(this));
  }
  setValue(t) {
    if (mn(this._$Ct))
      this._$Ct._$AI(t, this);
    else {
      const n = [...this._$Ct._$AH];
      n[this._$Ci] = t, this._$Ct._$AI(n, this, 0);
    }
  }
  disconnected() {
  }
  reconnected() {
  }
}
const Qe = /* @__PURE__ */ new WeakMap(), Te = pn(class extends _i {
  render(e) {
    return gt;
  }
  update(e, [t]) {
    var n;
    const r = t !== this.Y;
    return r && this.Y !== void 0 && this.rt(void 0), (r || this.lt !== this.ct) && (this.Y = t, this.ht = (n = e.options) == null ? void 0 : n.host, this.rt(this.ct = e.element)), gt;
  }
  rt(e) {
    if (this.isConnected || (e = void 0), typeof this.Y == "function") {
      const t = this.ht ?? globalThis;
      let n = Qe.get(t);
      n === void 0 && (n = /* @__PURE__ */ new WeakMap(), Qe.set(t, n)), n.get(this.Y) !== void 0 && this.Y.call(this.ht, void 0), n.set(this.Y, e), e !== void 0 && this.Y.call(this.ht, e);
    } else
      this.Y.value = e;
  }
  get lt() {
    var e, t;
    return typeof this.Y == "function" ? (e = Qe.get(this.ht ?? globalThis)) == null ? void 0 : e.get(this.Y) : (t = this.Y) == null ? void 0 : t.value;
  }
  disconnected() {
    this.lt === this.ct && this.rt(void 0);
  }
  reconnected() {
    this.rt(this.ct);
  }
});
function Li(e, t, n) {
  return ki().domain([n - t, n]).range([0, e]);
}
function Hi(e) {
  const t = jt(e).tickFormat((n) => `${n.getTime() / 1e3}`);
  return K`
    <g 
      class="axis axis--x" 
      ${Te((n) => {
    t(xe(n));
  })}>
    </g> 
  `;
}
function zi(e, t) {
  const n = jt(e).tickSize(-t).tickFormat("");
  return K`
    <g 
      class="axis-x-grid" 
      ${Te((r) => {
    n(xe(r));
  })}>
    </g> 
  `;
}
var ji = Object.defineProperty, be = (e, t, n, r) => {
  for (var i = void 0, o = e.length - 1, a; o >= 0; o--)
    (a = e[o]) && (i = a(t, n, i) || i);
  return i && ji(t, n, i), i;
};
class re extends Pe {
  constructor() {
    super(...arguments), this.value = 0, this.color = "", this.hide = !1, this.yAxis = 0, this.displayName = "";
  }
}
be([
  $({ type: Number })
], re.prototype, "value");
be([
  $({ type: String })
], re.prototype, "color");
be([
  $({ type: Boolean })
], re.prototype, "hide");
be([
  $({ type: Number, attribute: "y-axis" })
], re.prototype, "yAxis");
be([
  $({ type: String, attribute: "display-name", reflect: !0 })
], re.prototype, "displayName");
customElements.get("frc-line-chart-data") || customElements.define("frc-line-chart-data", re);
var Pi = Object.defineProperty, ie = (e, t, n, r) => {
  for (var i = void 0, o = e.length - 1, a; o >= 0; o--)
    (a = e[o]) && (i = a(t, n, i) || i);
  return i && Pi(t, n, i), i;
};
const Oi = Vt().domain(
  Array(8).fill(0).map((e, t) => t)
).range(Fi);
var ce;
const Q = (ce = class extends Pe {
  constructor() {
    super(...arguments), this.viewTime = 10, this.chartTitle = "", this.data = [], this.startTime = 0, this.elapsedTimeMs = 0;
  }
  firstUpdated() {
    this.startTime = Date.now(), requestAnimationFrame(() => {
      this.updateChart();
    });
  }
  getFilteredData(e) {
    const t = e.findIndex(
      ({ timeMs: n }) => this.elapsedTimeMs - n < Math.round(this.viewTime * 1e3)
    ) - 1;
    return t < 0 ? e : e.slice(t);
  }
  updateChart() {
    this.elapsedTimeMs = Date.now() - this.startTime;
    const e = [...this.children].filter(
      (t) => t.tagName.toLowerCase() === "frc-line-chart-data"
    );
    this.data = e.map((t, n) => {
      var r;
      const i = ((r = this.data[n]) == null ? void 0 : r.data) ?? [];
      return i.push({
        value: t.value ?? 0,
        timeMs: this.elapsedTimeMs
      }), {
        data: this.getFilteredData(i),
        color: t.color || Oi(n % 8),
        hide: t.hide ?? !1,
        yAxis: t.yAxis ?? 0,
        displayName: t.displayName || `Data ${n}`
      };
    }), requestAnimationFrame(() => {
      this.updateChart();
    });
  }
  static getYScale(e, t, n) {
    const { invert: r, lockMin: i, lockMax: o } = n;
    let { min: a, max: u } = n;
    (!i || !o) && t.forEach((s) => {
      s.data.forEach(({ value: h }) => {
        i || (a = Math.min(a, h)), o || (u = Math.max(u, h));
      });
    });
    const l = r ? [0, e] : [e, 0];
    return Zt().domain([a, u]).range(l);
  }
  getDimensions() {
    var e;
    const t = { top: 0, right: 40, bottom: 20, left: 40 }, n = ((e = this.chartContainer) == null ? void 0 : e.getBoundingClientRect()) ?? {
      width: 960,
      height: 500
    }, r = n.width, i = n.height, o = r - t.left - t.right, a = i - t.top - t.bottom;
    return {
      margin: t,
      svgWidth: r,
      svgHeight: i,
      width: o,
      height: a
    };
  }
  getYScales() {
    const { height: e } = this.getDimensions(), t = [...this.children].filter(
      (n) => n.tagName.toLowerCase() === "frc-line-chart-axis"
    );
    if (t.length === 0) {
      const n = {
        min: -1,
        max: 1,
        lockMin: !1,
        lockMax: !1,
        invert: !1,
        side: "left",
        hideGridLines: !1,
        hide: !1
      };
      return [
        {
          chartAxis: n,
          scale: ce.getYScale(
            e,
            this.data.filter(({ yAxis: r }) => r === 0),
            n
          )
        }
      ];
    }
    return t.map((n, r) => {
      const i = n, o = {
        min: i.min ?? -1,
        max: i.max ?? 1,
        lockMin: i.lockMin ?? !1,
        lockMax: i.lockMax ?? !1,
        invert: i.invert ?? !1,
        side: i.side ?? "left",
        hideGridLines: i.hideGridLines ?? !1,
        hide: i.hide ?? !1
      };
      return {
        chartAxis: o,
        scale: ce.getYScale(
          e,
          this.data.filter(({ yAxis: a }) => a === r),
          o
        )
      };
    });
  }
  static getPath(e, t, n) {
    return `M${e.map(
      (r) => [t(r).toFixed(3), n(r).toFixed(3)].join(",")
    ).join("L")}`;
  }
  getLegend() {
    const e = [...this.children].find(
      (n) => n.tagName.toLowerCase() === "frc-line-chart-legend"
    );
    if (!e)
      return {
        direction: "horizontal",
        position: "n",
        inside: !1,
        hide: !1
      };
    const t = e;
    return {
      direction: t.direction ?? "horizontal",
      position: t.position ?? "n",
      inside: t.inside ?? !1,
      hide: t.hide ?? !1
    };
  }
  render() {
    const { inside: e, position: t, direction: n, hide: r } = this.getLegend(), i = t === "w" || t === "e" || ["ne", "nw", "se", "sw"].includes(t) && n === "vertical", o = i && ["ne", "e", "se"].includes(t) || !i && ["sw", "s", "se"].includes(t), a = $e({
      display: e ? "block" : "flex",
      flexDirection: `${i ? "row" : "column"}${o ? "-reverse" : ""}`
    }), u = e ? {
      width: "100%",
      height: "100%"
    } : { flex: "1" };
    return De`
      <div class="chart-and-header">
        ${this.chartTitle ? De`<header>${this.chartTitle}</header>` : ""}
        <div class="chart-and-legend" style=${a}>
          ${r ? "" : this.renderLegend()}
          <div class="chart-container" style=${$e(u)}>
            ${this.renderChart()}
          </div>
        </div>
      </div>
    `;
  }
  renderLegend() {
    const { margin: e } = this.getDimensions(), { inside: t, direction: n, position: r } = this.getLegend(), i = ["n", "s", "e", "w"].includes(r) ? "center" : r === "nw" ? "flex-start" : r === "se" ? "flex-end" : r === "ne" ? n === "horizontal" ? "flex-end" : "flex-start" : n === "vertical" ? "flex-end" : "flex-start", o = (() => {
      if (!t)
        return {};
      const u = {
        alignItems: "center"
      };
      return r.includes("n") ? u.top = `${e.top}px` : r.includes("s") && (u.bottom = `${e.bottom}px`), r.includes("w") ? u.left = `${e.left}px` : r.includes("e") && (u.right = `${e.right}px`), ["n", "s", "e", "w"].includes(r) && (u.justifyContent = "center"), ["n", "s"].includes(r) && (u.left = "0", u.right = "0"), ["e", "w"].includes(r) && (u.top = "0", u.bottom = "0"), u;
    })(), a = $e({
      alignSelf: i,
      ...o
    });
    return De`
      <div
        class="legend ${t ? "inside" : "outside"} ${n === "horizontal" ? "horizontal" : "vertical"}"
        style=${a}
      >
        ${this.data.map(
      ({ color: u, displayName: l }) => De`
            <div class="legend-item">
              <div
                class="legend-item-box"
                style=${$e({
        background: u
      })}
              ></div>
              <span class="legend-item-label">${l}</span>
            </div>
          `
    )}
      </div>
    `;
  }
  renderChart() {
    const { margin: e, svgWidth: t, svgHeight: n, width: r, height: i } = this.getDimensions(), o = Li(
      r,
      Math.round(this.viewTime * 1e3),
      this.elapsedTimeMs
    ), a = this.getYScales().filter(
      ({ chartAxis: l }) => !l.hide
    ), u = a.map((l) => (s) => ce.getPath(
      s,
      (h) => o(new Date(h.timeMs)),
      (h) => l.scale(h.value)
    ));
    return K`
      <svg width=${t} height=${n} style="position: absolute">
        <g transform="translate(${e.left},${e.top})">
          <defs>
            <clipPath id="clip">
              <rect width=${r} height=${i}></rect>
            </clipPath>
          </defs>
          <g transform="translate(0,${i})">
            ${Hi(o)}
            ${zi(o, i)}
          </g>
          ${this.renderScales(a, "left")}
          ${this.renderScales(a, "right")}
          <rect class="chart-border" width=${r} height=${i}></rect>
          ${this.data.filter(
      ({ hide: l, yAxis: s }) => !l && s < a.length && s >= 0
    ).map(
      (l) => K`
            <g clip-path="url(#clip)">
              <path 
                class="data-path"
                ${Te((s) => {
        const h = xe(s), d = u[l.yAxis](l.data);
        h.attr("class", "line").attr("stroke", l.color).attr("d", d);
      })}>
              ></path>
            </g>
          `
    )}
        </g>
      </svg>
    `;
  }
  renderScales(e, t) {
    const { width: n } = this.getDimensions();
    return K`
     ${e.filter((r) => r.chartAxis.side === t).map((r, i) => {
      const o = r.chartAxis.side === "left" ? -i * 25 : n;
      return K`
              <g 
                class="axis axis--y" 
                transform="translate(${o}, 0)"
                ${Te((a) => {
        (r.chartAxis.side === "left" ? Mt : zn)(r.scale)(xe(a));
      })}>
              </g>
              ${r.chartAxis.hideGridLines ? "" : K`
                    <g 
                      class="axis-y-grid" 
                      ${Te((a) => {
        Mt(r.scale).tickSize(-n).tickFormat("")(xe(a));
      })}>
                    </g> 
                  `}
            `;
    })}
    `;
  }
}, ce.styles = fn`
    :host {
      display: inline-block;
      width: 700px;
      height: 400px;
      color: var(--frc-line-chart-text-color, black);
    }

    .chart-and-header {
      width: 100%;
      height: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    .chart-and-header header {
      font-size: 18px;
      font-family: sans-serif;
      margin: 3px 0;
    }

    .chart-and-legend {
      flex: 1;
      width: 100%;
      position: relative;
    }

    .legend-container {
      position: relative;
    }

    .legend {
      display: flex;
      z-index: 100;
    }

    .legend.outside {
      position: relative;
    }

    .legend.inside {
      position: absolute;
    }

    .legend-item {
      display: flex;
      flex-wrap: no-wrap;
      gap: 5px;
      align-items: center;
      padding: 5px;
    }

    .legend.horizontal {
      flex-direction: row;
    }

    .legend.vertical {
      flex-direction: column;
    }

    .legend-item-box {
      width: 15px;
      height: 15px;
    }

    .legend-item-label {
      font-family: sans-serif;
      font-size: 15px;
    }

    .line {
      fill: none;
      stroke-width: 1.5px;
    }

    .chart-border {
      fill: none;
      stroke-width: 1;
      stroke: var(--frc-line-chart-border-color, black);
    }

    .axis-x-grid line,
    .axis-y-grid line {
      stroke: var(--frc-line-chart-grid-color, #eee);
    }

    .axis--y {
      font-family: monospace;
    }

    svg {
      overflow: visible;
    }
  `, ce);
ie([
  $({ type: Number, attribute: "view-time" })
], Q.prototype, "viewTime");
ie([
  $({ type: String, attribute: "chart-title", reflect: !0 })
], Q.prototype, "chartTitle");
ie([
  hn()
], Q.prototype, "data");
ie([
  Ee(".data-path")
], Q.prototype, "path");
ie([
  Ee("svg")
], Q.prototype, "svg");
ie([
  Ee(".axis--x")
], Q.prototype, "xAxis");
ie([
  Ee(".chart-container")
], Q.prototype, "chartContainer");
let en = Q;
customElements.get("frc-line-chart") || customElements.define("frc-line-chart", en);
var Ei = Object.defineProperty, J = (e, t, n, r) => {
  for (var i = void 0, o = e.length - 1, a; o >= 0; o--)
    (a = e[o]) && (i = a(t, n, i) || i);
  return i && Ei(t, n, i), i;
};
class B extends Pe {
  constructor() {
    super(...arguments), this.min = -1, this.max = 1, this.lockMin = !1, this.lockMax = !1, this.invert = !1, this.side = "left", this.hideGridLines = !1, this.hide = !1;
  }
}
J([
  $({ type: Number })
], B.prototype, "min");
J([
  $({ type: Number })
], B.prototype, "max");
J([
  $({ type: Boolean, attribute: "lock-min" })
], B.prototype, "lockMin");
J([
  $({ type: Boolean, attribute: "lock-max" })
], B.prototype, "lockMax");
J([
  $({ type: Boolean })
], B.prototype, "invert");
J([
  $({ type: String, reflect: !0 })
], B.prototype, "side");
J([
  $({ type: Boolean, attribute: "hide-grid-lines" })
], B.prototype, "hideGridLines");
J([
  $({ type: Boolean })
], B.prototype, "hide");
customElements.get("frc-line-chart-axis") || customElements.define("frc-line-chart-axis", B);
var Vi = Object.defineProperty, Be = (e, t, n, r) => {
  for (var i = void 0, o = e.length - 1, a; o >= 0; o--)
    (a = e[o]) && (i = a(t, n, i) || i);
  return i && Vi(t, n, i), i;
};
class me extends Pe {
  constructor() {
    super(...arguments), this.position = "n", this.direction = "horizontal", this.hide = !1, this.inside = !1;
  }
}
Be([
  $({ type: String })
], me.prototype, "position");
Be([
  $({ type: String })
], me.prototype, "direction");
Be([
  $({ type: Boolean })
], me.prototype, "hide");
Be([
  $({ type: Boolean })
], me.prototype, "inside");
customElements.get("frc-line-chart-legend") || customElements.define("frc-line-chart-legend", me);
const Ji = Oe({
  tagName: "frc-line-chart",
  elementClass: en,
  react: je
}), Ri = Oe({
  tagName: "frc-line-chart-axis",
  elementClass: B,
  react: je
}), Ki = Oe({
  tagName: "frc-line-chart-data",
  elementClass: re,
  react: je
}), eo = Oe({
  tagName: "frc-line-chart-legend",
  elementClass: me,
  react: je
});
export {
  Ji as LineChart,
  Ri as LineChartAxis,
  Ki as LineChartData,
  eo as LineChartLegend,
  Ji as default
};
