// @frc-web-components/react/dist/components/NetworkAlerts.es.js
import p from "react";
import { U as v, M as o, H as q, O as t, o as f } from "../property-C4BPH-wS-7fb0a59c.js";
import { o as n } from "../when-CDK1Tt5Y-835602a5.js";
import "../icon-Dc8f99z4-878f257b.js";
var m = Object.defineProperty, l = (a, e, c, b) => {
  for (var i = void 0, s = a.length - 1, d; s >= 0; s--)
    (d = a[s]) && (i = d(e, c, i) || i);
  return i && m(e, c, i), i;
};
const T = "m40-120 440-760 440 760H40Zm138-80h604L480-720 178-200Zm302-40q17 0 28.5-11.5T520-280q0-17-11.5-28.5T480-320q-17 0-28.5 11.5T440-280q0 17 11.5 28.5T480-240Zm-40-120h80v-200h-80v200Zm40-100Z", g = "M480-280q17 0 28.5-11.5T520-320q0-17-11.5-28.5T480-360q-17 0-28.5 11.5T440-320q0 17 11.5 28.5T480-280Zm-40-160h80v-240h-80v240Zm40 360q-83 0-156-31.5T197-197q-54-54-85.5-127T80-480q0-83 31.5-156T197-763q54-54 127-85.5T480-880q83 0 156 31.5T763-763q54 54 85.5 127T880-480q0 83-31.5 156T763-197q-54 54-127 85.5T480-80Zm0-80q134 0 227-93t93-227q0-134-93-227t-227-93q-134 0-227 93t-93 227q0 134 93 227t227 93Zm0-320Z", x = "M440-280h80v-240h-80v240Zm40-320q17 0 28.5-11.5T520-640q0-17-11.5-28.5T480-680q-17 0-28.5 11.5T440-640q0 17 11.5 28.5T480-600Zm0 520q-83 0-156-31.5T197-197q-54-54-85.5-127T80-480q0-83 31.5-156T197-763q54-54 127-85.5T480-880q83 0 156 31.5T763-763q54 54 85.5 127T880-480q0 83-31.5 156T763-197q-54 54-127 85.5T480-80Zm0-80q134 0 227-93t93-227q0-134-93-227t-227-93q-134 0-227 93t-93 227q0 134 93 227t227 93Zm0-320Z", h = class extends q {
  constructor() {
    super(...arguments), this.errors = [], this.warnings = [], this.infos = [], this.level = "info", this.hideTitle = !1;
  }
  render() {
    return t`
      <div class="alerts-container">
        ${n(!this.hideTitle, () => t` <header>Alerts</header> `)}
        <div class="alerts-list">
          ${this.errors.map(
      (e) => t`
              <div class="alert">
                <div class="icon">
                  <frc-icon
                    color="red"
                    svg-path=${g}
                    view-box="0 -960 960 960"
                  ></frc-icon>
                </div>
                <label>${e}</label>
              </div>
            `
    )}
          ${n(
      this.level !== "error",
      () => t`
              ${this.warnings.map(
        (e) => t`
                  <div class="alert">
                    <div class="icon">
                      <frc-icon
                        color="yellow"
                        svg-path=${T}
                        view-box="0 -960 960 960"
                      ></frc-icon>
                    </div>
                    <label>${e}</label>
                  </div>
                `
      )}
            `
    )}
          ${n(
      this.level === "info",
      () => t`
              ${this.infos.map(
        (e) => t`
                  <div class="alert">
                    <div class="icon">
                      <frc-icon
                        color="green"
                        svg-path=${x}
                        view-box="0 -960 960 960"
                      ></frc-icon>
                    </div>
                    <label>${e}</label>
                  </div>
                `
      )}
            `
    )}
        </div>
      </div>
    `;
  }
};
h.styles = v`
    :host {
      font-family: sans-serif;
      display: inline-block;
      width: 300px;
      height: auto;
      box-sizing: border-box;
      color: var(--frc-network-alerts-text-color, black);
    }

    .alerts-container {
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;
    }

    header {
      background: var(--frc-network-alerts-header-color, lightblue);
      padding: 10px 0;
      text-align: center;
      font-weight: bold;
    }

    .alerts-list {
      padding: 8px 5px;
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      gap: 9px;
      flex: 1;
      overflow-y: auto;
    }

    .alert {
      display: flex;
      align-items: center;
      gap: 9px;
    }

    .icon {
      width: 24px;
      height: 24px;
    }

    .alert label {
      flex: 1;
    }

    .icon frc-icon {
      width: 100%;
      height: 100%;
    }
  `;
let r = h;
l([
  o({ type: Array })
], r.prototype, "errors");
l([
  o({ type: Array })
], r.prototype, "warnings");
l([
  o({ type: Array })
], r.prototype, "infos");
l([
  o({ type: String })
], r.prototype, "level");
l([
  o({ type: Boolean, attribute: "hide-title" })
], r.prototype, "hideTitle");
customElements.get("frc-network-alerts") || customElements.define("frc-network-alerts", r);
const k = f({
  tagName: "frc-network-alerts",
  elementClass: r,
  react: p
});
export {
  k as NetworkAlerts,
  k as default
};
