// @webbitjs/store/dist/index.es.js
var Q = (t, e, r) => {
  if (!e.has(t))
    throw TypeError("Cannot " + r);
};
var s = (t, e, r) => (Q(t, e, "read from private field"), r ? r.call(t) : e.get(t)), i = (t, e, r) => {
  if (e.has(t))
    throw TypeError("Cannot add the same private member more than once");
  e instanceof WeakSet ? e.add(t) : e.set(t, r);
}, h = (t, e, r, u) => (Q(t, e, "write to private field"), u ? u.call(t, r) : e.set(t, r), r);
var c = (t, e, r) => (Q(t, e, "access private method"), r);
var Le = 1 / 0, Me = "[object Symbol]", Ie = /[^\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\x7f]+/g, He = /[\xc0-\xd6\xd8-\xf6\xf8-\xff\u0100-\u017f]/g, $ = "\\ud800-\\udfff", le = "\\u0300-\\u036f\\ufe20-\\ufe23", be = "\\u20d0-\\u20f0", Se = "\\u2700-\\u27bf", xe = "a-z\\xdf-\\xf6\\xf8-\\xff", De = "\\xac\\xb1\\xd7\\xf7", ze = "\\x00-\\x2f\\x3a-\\x40\\x5b-\\x60\\x7b-\\xbf", Ze = "\\u2000-\\u206f", We = " \\t\\x0b\\f\\xa0\\ufeff\\n\\r\\u2028\\u2029\\u1680\\u180e\\u2000\\u2001\\u2002\\u2003\\u2004\\u2005\\u2006\\u2007\\u2008\\u2009\\u200a\\u202f\\u205f\\u3000", ge = "A-Z\\xc0-\\xd6\\xd8-\\xde", pe = "\\ufe0e\\ufe0f", ve = De + ze + Ze + We, re = "['’]", Ke = "[" + $ + "]", oe = "[" + ve + "]", G = "[" + le + be + "]", me = "\\d+", Fe = "[" + Se + "]", ye = "[" + xe + "]", je = "[^" + $ + ve + me + Se + xe + ge + "]", X = "\\ud83c[\\udffb-\\udfff]", Je = "(?:" + G + "|" + X + ")", Oe = "[^" + $ + "]", te = "(?:\\ud83c[\\udde6-\\uddff]){2}", ue = "[\\ud800-\\udbff][\\udc00-\\udfff]", E = "[" + ge + "]", Ee = "\\u200d", ie = "(?:" + ye + "|" + je + ")", Ge = "(?:" + E + "|" + je + ")", ne = "(?:" + re + "(?:d|ll|m|re|s|t|ve))?", ce = "(?:" + re + "(?:D|LL|M|RE|S|T|VE))?", Ae = Je + "?", Pe = "[" + pe + "]?", Ne = "(?:" + Ee + "(?:" + [Oe, te, ue].join("|") + ")" + Pe + Ae + ")*", Ce = Pe + Ae + Ne, Ye = "(?:" + [Fe, te, ue].join("|") + ")" + Ce, qe = "(?:" + [Oe + G + "?", G, te, ue, Ke].join("|") + ")", $e = RegExp(re, "g"), Be = RegExp(G, "g"), Qe = RegExp(X + "(?=" + X + ")|" + qe + Ce, "g"), Xe = RegExp([
  E + "?" + ye + "+" + ne + "(?=" + [oe, E, "$"].join("|") + ")",
  Ge + "+" + ce + "(?=" + [oe, E + ie, "$"].join("|") + ")",
  E + "?" + ie + "+" + ne,
  E + "+" + ce,
  me,
  Ye
].join("|"), "g"), _e = RegExp("[" + Ee + $ + le + be + pe + "]"), ke = /[a-z][A-Z]|[A-Z]{2,}[a-z]|[0-9][a-zA-Z]|[a-zA-Z][0-9]|[^a-zA-Z0-9 ]/, er = {
  // Latin-1 Supplement block.
  À: "A",
  Á: "A",
  Â: "A",
  Ã: "A",
  Ä: "A",
  Å: "A",
  à: "a",
  á: "a",
  â: "a",
  ã: "a",
  ä: "a",
  å: "a",
  Ç: "C",
  ç: "c",
  Ð: "D",
  ð: "d",
  È: "E",
  É: "E",
  Ê: "E",
  Ë: "E",
  è: "e",
  é: "e",
  ê: "e",
  ë: "e",
  Ì: "I",
  Í: "I",
  Î: "I",
  Ï: "I",
  ì: "i",
  í: "i",
  î: "i",
  ï: "i",
  Ñ: "N",
  ñ: "n",
  Ò: "O",
  Ó: "O",
  Ô: "O",
  Õ: "O",
  Ö: "O",
  Ø: "O",
  ò: "o",
  ó: "o",
  ô: "o",
  õ: "o",
  ö: "o",
  ø: "o",
  Ù: "U",
  Ú: "U",
  Û: "U",
  Ü: "U",
  ù: "u",
  ú: "u",
  û: "u",
  ü: "u",
  Ý: "Y",
  ý: "y",
  ÿ: "y",
  Æ: "Ae",
  æ: "ae",
  Þ: "Th",
  þ: "th",
  ß: "ss",
  // Latin Extended-A block.
  Ā: "A",
  Ă: "A",
  Ą: "A",
  ā: "a",
  ă: "a",
  ą: "a",
  Ć: "C",
  Ĉ: "C",
  Ċ: "C",
  Č: "C",
  ć: "c",
  ĉ: "c",
  ċ: "c",
  č: "c",
  Ď: "D",
  Đ: "D",
  ď: "d",
  đ: "d",
  Ē: "E",
  Ĕ: "E",
  Ė: "E",
  Ę: "E",
  Ě: "E",
  ē: "e",
  ĕ: "e",
  ė: "e",
  ę: "e",
  ě: "e",
  Ĝ: "G",
  Ğ: "G",
  Ġ: "G",
  Ģ: "G",
  ĝ: "g",
  ğ: "g",
  ġ: "g",
  ģ: "g",
  Ĥ: "H",
  Ħ: "H",
  ĥ: "h",
  ħ: "h",
  Ĩ: "I",
  Ī: "I",
  Ĭ: "I",
  Į: "I",
  İ: "I",
  ĩ: "i",
  ī: "i",
  ĭ: "i",
  į: "i",
  ı: "i",
  Ĵ: "J",
  ĵ: "j",
  Ķ: "K",
  ķ: "k",
  ĸ: "k",
  Ĺ: "L",
  Ļ: "L",
  Ľ: "L",
  Ŀ: "L",
  Ł: "L",
  ĺ: "l",
  ļ: "l",
  ľ: "l",
  ŀ: "l",
  ł: "l",
  Ń: "N",
  Ņ: "N",
  Ň: "N",
  Ŋ: "N",
  ń: "n",
  ņ: "n",
  ň: "n",
  ŋ: "n",
  Ō: "O",
  Ŏ: "O",
  Ő: "O",
  ō: "o",
  ŏ: "o",
  ő: "o",
  Ŕ: "R",
  Ŗ: "R",
  Ř: "R",
  ŕ: "r",
  ŗ: "r",
  ř: "r",
  Ś: "S",
  Ŝ: "S",
  Ş: "S",
  Š: "S",
  ś: "s",
  ŝ: "s",
  ş: "s",
  š: "s",
  Ţ: "T",
  Ť: "T",
  Ŧ: "T",
  ţ: "t",
  ť: "t",
  ŧ: "t",
  Ũ: "U",
  Ū: "U",
  Ŭ: "U",
  Ů: "U",
  Ű: "U",
  Ų: "U",
  ũ: "u",
  ū: "u",
  ŭ: "u",
  ů: "u",
  ű: "u",
  ų: "u",
  Ŵ: "W",
  ŵ: "w",
  Ŷ: "Y",
  ŷ: "y",
  Ÿ: "Y",
  Ź: "Z",
  Ż: "Z",
  Ž: "Z",
  ź: "z",
  ż: "z",
  ž: "z",
  Ĳ: "IJ",
  ĳ: "ij",
  Œ: "Oe",
  œ: "oe",
  ŉ: "'n",
  ſ: "ss"
}, rr = typeof global == "object" && global && global.Object === Object && global, tr = typeof self == "object" && self && self.Object === Object && self, ur = rr || tr || Function("return this")();
function sr(t, e, r, u) {
  var o = -1, n = t ? t.length : 0;
  for (u && n && (r = t[++o]); ++o < n; )
    r = e(r, t[o], o, t);
  return r;
}
function or(t) {
  return t.split("");
}
function ir(t) {
  return t.match(Ie) || [];
}
function nr(t) {
  return function(e) {
    return t == null ? void 0 : t[e];
  };
}
var cr = nr(er);
function Re(t) {
  return _e.test(t);
}
function ar(t) {
  return ke.test(t);
}
function hr(t) {
  return Re(t) ? fr(t) : or(t);
}
function fr(t) {
  return t.match(Qe) || [];
}
function dr(t) {
  return t.match(Xe) || [];
}
var lr = Object.prototype, br = lr.toString, ae = ur.Symbol, he = ae ? ae.prototype : void 0, fe = he ? he.toString : void 0;
function Sr(t, e, r) {
  var u = -1, o = t.length;
  e < 0 && (e = -e > o ? 0 : o + e), r = r > o ? o : r, r < 0 && (r += o), o = e > r ? 0 : r - e >>> 0, e >>>= 0;
  for (var n = Array(o); ++u < o; )
    n[u] = t[u + e];
  return n;
}
function xr(t) {
  if (typeof t == "string")
    return t;
  if (yr(t))
    return fe ? fe.call(t) : "";
  var e = t + "";
  return e == "0" && 1 / t == -Le ? "-0" : e;
}
function gr(t, e, r) {
  var u = t.length;
  return r = r === void 0 ? u : r, !e && r >= u ? t : Sr(t, e, r);
}
function pr(t) {
  return function(e) {
    e = B(e);
    var r = Re(e) ? hr(e) : void 0, u = r ? r[0] : e.charAt(0), o = r ? gr(r, 1).join("") : e.slice(1);
    return u[t]() + o;
  };
}
function vr(t) {
  return function(e) {
    return sr(Ar(Or(e).replace($e, "")), t, "");
  };
}
function mr(t) {
  return !!t && typeof t == "object";
}
function yr(t) {
  return typeof t == "symbol" || mr(t) && br.call(t) == Me;
}
function B(t) {
  return t == null ? "" : xr(t);
}
var Ue = vr(function(t, e, r) {
  return e = e.toLowerCase(), t + (r ? jr(e) : e);
});
function jr(t) {
  return Er(B(t).toLowerCase());
}
function Or(t) {
  return t = B(t), t && t.replace(He, cr).replace(Be, "");
}
var Er = pr("toUpperCase");
function Ar(t, e, r) {
  return t = B(t), e = r ? void 0 : e, e === void 0 ? ar(t) ? dr(t) : ir(t) : t.match(e) || [];
}
var P, S, I, x, C, p, H;
class Pr {
  constructor(e, r, u, o) {
    i(this, P, void 0);
    i(this, S, {});
    i(this, I, void 0);
    i(this, x, void 0);
    i(this, C, !1);
    i(this, p, void 0);
    i(this, H, void 0);
    h(this, H, e), h(this, p, r), h(this, P, u), h(this, I, o);
  }
  getKey() {
    return s(this, P);
  }
  hasValue() {
    return s(this, C);
  }
  getValue() {
    return s(this, x);
  }
  setValue(e) {
    h(this, x, e), h(this, C, !0);
  }
  removeValue() {
    h(this, x, void 0), h(this, C, !1);
  }
  getSourceValue() {
    return this.hasChildren() ? s(this, p) : s(this, x);
  }
  setSourceValue(e) {
    s(this, H).userUpdate(s(this, P), e);
  }
  getChildren() {
    return s(this, S);
  }
  hasChildren() {
    return Object.getOwnPropertyNames(s(this, S)).length > 0;
  }
  addChild(e, r) {
    const u = e.split("/").at(-1);
    typeof u > "u" || (s(this, S)[u] = r, !(typeof u > "u") && Object.defineProperty(s(this, p), u, {
      configurable: !0,
      enumerable: !0,
      set(o) {
        r.setSourceValue(o);
      },
      get() {
        return r.getSourceValue();
      }
    }));
  }
  removeChild(e) {
    const r = e.split("/").at(-1);
    typeof r < "u" && (delete s(this, S)[r], delete s(this, p)[r]);
  }
  getParent() {
    return s(this, I);
  }
  getJson(e = !0) {
    if (this.hasChildren()) {
      const r = {};
      return Object.entries(s(this, S)).forEach(([u, o]) => {
        const n = e ? Ue(u) : u;
        r[n] = o.getJson(e);
      }), r;
    }
    return s(this, x);
  }
}
P = new WeakMap(), S = new WeakMap(), I = new WeakMap(), x = new WeakMap(), C = new WeakMap(), p = new WeakMap(), H = new WeakMap();
function Cr(t) {
  return t.hasValue() ? !1 : Object.keys(t.getChildren()).length === 0;
}
var R, v, d, m, b, U, D, N, Ve, Y, we, z, _, q, Te, g, A;
class Rr {
  constructor(e) {
    i(this, N);
    i(this, Y);
    i(this, z);
    i(this, q);
    i(this, g);
    i(this, R, /* @__PURE__ */ new Map());
    i(this, v, /* @__PURE__ */ new Set());
    i(this, d, /* @__PURE__ */ new Map());
    i(this, m, void 0);
    i(this, b, {});
    i(this, U, /* @__PURE__ */ new Map());
    i(this, D, void 0);
    h(this, m, e), h(this, D, {
      clearSources: e.addClearSourcesHandler(() => {
        this.clearSources();
      }),
      sourcesChanged: e.addSourcesChangedHandler((r) => {
        Object.entries(r).forEach(([u, o]) => {
          this.updateSource(u, o, !1);
        }), Object.entries(r).forEach(([u]) => {
          c(this, g, A).call(this, u);
        });
      }),
      sourcesRemoved: e.addSourcesRemovedHandler((r) => {
        const u = r.filter((o) => s(this, d).has(o));
        u.forEach((o) => this.removeSource(o), !1), u.forEach((o) => {
          c(this, g, A).call(this, o);
        });
      })
    });
  }
  getSourceProvider() {
    return s(this, m);
  }
  getSource(e) {
    return s(this, d).get(e);
  }
  getSourceValue(e) {
    var r;
    return (r = this.getSource(e)) == null ? void 0 : r.getSourceValue();
  }
  updateSource(e, r, u = !0) {
    c(this, N, Ve).call(this, e), c(this, Y, we).call(this, e);
    const o = s(this, d).get(e);
    o == null || o.setValue(r), u && c(this, g, A).call(this, e);
  }
  clearSources() {
    const e = [];
    s(this, d).forEach((r, u) => {
      e.push(u), this.removeSource(u, !1);
    }), e.forEach((r) => c(this, g, A).call(this, r));
  }
  removeSource(e, r = !0) {
    const u = s(this, d).get(e);
    u && (u.removeValue(), c(this, z, _).call(this, u), r && c(this, g, A).call(this, e), s(this, v).delete(e));
  }
  subscribe(e, r, u) {
    var n;
    typeof s(this, b)[e] > "u" && (s(this, b)[e] = /* @__PURE__ */ new Map());
    const o = Symbol("SourceSubscriber");
    return s(this, b)[e].set(o, r), u && ((n = this.getSource(e)) != null && n.hasValue() || s(this, m).requestSource(e), r(this.getSourceValue(e), e, e)), () => {
      s(this, b)[e].delete(o);
    };
  }
  subscribeAll(e, r) {
    const u = Symbol("SourceSubscriberAll");
    return s(this, U).set(u, e), r && s(this, v).forEach((o) => {
      e(this.getSourceValue(o), o);
    }), () => {
      s(this, U).delete(u);
    };
  }
  unsubscribeFromProvider() {
    const { clearSources: e, sourcesChanged: r, sourcesRemoved: u } = s(this, D);
    e(), r(), u();
  }
}
R = new WeakMap(), v = new WeakMap(), d = new WeakMap(), m = new WeakMap(), b = new WeakMap(), U = new WeakMap(), D = new WeakMap(), N = new WeakSet(), Ve = function(e) {
  s(this, v).has(e) || s(this, v).add(e);
}, Y = new WeakSet(), we = function(e) {
  const r = e.split("/");
  let u;
  r.forEach((o, n) => {
    const a = r.slice(0, n + 1).join("/");
    if (!s(this, d).has(a)) {
      const O = s(this, d).get(u), se = new Pr(
        s(this, m),
        c(this, q, Te).call(this, a),
        a,
        O
      );
      s(this, d).set(a, se), O == null || O.addChild(a, se);
    }
    u = a;
  });
}, z = new WeakSet(), _ = function(e) {
  const r = Cr(e), u = e.getParent();
  r && u && (u.removeChild(e.getKey()), s(this, d).delete(e.getKey()), c(this, z, _).call(this, u));
}, q = new WeakSet(), Te = function(e) {
  return s(this, R).has(e) || s(this, R).set(e, {}), s(this, R).get(e);
}, g = new WeakSet(), A = function(e) {
  const r = e.split("/"), u = this.getSourceValue(e);
  r.forEach((o, n) => {
    const a = r.slice(0, n + 1).join("/");
    a in s(this, b) && s(this, b)[a].forEach((O) => {
      O(this.getSourceValue(a), a, e);
    });
  }), s(this, U).forEach((o) => {
    o(u, e);
  });
};
var f, Z, W, K;
class Vr {
  constructor() {
    i(this, f, {});
    i(this, Z, []);
    i(this, W, void 0);
    i(this, K, []);
  }
  hasSourceProvider(e) {
    return e in s(this, f);
  }
  addSourceProvider(e, r) {
    if (this.hasSourceProvider(e))
      throw new Error("A source provider with that name has already been added.");
    const u = new Rr(r);
    s(this, f)[e] = u, s(this, Z).forEach((o) => {
      o(e);
    });
  }
  sourceProviderAdded(e) {
    s(this, Z).push(e);
  }
  removeSourceProvider(e) {
    if (!this.hasSourceProvider(e))
      return;
    this.getSourceProvider(e).disconnect(), s(this, f)[e].unsubscribeFromProvider(), delete s(this, f)[e];
  }
  getSourceProvider(e) {
    return s(this, f)[e].getSourceProvider();
  }
  getSourceProviderNames() {
    return Object.keys(s(this, f));
  }
  setDefaultSourceProvider(e) {
    h(this, W, e), s(this, K).forEach((r) => {
      r(e);
    });
  }
  defaultSourceProviderSet(e) {
    if (typeof e != "function")
      throw new Error("listener is not a function");
    s(this, K).push(e);
  }
  getDefaultSourceProvider() {
    return s(this, W);
  }
  getSource(e, r) {
    var u;
    return (u = s(this, f)[e]) == null ? void 0 : u.getSource(r);
  }
  getSourceValue(e, r) {
    var u;
    return (u = s(this, f)[e]) == null ? void 0 : u.getSourceValue(r);
  }
  subscribe(e, r, u, o) {
    var n;
    return (n = s(this, f)[e]) == null ? void 0 : n.subscribe(r, u, o);
  }
  subscribeAll(e, r, u) {
    var o;
    return (o = s(this, f)[e]) == null ? void 0 : o.subscribeAll(r, u);
  }
}
f = new WeakMap(), Z = new WeakMap(), W = new WeakMap(), K = new WeakMap();
const de = () => {
}, wr = (t) => t.split("/").map((e) => Ue(e)).join("/");
var y, l, V, w, T, L, j, M, F, k, J, ee;
class Tr {
  constructor(e = {}, r = 0) {
    i(this, j);
    i(this, F);
    i(this, J);
    i(this, y, void 0);
    i(this, l, {});
    i(this, V, void 0);
    i(this, w, /* @__PURE__ */ new Map());
    i(this, T, /* @__PURE__ */ new Map());
    i(this, L, /* @__PURE__ */ new Map());
    r && h(this, y, setInterval(c(this, j, M).bind(this), r)), Object.entries(e).forEach(([u, o]) => {
      this.updateSource(u, o);
    });
  }
  /**
   * Updates the value of a source in the store. If the source doesn't
   * exist then it is added. Should only be called internally by the
   * source provider.
   *
   * @protected
   * @param {string} key - The source's key. This is a string separated
   * by '/'.
   * @param {*} value - The new value.
   */
  updateSource(e, r) {
    clearTimeout(s(this, V)), s(this, l)[e] === void 0 ? s(this, l)[e] = {
      first: {
        type: "change",
        value: r
      }
    } : s(this, l)[e].last = {
      type: "change",
      value: r
    }, s(this, y) || c(this, j, M).call(this);
  }
  /**
   * Removes an existing source from the store. If the source
   * doesn't exist this does nothing. Should only be called
   * internally by the source provider.
   *
   * @protected
   * @param {string} key - The source's key. This is a string separated
   * by '/'.
   */
  removeSource(e) {
    s(this, l)[e] === void 0 ? s(this, l)[e] = {
      first: {
        type: "removal"
      }
    } : s(this, l)[e].last = {
      type: "removal"
    }, s(this, y) || c(this, j, M).call(this);
  }
  /**
   * Removes all sources in the store for this provider. Should only be
   * called internally by the source provider.
   *
   * @protected
   * @param {function} callback - An optional callback. Called when sources
   * have been cleared.
   */
  clearSources(e = de) {
    c(this, j, M).call(this, () => {
      s(this, w).forEach((r) => r()), e();
    });
  }
  /**
   * Removes all sources in the store for this provider after a period of time.
   * If a source is set or this function is called before that period of time
   * ends, sources will not be cleared. This is useful for preventing sources
   * from being cleared on an unreliable network. Should only be called internally
   * by the source provider.
   *
   * @protected
   * @param {number} timeout - The period of time before clearing the sources
   * in milliseconds.
   * @param {function} callback - An optional callback. Called when sources
   * have been cleared.
   */
  clearSourcesWithTimeout(e, r) {
    clearTimeout(s(this, V)), h(this, V, setTimeout(() => {
      this.clearSources(r);
    }, e));
  }
  /**
   * Called when a source's value is modified by the user. This method
   * should be overridden by the child class to handle these updates.
   * This method should not be called directly.
   *
   * @protected
   * @param {string} key - The source's key. This is a string separated
   * by '/'.
   * @param {*} value - The source's updated value.
   */
  userUpdate(e, r) {
    this.updateSource(e, r);
  }
  /**
   * This optional method is called when a key is subscribed to that the store does not
   * already have. A SourceProvider can override this method to update the store with
   * the requested source.
   * @Override
   * @param {string} key
   */
  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-empty-function
  requestSource(e) {
  }
  /**
   * Helper function to get the type of a variable represented
   * by a string.
   *
   * @param {*} value
   * @returns {string} - The value's type.
   */
  static getType(e) {
    return typeof e == "string" ? "string" : typeof e == "number" ? "number" : typeof e == "boolean" ? "boolean" : e instanceof Array ? "Array" : e === null ? "null" : null;
  }
  addClearSourcesHandler(e) {
    const r = Symbol("clearSources");
    return s(this, w).set(r, e), () => {
      s(this, w).delete(r);
    };
  }
  addSourcesChangedHandler(e) {
    const r = Symbol("sourcesChanged");
    return s(this, T).set(r, e), () => {
      s(this, T).delete(r);
    };
  }
  addSourcesRemovedHandler(e) {
    const r = Symbol("sourcesRemoved");
    return s(this, L).set(r, e), () => {
      s(this, L).delete(r);
    };
  }
  disconnect() {
    clearTimeout(s(this, y));
  }
}
y = new WeakMap(), l = new WeakMap(), V = new WeakMap(), w = new WeakMap(), T = new WeakMap(), L = new WeakMap(), j = new WeakSet(), M = function(e = de) {
  if (Object.keys(s(this, l)).length === 0) {
    e();
    return;
  }
  const r = { ...s(this, l) };
  h(this, l, {});
  const u = {}, o = {};
  Object.entries(r).forEach(([n, a]) => {
    u[n] = a.first, typeof a.last < "u" && (o[n] = a.last);
  }), c(this, F, k).call(this, u), c(this, J, ee).call(this, u), Object.keys(o).length > 0 ? setTimeout(() => {
    c(this, F, k).call(this, o), c(this, J, ee).call(this, o), e();
  }) : e();
}, F = new WeakSet(), k = function(e) {
  const r = {};
  Object.entries(e).forEach(([u, { type: o }]) => {
    o === "change" && (r[u] = e[u].value);
  }), Object.keys(r).length > 0 && s(this, T).forEach((u) => u(r));
}, J = new WeakSet(), ee = function(e) {
  const r = [];
  Object.entries(e).forEach(([u, { type: o }]) => {
    o === "removal" && r.push(u);
  }), r.length > 0 && s(this, L).forEach((u) => u(r));
};
export {
  Pr as Source,
  Tr as SourceProvider,
  Vr as Store,
  Vr as default,
  wr as normalizeKey
};
