// @frc-web-components/react/dist/networktables/index.es.js
import { u as c } from "../NT4Provider-5cc60b98.js";
import { N as E } from "../NT4Provider-5cc60b98.js";
import { useEffect as a, useState as l, useCallback as f } from "react";
import "react/jsx-runtime";
import "@frc-web-components/fwc";
function b(e, t, s) {
  const { store: o } = c();
  a(() => o.subscribe(
    "NetworkTables",
    e,
    (r) => {
      t(e, r);
    },
    s
  ), []);
}
function d(e, t) {
  const { nt4Provider: s } = c(), [o, r] = l(t), i = f((n) => {
    s.userUpdate(e, n);
  }, []);
  return b(
    e,
    (n, u) => {
      r(u === void 0 ? t : u);
    },
    !0
  ), [o, i];
}
function T(e, t) {
  const { store: s } = c();
  a(() => s.subscribeAll(
    "NetworkTables",
    (o, r) => {
      e(r, o);
    },
    t
  ), []);
}
function g(e, t, s = !0) {
  const { store: o } = c(), [r, i] = l(t);
  return b(
    e,
    () => {
      var u;
      const n = (u = o.getSource("NetworkTables", e)) == null ? void 0 : u.getJson(s);
      i(n);
    },
    !0
  ), r;
}
export {
  E as NT4Provider,
  d as useEntry,
  T as useGlobalListener,
  g as useJson,
  b as useKeyListener,
  c as useNt4
};
