// @frc-web-components/fwc/dist/components.es.js
import { axisDashboardConfig as r } from "./components/axis.es.js";
import { Axis as Dr } from "./components/axis.es.js";
import { barDashboardConfig as o } from "./components/bar.es.js";
import { Bar as ur } from "./components/bar.es.js";
import { differentialDrivebaseDashboardConfig as e, mecanumDrivebaseDashboardConfig as a, swerveDrivebaseDashboardConfig as i } from "./components/drivebases.es.js";
import { DifferentialDrivebase as wr, MecanumDrivebase as Lr, SwerveDrivebase as Pr } from "./components/drivebases.es.js";
import { fieldDashboardConfig as f, fieldRobotDashboardConfig as t, fieldPathDashboardConfig as s } from "./components/field.es.js";
import { default as Fr, FieldPath as Sr, FieldRobot as Br, FieldWrapper as yr, fieldConfigs as Gr } from "./components/field.es.js";
import { b as Rr, c as Nr, t as Tr, u as Ur } from "./units-DyibxdHs.js";
import { Field3dObject as jr, Field3dPoseVisualizer as Ir, Field3dUrdf as Vr, FieldObject as kr, Urdf as Er, field3dConfigs as Or, getField3dDashboardConfigs as zr, objectConfigs as qr, urdfConfigs as Hr } from "./components/field3d.es.js";
import { a as m } from "./icon-Dc8f99z4.js";
import { I as Kr, i as Qr } from "./icon-Dc8f99z4.js";
import { threeAxisAccelerometerDashboardConfig as n } from "./components/3-axis-accelerometer.es.js";
import { ThreeAxisAccelerometer as Yr } from "./components/3-axis-accelerometer.es.js";
import { accelerometerDashboardConfig as d } from "./components/accelerometer.es.js";
import { Accelerometer as _r } from "./components/accelerometer.es.js";
import { basicFmsInfoDashboardConfig as b } from "./components/basic-fms-info.es.js";
import { BasicFmsInfo as ro } from "./components/basic-fms-info.es.js";
import { booleanBoxDashboardConfig as p } from "./components/boolean-box.es.js";
import { BooleanBox as eo } from "./components/boolean-box.es.js";
import { gaugeDashboardConfig as l } from "./components/gauge.es.js";
import { Gauge as io } from "./components/gauge.es.js";
import { gyroDashboardConfig as c } from "./components/gyro.es.js";
import { Gyro as to } from "./components/gyro.es.js";
import { numberBarDashboardConfig as g } from "./components/number-bar.es.js";
import { NumberBar as mo } from "./components/number-bar.es.js";
import { sendableChooserOldDashboardConfig as C, sendableChooserDashboardConfig as h } from "./components/sendable-chooser.es.js";
import { SendableChooser as bo, SendableChooserWrapper as po } from "./components/sendable-chooser.es.js";
import { lineChartDashboardConfigs as D } from "./components/line-chart.es.js";
import { ChartLegendPositions as co, default as go, LineChartAxis as Co, LineChartData as ho, LineChartLegend as Do, lineChartAxisDashboardConfig as xo, lineChartDashboardConfig as uo, lineChartDataDashboardConfig as vo, lineChartLegendDashboardConfig as wo } from "./components/line-chart.es.js";
import { Pdp as Po, elementName as Ao } from "./components/pdp.es.js";
import { voltageViewDashboardConfig as x } from "./components/voltage-view.es.js";
import { VoltageView as So } from "./components/voltage-view.es.js";
import { numberSliderDashboardConfig as u } from "./components/number-slider.es.js";
import { NumberSlider as yo } from "./components/number-slider.es.js";
import { toggleSwitchDashboardConfig as v } from "./components/toggle-switch.es.js";
import { ToggleSwitch as Mo } from "./components/toggle-switch.es.js";
import { toggleButtonDashboardConfig as w } from "./components/toggle-button.es.js";
import { ToggleButton as No } from "./components/toggle-button.es.js";
import { toggleGroupDashboardConfig as L } from "./components/toggle-group.es.js";
import { ToggleGroup as Uo } from "./components/toggle-group.es.js";
import { relayDashboardConfig as P } from "./components/relay.es.js";
import { Relay as jo } from "./components/relay.es.js";
import { m as A } from "./wrapper-BBKTpE7c.js";
import { M as Vo, b as ko, g as Eo, a as Oo } from "./wrapper-BBKTpE7c.js";
import { encoderDashboardConfig as F } from "./components/encoder.es.js";
import { Encoder as qo } from "./components/encoder.es.js";
import { networkAlertsDashboardConfig as S } from "./components/network-alerts.es.js";
import { NetworkAlerts as Jo } from "./components/network-alerts.es.js";
import { pidCommandDashboardConfig as B, pidControllerDashboardConfig as y, profiledPidControllerDashboardConfig as G } from "./components/pid.es.js";
import { PidCommand as Qo, PidController as Xo, ProfiledPidController as Yo } from "./components/pid.es.js";
import { scoringGridDashboardConfig as M } from "./components/scoring-grid.es.js";
import { ScoringGrid as _o } from "./components/scoring-grid.es.js";
import { robotCommandDashboardConfig as R, robotSubsystemDashboardConfig as N } from "./components/command-based.es.js";
import { RobotCommand as re, RobotSubsystem as oe } from "./components/command-based.es.js";
import { cameraDashboardConfig as T } from "./components/canvas.es.js";
import { CameraWrapper as ae, default as ie, CanvasCircle as fe, CanvasGroup as te, CanvasLine as se, CanvasMjpgStream as me, CanvasNGon as ne, CanvasRect as de, CanvasText as be } from "./components/canvas.es.js";
import { labelDashboardConfig as U, numberLabelDashboardConfig as W } from "./components/label.es.js";
import { Label as le, NumberLabel as ce } from "./components/label.es.js";
import { preferencesDashboardConfig as j } from "./components/preferences.es.js";
import { Preferences as Ce, PreferencesWrapper as he } from "./components/preferences.es.js";
import { formElementDashboardConfigs as I } from "./components/form-elements.es.js";
const gr = {
  "frc-axis": r,
  "frc-bar": o,
  "frc-differential-drivebase": e,
  "frc-mecanum-drivebase": a,
  "frc-swerve-drivebase": i,
  "frc-icon": m,
  "frc-3-axis-accelerometer": n,
  "frc-accelerometer": d,
  "frc-basic-fms-info": b,
  "frc-boolean-box": p,
  "frc-gauge": l,
  "frc-gyro": c,
  "frc-number-bar": g,
  "frc-sendable-chooser": C,
  "frc-sendable-chooser-wrapper": h,
  "frc-voltage-view": x,
  "frc-number-slider": u,
  "frc-toggle-switch": v,
  "frc-toggle-button": w,
  "frc-toggle-group": L,
  "frc-relay": P,
  "frc-encoder": F,
  "frc-network-alerts": S,
  "frc-pid-command": B,
  "frc-pid-controller": y,
  "frc-profiled-pid-controller": G,
  "frc-scoring-grid": M,
  "frc-robot-command": R,
  "frc-robot-subsystem": N,
  "frc-mechanism2d-wrapper": A,
  "frc-field-wrapper": f,
  "frc-field-robot": t,
  "frc-field-path": s,
  "frc-camera-wrapper": T,
  "frc-label": U,
  "frc-number-label": W,
  "frc-preferences-wrapper": j,
  ...D,
  ...I
};
export {
  _r as Accelerometer,
  Dr as Axis,
  ur as Bar,
  ro as BasicFmsInfo,
  eo as BooleanBox,
  ae as CameraWrapper,
  ie as Canvas,
  fe as CanvasCircle,
  te as CanvasGroup,
  se as CanvasLine,
  me as CanvasMjpgStream,
  ne as CanvasNGon,
  de as CanvasRect,
  be as CanvasText,
  co as ChartLegendPositions,
  wr as DifferentialDrivebase,
  qo as Encoder,
  Fr as Field,
  jr as Field3dObject,
  Ir as Field3dPoseVisualizer,
  Vr as Field3dUrdf,
  kr as FieldObject,
  Sr as FieldPath,
  Br as FieldRobot,
  yr as FieldWrapper,
  io as Gauge,
  to as Gyro,
  Kr as Icon,
  le as Label,
  go as LineChart,
  Co as LineChartAxis,
  ho as LineChartData,
  Do as LineChartLegend,
  Lr as MecanumDrivebase,
  Vo as Mechanism2d,
  ko as Mechanism2dWrapper,
  Jo as NetworkAlerts,
  mo as NumberBar,
  ce as NumberLabel,
  yo as NumberSlider,
  Po as Pdp,
  Qo as PidCommand,
  Xo as PidController,
  Ce as Preferences,
  he as PreferencesWrapper,
  Yo as ProfiledPidController,
  jo as Relay,
  re as RobotCommand,
  oe as RobotSubsystem,
  _o as ScoringGrid,
  bo as SendableChooser,
  po as SendableChooserWrapper,
  Pr as SwerveDrivebase,
  Yr as ThreeAxisAccelerometer,
  No as ToggleButton,
  Uo as ToggleGroup,
  Mo as ToggleSwitch,
  Er as Urdf,
  So as VoltageView,
  d as accelerometerDashboardConfig,
  r as axisDashboardConfig,
  o as barDashboardConfig,
  Rr as baseUnit,
  b as basicFmsInfoDashboardConfig,
  p as booleanBoxDashboardConfig,
  T as cameraDashboardConfig,
  Nr as convertUnit,
  gr as dashboardElementConfigs,
  e as differentialDrivebaseDashboardConfig,
  Ao as elementName,
  F as encoderDashboardConfig,
  Or as field3dConfigs,
  Gr as fieldConfigs,
  f as fieldDashboardConfig,
  s as fieldPathDashboardConfig,
  t as fieldRobotDashboardConfig,
  l as gaugeDashboardConfig,
  zr as getField3dDashboardConfigs,
  Eo as getMechanism2dLines,
  Oo as getMechanism2dRoots,
  c as gyroDashboardConfig,
  m as iconDashboardConfig,
  Qr as iconset,
  U as labelDashboardConfig,
  xo as lineChartAxisDashboardConfig,
  uo as lineChartDashboardConfig,
  D as lineChartDashboardConfigs,
  vo as lineChartDataDashboardConfig,
  wo as lineChartLegendDashboardConfig,
  a as mecanumDrivebaseDashboardConfig,
  A as mechanism2dDashboardConfig,
  S as networkAlertsDashboardConfig,
  g as numberBarDashboardConfig,
  W as numberLabelDashboardConfig,
  u as numberSliderDashboardConfig,
  qr as objectConfigs,
  B as pidCommandDashboardConfig,
  y as pidControllerDashboardConfig,
  j as preferencesDashboardConfig,
  G as profiledPidControllerDashboardConfig,
  P as relayDashboardConfig,
  R as robotCommandDashboardConfig,
  N as robotSubsystemDashboardConfig,
  M as scoringGridDashboardConfig,
  h as sendableChooserDashboardConfig,
  C as sendableChooserOldDashboardConfig,
  i as swerveDrivebaseDashboardConfig,
  n as threeAxisAccelerometerDashboardConfig,
  Tr as toBaseUnitConversions,
  w as toggleButtonDashboardConfig,
  L as toggleGroupDashboardConfig,
  v as toggleSwitchDashboardConfig,
  Ur as unitAliases,
  Hr as urdfConfigs,
  x as voltageViewDashboardConfig
};
