// @frc-web-components/fwc/dist/fwc.es.js
import { G as fs } from "./gamepad-provider-D67lJPD2.js";
import { a as ds, N as gs, b as bs } from "./nt4-provider-CjF_GyI0.js";
import { Axis as vs, axisDashboardConfig as ys } from "./components/axis.es.js";
import { Bar as xs, barDashboardConfig as Es } from "./components/bar.es.js";
import { DifferentialDrivebase as Ss, MecanumDrivebase as Ts, SwerveDrivebase as ks, differentialDrivebaseDashboardConfig as Cs, mecanumDrivebaseDashboardConfig as Rs, swerveDrivebaseDashboardConfig as As } from "./components/drivebases.es.js";
import { default as Ps, FieldPath as Ls, FieldRobot as Ds, FieldWrapper as Ws, fieldConfigs as Ms, fieldDashboardConfig as js, fieldPathDashboardConfig as Is, fieldRobotDashboardConfig as Bs } from "./components/field.es.js";
import { b as Vs, c as Us, t as zs, u as Fs } from "./units-DyibxdHs.js";
import { Field3dObject as $s, Field3dPoseVisualizer as Gs, Field3dUrdf as Xs, FieldObject as Qs, Urdf as Hs, field3dConfigs as Js, getField3dDashboardConfigs as Zs, objectConfigs as Ys, urdfConfigs as tn } from "./components/field3d.es.js";
import { I as sn, a as nn, i as rn } from "./icon-Dc8f99z4.js";
import { ThreeAxisAccelerometer as _n, threeAxisAccelerometerDashboardConfig as on } from "./components/3-axis-accelerometer.es.js";
import { Accelerometer as ln, accelerometerDashboardConfig as hn } from "./components/accelerometer.es.js";
import { BasicFmsInfo as fn, basicFmsInfoDashboardConfig as cn } from "./components/basic-fms-info.es.js";
import { BooleanBox as gn, booleanBoxDashboardConfig as bn } from "./components/boolean-box.es.js";
import { Gauge as vn, gaugeDashboardConfig as yn } from "./components/gauge.es.js";
import { Gyro as xn, gyroDashboardConfig as En } from "./components/gyro.es.js";
import { NumberBar as Sn, numberBarDashboardConfig as Tn } from "./components/number-bar.es.js";
import { SendableChooser as Cn, SendableChooserWrapper as Rn, sendableChooserDashboardConfig as An, sendableChooserOldDashboardConfig as Nn } from "./components/sendable-chooser.es.js";
import { ChartLegendPositions as Ln, default as Dn, LineChartAxis as Wn, LineChartData as Mn, LineChartLegend as jn, lineChartAxisDashboardConfig as In, lineChartDashboardConfig as Bn, lineChartDashboardConfigs as Kn, lineChartDataDashboardConfig as Vn, lineChartLegendDashboardConfig as Un } from "./components/line-chart.es.js";
import { Pdp as Fn, elementName as qn } from "./components/pdp.es.js";
import { VoltageView as Gn, voltageViewDashboardConfig as Xn } from "./components/voltage-view.es.js";
import { NumberSlider as Hn, numberSliderDashboardConfig as Jn } from "./components/number-slider.es.js";
import { ToggleSwitch as Yn, toggleSwitchDashboardConfig as tr } from "./components/toggle-switch.es.js";
import { ToggleButton as ir, toggleButtonDashboardConfig as sr } from "./components/toggle-button.es.js";
import { ToggleGroup as rr, toggleGroupDashboardConfig as ar } from "./components/toggle-group.es.js";
import { Relay as or, relayDashboardConfig as ur } from "./components/relay.es.js";
import { M as hr, b as pr, g as fr, a as cr, m as dr } from "./wrapper-BBKTpE7c.js";
import { Encoder as br, encoderDashboardConfig as mr } from "./components/encoder.es.js";
import { NetworkAlerts as yr, networkAlertsDashboardConfig as wr } from "./components/network-alerts.es.js";
import { PidCommand as Er, PidController as Or, ProfiledPidController as Sr, pidCommandDashboardConfig as Tr, pidControllerDashboardConfig as kr, profiledPidControllerDashboardConfig as Cr } from "./components/pid.es.js";
import { ScoringGrid as Ar, scoringGridDashboardConfig as Nr } from "./components/scoring-grid.es.js";
import { RobotCommand as Lr, RobotSubsystem as Dr, robotCommandDashboardConfig as Wr, robotSubsystemDashboardConfig as Mr } from "./components/command-based.es.js";
import { CameraWrapper as Ir, default as Br, CanvasCircle as Kr, CanvasGroup as Vr, CanvasLine as Ur, CanvasMjpgStream as zr, CanvasNGon as Fr, CanvasRect as qr, CanvasText as $r, cameraDashboardConfig as Gr } from "./components/canvas.es.js";
import { Label as Qr, NumberLabel as Hr, labelDashboardConfig as Jr, numberLabelDashboardConfig as Zr } from "./components/label.es.js";
import { Preferences as ta, PreferencesWrapper as ea, preferencesDashboardConfig as ia } from "./components/preferences.es.js";
import { dashboardElementConfigs as na } from "./components.es.js";
import { DashboardThemes as aa, darkTheme as _a } from "./themes.es.js";
import { V as Oe, T as Ei } from "./index.es-3BKgpzRw.js";
var ce = (a, t, i) => {
  if (!t.has(a))
    throw TypeError("Cannot " + i);
}, w = (a, t, i) => (ce(a, t, "read from private field"), i ? i.call(a) : t.get(a)), P = (a, t, i) => {
  if (t.has(a))
    throw TypeError("Cannot add the same private member more than once");
  t instanceof WeakSet ? t.add(a) : t.set(a, i);
}, L = (a, t, i, e) => (ce(a, t, "write to private field"), t.set(a, i), i), D = (a, t, i) => (ce(a, t, "access private method"), i), Oi = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, Rt = { exports: {} };
Rt.exports;
(function(a, t) {
  (function(i, e) {
    var r = {};
    i.PubSub ? (r = i.PubSub, console.warn("PubSub already loaded, using existing version")) : (i.PubSub = r, e(r)), a !== void 0 && a.exports && (t = a.exports = r), t.PubSub = r, a.exports = t = r;
  })(typeof window == "object" && window || Oi, function(i) {
    var e = {}, r = -1, _ = "*";
    function h(x) {
      var o;
      for (o in x)
        if (Object.prototype.hasOwnProperty.call(x, o))
          return !0;
      return !1;
    }
    function u(x) {
      return function() {
        throw x;
      };
    }
    function l(x, o, n) {
      try {
        x(o, n);
      } catch (p) {
        setTimeout(u(p), 0);
      }
    }
    function O(x, o, n) {
      x(o, n);
    }
    function v(x, o, n, p) {
      var T = e[o], E = p ? O : l, k;
      if (Object.prototype.hasOwnProperty.call(e, o))
        for (k in T)
          Object.prototype.hasOwnProperty.call(T, k) && E(T[k], x, n);
    }
    function y(x, o, n) {
      return function() {
        var p = String(x), T = p.lastIndexOf(".");
        for (v(x, x, o, n); T !== -1; )
          p = p.substr(0, T), T = p.lastIndexOf("."), v(x, p, o, n);
        v(x, _, o, n);
      };
    }
    function g(x) {
      var o = String(x), n = !!(Object.prototype.hasOwnProperty.call(e, o) && h(e[o]));
      return n;
    }
    function C(x) {
      for (var o = String(x), n = g(o) || g(_), p = o.lastIndexOf("."); !n && p !== -1; )
        o = o.substr(0, p), p = o.lastIndexOf("."), n = g(o);
      return n;
    }
    function A(x, o, n, p) {
      x = typeof x == "symbol" ? x.toString() : x;
      var T = y(x, o, p), E = C(x);
      return E ? (n === !0 ? T() : setTimeout(T, 0), !0) : !1;
    }
    i.publish = function(x, o) {
      return A(x, o, !1, i.immediateExceptions);
    }, i.publishSync = function(x, o) {
      return A(x, o, !0, i.immediateExceptions);
    }, i.subscribe = function(x, o) {
      if (typeof o != "function")
        return !1;
      x = typeof x == "symbol" ? x.toString() : x, Object.prototype.hasOwnProperty.call(e, x) || (e[x] = {});
      var n = "uid_" + String(++r);
      return e[x][n] = o, n;
    }, i.subscribeAll = function(x) {
      return i.subscribe(_, x);
    }, i.subscribeOnce = function(x, o) {
      var n = i.subscribe(x, function() {
        i.unsubscribe(n), o.apply(this, arguments);
      });
      return i;
    }, i.clearAllSubscriptions = function() {
      e = {};
    }, i.clearSubscriptions = function(x) {
      var o;
      for (o in e)
        Object.prototype.hasOwnProperty.call(e, o) && o.indexOf(x) === 0 && delete e[o];
    }, i.countSubscriptions = function(x) {
      var o, n, p = 0;
      for (o in e)
        if (Object.prototype.hasOwnProperty.call(e, o) && o.indexOf(x) === 0) {
          for (n in e[o])
            p++;
          break;
        }
      return p;
    }, i.getSubscriptions = function(x) {
      var o, n = [];
      for (o in e)
        Object.prototype.hasOwnProperty.call(e, o) && o.indexOf(x) === 0 && n.push(o);
      return n;
    }, i.unsubscribe = function(x) {
      var o = function(f) {
        var S;
        for (S in e)
          if (Object.prototype.hasOwnProperty.call(e, S) && S.indexOf(f) === 0)
            return !0;
        return !1;
      }, n = typeof x == "string" && (Object.prototype.hasOwnProperty.call(e, x) || o(x)), p = !n && typeof x == "string", T = typeof x == "function", E = !1, k, c, b;
      if (n) {
        i.clearSubscriptions(x);
        return;
      }
      for (k in e)
        if (Object.prototype.hasOwnProperty.call(e, k)) {
          if (c = e[k], p && c[x]) {
            delete c[x], E = x;
            break;
          }
          if (T)
            for (b in c)
              Object.prototype.hasOwnProperty.call(c, b) && c[b] === x && (delete c[b], E = !0);
        }
      return E;
    };
  });
})(Rt, Rt.exports);
var q = Rt.exports, At = function() {
  return At = Object.assign || function(a) {
    for (var t, i = 1, e = arguments.length; i < e; i++) {
      t = arguments[i];
      for (var r in t)
        Object.prototype.hasOwnProperty.call(t, r) && (a[r] = t[r]);
    }
    return a;
  }, At.apply(this, arguments);
};
function Si(a) {
  return a.toLowerCase();
}
var Ti = [/([a-z0-9])([A-Z])/g, /([A-Z])([A-Z][a-z])/g], ki = /[^A-Z0-9]+/gi;
function Ci(a, t) {
  t === void 0 && (t = {});
  for (var i = t.splitRegexp, e = i === void 0 ? Ti : i, r = t.stripRegexp, _ = r === void 0 ? ki : r, h = t.transform, u = h === void 0 ? Si : h, l = t.delimiter, O = l === void 0 ? " " : l, v = Se(Se(a, e, "$1\0$2"), _, "\0"), y = 0, g = v.length; v.charAt(y) === "\0"; )
    y++;
  for (; v.charAt(g - 1) === "\0"; )
    g--;
  return v.slice(y, g).split("\0").map(u).join(O);
}
function Se(a, t, i) {
  return t instanceof RegExp ? a.replace(t, i) : t.reduce(function(e, r) {
    return e.replace(r, i);
  }, a);
}
function _i(a, t) {
  var i = a.charAt(0), e = a.substr(1).toLowerCase();
  return t > 0 && i >= "0" && i <= "9" ? "_" + i + e : "" + i.toUpperCase() + e;
}
function Ri(a) {
  return a.charAt(0).toUpperCase() + a.slice(1).toLowerCase();
}
function Ai(a, t) {
  return t === void 0 && (t = {}), Ci(a, At({ delimiter: "", transform: _i }, t));
}
function Ni(a, t) {
  return t === 0 ? a.toLowerCase() : _i(a, t);
}
function Pi(a, t) {
  return t === 0 ? a.toLowerCase() : Ri(a);
}
function Li(a, t) {
  return t === void 0 && (t = {}), Ai(a, At({ transform: Ni }, t));
}
function Di(a) {
  switch (a) {
    case "String":
      return "";
    case "Boolean":
      return !1;
    case "Number":
      return 0;
    case "Array":
      return [];
    case "SourceProvider":
      return;
    case "Store":
      return;
    case "Object":
    default:
      return {};
  }
}
const Wi = (a, {
  property: t = a,
  description: i = "",
  type: e = "String",
  defaultValue: r,
  attribute: _ = a.toLowerCase(),
  reflect: h = !1,
  primary: u = !1,
  changeEvent: l,
  ...O
} = {}) => {
  const v = typeof e == "function" ? e.name : e, y = r ?? Di(v);
  return {
    property: t,
    description: i,
    type: v,
    defaultValue: y,
    attribute: _,
    reflect: !!h,
    primary: !!u,
    changeEvent: l,
    ...O
  };
}, oi = ({
  description: a = "",
  group: t = "default",
  defaultSourceKey: i,
  defaultSourceProvider: e,
  dashboard: r,
  properties: _ = {},
  events: h = [],
  slots: u = [],
  cssProperties: l = [],
  cssParts: O = [],
  ...v
} = {}) => ({
  description: a,
  group: t,
  defaultSourceKey: i,
  defaultSourceProvider: e,
  dashboard: {
    topLevel: !0,
    ...r
  },
  properties: Object.fromEntries(
    Object.entries(_).map(([y, g]) => [y, Wi(y, g)])
  ),
  events: h,
  slots: u,
  cssProperties: l,
  cssParts: O,
  ...v
}), xt = (a) => typeof a == "string" ? "String" : typeof a == "number" ? "Number" : typeof a == "boolean" ? "Boolean" : a instanceof Array ? "Array" : a instanceof Object ? "Object" : null, J = (a, t) => {
  if (a === t)
    return !0;
  if (typeof a != typeof t)
    return !1;
  if (a instanceof Array && t instanceof Array) {
    if (a.length !== t.length)
      return !1;
    for (let i = 0; i < a.length; i++)
      if (a[i] !== t[i])
        return !1;
    return !0;
  }
  return !1;
};
function ee() {
}
class de {
  static prop2PropValue(t) {
    if (t === null)
      return null;
    if (t instanceof Array)
      return t;
    if (typeof t == "string")
      try {
        const i = JSON.parse(t);
        return i instanceof Array ? i : [];
      } catch {
        return [];
      }
    return [];
  }
  static prop2AttrValue(t) {
    const i = de.prop2PropValue(t);
    return i === null ? null : JSON.stringify(i);
  }
  static attr2PropValue(t) {
    if (t === null)
      return null;
    try {
      const i = JSON.parse(t);
      return i instanceof Array ? i : [];
    } catch {
      return [];
    }
  }
}
class Mi {
  static prop2PropValue(t) {
    return t === null ? null : !!t;
  }
  static prop2AttrValue(t) {
    return t ? "" : null;
  }
  static attr2PropValue(t) {
    return t !== null;
  }
}
class ji {
  static prop2PropValue(t) {
    return t === null ? null : parseFloat(t);
  }
  static prop2AttrValue(t) {
    return t === null ? null : parseFloat(t).toString();
  }
  static attr2PropValue(t, i) {
    return t === null ? null : parseFloat(t);
  }
}
class ge {
  static prop2PropValue(t) {
    if (t === null)
      return null;
    if (t instanceof Object)
      return t;
    if (typeof t == "string")
      try {
        const i = JSON.parse(t);
        return i instanceof Object ? i : {};
      } catch {
        return {};
      }
    return {};
  }
  static prop2AttrValue(t) {
    const i = ge.prop2PropValue(t);
    return i === null ? null : JSON.stringify(i);
  }
  static attr2PropValue(t) {
    if (t === null)
      return null;
    try {
      const i = JSON.parse(t);
      return i instanceof Object ? i : {};
    } catch {
      return {};
    }
  }
}
class Ii {
  static prop2PropValue(t) {
    return t === null ? null : (t == null ? void 0 : t.toString()) || "";
  }
  static prop2AttrValue(t) {
    return t === null ? null : (t == null ? void 0 : t.toString()) || "";
  }
  static attr2PropValue(t, i) {
    return t === null ? null : t;
  }
}
const be = {
  Array: de,
  Boolean: Mi,
  Number: ji,
  Object: ge,
  String: Ii
}, Z = (a, t) => {
  var i;
  return ((i = be[t]) == null ? void 0 : i.prop2PropValue(a)) ?? null;
}, Te = (a, t) => {
  var i;
  return ((i = be[t]) == null ? void 0 : i.prop2AttrValue(a)) ?? null;
}, ke = (a, t) => {
  var i;
  return ((i = be[t]) == null ? void 0 : i.attr2PropValue(a)) ?? null;
};
var I, K, U, z, ut, ie, ui, Et, Mt, Ot, se;
const li = class ne {
  constructor(t, i) {
    P(this, ie), P(this, Et), P(this, Ot), P(this, I, void 0), P(this, K, void 0), P(this, U, !1), P(this, z, void 0), P(this, ut, void 0), ne.UNIQUE_ID += 1, L(this, ut, Symbol(`WEBBIT_PROPERTY_CHANGE_${ne.UNIQUE_ID}`)), L(this, I, t), L(this, K, i), L(this, U, !1), L(this, z, this.value), D(this, ie, ui).call(this);
  }
  get value() {
    const {
      reflect: t,
      attribute: i,
      property: e,
      type: r
    } = w(this, K);
    return typeof e == "string" && e in w(this, I) ? w(this, I)[e] : ke(i && t ? w(this, I).getAttribute(i) : i ? w(this, I).getAttribute(i) : null, r);
  }
  set value(t) {
    const { attribute: i, property: e, type: r } = w(this, K), _ = xt(t), h = this.value;
    if (t === null || typeof t > "u") {
      if (t === h)
        return;
    } else if (J(Z(h, _), t))
      return;
    if (e) {
      const u = Z(t, r), l = Z(u, _);
      J(t, l) && (w(this, I)[e] = u);
    } else if (i) {
      const u = Te(t, r), l = ke(u, _);
      J(t, l) && (u === null ? w(this, I).removeAttribute(i) : w(this, I).setAttribute(i, u));
    }
  }
  isAttribute() {
    return typeof w(this, K).attribute == "string";
  }
  getStoredAttribute() {
    const { type: t } = w(this, K);
    return Te(w(this, z), t);
  }
  connect() {
    if (!w(this, U)) {
      const t = this.value;
      L(this, z, typeof t < "u" ? t : w(this, K).defaultValue), L(this, U, !0);
    }
  }
  disconnect() {
    w(this, U) && (L(this, U, !1), D(this, Ot, se).call(this));
  }
  isConnected() {
    return w(this, U);
  }
  update(t) {
    this.connect(), this.value = t;
  }
  subscribe(t, i = !1) {
    q.subscribe(w(this, ut), (e, r) => {
      (w(this, U) || i) && t(r);
    });
  }
  getProperty() {
    return w(this, K);
  }
  setStoredValue(t) {
    L(this, z, t), this.isConnected() || D(this, Ot, se).call(this);
  }
  getStoredValue() {
    return w(this, z);
  }
};
I = /* @__PURE__ */ new WeakMap(), K = /* @__PURE__ */ new WeakMap(), U = /* @__PURE__ */ new WeakMap(), z = /* @__PURE__ */ new WeakMap(), ut = /* @__PURE__ */ new WeakMap(), ie = /* @__PURE__ */ new WeakSet(), ui = function() {
  const { changeEvent: a, attribute: t } = w(this, K);
  if (a) {
    const i = () => {
      D(this, Et, Mt).call(this);
    };
    w(this, I).addEventListener(a, i, !1);
  } else
    t && new MutationObserver(() => {
      D(this, Et, Mt).call(this);
    }).observe(w(this, I), {
      attributes: !0,
      attributeFilter: [t]
    });
}, Et = /* @__PURE__ */ new WeakSet(), Mt = function() {
  const { value: a } = this;
  this.isConnected() || L(this, z, a), q.publish(w(this, ut), a);
}, Ot = /* @__PURE__ */ new WeakSet(), se = function() {
  this.value = w(this, z);
}, li.UNIQUE_ID = 0;
let Bi = li;
function St(a) {
  return Li(a, { transform: Pi });
}
function Ce(a, t) {
  const i = Object.entries(a).find(([_]) => St(_) === t);
  if (!i)
    return;
  const [e, r] = i;
  return { name: e, source: r };
}
var j, F, Q, B, nt, lt, tt, rt, at, G, _t, re, hi, ae, Re, ht, Nt, _e, Ae, oe, Ne, ue, pi;
const fi = class le {
  constructor(t, i, e) {
    P(this, re), P(this, ae), P(this, ht), P(this, _e), P(this, oe), P(this, ue), P(this, j, void 0), P(this, F, void 0), P(this, Q, void 0), P(this, B, void 0), P(this, nt, void 0), P(this, lt, void 0), P(this, tt, !1), P(this, rt, void 0), P(this, at, {}), P(this, G, ee), P(this, _t, void 0), le.UNIQUE_ID += 1, L(this, _t, Symbol(`WEBBIT_ANY_PROPERTY_CHANGE_${le.UNIQUE_ID}`)), L(this, j, t), L(this, F, i), L(this, Q, oi(e || {
      name: t.tagName.toLowerCase()
    }));
    const r = Object.entries(w(this, Q).properties).filter(([, _]) => !["SourceProvider", "Store", "Source"].includes(_.type)).map(([_, h]) => ({ ...h, name: _ }));
    L(this, B, new Map(
      r.map((_) => {
        const h = new Bi(w(this, j), _);
        return h.subscribe((u) => {
          h.isConnected() && D(this, ue, pi).call(this, _, u), q.publish(w(this, _t), {
            property: _,
            value: u,
            connected: w(this, tt)
          });
        }, !0), [_.name, h];
      })
    )), L(this, nt, r.find(({ primary: _ }) => _)), L(this, lt, w(this, nt) ? w(this, B).get(w(this, nt).name) : void 0), L(this, rt, D(this, re, hi).call(this)), w(this, F).defaultSourceProviderSet((_) => {
      this.sourceProvider === null && (this.sourceProvider = _);
    }), this.connect();
  }
  get sourceProvider() {
    return w(this, j).getAttribute("source-provider") ?? void 0;
  }
  set sourceProvider(t) {
    typeof t < "u" ? w(this, j).setAttribute("source-provider", t) : w(this, j).removeAttribute("source-provider");
  }
  get sourceKey() {
    return w(this, j).getAttribute("source-key") ?? void 0;
  }
  set sourceKey(t) {
    typeof t < "u" ? w(this, j).setAttribute("source-key", t) : w(this, j).removeAttribute("source-key");
  }
  get source() {
    if (typeof this.sourceProvider < "u" && typeof this.sourceKey < "u")
      return w(this, F).getSource(this.sourceProvider, this.sourceKey);
  }
  get hasSource() {
    return typeof this.source < "u";
  }
  subscribe(t) {
    const i = q.subscribe(w(this, _t), (e, r) => {
      t(r);
    });
    return () => q.unsubscribe(i);
  }
  getPropertyHandler(t) {
    return w(this, B).get(t);
  }
  connect() {
    L(this, tt, !0), w(this, rt).connect(), D(this, ht, Nt).call(this);
  }
  disconnect() {
    L(this, tt, !1), w(this, rt).disconnect(), D(this, ht, Nt).call(this);
  }
  setDefaultPropertyValue(t, i) {
    w(this, at)[t] = i;
  }
  getDefaultPropertyValue(t) {
    return w(this, at)[t];
  }
  getDefaultPropertyValues() {
    return w(this, at);
  }
  getConfig() {
    return w(this, Q);
  }
};
j = /* @__PURE__ */ new WeakMap(), F = /* @__PURE__ */ new WeakMap(), Q = /* @__PURE__ */ new WeakMap(), B = /* @__PURE__ */ new WeakMap(), nt = /* @__PURE__ */ new WeakMap(), lt = /* @__PURE__ */ new WeakMap(), tt = /* @__PURE__ */ new WeakMap(), rt = /* @__PURE__ */ new WeakMap(), at = /* @__PURE__ */ new WeakMap(), G = /* @__PURE__ */ new WeakMap(), _t = /* @__PURE__ */ new WeakMap(), re = /* @__PURE__ */ new WeakSet(), hi = function() {
  const a = new MutationObserver(() => {
    D(this, ht, Nt).call(this);
  });
  return {
    connect: () => {
      a.observe(w(this, j), {
        attributes: !0,
        attributeFilter: ["source-provider", "source-key"]
      });
    },
    disconnect: () => {
      a.disconnect();
    }
  };
}, ae = /* @__PURE__ */ new WeakSet(), Re = function() {
  const a = w(this, j).closest("[default-source-provider]");
  return (a == null ? void 0 : a.getAttribute("default-source-provider")) ?? void 0;
}, ht = /* @__PURE__ */ new WeakSet(), Nt = function() {
  if (D(this, _e, Ae).call(this), !w(this, tt)) {
    w(this, G).call(this), L(this, G, ee), w(this, B).forEach((a) => a.disconnect());
    return;
  }
  this.sourceKey || (this.sourceKey = w(this, Q).defaultSourceKey), this.sourceProvider || (this.sourceProvider = w(this, Q).defaultSourceProvider || D(this, ae, Re).call(this) || w(this, F).getDefaultSourceProvider()), !this.sourceKey || !this.sourceProvider ? (w(this, G).call(this), L(this, G, ee), w(this, B).forEach((a) => a.disconnect())) : (w(this, G).call(this), w(this, B).forEach((a) => {
    a.disconnect();
  }), L(this, G, w(this, F).subscribe(
    this.sourceProvider,
    this.sourceKey,
    (a, t, i) => {
      D(this, oe, Ne).call(this, t, i);
    },
    !0
  )));
}, _e = /* @__PURE__ */ new WeakSet(), Ae = function() {
  const a = this.sourceProvider ? w(this, F).getSourceProvider(this.sourceProvider) : void 0;
  Object.entries(w(this, Q).properties).forEach(([, { property: t, type: i }]) => {
    t && (i === "SourceProvider" ? w(this, j)[t] = a : i === "Store" ? w(this, j)[t] = w(this, F) : i === "Source" && (w(this, j)[t] = this.source));
  });
}, oe = /* @__PURE__ */ new WeakSet(), Ne = function(a, t) {
  var i, e;
  if (typeof this.source > "u")
    w(this, B).forEach((r) => {
      r.disconnect();
    });
  else if (this.source.hasChildren()) {
    const r = this.source.getChildren();
    if (a === t) {
      const _ = Object.entries(r);
      [...w(this, B).entries()].forEach(([h, u]) => {
        const l = _.find(([O]) => St(O) === h);
        if (l) {
          const [, O] = l;
          u.update(O.getSourceValue());
        }
      });
    } else {
      const _ = St(t.replace(`${a}/`, ""));
      if (w(this, B).has(_)) {
        const h = w(this, B).get(_), u = (e = (i = Ce(r, _)) == null ? void 0 : i.source) == null ? void 0 : e.getSourceValue();
        typeof u > "u" ? h == null || h.disconnect() : h == null || h.update(u);
      }
    }
  } else
    w(this, lt) && w(this, lt).update(this.source.getSourceValue());
}, ue = /* @__PURE__ */ new WeakSet(), pi = function({ name: a, primary: t }, i) {
  if (typeof this.source > "u")
    return;
  const e = xt(i), r = this.source.getChildren(), _ = Object.entries(r);
  if (this.source.hasChildren())
    if (i === null) {
      const h = _.find(([u]) => St(u) === a);
      if (h) {
        const [, u] = h;
        u.getSourceValue() !== null && u.setSourceValue(i);
      }
    } else {
      const h = Ce(r, a);
      if (!h)
        return;
      const u = xt(h.source.getSourceValue()) || e, l = Z(i, u), O = Z(l, e);
      J(i, O) && !J(h.source.getSourceValue(), l) && h.source.setSourceValue(i);
    }
  else {
    if (!t)
      return;
    if (i === null)
      this.source.getSourceValue() !== null && this.source.setSourceValue(i);
    else {
      const h = xt(this.source.getSourceValue()) || e, u = Z(i, h), l = Z(u, e);
      J(i, l) && !J(this.source.getSourceValue(), u) && this.source.setSourceValue(u);
    }
  }
}, fi.UNIQUE_ID = 0;
let Ki = fi;
var Pe = function(a) {
  var t = a, i, e = {
    a: 0,
    b: 0,
    c: 0
  }, r = [], _ = /(\[[^\]]+\])/g, h = /(#[^\#\s\+>~\.\[:\)]+)/g, u = /(\.[^\s\+>~\.\[:\)]+)/g, l = /(::[^\s\+>~\.\[:]+|:first-line|:first-letter|:before|:after)/gi, O = /(:(?!not|global|local)[\w-]+\([^\)]*\))/gi, v = /(:(?!not|global|local)[^\s\+>~\.\[:]+)/g, y = /([^\s\+>~\.\[:]+)/g;
  return i = function(g, C) {
    var A, x, o, n, p, T;
    if (g.test(t))
      for (A = t.match(g), x = 0, o = A.length; x < o; x += 1)
        e[C] += 1, n = A[x], p = t.indexOf(n), T = n.length, r.push({
          selector: a.substr(p, T),
          type: C,
          index: p,
          length: T
        }), t = t.replace(n, Array(T + 1).join(" "));
  }, function() {
    var g = function(o) {
      var n, p, T, E;
      if (o.test(t))
        for (n = t.match(o), p = 0, T = n.length; p < T; p += 1)
          E = n[p], t = t.replace(E, Array(E.length + 1).join("A"));
    }, C = /\\[0-9A-Fa-f]{6}\s?/g, A = /\\[0-9A-Fa-f]{1,5}\s/g, x = /\\./g;
    g(C), g(A), g(x);
  }(), function() {
    var g = /{[^]*/gm, C, A, x, o;
    if (g.test(t))
      for (C = t.match(g), A = 0, x = C.length; A < x; A += 1)
        o = C[A], t = t.replace(o, Array(o.length + 1).join(" "));
  }(), i(_, "b"), i(h, "a"), i(u, "b"), i(l, "c"), i(O, "b"), i(v, "b"), t = t.replace(/[\*\s\+>~]/g, " "), t = t.replace(/[#\.]/g, " "), t = t.replace(/:not/g, "    "), t = t.replace(/:local/g, "      "), t = t.replace(/:global/g, "       "), t = t.replace(/[\(\)]/g, " "), i(y, "c"), r.sort(function(g, C) {
    return g.index - C.index;
  }), {
    selector: a,
    specificity: "0," + e.a.toString() + "," + e.b.toString() + "," + e.c.toString(),
    specificityArray: [0, e.a, e.b, e.c],
    parts: r
  };
}, Le = function(a, t) {
  var i, e, r;
  if (typeof a == "string") {
    if (a.indexOf(",") !== -1)
      throw "Invalid CSS selector";
    i = Pe(a).specificityArray;
  } else if (Array.isArray(a)) {
    if (a.filter(function(_) {
      return typeof _ == "number";
    }).length !== 4)
      throw "Invalid specificity array";
    i = a;
  } else
    throw "Invalid CSS selector or specificity array";
  if (typeof t == "string") {
    if (t.indexOf(",") !== -1)
      throw "Invalid CSS selector";
    e = Pe(t).specificityArray;
  } else if (Array.isArray(t)) {
    if (t.filter(function(_) {
      return typeof _ == "number";
    }).length !== 4)
      throw "Invalid specificity array";
    e = t;
  } else
    throw "Invalid CSS selector or specificity array";
  for (r = 0; r < 4; r += 1) {
    if (i[r] < e[r])
      return -1;
    if (i[r] > e[r])
      return 1;
  }
  return 0;
};
function Vi(a, t) {
  const i = [...a.entries()].find(([e]) => t.matches(e));
  return i == null ? void 0 : i[1];
}
function ci(a, t, i) {
  const e = a, r = Vi(i, e);
  return typeof r > "u" || !e.hasAttribute("source-provider") && typeof r.defaultSourceProvider > "u" && typeof t > "u" ? NodeFilter.FILTER_SKIP : NodeFilter.FILTER_ACCEPT;
}
function Ui(a, t) {
  return (i) => ci(i, a, t);
}
function zi(a, t, i) {
  return document.createNodeIterator(
    a,
    NodeFilter.SHOW_ELEMENT,
    Ui(t, i)
  );
}
var X, V, H, et, pt, ft, ot, dt, Tt, jt, Pt, he, pe, di, fe, De, kt, It;
const gi = class Ct {
  constructor(t, i, e = {}, r = "default") {
    P(this, ot), P(this, Tt), P(this, Pt), P(this, pe), P(this, fe), P(this, kt), P(this, X, void 0), P(this, V, /* @__PURE__ */ new Map()), P(this, H, /* @__PURE__ */ new Map()), P(this, et, void 0), P(this, pt, void 0), P(this, ft, void 0), Ct.UNIQUE_ID += 1, L(this, pt, Symbol(`WEBBIT_CONNECTOR_ELEMENT_CONNECTED_${Ct.UNIQUE_ID}`)), L(this, ft, Symbol(`WEBBIT_CONNECTOR_ELEMENT_DISCONNECTED_${Ct.UNIQUE_ID}`)), L(this, X, i), w(this, X).defaultSourceProviderSet((_) => {
      w(this, H).forEach((h, u) => {
        h.sourceProvider === null && this.setSourceProvider(u, _);
      });
    }), L(this, et, t), D(this, pe, di).call(this), this.addElementConfigs(e, r);
  }
  addElementConfigs(t = {}, i = "default") {
    Object.entries(t).forEach(([e, r]) => {
      w(this, V).set(e, oi({
        group: i,
        ...r
      }));
    }), D(this, Pt, he).call(this, w(this, et));
  }
  subscribeElementConnected(t) {
    q.subscribe(w(this, pt), (i, e) => {
      t(e);
    });
  }
  subscribeElementDisconnected(t) {
    q.subscribe(w(this, ft), (i, e) => {
      t(e);
    });
  }
  getRootElement() {
    return w(this, et);
  }
  /**
   * Gets the element config with the most specific selector that matches the element
   * @param element
   * @returns The element config or undefined if none is found
   */
  getMatchingElementConfig(t) {
    if (!t)
      return;
    const i = [
      ...w(this, V).keys()
    ].filter((e) => t.matches(e)).sort(Le).at(-1);
    return i ? this.getElementConfig(i) : void 0;
  }
  /**
   * Gets the most specific selector that matches the element
   * @param element
   * @returns The selector or undefined if none is found
   */
  getMatchingElementSelector(t) {
    return t ? [
      ...w(this, V).keys()
    ].filter((i) => t.matches(i)).sort(Le).at(-1) : void 0;
  }
  getElementConfigSelectors() {
    return [...w(this, V).keys()];
  }
  getElementConfig(t) {
    return w(this, V).get(t);
  }
  hasElementConfig(t) {
    return w(this, V).has(t);
  }
  setDefaultPropertyValue(t, i, e) {
    const r = this.getElementWebbit(t);
    r && r.setDefaultPropertyValue(i, e);
  }
  setSourceProvider(t, i) {
    const e = this.getElementWebbit(t);
    e && (e.sourceProvider = i);
  }
  setSourceKey(t, i) {
    const e = this.getElementWebbit(t);
    e && (e.sourceKey = i);
  }
  getDefaultPropertyValue(t, i) {
    const e = this.getElementWebbit(t);
    return e ? e.getDefaultPropertyValue(i) : void 0;
  }
  getDefaultPropertyValues(t) {
    const i = this.getElementWebbit(t);
    return i ? i.getDefaultPropertyValues() : {};
  }
  getSourceProvider(t) {
    const i = this.getElementWebbit(t);
    return i ? i.sourceProvider : w(this, X).getDefaultSourceProvider();
  }
  getSourceKey(t) {
    const i = this.getElementWebbit(t);
    return i ? i.sourceKey : void 0;
  }
  getElementWebbit(t) {
    return w(this, H).get(t);
  }
  hasElement(t) {
    return w(this, H).has(t);
  }
  getStore() {
    return w(this, X);
  }
};
X = /* @__PURE__ */ new WeakMap(), V = /* @__PURE__ */ new WeakMap(), H = /* @__PURE__ */ new WeakMap(), et = /* @__PURE__ */ new WeakMap(), pt = /* @__PURE__ */ new WeakMap(), ft = /* @__PURE__ */ new WeakMap(), ot = /* @__PURE__ */ new WeakSet(), dt = function(a) {
  const t = this.getMatchingElementConfig(a);
  if (this.hasElement(a) || !t)
    return;
  const i = new Ki(a, w(this, X), t);
  w(this, H).set(a, i), q.publish(w(this, pt), { element: a });
}, Tt = /* @__PURE__ */ new WeakSet(), jt = function(a) {
  const t = this.getElementWebbit(a);
  t && (t.disconnect(), w(this, H).delete(a), q.publish(w(this, ft), { element: a }));
}, Pt = /* @__PURE__ */ new WeakSet(), he = function(a) {
  const t = D(this, fe, De).call(this, a);
  for (; t.nextNode(); )
    D(this, ot, dt).call(this, t.referenceNode);
}, pe = /* @__PURE__ */ new WeakSet(), di = function() {
  new MutationObserver((a) => {
    a.forEach((t) => {
      if (t.type === "childList") {
        const i = t.addedNodes || [], e = t.removedNodes || [];
        i.forEach((r) => {
          r instanceof HTMLElement && (D(this, kt, It).call(this, r) && D(this, ot, dt).call(this, r), D(this, Pt, he).call(this, r));
        }), e.forEach((r) => {
          D(this, Tt, jt).call(this, r), [...w(this, H).keys()].forEach((_) => {
            r.contains(_) && D(this, Tt, jt).call(this, _);
          });
        });
      } else
        t.type === "attributes" && t.target instanceof HTMLElement && D(this, kt, It).call(this, t.target) && D(this, ot, dt).call(this, t.target);
    });
  }).observe(w(this, et), {
    childList: !0,
    subtree: !0,
    attributes: !0,
    attributeFilter: ["source-key", "source-provider"]
  });
}, fe = /* @__PURE__ */ new WeakSet(), De = function(a) {
  return zi(
    a,
    this.getStore().getDefaultSourceProvider(),
    w(this, V)
  );
}, kt = /* @__PURE__ */ new WeakSet(), It = function(a) {
  return ci(
    a,
    w(this, X).getDefaultSourceProvider(),
    w(this, V)
  ) === NodeFilter.FILTER_ACCEPT;
}, gi.UNIQUE_ID = 0;
let Fi = gi;
var Bt = { exports: {} }, it = {}, gt = { exports: {} }, Kt = {}, Vt = {}, We;
function me() {
  if (We) return Vt;
  We = 1;
  function a(e) {
    this.__parent = e, this.__character_count = 0, this.__indent_count = -1, this.__alignment_count = 0, this.__wrap_point_index = 0, this.__wrap_point_character_count = 0, this.__wrap_point_indent_count = -1, this.__wrap_point_alignment_count = 0, this.__items = [];
  }
  a.prototype.clone_empty = function() {
    var e = new a(this.__parent);
    return e.set_indent(this.__indent_count, this.__alignment_count), e;
  }, a.prototype.item = function(e) {
    return e < 0 ? this.__items[this.__items.length + e] : this.__items[e];
  }, a.prototype.has_match = function(e) {
    for (var r = this.__items.length - 1; r >= 0; r--)
      if (this.__items[r].match(e))
        return !0;
    return !1;
  }, a.prototype.set_indent = function(e, r) {
    this.is_empty() && (this.__indent_count = e || 0, this.__alignment_count = r || 0, this.__character_count = this.__parent.get_indent_size(this.__indent_count, this.__alignment_count));
  }, a.prototype._set_wrap_point = function() {
    this.__parent.wrap_line_length && (this.__wrap_point_index = this.__items.length, this.__wrap_point_character_count = this.__character_count, this.__wrap_point_indent_count = this.__parent.next_line.__indent_count, this.__wrap_point_alignment_count = this.__parent.next_line.__alignment_count);
  }, a.prototype._should_wrap = function() {
    return this.__wrap_point_index && this.__character_count > this.__parent.wrap_line_length && this.__wrap_point_character_count > this.__parent.next_line.__character_count;
  }, a.prototype._allow_wrap = function() {
    if (this._should_wrap()) {
      this.__parent.add_new_line();
      var e = this.__parent.current_line;
      return e.set_indent(this.__wrap_point_indent_count, this.__wrap_point_alignment_count), e.__items = this.__items.slice(this.__wrap_point_index), this.__items = this.__items.slice(0, this.__wrap_point_index), e.__character_count += this.__character_count - this.__wrap_point_character_count, this.__character_count = this.__wrap_point_character_count, e.__items[0] === " " && (e.__items.splice(0, 1), e.__character_count -= 1), !0;
    }
    return !1;
  }, a.prototype.is_empty = function() {
    return this.__items.length === 0;
  }, a.prototype.last = function() {
    return this.is_empty() ? null : this.__items[this.__items.length - 1];
  }, a.prototype.push = function(e) {
    this.__items.push(e);
    var r = e.lastIndexOf(`
`);
    r !== -1 ? this.__character_count = e.length - r : this.__character_count += e.length;
  }, a.prototype.pop = function() {
    var e = null;
    return this.is_empty() || (e = this.__items.pop(), this.__character_count -= e.length), e;
  }, a.prototype._remove_indent = function() {
    this.__indent_count > 0 && (this.__indent_count -= 1, this.__character_count -= this.__parent.indent_size);
  }, a.prototype._remove_wrap_indent = function() {
    this.__wrap_point_indent_count > 0 && (this.__wrap_point_indent_count -= 1);
  }, a.prototype.trim = function() {
    for (; this.last() === " "; )
      this.__items.pop(), this.__character_count -= 1;
  }, a.prototype.toString = function() {
    var e = "";
    return this.is_empty() ? this.__parent.indent_empty_lines && (e = this.__parent.get_indent_string(this.__indent_count)) : (e = this.__parent.get_indent_string(this.__indent_count, this.__alignment_count), e += this.__items.join("")), e;
  };
  function t(e, r) {
    this.__cache = [""], this.__indent_size = e.indent_size, this.__indent_string = e.indent_char, e.indent_with_tabs || (this.__indent_string = new Array(e.indent_size + 1).join(e.indent_char)), r = r || "", e.indent_level > 0 && (r = new Array(e.indent_level + 1).join(this.__indent_string)), this.__base_string = r, this.__base_string_length = r.length;
  }
  t.prototype.get_indent_size = function(e, r) {
    var _ = this.__base_string_length;
    return r = r || 0, e < 0 && (_ = 0), _ += e * this.__indent_size, _ += r, _;
  }, t.prototype.get_indent_string = function(e, r) {
    var _ = this.__base_string;
    return r = r || 0, e < 0 && (e = 0, _ = ""), r += e * this.__indent_size, this.__ensure_cache(r), _ += this.__cache[r], _;
  }, t.prototype.__ensure_cache = function(e) {
    for (; e >= this.__cache.length; )
      this.__add_column();
  }, t.prototype.__add_column = function() {
    var e = this.__cache.length, r = 0, _ = "";
    this.__indent_size && e >= this.__indent_size && (r = Math.floor(e / this.__indent_size), e -= r * this.__indent_size, _ = new Array(r + 1).join(this.__indent_string)), e && (_ += new Array(e + 1).join(" ")), this.__cache.push(_);
  };
  function i(e, r) {
    this.__indent_cache = new t(e, r), this.raw = !1, this._end_with_newline = e.end_with_newline, this.indent_size = e.indent_size, this.wrap_line_length = e.wrap_line_length, this.indent_empty_lines = e.indent_empty_lines, this.__lines = [], this.previous_line = null, this.current_line = null, this.next_line = new a(this), this.space_before_token = !1, this.non_breaking_space = !1, this.previous_token_wrapped = !1, this.__add_outputline();
  }
  return i.prototype.__add_outputline = function() {
    this.previous_line = this.current_line, this.current_line = this.next_line.clone_empty(), this.__lines.push(this.current_line);
  }, i.prototype.get_line_number = function() {
    return this.__lines.length;
  }, i.prototype.get_indent_string = function(e, r) {
    return this.__indent_cache.get_indent_string(e, r);
  }, i.prototype.get_indent_size = function(e, r) {
    return this.__indent_cache.get_indent_size(e, r);
  }, i.prototype.is_empty = function() {
    return !this.previous_line && this.current_line.is_empty();
  }, i.prototype.add_new_line = function(e) {
    return this.is_empty() || !e && this.just_added_newline() ? !1 : (this.raw || this.__add_outputline(), !0);
  }, i.prototype.get_code = function(e) {
    this.trim(!0);
    var r = this.current_line.pop();
    r && (r[r.length - 1] === `
` && (r = r.replace(/\n+$/g, "")), this.current_line.push(r)), this._end_with_newline && this.__add_outputline();
    var _ = this.__lines.join(`
`);
    return e !== `
` && (_ = _.replace(/[\n]/g, e)), _;
  }, i.prototype.set_wrap_point = function() {
    this.current_line._set_wrap_point();
  }, i.prototype.set_indent = function(e, r) {
    return e = e || 0, r = r || 0, this.next_line.set_indent(e, r), this.__lines.length > 1 ? (this.current_line.set_indent(e, r), !0) : (this.current_line.set_indent(), !1);
  }, i.prototype.add_raw_token = function(e) {
    for (var r = 0; r < e.newlines; r++)
      this.__add_outputline();
    this.current_line.set_indent(-1), this.current_line.push(e.whitespace_before), this.current_line.push(e.text), this.space_before_token = !1, this.non_breaking_space = !1, this.previous_token_wrapped = !1;
  }, i.prototype.add_token = function(e) {
    this.__add_space_before_token(), this.current_line.push(e), this.space_before_token = !1, this.non_breaking_space = !1, this.previous_token_wrapped = this.current_line._allow_wrap();
  }, i.prototype.__add_space_before_token = function() {
    this.space_before_token && !this.just_added_newline() && (this.non_breaking_space || this.set_wrap_point(), this.current_line.push(" "));
  }, i.prototype.remove_indent = function(e) {
    for (var r = this.__lines.length; e < r; )
      this.__lines[e]._remove_indent(), e++;
    this.current_line._remove_wrap_indent();
  }, i.prototype.trim = function(e) {
    for (e = e === void 0 ? !1 : e, this.current_line.trim(); e && this.__lines.length > 1 && this.current_line.is_empty(); )
      this.__lines.pop(), this.current_line = this.__lines[this.__lines.length - 1], this.current_line.trim();
    this.previous_line = this.__lines.length > 1 ? this.__lines[this.__lines.length - 2] : null;
  }, i.prototype.just_added_newline = function() {
    return this.current_line.is_empty();
  }, i.prototype.just_added_blankline = function() {
    return this.is_empty() || this.current_line.is_empty() && this.previous_line.is_empty();
  }, i.prototype.ensure_empty_line_above = function(e, r) {
    for (var _ = this.__lines.length - 2; _ >= 0; ) {
      var h = this.__lines[_];
      if (h.is_empty())
        break;
      if (h.item(0).indexOf(e) !== 0 && h.item(-1) !== r) {
        this.__lines.splice(_ + 1, 0, new a(this)), this.previous_line = this.__lines[this.__lines.length - 2];
        break;
      }
      _--;
    }
  }, Vt.Output = i, Vt;
}
var Ut = {}, Me;
function bi() {
  if (Me) return Ut;
  Me = 1;
  function a(t, i, e, r) {
    this.type = t, this.text = i, this.comments_before = null, this.newlines = e || 0, this.whitespace_before = r || "", this.parent = null, this.next = null, this.previous = null, this.opened = null, this.closed = null, this.directives = null;
  }
  return Ut.Token = a, Ut;
}
var zt = {}, je;
function mi() {
  return je || (je = 1, function(a) {
    var t = "\\x23\\x24\\x40\\x41-\\x5a\\x5f\\x61-\\x7a", i = "\\x24\\x30-\\x39\\x41-\\x5a\\x5f\\x61-\\x7a", e = "\\xaa\\xb5\\xba\\xc0-\\xd6\\xd8-\\xf6\\xf8-\\u02c1\\u02c6-\\u02d1\\u02e0-\\u02e4\\u02ec\\u02ee\\u0370-\\u0374\\u0376\\u0377\\u037a-\\u037d\\u0386\\u0388-\\u038a\\u038c\\u038e-\\u03a1\\u03a3-\\u03f5\\u03f7-\\u0481\\u048a-\\u0527\\u0531-\\u0556\\u0559\\u0561-\\u0587\\u05d0-\\u05ea\\u05f0-\\u05f2\\u0620-\\u064a\\u066e\\u066f\\u0671-\\u06d3\\u06d5\\u06e5\\u06e6\\u06ee\\u06ef\\u06fa-\\u06fc\\u06ff\\u0710\\u0712-\\u072f\\u074d-\\u07a5\\u07b1\\u07ca-\\u07ea\\u07f4\\u07f5\\u07fa\\u0800-\\u0815\\u081a\\u0824\\u0828\\u0840-\\u0858\\u08a0\\u08a2-\\u08ac\\u0904-\\u0939\\u093d\\u0950\\u0958-\\u0961\\u0971-\\u0977\\u0979-\\u097f\\u0985-\\u098c\\u098f\\u0990\\u0993-\\u09a8\\u09aa-\\u09b0\\u09b2\\u09b6-\\u09b9\\u09bd\\u09ce\\u09dc\\u09dd\\u09df-\\u09e1\\u09f0\\u09f1\\u0a05-\\u0a0a\\u0a0f\\u0a10\\u0a13-\\u0a28\\u0a2a-\\u0a30\\u0a32\\u0a33\\u0a35\\u0a36\\u0a38\\u0a39\\u0a59-\\u0a5c\\u0a5e\\u0a72-\\u0a74\\u0a85-\\u0a8d\\u0a8f-\\u0a91\\u0a93-\\u0aa8\\u0aaa-\\u0ab0\\u0ab2\\u0ab3\\u0ab5-\\u0ab9\\u0abd\\u0ad0\\u0ae0\\u0ae1\\u0b05-\\u0b0c\\u0b0f\\u0b10\\u0b13-\\u0b28\\u0b2a-\\u0b30\\u0b32\\u0b33\\u0b35-\\u0b39\\u0b3d\\u0b5c\\u0b5d\\u0b5f-\\u0b61\\u0b71\\u0b83\\u0b85-\\u0b8a\\u0b8e-\\u0b90\\u0b92-\\u0b95\\u0b99\\u0b9a\\u0b9c\\u0b9e\\u0b9f\\u0ba3\\u0ba4\\u0ba8-\\u0baa\\u0bae-\\u0bb9\\u0bd0\\u0c05-\\u0c0c\\u0c0e-\\u0c10\\u0c12-\\u0c28\\u0c2a-\\u0c33\\u0c35-\\u0c39\\u0c3d\\u0c58\\u0c59\\u0c60\\u0c61\\u0c85-\\u0c8c\\u0c8e-\\u0c90\\u0c92-\\u0ca8\\u0caa-\\u0cb3\\u0cb5-\\u0cb9\\u0cbd\\u0cde\\u0ce0\\u0ce1\\u0cf1\\u0cf2\\u0d05-\\u0d0c\\u0d0e-\\u0d10\\u0d12-\\u0d3a\\u0d3d\\u0d4e\\u0d60\\u0d61\\u0d7a-\\u0d7f\\u0d85-\\u0d96\\u0d9a-\\u0db1\\u0db3-\\u0dbb\\u0dbd\\u0dc0-\\u0dc6\\u0e01-\\u0e30\\u0e32\\u0e33\\u0e40-\\u0e46\\u0e81\\u0e82\\u0e84\\u0e87\\u0e88\\u0e8a\\u0e8d\\u0e94-\\u0e97\\u0e99-\\u0e9f\\u0ea1-\\u0ea3\\u0ea5\\u0ea7\\u0eaa\\u0eab\\u0ead-\\u0eb0\\u0eb2\\u0eb3\\u0ebd\\u0ec0-\\u0ec4\\u0ec6\\u0edc-\\u0edf\\u0f00\\u0f40-\\u0f47\\u0f49-\\u0f6c\\u0f88-\\u0f8c\\u1000-\\u102a\\u103f\\u1050-\\u1055\\u105a-\\u105d\\u1061\\u1065\\u1066\\u106e-\\u1070\\u1075-\\u1081\\u108e\\u10a0-\\u10c5\\u10c7\\u10cd\\u10d0-\\u10fa\\u10fc-\\u1248\\u124a-\\u124d\\u1250-\\u1256\\u1258\\u125a-\\u125d\\u1260-\\u1288\\u128a-\\u128d\\u1290-\\u12b0\\u12b2-\\u12b5\\u12b8-\\u12be\\u12c0\\u12c2-\\u12c5\\u12c8-\\u12d6\\u12d8-\\u1310\\u1312-\\u1315\\u1318-\\u135a\\u1380-\\u138f\\u13a0-\\u13f4\\u1401-\\u166c\\u166f-\\u167f\\u1681-\\u169a\\u16a0-\\u16ea\\u16ee-\\u16f0\\u1700-\\u170c\\u170e-\\u1711\\u1720-\\u1731\\u1740-\\u1751\\u1760-\\u176c\\u176e-\\u1770\\u1780-\\u17b3\\u17d7\\u17dc\\u1820-\\u1877\\u1880-\\u18a8\\u18aa\\u18b0-\\u18f5\\u1900-\\u191c\\u1950-\\u196d\\u1970-\\u1974\\u1980-\\u19ab\\u19c1-\\u19c7\\u1a00-\\u1a16\\u1a20-\\u1a54\\u1aa7\\u1b05-\\u1b33\\u1b45-\\u1b4b\\u1b83-\\u1ba0\\u1bae\\u1baf\\u1bba-\\u1be5\\u1c00-\\u1c23\\u1c4d-\\u1c4f\\u1c5a-\\u1c7d\\u1ce9-\\u1cec\\u1cee-\\u1cf1\\u1cf5\\u1cf6\\u1d00-\\u1dbf\\u1e00-\\u1f15\\u1f18-\\u1f1d\\u1f20-\\u1f45\\u1f48-\\u1f4d\\u1f50-\\u1f57\\u1f59\\u1f5b\\u1f5d\\u1f5f-\\u1f7d\\u1f80-\\u1fb4\\u1fb6-\\u1fbc\\u1fbe\\u1fc2-\\u1fc4\\u1fc6-\\u1fcc\\u1fd0-\\u1fd3\\u1fd6-\\u1fdb\\u1fe0-\\u1fec\\u1ff2-\\u1ff4\\u1ff6-\\u1ffc\\u2071\\u207f\\u2090-\\u209c\\u2102\\u2107\\u210a-\\u2113\\u2115\\u2119-\\u211d\\u2124\\u2126\\u2128\\u212a-\\u212d\\u212f-\\u2139\\u213c-\\u213f\\u2145-\\u2149\\u214e\\u2160-\\u2188\\u2c00-\\u2c2e\\u2c30-\\u2c5e\\u2c60-\\u2ce4\\u2ceb-\\u2cee\\u2cf2\\u2cf3\\u2d00-\\u2d25\\u2d27\\u2d2d\\u2d30-\\u2d67\\u2d6f\\u2d80-\\u2d96\\u2da0-\\u2da6\\u2da8-\\u2dae\\u2db0-\\u2db6\\u2db8-\\u2dbe\\u2dc0-\\u2dc6\\u2dc8-\\u2dce\\u2dd0-\\u2dd6\\u2dd8-\\u2dde\\u2e2f\\u3005-\\u3007\\u3021-\\u3029\\u3031-\\u3035\\u3038-\\u303c\\u3041-\\u3096\\u309d-\\u309f\\u30a1-\\u30fa\\u30fc-\\u30ff\\u3105-\\u312d\\u3131-\\u318e\\u31a0-\\u31ba\\u31f0-\\u31ff\\u3400-\\u4db5\\u4e00-\\u9fcc\\ua000-\\ua48c\\ua4d0-\\ua4fd\\ua500-\\ua60c\\ua610-\\ua61f\\ua62a\\ua62b\\ua640-\\ua66e\\ua67f-\\ua697\\ua6a0-\\ua6ef\\ua717-\\ua71f\\ua722-\\ua788\\ua78b-\\ua78e\\ua790-\\ua793\\ua7a0-\\ua7aa\\ua7f8-\\ua801\\ua803-\\ua805\\ua807-\\ua80a\\ua80c-\\ua822\\ua840-\\ua873\\ua882-\\ua8b3\\ua8f2-\\ua8f7\\ua8fb\\ua90a-\\ua925\\ua930-\\ua946\\ua960-\\ua97c\\ua984-\\ua9b2\\ua9cf\\uaa00-\\uaa28\\uaa40-\\uaa42\\uaa44-\\uaa4b\\uaa60-\\uaa76\\uaa7a\\uaa80-\\uaaaf\\uaab1\\uaab5\\uaab6\\uaab9-\\uaabd\\uaac0\\uaac2\\uaadb-\\uaadd\\uaae0-\\uaaea\\uaaf2-\\uaaf4\\uab01-\\uab06\\uab09-\\uab0e\\uab11-\\uab16\\uab20-\\uab26\\uab28-\\uab2e\\uabc0-\\uabe2\\uac00-\\ud7a3\\ud7b0-\\ud7c6\\ud7cb-\\ud7fb\\uf900-\\ufa6d\\ufa70-\\ufad9\\ufb00-\\ufb06\\ufb13-\\ufb17\\ufb1d\\ufb1f-\\ufb28\\ufb2a-\\ufb36\\ufb38-\\ufb3c\\ufb3e\\ufb40\\ufb41\\ufb43\\ufb44\\ufb46-\\ufbb1\\ufbd3-\\ufd3d\\ufd50-\\ufd8f\\ufd92-\\ufdc7\\ufdf0-\\ufdfb\\ufe70-\\ufe74\\ufe76-\\ufefc\\uff21-\\uff3a\\uff41-\\uff5a\\uff66-\\uffbe\\uffc2-\\uffc7\\uffca-\\uffcf\\uffd2-\\uffd7\\uffda-\\uffdc", r = "\\u0300-\\u036f\\u0483-\\u0487\\u0591-\\u05bd\\u05bf\\u05c1\\u05c2\\u05c4\\u05c5\\u05c7\\u0610-\\u061a\\u0620-\\u0649\\u0672-\\u06d3\\u06e7-\\u06e8\\u06fb-\\u06fc\\u0730-\\u074a\\u0800-\\u0814\\u081b-\\u0823\\u0825-\\u0827\\u0829-\\u082d\\u0840-\\u0857\\u08e4-\\u08fe\\u0900-\\u0903\\u093a-\\u093c\\u093e-\\u094f\\u0951-\\u0957\\u0962-\\u0963\\u0966-\\u096f\\u0981-\\u0983\\u09bc\\u09be-\\u09c4\\u09c7\\u09c8\\u09d7\\u09df-\\u09e0\\u0a01-\\u0a03\\u0a3c\\u0a3e-\\u0a42\\u0a47\\u0a48\\u0a4b-\\u0a4d\\u0a51\\u0a66-\\u0a71\\u0a75\\u0a81-\\u0a83\\u0abc\\u0abe-\\u0ac5\\u0ac7-\\u0ac9\\u0acb-\\u0acd\\u0ae2-\\u0ae3\\u0ae6-\\u0aef\\u0b01-\\u0b03\\u0b3c\\u0b3e-\\u0b44\\u0b47\\u0b48\\u0b4b-\\u0b4d\\u0b56\\u0b57\\u0b5f-\\u0b60\\u0b66-\\u0b6f\\u0b82\\u0bbe-\\u0bc2\\u0bc6-\\u0bc8\\u0bca-\\u0bcd\\u0bd7\\u0be6-\\u0bef\\u0c01-\\u0c03\\u0c46-\\u0c48\\u0c4a-\\u0c4d\\u0c55\\u0c56\\u0c62-\\u0c63\\u0c66-\\u0c6f\\u0c82\\u0c83\\u0cbc\\u0cbe-\\u0cc4\\u0cc6-\\u0cc8\\u0cca-\\u0ccd\\u0cd5\\u0cd6\\u0ce2-\\u0ce3\\u0ce6-\\u0cef\\u0d02\\u0d03\\u0d46-\\u0d48\\u0d57\\u0d62-\\u0d63\\u0d66-\\u0d6f\\u0d82\\u0d83\\u0dca\\u0dcf-\\u0dd4\\u0dd6\\u0dd8-\\u0ddf\\u0df2\\u0df3\\u0e34-\\u0e3a\\u0e40-\\u0e45\\u0e50-\\u0e59\\u0eb4-\\u0eb9\\u0ec8-\\u0ecd\\u0ed0-\\u0ed9\\u0f18\\u0f19\\u0f20-\\u0f29\\u0f35\\u0f37\\u0f39\\u0f41-\\u0f47\\u0f71-\\u0f84\\u0f86-\\u0f87\\u0f8d-\\u0f97\\u0f99-\\u0fbc\\u0fc6\\u1000-\\u1029\\u1040-\\u1049\\u1067-\\u106d\\u1071-\\u1074\\u1082-\\u108d\\u108f-\\u109d\\u135d-\\u135f\\u170e-\\u1710\\u1720-\\u1730\\u1740-\\u1750\\u1772\\u1773\\u1780-\\u17b2\\u17dd\\u17e0-\\u17e9\\u180b-\\u180d\\u1810-\\u1819\\u1920-\\u192b\\u1930-\\u193b\\u1951-\\u196d\\u19b0-\\u19c0\\u19c8-\\u19c9\\u19d0-\\u19d9\\u1a00-\\u1a15\\u1a20-\\u1a53\\u1a60-\\u1a7c\\u1a7f-\\u1a89\\u1a90-\\u1a99\\u1b46-\\u1b4b\\u1b50-\\u1b59\\u1b6b-\\u1b73\\u1bb0-\\u1bb9\\u1be6-\\u1bf3\\u1c00-\\u1c22\\u1c40-\\u1c49\\u1c5b-\\u1c7d\\u1cd0-\\u1cd2\\u1d00-\\u1dbe\\u1e01-\\u1f15\\u200c\\u200d\\u203f\\u2040\\u2054\\u20d0-\\u20dc\\u20e1\\u20e5-\\u20f0\\u2d81-\\u2d96\\u2de0-\\u2dff\\u3021-\\u3028\\u3099\\u309a\\ua640-\\ua66d\\ua674-\\ua67d\\ua69f\\ua6f0-\\ua6f1\\ua7f8-\\ua800\\ua806\\ua80b\\ua823-\\ua827\\ua880-\\ua881\\ua8b4-\\ua8c4\\ua8d0-\\ua8d9\\ua8f3-\\ua8f7\\ua900-\\ua909\\ua926-\\ua92d\\ua930-\\ua945\\ua980-\\ua983\\ua9b3-\\ua9c0\\uaa00-\\uaa27\\uaa40-\\uaa41\\uaa4c-\\uaa4d\\uaa50-\\uaa59\\uaa7b\\uaae0-\\uaae9\\uaaf2-\\uaaf3\\uabc0-\\uabe1\\uabec\\uabed\\uabf0-\\uabf9\\ufb20-\\ufb28\\ufe00-\\ufe0f\\ufe20-\\ufe26\\ufe33\\ufe34\\ufe4d-\\ufe4f\\uff10-\\uff19\\uff3f", _ = "\\\\u[0-9a-fA-F]{4}|\\\\u\\{[0-9a-fA-F]+\\}", h = "(?:" + _ + "|[" + t + e + "])", u = "(?:" + _ + "|[" + i + e + r + "])*";
    a.identifier = new RegExp(h + u, "g"), a.identifierStart = new RegExp(h), a.identifierMatch = new RegExp("(?:" + _ + "|[" + i + e + r + "])+"), a.newline = /[\n\r\u2028\u2029]/, a.lineBreak = new RegExp(`\r
|` + a.newline.source), a.allLineBreaks = new RegExp(a.lineBreak.source, "g");
  }(zt)), zt;
}
var Ft = {}, st = {}, Ie;
function ve() {
  if (Ie) return st;
  Ie = 1;
  function a(e, r) {
    this.raw_options = t(e, r), this.disabled = this._get_boolean("disabled"), this.eol = this._get_characters("eol", "auto"), this.end_with_newline = this._get_boolean("end_with_newline"), this.indent_size = this._get_number("indent_size", 4), this.indent_char = this._get_characters("indent_char", " "), this.indent_level = this._get_number("indent_level"), this.preserve_newlines = this._get_boolean("preserve_newlines", !0), this.max_preserve_newlines = this._get_number("max_preserve_newlines", 32786), this.preserve_newlines || (this.max_preserve_newlines = 0), this.indent_with_tabs = this._get_boolean("indent_with_tabs", this.indent_char === "	"), this.indent_with_tabs && (this.indent_char = "	", this.indent_size === 1 && (this.indent_size = 4)), this.wrap_line_length = this._get_number("wrap_line_length", this._get_number("max_char")), this.indent_empty_lines = this._get_boolean("indent_empty_lines"), this.templating = this._get_selection_list("templating", ["auto", "none", "angular", "django", "erb", "handlebars", "php", "smarty"], ["auto"]);
  }
  a.prototype._get_array = function(e, r) {
    var _ = this.raw_options[e], h = r || [];
    return typeof _ == "object" ? _ !== null && typeof _.concat == "function" && (h = _.concat()) : typeof _ == "string" && (h = _.split(/[^a-zA-Z0-9_\/\-]+/)), h;
  }, a.prototype._get_boolean = function(e, r) {
    var _ = this.raw_options[e], h = _ === void 0 ? !!r : !!_;
    return h;
  }, a.prototype._get_characters = function(e, r) {
    var _ = this.raw_options[e], h = r || "";
    return typeof _ == "string" && (h = _.replace(/\\r/, "\r").replace(/\\n/, `
`).replace(/\\t/, "	")), h;
  }, a.prototype._get_number = function(e, r) {
    var _ = this.raw_options[e];
    r = parseInt(r, 10), isNaN(r) && (r = 0);
    var h = parseInt(_, 10);
    return isNaN(h) && (h = r), h;
  }, a.prototype._get_selection = function(e, r, _) {
    var h = this._get_selection_list(e, r, _);
    if (h.length !== 1)
      throw new Error(
        "Invalid Option Value: The option '" + e + `' can only be one of the following values:
` + r + `
You passed in: '` + this.raw_options[e] + "'"
      );
    return h[0];
  }, a.prototype._get_selection_list = function(e, r, _) {
    if (!r || r.length === 0)
      throw new Error("Selection list cannot be empty.");
    if (_ = _ || [r[0]], !this._is_valid_selection(_, r))
      throw new Error("Invalid Default Value!");
    var h = this._get_array(e, _);
    if (!this._is_valid_selection(h, r))
      throw new Error(
        "Invalid Option Value: The option '" + e + `' can contain only the following values:
` + r + `
You passed in: '` + this.raw_options[e] + "'"
      );
    return h;
  }, a.prototype._is_valid_selection = function(e, r) {
    return e.length && r.length && !e.some(function(_) {
      return r.indexOf(_) === -1;
    });
  };
  function t(e, r) {
    var _ = {};
    e = i(e);
    var h;
    for (h in e)
      h !== r && (_[h] = e[h]);
    if (r && e[r])
      for (h in e[r])
        _[h] = e[r][h];
    return _;
  }
  function i(e) {
    var r = {}, _;
    for (_ in e) {
      var h = _.replace(/-/g, "_");
      r[h] = e[_];
    }
    return r;
  }
  return st.Options = a, st.normalizeOpts = i, st.mergeOpts = t, st;
}
var Be;
function vi() {
  if (Be) return Ft;
  Be = 1;
  var a = ve().Options, t = ["before-newline", "after-newline", "preserve-newline"];
  function i(e) {
    a.call(this, e, "js");
    var r = this.raw_options.brace_style || null;
    r === "expand-strict" ? this.raw_options.brace_style = "expand" : r === "collapse-preserve-inline" ? this.raw_options.brace_style = "collapse,preserve-inline" : this.raw_options.braces_on_own_line !== void 0 && (this.raw_options.brace_style = this.raw_options.braces_on_own_line ? "expand" : "collapse");
    var _ = this._get_selection_list("brace_style", ["collapse", "expand", "end-expand", "none", "preserve-inline"]);
    this.brace_preserve_inline = !1, this.brace_style = "collapse";
    for (var h = 0; h < _.length; h++)
      _[h] === "preserve-inline" ? this.brace_preserve_inline = !0 : this.brace_style = _[h];
    this.unindent_chained_methods = this._get_boolean("unindent_chained_methods"), this.break_chained_methods = this._get_boolean("break_chained_methods"), this.space_in_paren = this._get_boolean("space_in_paren"), this.space_in_empty_paren = this._get_boolean("space_in_empty_paren"), this.jslint_happy = this._get_boolean("jslint_happy"), this.space_after_anon_function = this._get_boolean("space_after_anon_function"), this.space_after_named_function = this._get_boolean("space_after_named_function"), this.keep_array_indentation = this._get_boolean("keep_array_indentation"), this.space_before_conditional = this._get_boolean("space_before_conditional", !0), this.unescape_strings = this._get_boolean("unescape_strings"), this.e4x = this._get_boolean("e4x"), this.comma_first = this._get_boolean("comma_first"), this.operator_position = this._get_selection("operator_position", t), this.test_output_raw = this._get_boolean("test_output_raw"), this.jslint_happy && (this.space_after_anon_function = !0);
  }
  return i.prototype = new a(), Ft.Options = i, Ft;
}
var Y = {}, qt = {}, Ke;
function ye() {
  if (Ke) return qt;
  Ke = 1;
  var a = RegExp.prototype.hasOwnProperty("sticky");
  function t(i) {
    this.__input = i || "", this.__input_length = this.__input.length, this.__position = 0;
  }
  return t.prototype.restart = function() {
    this.__position = 0;
  }, t.prototype.back = function() {
    this.__position > 0 && (this.__position -= 1);
  }, t.prototype.hasNext = function() {
    return this.__position < this.__input_length;
  }, t.prototype.next = function() {
    var i = null;
    return this.hasNext() && (i = this.__input.charAt(this.__position), this.__position += 1), i;
  }, t.prototype.peek = function(i) {
    var e = null;
    return i = i || 0, i += this.__position, i >= 0 && i < this.__input_length && (e = this.__input.charAt(i)), e;
  }, t.prototype.__match = function(i, e) {
    i.lastIndex = e;
    var r = i.exec(this.__input);
    return r && !(a && i.sticky) && r.index !== e && (r = null), r;
  }, t.prototype.test = function(i, e) {
    return e = e || 0, e += this.__position, e >= 0 && e < this.__input_length ? !!this.__match(i, e) : !1;
  }, t.prototype.testChar = function(i, e) {
    var r = this.peek(e);
    return i.lastIndex = 0, r !== null && i.test(r);
  }, t.prototype.match = function(i) {
    var e = this.__match(i, this.__position);
    return e ? this.__position += e[0].length : e = null, e;
  }, t.prototype.read = function(i, e, r) {
    var _ = "", h;
    return i && (h = this.match(i), h && (_ += h[0])), e && (h || !i) && (_ += this.readUntil(e, r)), _;
  }, t.prototype.readUntil = function(i, e) {
    var r = "", _ = this.__position;
    i.lastIndex = this.__position;
    var h = i.exec(this.__input);
    return h ? (_ = h.index, e && (_ += h[0].length)) : _ = this.__input_length, r = this.__input.substring(this.__position, _), this.__position = _, r;
  }, t.prototype.readUntilAfter = function(i) {
    return this.readUntil(i, !0);
  }, t.prototype.get_regexp = function(i, e) {
    var r = null, _ = "g";
    return e && a && (_ = "y"), typeof i == "string" && i !== "" ? r = new RegExp(i, _) : i && (r = new RegExp(i.source, _)), r;
  }, t.prototype.get_literal_regexp = function(i) {
    return RegExp(i.replace(/[-\/\\^$*+?.()|[\]{}]/g, "\\$&"));
  }, t.prototype.peekUntilAfter = function(i) {
    var e = this.__position, r = this.readUntilAfter(i);
    return this.__position = e, r;
  }, t.prototype.lookBack = function(i) {
    var e = this.__position - 1;
    return e >= i.length && this.__input.substring(e - i.length, e).toLowerCase() === i;
  }, qt.InputScanner = t, qt;
}
var bt = {}, $t = {}, Ve;
function qi() {
  if (Ve) return $t;
  Ve = 1;
  function a(t) {
    this.__tokens = [], this.__tokens_length = this.__tokens.length, this.__position = 0, this.__parent_token = t;
  }
  return a.prototype.restart = function() {
    this.__position = 0;
  }, a.prototype.isEmpty = function() {
    return this.__tokens_length === 0;
  }, a.prototype.hasNext = function() {
    return this.__position < this.__tokens_length;
  }, a.prototype.next = function() {
    var t = null;
    return this.hasNext() && (t = this.__tokens[this.__position], this.__position += 1), t;
  }, a.prototype.peek = function(t) {
    var i = null;
    return t = t || 0, t += this.__position, t >= 0 && t < this.__tokens_length && (i = this.__tokens[t]), i;
  }, a.prototype.add = function(t) {
    this.__parent_token && (t.parent = this.__parent_token), this.__tokens.push(t), this.__tokens_length += 1;
  }, $t.TokenStream = a, $t;
}
var Gt = {}, Xt = {}, Ue;
function Dt() {
  if (Ue) return Xt;
  Ue = 1;
  function a(t, i) {
    this._input = t, this._starting_pattern = null, this._match_pattern = null, this._until_pattern = null, this._until_after = !1, i && (this._starting_pattern = this._input.get_regexp(i._starting_pattern, !0), this._match_pattern = this._input.get_regexp(i._match_pattern, !0), this._until_pattern = this._input.get_regexp(i._until_pattern), this._until_after = i._until_after);
  }
  return a.prototype.read = function() {
    var t = this._input.read(this._starting_pattern);
    return (!this._starting_pattern || t) && (t += this._input.read(this._match_pattern, this._until_pattern, this._until_after)), t;
  }, a.prototype.read_match = function() {
    return this._input.match(this._match_pattern);
  }, a.prototype.until_after = function(t) {
    var i = this._create();
    return i._until_after = !0, i._until_pattern = this._input.get_regexp(t), i._update(), i;
  }, a.prototype.until = function(t) {
    var i = this._create();
    return i._until_after = !1, i._until_pattern = this._input.get_regexp(t), i._update(), i;
  }, a.prototype.starting_with = function(t) {
    var i = this._create();
    return i._starting_pattern = this._input.get_regexp(t, !0), i._update(), i;
  }, a.prototype.matching = function(t) {
    var i = this._create();
    return i._match_pattern = this._input.get_regexp(t, !0), i._update(), i;
  }, a.prototype._create = function() {
    return new a(this._input, this);
  }, a.prototype._update = function() {
  }, Xt.Pattern = a, Xt;
}
var ze;
function $i() {
  if (ze) return Gt;
  ze = 1;
  var a = Dt().Pattern;
  function t(i, e) {
    a.call(this, i, e), e ? this._line_regexp = this._input.get_regexp(e._line_regexp) : this.__set_whitespace_patterns("", ""), this.newline_count = 0, this.whitespace_before_token = "";
  }
  return t.prototype = new a(), t.prototype.__set_whitespace_patterns = function(i, e) {
    i += "\\t ", e += "\\n\\r", this._match_pattern = this._input.get_regexp(
      "[" + i + e + "]+",
      !0
    ), this._newline_regexp = this._input.get_regexp(
      "\\r\\n|[" + e + "]"
    );
  }, t.prototype.read = function() {
    this.newline_count = 0, this.whitespace_before_token = "";
    var i = this._input.read(this._match_pattern);
    if (i === " ")
      this.whitespace_before_token = " ";
    else if (i) {
      var e = this.__split(this._newline_regexp, i);
      this.newline_count = e.length - 1, this.whitespace_before_token = e[this.newline_count];
    }
    return i;
  }, t.prototype.matching = function(i, e) {
    var r = this._create();
    return r.__set_whitespace_patterns(i, e), r._update(), r;
  }, t.prototype._create = function() {
    return new t(this._input, this);
  }, t.prototype.__split = function(i, e) {
    i.lastIndex = 0;
    for (var r = 0, _ = [], h = i.exec(e); h; )
      _.push(e.substring(r, h.index)), r = h.index + h[0].length, h = i.exec(e);
    return r < e.length ? _.push(e.substring(r, e.length)) : _.push(""), _;
  }, Gt.WhitespacePattern = t, Gt;
}
var Fe;
function Lt() {
  if (Fe) return bt;
  Fe = 1;
  var a = ye().InputScanner, t = bi().Token, i = qi().TokenStream, e = $i().WhitespacePattern, r = {
    START: "TK_START",
    RAW: "TK_RAW",
    EOF: "TK_EOF"
  }, _ = function(h, u) {
    this._input = new a(h), this._options = u || {}, this.__tokens = null, this._patterns = {}, this._patterns.whitespace = new e(this._input);
  };
  return _.prototype.tokenize = function() {
    this._input.restart(), this.__tokens = new i(), this._reset();
    for (var h, u = new t(r.START, ""), l = null, O = [], v = new i(); u.type !== r.EOF; ) {
      for (h = this._get_next_token(u, l); this._is_comment(h); )
        v.add(h), h = this._get_next_token(u, l);
      v.isEmpty() || (h.comments_before = v, v = new i()), h.parent = l, this._is_opening(h) ? (O.push(l), l = h) : l && this._is_closing(h, l) && (h.opened = l, l.closed = h, l = O.pop(), h.parent = l), h.previous = u, u.next = h, this.__tokens.add(h), u = h;
    }
    return this.__tokens;
  }, _.prototype._is_first_token = function() {
    return this.__tokens.isEmpty();
  }, _.prototype._reset = function() {
  }, _.prototype._get_next_token = function(h, u) {
    this._readWhitespace();
    var l = this._input.read(/.+/g);
    return l ? this._create_token(r.RAW, l) : this._create_token(r.EOF, "");
  }, _.prototype._is_comment = function(h) {
    return !1;
  }, _.prototype._is_opening = function(h) {
    return !1;
  }, _.prototype._is_closing = function(h, u) {
    return !1;
  }, _.prototype._create_token = function(h, u) {
    var l = new t(
      h,
      u,
      this._patterns.whitespace.newline_count,
      this._patterns.whitespace.whitespace_before_token
    );
    return l;
  }, _.prototype._readWhitespace = function() {
    return this._patterns.whitespace.read();
  }, bt.Tokenizer = _, bt.TOKEN = r, bt;
}
var Qt = {}, qe;
function we() {
  if (qe) return Qt;
  qe = 1;
  function a(t, i) {
    t = typeof t == "string" ? t : t.source, i = typeof i == "string" ? i : i.source, this.__directives_block_pattern = new RegExp(t + / beautify( \w+[:]\w+)+ /.source + i, "g"), this.__directive_pattern = / (\w+)[:](\w+)/g, this.__directives_end_ignore_pattern = new RegExp(t + /\sbeautify\signore:end\s/.source + i, "g");
  }
  return a.prototype.get_directives = function(t) {
    if (!t.match(this.__directives_block_pattern))
      return null;
    var i = {};
    this.__directive_pattern.lastIndex = 0;
    for (var e = this.__directive_pattern.exec(t); e; )
      i[e[1]] = e[2], e = this.__directive_pattern.exec(t);
    return i;
  }, a.prototype.readIgnored = function(t) {
    return t.readUntilAfter(this.__directives_end_ignore_pattern);
  }, Qt.Directives = a, Qt;
}
var Ht = {}, $e;
function yi() {
  if ($e) return Ht;
  $e = 1;
  var a = Dt().Pattern, t = {
    django: !1,
    erb: !1,
    handlebars: !1,
    php: !1,
    smarty: !1,
    angular: !1
  };
  function i(e, r) {
    a.call(this, e, r), this.__template_pattern = null, this._disabled = Object.assign({}, t), this._excluded = Object.assign({}, t), r && (this.__template_pattern = this._input.get_regexp(r.__template_pattern), this._excluded = Object.assign(this._excluded, r._excluded), this._disabled = Object.assign(this._disabled, r._disabled));
    var _ = new a(e);
    this.__patterns = {
      handlebars_comment: _.starting_with(/{{!--/).until_after(/--}}/),
      handlebars_unescaped: _.starting_with(/{{{/).until_after(/}}}/),
      handlebars: _.starting_with(/{{/).until_after(/}}/),
      php: _.starting_with(/<\?(?:[= ]|php)/).until_after(/\?>/),
      erb: _.starting_with(/<%[^%]/).until_after(/[^%]%>/),
      // django coflicts with handlebars a bit.
      django: _.starting_with(/{%/).until_after(/%}/),
      django_value: _.starting_with(/{{/).until_after(/}}/),
      django_comment: _.starting_with(/{#/).until_after(/#}/),
      smarty: _.starting_with(/{(?=[^}{\s\n])/).until_after(/[^\s\n]}/),
      smarty_comment: _.starting_with(/{\*/).until_after(/\*}/),
      smarty_literal: _.starting_with(/{literal}/).until_after(/{\/literal}/)
    };
  }
  return i.prototype = new a(), i.prototype._create = function() {
    return new i(this._input, this);
  }, i.prototype._update = function() {
    this.__set_templated_pattern();
  }, i.prototype.disable = function(e) {
    var r = this._create();
    return r._disabled[e] = !0, r._update(), r;
  }, i.prototype.read_options = function(e) {
    var r = this._create();
    for (var _ in t)
      r._disabled[_] = e.templating.indexOf(_) === -1;
    return r._update(), r;
  }, i.prototype.exclude = function(e) {
    var r = this._create();
    return r._excluded[e] = !0, r._update(), r;
  }, i.prototype.read = function() {
    var e = "";
    this._match_pattern ? e = this._input.read(this._starting_pattern) : e = this._input.read(this._starting_pattern, this.__template_pattern);
    for (var r = this._read_template(); r; )
      this._match_pattern ? r += this._input.read(this._match_pattern) : r += this._input.readUntil(this.__template_pattern), e += r, r = this._read_template();
    return this._until_after && (e += this._input.readUntilAfter(this._until_pattern)), e;
  }, i.prototype.__set_templated_pattern = function() {
    var e = [];
    this._disabled.php || e.push(this.__patterns.php._starting_pattern.source), this._disabled.handlebars || e.push(this.__patterns.handlebars._starting_pattern.source), this._disabled.erb || e.push(this.__patterns.erb._starting_pattern.source), this._disabled.django || (e.push(this.__patterns.django._starting_pattern.source), e.push(this.__patterns.django_value._starting_pattern.source), e.push(this.__patterns.django_comment._starting_pattern.source)), this._disabled.smarty || e.push(this.__patterns.smarty._starting_pattern.source), this._until_pattern && e.push(this._until_pattern.source), this.__template_pattern = this._input.get_regexp("(?:" + e.join("|") + ")");
  }, i.prototype._read_template = function() {
    var e = "", r = this._input.peek();
    if (r === "<") {
      var _ = this._input.peek(1);
      !this._disabled.php && !this._excluded.php && _ === "?" && (e = e || this.__patterns.php.read()), !this._disabled.erb && !this._excluded.erb && _ === "%" && (e = e || this.__patterns.erb.read());
    } else r === "{" && (!this._disabled.handlebars && !this._excluded.handlebars && (e = e || this.__patterns.handlebars_comment.read(), e = e || this.__patterns.handlebars_unescaped.read(), e = e || this.__patterns.handlebars.read()), this._disabled.django || (!this._excluded.django && !this._excluded.handlebars && (e = e || this.__patterns.django_value.read()), this._excluded.django || (e = e || this.__patterns.django_comment.read(), e = e || this.__patterns.django.read())), this._disabled.smarty || this._disabled.django && this._disabled.handlebars && (e = e || this.__patterns.smarty_comment.read(), e = e || this.__patterns.smarty_literal.read(), e = e || this.__patterns.smarty.read()));
    return e;
  }, Ht.TemplatablePattern = i, Ht;
}
var Ge;
function mt() {
  if (Ge) return Y;
  Ge = 1;
  var a = ye().InputScanner, t = Lt().Tokenizer, i = Lt().TOKEN, e = we().Directives, r = mi(), _ = Dt().Pattern, h = yi().TemplatablePattern;
  function u(c, b) {
    return b.indexOf(c) !== -1;
  }
  var l = {
    START_EXPR: "TK_START_EXPR",
    END_EXPR: "TK_END_EXPR",
    START_BLOCK: "TK_START_BLOCK",
    END_BLOCK: "TK_END_BLOCK",
    WORD: "TK_WORD",
    RESERVED: "TK_RESERVED",
    SEMICOLON: "TK_SEMICOLON",
    STRING: "TK_STRING",
    EQUALS: "TK_EQUALS",
    OPERATOR: "TK_OPERATOR",
    COMMA: "TK_COMMA",
    BLOCK_COMMENT: "TK_BLOCK_COMMENT",
    COMMENT: "TK_COMMENT",
    DOT: "TK_DOT",
    UNKNOWN: "TK_UNKNOWN",
    START: i.START,
    RAW: i.RAW,
    EOF: i.EOF
  }, O = new e(/\/\*/, /\*\//), v = /0[xX][0123456789abcdefABCDEF_]*n?|0[oO][01234567_]*n?|0[bB][01_]*n?|\d[\d_]*n|(?:\.\d[\d_]*|\d[\d_]*\.?[\d_]*)(?:[eE][+-]?[\d_]+)?/, y = /[0-9]/, g = /[^\d\.]/, C = ">>> === !== &&= ??= ||= << && >= ** != == <= >> || ?? |> < / - + > : & % ? ^ | *".split(" "), A = ">>>= ... >>= <<= === >>> !== **= &&= ??= ||= => ^= :: /= << <= == && -= >= >> != -- += ** || ?? ++ %= &= *= |= |> = ! ? > < : / ^ - + * & % ~ |";
  A = A.replace(/[-[\]{}()*+?.,\\^$|#]/g, "\\$&"), A = "\\?\\.(?!\\d) " + A, A = A.replace(/ /g, "|");
  var x = new RegExp(A), o = "continue,try,throw,return,var,let,const,if,switch,case,default,for,while,break,function,import,export".split(","), n = o.concat(["do", "in", "of", "else", "get", "set", "new", "catch", "finally", "typeof", "yield", "async", "await", "from", "as", "class", "extends"]), p = new RegExp("^(?:" + n.join("|") + ")$"), T, E = function(c, b) {
    t.call(this, c, b), this._patterns.whitespace = this._patterns.whitespace.matching(
      /\u00A0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000\ufeff/.source,
      /\u2028\u2029/.source
    );
    var f = new _(this._input), S = new h(this._input).read_options(this._options);
    this.__patterns = {
      template: S,
      identifier: S.starting_with(r.identifier).matching(r.identifierMatch),
      number: f.matching(v),
      punct: f.matching(x),
      // comment ends just before nearest linefeed or end of file
      comment: f.starting_with(/\/\//).until(/[\n\r\u2028\u2029]/),
      //  /* ... */ comment ends with nearest */ or end of file
      block_comment: f.starting_with(/\/\*/).until_after(/\*\//),
      html_comment_start: f.matching(/<!--/),
      html_comment_end: f.matching(/-->/),
      include: f.starting_with(/#include/).until_after(r.lineBreak),
      shebang: f.starting_with(/#!/).until_after(r.lineBreak),
      xml: f.matching(/[\s\S]*?<(\/?)([-a-zA-Z:0-9_.]+|{[^}]+?}|!\[CDATA\[[^\]]*?\]\]|)(\s*{[^}]+?}|\s+[-a-zA-Z:0-9_.]+|\s+[-a-zA-Z:0-9_.]+\s*=\s*('[^']*'|"[^"]*"|{([^{}]|{[^}]+?})+?}))*\s*(\/?)\s*>/),
      single_quote: S.until(/['\\\n\r\u2028\u2029]/),
      double_quote: S.until(/["\\\n\r\u2028\u2029]/),
      template_text: S.until(/[`\\$]/),
      template_expression: S.until(/[`}\\]/)
    };
  };
  E.prototype = new t(), E.prototype._is_comment = function(c) {
    return c.type === l.COMMENT || c.type === l.BLOCK_COMMENT || c.type === l.UNKNOWN;
  }, E.prototype._is_opening = function(c) {
    return c.type === l.START_BLOCK || c.type === l.START_EXPR;
  }, E.prototype._is_closing = function(c, b) {
    return (c.type === l.END_BLOCK || c.type === l.END_EXPR) && b && (c.text === "]" && b.text === "[" || c.text === ")" && b.text === "(" || c.text === "}" && b.text === "{");
  }, E.prototype._reset = function() {
    T = !1;
  }, E.prototype._get_next_token = function(c, b) {
    var f = null;
    this._readWhitespace();
    var S = this._input.peek();
    return S === null ? this._create_token(l.EOF, "") : (f = f || this._read_non_javascript(S), f = f || this._read_string(S), f = f || this._read_pair(S, this._input.peek(1)), f = f || this._read_word(c), f = f || this._read_singles(S), f = f || this._read_comment(S), f = f || this._read_regexp(S, c), f = f || this._read_xml(S, c), f = f || this._read_punctuation(), f = f || this._create_token(l.UNKNOWN, this._input.next()), f);
  }, E.prototype._read_word = function(c) {
    var b;
    if (b = this.__patterns.identifier.read(), b !== "")
      return b = b.replace(r.allLineBreaks, `
`), !(c.type === l.DOT || c.type === l.RESERVED && (c.text === "set" || c.text === "get")) && p.test(b) ? (b === "in" || b === "of") && (c.type === l.WORD || c.type === l.STRING) ? this._create_token(l.OPERATOR, b) : this._create_token(l.RESERVED, b) : this._create_token(l.WORD, b);
    if (b = this.__patterns.number.read(), b !== "")
      return this._create_token(l.WORD, b);
  }, E.prototype._read_singles = function(c) {
    var b = null;
    return c === "(" || c === "[" ? b = this._create_token(l.START_EXPR, c) : c === ")" || c === "]" ? b = this._create_token(l.END_EXPR, c) : c === "{" ? b = this._create_token(l.START_BLOCK, c) : c === "}" ? b = this._create_token(l.END_BLOCK, c) : c === ";" ? b = this._create_token(l.SEMICOLON, c) : c === "." && g.test(this._input.peek(1)) ? b = this._create_token(l.DOT, c) : c === "," && (b = this._create_token(l.COMMA, c)), b && this._input.next(), b;
  }, E.prototype._read_pair = function(c, b) {
    var f = null;
    return c === "#" && b === "{" && (f = this._create_token(l.START_BLOCK, c + b)), f && (this._input.next(), this._input.next()), f;
  }, E.prototype._read_punctuation = function() {
    var c = this.__patterns.punct.read();
    if (c !== "")
      return c === "=" ? this._create_token(l.EQUALS, c) : c === "?." ? this._create_token(l.DOT, c) : this._create_token(l.OPERATOR, c);
  }, E.prototype._read_non_javascript = function(c) {
    var b = "";
    if (c === "#") {
      if (this._is_first_token() && (b = this.__patterns.shebang.read(), b))
        return this._create_token(l.UNKNOWN, b.trim() + `
`);
      if (b = this.__patterns.include.read(), b)
        return this._create_token(l.UNKNOWN, b.trim() + `
`);
      c = this._input.next();
      var f = "#";
      if (this._input.hasNext() && this._input.testChar(y)) {
        do
          c = this._input.next(), f += c;
        while (this._input.hasNext() && c !== "#" && c !== "=");
        return c === "#" || (this._input.peek() === "[" && this._input.peek(1) === "]" ? (f += "[]", this._input.next(), this._input.next()) : this._input.peek() === "{" && this._input.peek(1) === "}" && (f += "{}", this._input.next(), this._input.next())), this._create_token(l.WORD, f);
      }
      this._input.back();
    } else if (c === "<" && this._is_first_token()) {
      if (b = this.__patterns.html_comment_start.read(), b) {
        for (; this._input.hasNext() && !this._input.testChar(r.newline); )
          b += this._input.next();
        return T = !0, this._create_token(l.COMMENT, b);
      }
    } else if (T && c === "-" && (b = this.__patterns.html_comment_end.read(), b))
      return T = !1, this._create_token(l.COMMENT, b);
    return null;
  }, E.prototype._read_comment = function(c) {
    var b = null;
    if (c === "/") {
      var f = "";
      if (this._input.peek(1) === "*") {
        f = this.__patterns.block_comment.read();
        var S = O.get_directives(f);
        S && S.ignore === "start" && (f += O.readIgnored(this._input)), f = f.replace(r.allLineBreaks, `
`), b = this._create_token(l.BLOCK_COMMENT, f), b.directives = S;
      } else this._input.peek(1) === "/" && (f = this.__patterns.comment.read(), b = this._create_token(l.COMMENT, f));
    }
    return b;
  }, E.prototype._read_string = function(c) {
    if (c === "`" || c === "'" || c === '"') {
      var b = this._input.next();
      return this.has_char_escapes = !1, c === "`" ? b += this._read_string_recursive("`", !0, "${") : b += this._read_string_recursive(c), this.has_char_escapes && this._options.unescape_strings && (b = k(b)), this._input.peek() === c && (b += this._input.next()), b = b.replace(r.allLineBreaks, `
`), this._create_token(l.STRING, b);
    }
    return null;
  }, E.prototype._allow_regexp_or_xml = function(c) {
    return c.type === l.RESERVED && u(c.text, ["return", "case", "throw", "else", "do", "typeof", "yield"]) || c.type === l.END_EXPR && c.text === ")" && c.opened.previous.type === l.RESERVED && u(c.opened.previous.text, ["if", "while", "for"]) || u(c.type, [
      l.COMMENT,
      l.START_EXPR,
      l.START_BLOCK,
      l.START,
      l.END_BLOCK,
      l.OPERATOR,
      l.EQUALS,
      l.EOF,
      l.SEMICOLON,
      l.COMMA
    ]);
  }, E.prototype._read_regexp = function(c, b) {
    if (c === "/" && this._allow_regexp_or_xml(b)) {
      for (var f = this._input.next(), S = !1, s = !1; this._input.hasNext() && (S || s || this._input.peek() !== c) && !this._input.testChar(r.newline); )
        f += this._input.peek(), S ? S = !1 : (S = this._input.peek() === "\\", this._input.peek() === "[" ? s = !0 : this._input.peek() === "]" && (s = !1)), this._input.next();
      return this._input.peek() === c && (f += this._input.next(), f += this._input.read(r.identifier)), this._create_token(l.STRING, f);
    }
    return null;
  }, E.prototype._read_xml = function(c, b) {
    if (this._options.e4x && c === "<" && this._allow_regexp_or_xml(b)) {
      var f = "", S = this.__patterns.xml.read_match();
      if (S) {
        for (var s = S[2].replace(/^{\s+/, "{").replace(/\s+}$/, "}"), d = s.indexOf("{") === 0, m = 0; S; ) {
          var R = !!S[1], N = S[2], W = !!S[S.length - 1] || N.slice(0, 8) === "![CDATA[";
          if (!W && (N === s || d && N.replace(/^{\s+/, "{").replace(/\s+}$/, "}")) && (R ? --m : ++m), f += S[0], m <= 0)
            break;
          S = this.__patterns.xml.read_match();
        }
        return S || (f += this._input.match(/[\s\S]*/g)[0]), f = f.replace(r.allLineBreaks, `
`), this._create_token(l.STRING, f);
      }
    }
    return null;
  };
  function k(c) {
    for (var b = "", f = 0, S = new a(c), s = null; S.hasNext(); )
      if (s = S.match(/([\s]|[^\\]|\\\\)+/g), s && (b += s[0]), S.peek() === "\\") {
        if (S.next(), S.peek() === "x")
          s = S.match(/x([0-9A-Fa-f]{2})/g);
        else if (S.peek() === "u")
          s = S.match(/u([0-9A-Fa-f]{4})/g), s || (s = S.match(/u\{([0-9A-Fa-f]+)\}/g));
        else {
          b += "\\", S.hasNext() && (b += S.next());
          continue;
        }
        if (!s || (f = parseInt(s[1], 16), f > 126 && f <= 255 && s[0].indexOf("x") === 0))
          return c;
        f >= 0 && f < 32 || f > 1114111 ? b += "\\" + s[0] : f === 34 || f === 39 || f === 92 ? b += "\\" + String.fromCharCode(f) : b += String.fromCharCode(f);
      }
    return b;
  }
  return E.prototype._read_string_recursive = function(c, b, f) {
    var S, s;
    c === "'" ? s = this.__patterns.single_quote : c === '"' ? s = this.__patterns.double_quote : c === "`" ? s = this.__patterns.template_text : c === "}" && (s = this.__patterns.template_expression);
    for (var d = s.read(), m = ""; this._input.hasNext(); ) {
      if (m = this._input.next(), m === c || !b && r.newline.test(m)) {
        this._input.back();
        break;
      } else m === "\\" && this._input.hasNext() ? (S = this._input.peek(), S === "x" || S === "u" ? this.has_char_escapes = !0 : S === "\r" && this._input.peek(1) === `
` && this._input.next(), m += this._input.next()) : f && (f === "${" && m === "$" && this._input.peek() === "{" && (m += this._input.next()), f === m && (c === "`" ? m += this._read_string_recursive("}", b, "`") : m += this._read_string_recursive("`", b, "${"), this._input.hasNext() && (m += this._input.next())));
      m += s.read(), d += m;
    }
    return d;
  }, Y.Tokenizer = E, Y.TOKEN = l, Y.positionable_operators = C.slice(), Y.line_starters = o.slice(), Y;
}
var Xe;
function Gi() {
  if (Xe) return Kt;
  Xe = 1;
  var a = me().Output, t = bi().Token, i = mi(), e = vi().Options, r = mt().Tokenizer, _ = mt().line_starters, h = mt().positionable_operators, u = mt().TOKEN;
  function l(s, d) {
    return d.indexOf(s) !== -1;
  }
  function O(s) {
    return s.replace(/^\s+/g, "");
  }
  function v(s) {
    for (var d = {}, m = 0; m < s.length; m++)
      d[s[m].replace(/-/g, "_")] = s[m];
    return d;
  }
  function y(s, d) {
    return s && s.type === u.RESERVED && s.text === d;
  }
  function g(s, d) {
    return s && s.type === u.RESERVED && l(s.text, d);
  }
  var C = ["case", "return", "do", "if", "throw", "else", "await", "break", "continue", "async"], A = ["before-newline", "after-newline", "preserve-newline"], x = v(A), o = [x.before_newline, x.preserve_newline], n = {
    BlockStatement: "BlockStatement",
    // 'BLOCK'
    Statement: "Statement",
    // 'STATEMENT'
    ObjectLiteral: "ObjectLiteral",
    // 'OBJECT',
    ArrayLiteral: "ArrayLiteral",
    //'[EXPRESSION]',
    ForInitializer: "ForInitializer",
    //'(FOR-EXPRESSION)',
    Conditional: "Conditional",
    //'(COND-EXPRESSION)',
    Expression: "Expression"
    //'(EXPRESSION)'
  };
  function p(s, d) {
    d.multiline_frame || d.mode === n.ForInitializer || d.mode === n.Conditional || s.remove_indent(d.start_line_index);
  }
  function T(s) {
    s = s.replace(i.allLineBreaks, `
`);
    for (var d = [], m = s.indexOf(`
`); m !== -1; )
      d.push(s.substring(0, m)), s = s.substring(m + 1), m = s.indexOf(`
`);
    return s.length && d.push(s), d;
  }
  function E(s) {
    return s === n.ArrayLiteral;
  }
  function k(s) {
    return l(s, [n.Expression, n.ForInitializer, n.Conditional]);
  }
  function c(s, d) {
    for (var m = 0; m < s.length; m++) {
      var R = s[m].trim();
      if (R.charAt(0) !== d)
        return !1;
    }
    return !0;
  }
  function b(s, d) {
    for (var m = 0, R = s.length, N; m < R; m++)
      if (N = s[m], N && N.indexOf(d) !== 0)
        return !1;
    return !0;
  }
  function f(s, d) {
    d = d || {}, this._source_text = s || "", this._output = null, this._tokens = null, this._last_last_text = null, this._flags = null, this._previous_flags = null, this._flag_store = null, this._options = new e(d);
  }
  f.prototype.create_flags = function(s, d) {
    var m = 0;
    s && (m = s.indentation_level, !this._output.just_added_newline() && s.line_indent_level > m && (m = s.line_indent_level));
    var R = {
      mode: d,
      parent: s,
      last_token: s ? s.last_token : new t(u.START_BLOCK, ""),
      // last token text
      last_word: s ? s.last_word : "",
      // last TOKEN.WORD passed
      declaration_statement: !1,
      declaration_assignment: !1,
      multiline_frame: !1,
      inline_frame: !1,
      if_block: !1,
      else_block: !1,
      class_start_block: !1,
      // class A { INSIDE HERE } or class B extends C { INSIDE HERE }
      do_block: !1,
      do_while: !1,
      import_block: !1,
      in_case_statement: !1,
      // switch(..){ INSIDE HERE }
      in_case: !1,
      // we're on the exact line with "case 0:"
      case_body: !1,
      // the indented case-action block
      case_block: !1,
      // the indented case-action block is wrapped with {}
      indentation_level: m,
      alignment: 0,
      line_indent_level: s ? s.line_indent_level : m,
      start_line_index: this._output.get_line_number(),
      ternary_depth: 0
    };
    return R;
  }, f.prototype._reset = function(s) {
    var d = s.match(/^[\t ]*/)[0];
    this._last_last_text = "", this._output = new a(this._options, d), this._output.raw = this._options.test_output_raw, this._flag_store = [], this.set_mode(n.BlockStatement);
    var m = new r(s, this._options);
    return this._tokens = m.tokenize(), s;
  }, f.prototype.beautify = function() {
    if (this._options.disabled)
      return this._source_text;
    var s, d = this._reset(this._source_text), m = this._options.eol;
    this._options.eol === "auto" && (m = `
`, d && i.lineBreak.test(d || "") && (m = d.match(i.lineBreak)[0]));
    for (var R = this._tokens.next(); R; )
      this.handle_token(R), this._last_last_text = this._flags.last_token.text, this._flags.last_token = R, R = this._tokens.next();
    return s = this._output.get_code(m), s;
  }, f.prototype.handle_token = function(s, d) {
    s.type === u.START_EXPR ? this.handle_start_expr(s) : s.type === u.END_EXPR ? this.handle_end_expr(s) : s.type === u.START_BLOCK ? this.handle_start_block(s) : s.type === u.END_BLOCK ? this.handle_end_block(s) : s.type === u.WORD ? this.handle_word(s) : s.type === u.RESERVED ? this.handle_word(s) : s.type === u.SEMICOLON ? this.handle_semicolon(s) : s.type === u.STRING ? this.handle_string(s) : s.type === u.EQUALS ? this.handle_equals(s) : s.type === u.OPERATOR ? this.handle_operator(s) : s.type === u.COMMA ? this.handle_comma(s) : s.type === u.BLOCK_COMMENT ? this.handle_block_comment(s, d) : s.type === u.COMMENT ? this.handle_comment(s, d) : s.type === u.DOT ? this.handle_dot(s) : s.type === u.EOF ? this.handle_eof(s) : s.type === u.UNKNOWN ? this.handle_unknown(s, d) : this.handle_unknown(s, d);
  }, f.prototype.handle_whitespace_and_comments = function(s, d) {
    var m = s.newlines, R = this._options.keep_array_indentation && E(this._flags.mode);
    if (s.comments_before)
      for (var N = s.comments_before.next(); N; )
        this.handle_whitespace_and_comments(N, d), this.handle_token(N, d), N = s.comments_before.next();
    if (R)
      for (var W = 0; W < m; W += 1)
        this.print_newline(W > 0, d);
    else if (this._options.max_preserve_newlines && m > this._options.max_preserve_newlines && (m = this._options.max_preserve_newlines), this._options.preserve_newlines && m > 1) {
      this.print_newline(!1, d);
      for (var M = 1; M < m; M += 1)
        this.print_newline(!0, d);
    }
  };
  var S = ["async", "break", "continue", "return", "throw", "yield"];
  return f.prototype.allow_wrap_or_preserved_newline = function(s, d) {
    if (d = d === void 0 ? !1 : d, !this._output.just_added_newline()) {
      var m = this._options.preserve_newlines && s.newlines || d, R = l(this._flags.last_token.text, h) || l(s.text, h);
      if (R) {
        var N = l(this._flags.last_token.text, h) && l(this._options.operator_position, o) || l(s.text, h);
        m = m && N;
      }
      if (m)
        this.print_newline(!1, !0);
      else if (this._options.wrap_line_length) {
        if (g(this._flags.last_token, S))
          return;
        this._output.set_wrap_point();
      }
    }
  }, f.prototype.print_newline = function(s, d) {
    if (!d && this._flags.last_token.text !== ";" && this._flags.last_token.text !== "," && this._flags.last_token.text !== "=" && (this._flags.last_token.type !== u.OPERATOR || this._flags.last_token.text === "--" || this._flags.last_token.text === "++"))
      for (var m = this._tokens.peek(); this._flags.mode === n.Statement && !(this._flags.if_block && y(m, "else")) && !this._flags.do_block; )
        this.restore_mode();
    this._output.add_new_line(s) && (this._flags.multiline_frame = !0);
  }, f.prototype.print_token_line_indentation = function(s) {
    this._output.just_added_newline() && (this._options.keep_array_indentation && s.newlines && (s.text === "[" || E(this._flags.mode)) ? (this._output.current_line.set_indent(-1), this._output.current_line.push(s.whitespace_before), this._output.space_before_token = !1) : this._output.set_indent(this._flags.indentation_level, this._flags.alignment) && (this._flags.line_indent_level = this._flags.indentation_level));
  }, f.prototype.print_token = function(s) {
    if (this._output.raw) {
      this._output.add_raw_token(s);
      return;
    }
    if (this._options.comma_first && s.previous && s.previous.type === u.COMMA && this._output.just_added_newline() && this._output.previous_line.last() === ",") {
      var d = this._output.previous_line.pop();
      this._output.previous_line.is_empty() && (this._output.previous_line.push(d), this._output.trim(!0), this._output.current_line.pop(), this._output.trim()), this.print_token_line_indentation(s), this._output.add_token(","), this._output.space_before_token = !0;
    }
    this.print_token_line_indentation(s), this._output.non_breaking_space = !0, this._output.add_token(s.text), this._output.previous_token_wrapped && (this._flags.multiline_frame = !0);
  }, f.prototype.indent = function() {
    this._flags.indentation_level += 1, this._output.set_indent(this._flags.indentation_level, this._flags.alignment);
  }, f.prototype.deindent = function() {
    this._flags.indentation_level > 0 && (!this._flags.parent || this._flags.indentation_level > this._flags.parent.indentation_level) && (this._flags.indentation_level -= 1, this._output.set_indent(this._flags.indentation_level, this._flags.alignment));
  }, f.prototype.set_mode = function(s) {
    this._flags ? (this._flag_store.push(this._flags), this._previous_flags = this._flags) : this._previous_flags = this.create_flags(null, s), this._flags = this.create_flags(this._previous_flags, s), this._output.set_indent(this._flags.indentation_level, this._flags.alignment);
  }, f.prototype.restore_mode = function() {
    this._flag_store.length > 0 && (this._previous_flags = this._flags, this._flags = this._flag_store.pop(), this._previous_flags.mode === n.Statement && p(this._output, this._previous_flags), this._output.set_indent(this._flags.indentation_level, this._flags.alignment));
  }, f.prototype.start_of_object_property = function() {
    return this._flags.parent.mode === n.ObjectLiteral && this._flags.mode === n.Statement && (this._flags.last_token.text === ":" && this._flags.ternary_depth === 0 || g(this._flags.last_token, ["get", "set"]));
  }, f.prototype.start_of_statement = function(s) {
    var d = !1;
    return d = d || g(this._flags.last_token, ["var", "let", "const"]) && s.type === u.WORD, d = d || y(this._flags.last_token, "do"), d = d || !(this._flags.parent.mode === n.ObjectLiteral && this._flags.mode === n.Statement) && g(this._flags.last_token, S) && !s.newlines, d = d || y(this._flags.last_token, "else") && !(y(s, "if") && !s.comments_before), d = d || this._flags.last_token.type === u.END_EXPR && (this._previous_flags.mode === n.ForInitializer || this._previous_flags.mode === n.Conditional), d = d || this._flags.last_token.type === u.WORD && this._flags.mode === n.BlockStatement && !this._flags.in_case && !(s.text === "--" || s.text === "++") && this._last_last_text !== "function" && s.type !== u.WORD && s.type !== u.RESERVED, d = d || this._flags.mode === n.ObjectLiteral && (this._flags.last_token.text === ":" && this._flags.ternary_depth === 0 || g(this._flags.last_token, ["get", "set"])), d ? (this.set_mode(n.Statement), this.indent(), this.handle_whitespace_and_comments(s, !0), this.start_of_object_property() || this.allow_wrap_or_preserved_newline(
      s,
      g(s, ["do", "for", "if", "while"])
    ), !0) : !1;
  }, f.prototype.handle_start_expr = function(s) {
    this.start_of_statement(s) || this.handle_whitespace_and_comments(s);
    var d = n.Expression;
    if (s.text === "[") {
      if (this._flags.last_token.type === u.WORD || this._flags.last_token.text === ")") {
        g(this._flags.last_token, _) && (this._output.space_before_token = !0), this.print_token(s), this.set_mode(d), this.indent(), this._options.space_in_paren && (this._output.space_before_token = !0);
        return;
      }
      d = n.ArrayLiteral, E(this._flags.mode) && (this._flags.last_token.text === "[" || this._flags.last_token.text === "," && (this._last_last_text === "]" || this._last_last_text === "}")) && (this._options.keep_array_indentation || this.print_newline()), l(this._flags.last_token.type, [u.START_EXPR, u.END_EXPR, u.WORD, u.OPERATOR, u.DOT]) || (this._output.space_before_token = !0);
    } else {
      if (this._flags.last_token.type === u.RESERVED)
        this._flags.last_token.text === "for" ? (this._output.space_before_token = this._options.space_before_conditional, d = n.ForInitializer) : l(this._flags.last_token.text, ["if", "while", "switch"]) ? (this._output.space_before_token = this._options.space_before_conditional, d = n.Conditional) : l(this._flags.last_word, ["await", "async"]) ? this._output.space_before_token = !0 : this._flags.last_token.text === "import" && s.whitespace_before === "" ? this._output.space_before_token = !1 : (l(this._flags.last_token.text, _) || this._flags.last_token.text === "catch") && (this._output.space_before_token = !0);
      else if (this._flags.last_token.type === u.EQUALS || this._flags.last_token.type === u.OPERATOR)
        this.start_of_object_property() || this.allow_wrap_or_preserved_newline(s);
      else if (this._flags.last_token.type === u.WORD) {
        this._output.space_before_token = !1;
        var m = this._tokens.peek(-3);
        if (this._options.space_after_named_function && m) {
          var R = this._tokens.peek(-4);
          g(m, ["async", "function"]) || m.text === "*" && g(R, ["async", "function"]) ? this._output.space_before_token = !0 : this._flags.mode === n.ObjectLiteral ? (m.text === "{" || m.text === "," || m.text === "*" && (R.text === "{" || R.text === ",")) && (this._output.space_before_token = !0) : this._flags.parent && this._flags.parent.class_start_block && (this._output.space_before_token = !0);
        }
      } else
        this.allow_wrap_or_preserved_newline(s);
      (this._flags.last_token.type === u.RESERVED && (this._flags.last_word === "function" || this._flags.last_word === "typeof") || this._flags.last_token.text === "*" && (l(this._last_last_text, ["function", "yield"]) || this._flags.mode === n.ObjectLiteral && l(this._last_last_text, ["{", ","]))) && (this._output.space_before_token = this._options.space_after_anon_function);
    }
    this._flags.last_token.text === ";" || this._flags.last_token.type === u.START_BLOCK ? this.print_newline() : (this._flags.last_token.type === u.END_EXPR || this._flags.last_token.type === u.START_EXPR || this._flags.last_token.type === u.END_BLOCK || this._flags.last_token.text === "." || this._flags.last_token.type === u.COMMA) && this.allow_wrap_or_preserved_newline(s, s.newlines), this.print_token(s), this.set_mode(d), this._options.space_in_paren && (this._output.space_before_token = !0), this.indent();
  }, f.prototype.handle_end_expr = function(s) {
    for (; this._flags.mode === n.Statement; )
      this.restore_mode();
    this.handle_whitespace_and_comments(s), this._flags.multiline_frame && this.allow_wrap_or_preserved_newline(
      s,
      s.text === "]" && E(this._flags.mode) && !this._options.keep_array_indentation
    ), this._options.space_in_paren && (this._flags.last_token.type === u.START_EXPR && !this._options.space_in_empty_paren ? (this._output.trim(), this._output.space_before_token = !1) : this._output.space_before_token = !0), this.deindent(), this.print_token(s), this.restore_mode(), p(this._output, this._previous_flags), this._flags.do_while && this._previous_flags.mode === n.Conditional && (this._previous_flags.mode = n.Expression, this._flags.do_block = !1, this._flags.do_while = !1);
  }, f.prototype.handle_start_block = function(s) {
    this.handle_whitespace_and_comments(s);
    var d = this._tokens.peek(), m = this._tokens.peek(1);
    this._flags.last_word === "switch" && this._flags.last_token.type === u.END_EXPR ? (this.set_mode(n.BlockStatement), this._flags.in_case_statement = !0) : this._flags.case_body ? this.set_mode(n.BlockStatement) : m && (l(m.text, [":", ","]) && l(d.type, [u.STRING, u.WORD, u.RESERVED]) || l(d.text, ["get", "set", "..."]) && l(m.type, [u.WORD, u.RESERVED])) ? l(this._last_last_text, ["class", "interface"]) && !l(m.text, [":", ","]) ? this.set_mode(n.BlockStatement) : this.set_mode(n.ObjectLiteral) : this._flags.last_token.type === u.OPERATOR && this._flags.last_token.text === "=>" ? this.set_mode(n.BlockStatement) : l(this._flags.last_token.type, [u.EQUALS, u.START_EXPR, u.COMMA, u.OPERATOR]) || g(this._flags.last_token, ["return", "throw", "import", "default"]) ? this.set_mode(n.ObjectLiteral) : this.set_mode(n.BlockStatement), this._flags.last_token && g(this._flags.last_token.previous, ["class", "extends"]) && (this._flags.class_start_block = !0);
    var R = !d.comments_before && d.text === "}", N = R && this._flags.last_word === "function" && this._flags.last_token.type === u.END_EXPR;
    if (this._options.brace_preserve_inline) {
      var W = 0, M = null;
      this._flags.inline_frame = !0;
      do
        if (W += 1, M = this._tokens.peek(W - 1), M.newlines) {
          this._flags.inline_frame = !1;
          break;
        }
      while (M.type !== u.EOF && !(M.type === u.END_BLOCK && M.opened === s));
    }
    (this._options.brace_style === "expand" || this._options.brace_style === "none" && s.newlines) && !this._flags.inline_frame ? this._flags.last_token.type !== u.OPERATOR && (N || this._flags.last_token.type === u.EQUALS || g(this._flags.last_token, C) && this._flags.last_token.text !== "else") ? this._output.space_before_token = !0 : this.print_newline(!1, !0) : (E(this._previous_flags.mode) && (this._flags.last_token.type === u.START_EXPR || this._flags.last_token.type === u.COMMA) && ((this._flags.last_token.type === u.COMMA || this._options.space_in_paren) && (this._output.space_before_token = !0), (this._flags.last_token.type === u.COMMA || this._flags.last_token.type === u.START_EXPR && this._flags.inline_frame) && (this.allow_wrap_or_preserved_newline(s), this._previous_flags.multiline_frame = this._previous_flags.multiline_frame || this._flags.multiline_frame, this._flags.multiline_frame = !1)), this._flags.last_token.type !== u.OPERATOR && this._flags.last_token.type !== u.START_EXPR && (l(this._flags.last_token.type, [u.START_BLOCK, u.SEMICOLON]) && !this._flags.inline_frame ? this.print_newline() : this._output.space_before_token = !0)), this.print_token(s), this.indent(), !R && !(this._options.brace_preserve_inline && this._flags.inline_frame) && this.print_newline();
  }, f.prototype.handle_end_block = function(s) {
    for (this.handle_whitespace_and_comments(s); this._flags.mode === n.Statement; )
      this.restore_mode();
    var d = this._flags.last_token.type === u.START_BLOCK;
    this._flags.inline_frame && !d ? this._output.space_before_token = !0 : this._options.brace_style === "expand" ? d || this.print_newline() : d || (E(this._flags.mode) && this._options.keep_array_indentation ? (this._options.keep_array_indentation = !1, this.print_newline(), this._options.keep_array_indentation = !0) : this.print_newline()), this.restore_mode(), this.print_token(s);
  }, f.prototype.handle_word = function(s) {
    if (s.type === u.RESERVED) {
      if (l(s.text, ["set", "get"]) && this._flags.mode !== n.ObjectLiteral)
        s.type = u.WORD;
      else if (s.text === "import" && l(this._tokens.peek().text, ["(", "."]))
        s.type = u.WORD;
      else if (l(s.text, ["as", "from"]) && !this._flags.import_block)
        s.type = u.WORD;
      else if (this._flags.mode === n.ObjectLiteral) {
        var d = this._tokens.peek();
        d.text === ":" && (s.type = u.WORD);
      }
    }
    if (this.start_of_statement(s) ? g(this._flags.last_token, ["var", "let", "const"]) && s.type === u.WORD && (this._flags.declaration_statement = !0) : s.newlines && !k(this._flags.mode) && (this._flags.last_token.type !== u.OPERATOR || this._flags.last_token.text === "--" || this._flags.last_token.text === "++") && this._flags.last_token.type !== u.EQUALS && (this._options.preserve_newlines || !g(this._flags.last_token, ["var", "let", "const", "set", "get"])) ? (this.handle_whitespace_and_comments(s), this.print_newline()) : this.handle_whitespace_and_comments(s), this._flags.do_block && !this._flags.do_while)
      if (y(s, "while")) {
        this._output.space_before_token = !0, this.print_token(s), this._output.space_before_token = !0, this._flags.do_while = !0;
        return;
      } else
        this.print_newline(), this._flags.do_block = !1;
    if (this._flags.if_block)
      if (!this._flags.else_block && y(s, "else"))
        this._flags.else_block = !0;
      else {
        for (; this._flags.mode === n.Statement; )
          this.restore_mode();
        this._flags.if_block = !1, this._flags.else_block = !1;
      }
    if (this._flags.in_case_statement && g(s, ["case", "default"])) {
      this.print_newline(), !this._flags.case_block && (this._flags.case_body || this._options.jslint_happy) && this.deindent(), this._flags.case_body = !1, this.print_token(s), this._flags.in_case = !0;
      return;
    }
    if ((this._flags.last_token.type === u.COMMA || this._flags.last_token.type === u.START_EXPR || this._flags.last_token.type === u.EQUALS || this._flags.last_token.type === u.OPERATOR) && !this.start_of_object_property() && !// start of object property is different for numeric values with +/- prefix operators
    (l(this._flags.last_token.text, ["+", "-"]) && this._last_last_text === ":" && this._flags.parent.mode === n.ObjectLiteral) && this.allow_wrap_or_preserved_newline(s), y(s, "function")) {
      (l(this._flags.last_token.text, ["}", ";"]) || this._output.just_added_newline() && !(l(this._flags.last_token.text, ["(", "[", "{", ":", "=", ","]) || this._flags.last_token.type === u.OPERATOR)) && !this._output.just_added_blankline() && !s.comments_before && (this.print_newline(), this.print_newline(!0)), this._flags.last_token.type === u.RESERVED || this._flags.last_token.type === u.WORD ? g(this._flags.last_token, ["get", "set", "new", "export"]) || g(this._flags.last_token, S) ? this._output.space_before_token = !0 : y(this._flags.last_token, "default") && this._last_last_text === "export" ? this._output.space_before_token = !0 : this._flags.last_token.text === "declare" ? this._output.space_before_token = !0 : this.print_newline() : this._flags.last_token.type === u.OPERATOR || this._flags.last_token.text === "=" ? this._output.space_before_token = !0 : !this._flags.multiline_frame && (k(this._flags.mode) || E(this._flags.mode)) || this.print_newline(), this.print_token(s), this._flags.last_word = s.text;
      return;
    }
    var m = "NONE";
    if (this._flags.last_token.type === u.END_BLOCK ? this._previous_flags.inline_frame ? m = "SPACE" : g(s, ["else", "catch", "finally", "from"]) ? this._options.brace_style === "expand" || this._options.brace_style === "end-expand" || this._options.brace_style === "none" && s.newlines ? m = "NEWLINE" : (m = "SPACE", this._output.space_before_token = !0) : m = "NEWLINE" : this._flags.last_token.type === u.SEMICOLON && this._flags.mode === n.BlockStatement ? m = "NEWLINE" : this._flags.last_token.type === u.SEMICOLON && k(this._flags.mode) ? m = "SPACE" : this._flags.last_token.type === u.STRING ? m = "NEWLINE" : this._flags.last_token.type === u.RESERVED || this._flags.last_token.type === u.WORD || this._flags.last_token.text === "*" && (l(this._last_last_text, ["function", "yield"]) || this._flags.mode === n.ObjectLiteral && l(this._last_last_text, ["{", ","])) ? m = "SPACE" : this._flags.last_token.type === u.START_BLOCK ? this._flags.inline_frame ? m = "SPACE" : m = "NEWLINE" : this._flags.last_token.type === u.END_EXPR && (this._output.space_before_token = !0, m = "NEWLINE"), g(s, _) && this._flags.last_token.text !== ")" && (this._flags.inline_frame || this._flags.last_token.text === "else" || this._flags.last_token.text === "export" ? m = "SPACE" : m = "NEWLINE"), g(s, ["else", "catch", "finally"]))
      if ((!(this._flags.last_token.type === u.END_BLOCK && this._previous_flags.mode === n.BlockStatement) || this._options.brace_style === "expand" || this._options.brace_style === "end-expand" || this._options.brace_style === "none" && s.newlines) && !this._flags.inline_frame)
        this.print_newline();
      else {
        this._output.trim(!0);
        var R = this._output.current_line;
        R.last() !== "}" && this.print_newline(), this._output.space_before_token = !0;
      }
    else m === "NEWLINE" ? g(this._flags.last_token, C) ? this._output.space_before_token = !0 : this._flags.last_token.text === "declare" && g(s, ["var", "let", "const"]) ? this._output.space_before_token = !0 : this._flags.last_token.type !== u.END_EXPR ? (this._flags.last_token.type !== u.START_EXPR || !g(s, ["var", "let", "const"])) && this._flags.last_token.text !== ":" && (y(s, "if") && y(s.previous, "else") ? this._output.space_before_token = !0 : this.print_newline()) : g(s, _) && this._flags.last_token.text !== ")" && this.print_newline() : this._flags.multiline_frame && E(this._flags.mode) && this._flags.last_token.text === "," && this._last_last_text === "}" ? this.print_newline() : m === "SPACE" && (this._output.space_before_token = !0);
    s.previous && (s.previous.type === u.WORD || s.previous.type === u.RESERVED) && (this._output.space_before_token = !0), this.print_token(s), this._flags.last_word = s.text, s.type === u.RESERVED && (s.text === "do" ? this._flags.do_block = !0 : s.text === "if" ? this._flags.if_block = !0 : s.text === "import" ? this._flags.import_block = !0 : this._flags.import_block && y(s, "from") && (this._flags.import_block = !1));
  }, f.prototype.handle_semicolon = function(s) {
    this.start_of_statement(s) ? this._output.space_before_token = !1 : this.handle_whitespace_and_comments(s);
    for (var d = this._tokens.peek(); this._flags.mode === n.Statement && !(this._flags.if_block && y(d, "else")) && !this._flags.do_block; )
      this.restore_mode();
    this._flags.import_block && (this._flags.import_block = !1), this.print_token(s);
  }, f.prototype.handle_string = function(s) {
    s.text.startsWith("`") && s.newlines === 0 && s.whitespace_before === "" && (s.previous.text === ")" || this._flags.last_token.type === u.WORD) || (this.start_of_statement(s) ? this._output.space_before_token = !0 : (this.handle_whitespace_and_comments(s), this._flags.last_token.type === u.RESERVED || this._flags.last_token.type === u.WORD || this._flags.inline_frame ? this._output.space_before_token = !0 : this._flags.last_token.type === u.COMMA || this._flags.last_token.type === u.START_EXPR || this._flags.last_token.type === u.EQUALS || this._flags.last_token.type === u.OPERATOR ? this.start_of_object_property() || this.allow_wrap_or_preserved_newline(s) : s.text.startsWith("`") && this._flags.last_token.type === u.END_EXPR && (s.previous.text === "]" || s.previous.text === ")") && s.newlines === 0 ? this._output.space_before_token = !0 : this.print_newline())), this.print_token(s);
  }, f.prototype.handle_equals = function(s) {
    this.start_of_statement(s) || this.handle_whitespace_and_comments(s), this._flags.declaration_statement && (this._flags.declaration_assignment = !0), this._output.space_before_token = !0, this.print_token(s), this._output.space_before_token = !0;
  }, f.prototype.handle_comma = function(s) {
    this.handle_whitespace_and_comments(s, !0), this.print_token(s), this._output.space_before_token = !0, this._flags.declaration_statement ? (k(this._flags.parent.mode) && (this._flags.declaration_assignment = !1), this._flags.declaration_assignment ? (this._flags.declaration_assignment = !1, this.print_newline(!1, !0)) : this._options.comma_first && this.allow_wrap_or_preserved_newline(s)) : this._flags.mode === n.ObjectLiteral || this._flags.mode === n.Statement && this._flags.parent.mode === n.ObjectLiteral ? (this._flags.mode === n.Statement && this.restore_mode(), this._flags.inline_frame || this.print_newline()) : this._options.comma_first && this.allow_wrap_or_preserved_newline(s);
  }, f.prototype.handle_operator = function(s) {
    var d = s.text === "*" && (g(this._flags.last_token, ["function", "yield"]) || l(this._flags.last_token.type, [u.START_BLOCK, u.COMMA, u.END_BLOCK, u.SEMICOLON])), m = l(s.text, ["-", "+"]) && (l(this._flags.last_token.type, [u.START_BLOCK, u.START_EXPR, u.EQUALS, u.OPERATOR]) || l(this._flags.last_token.text, _) || this._flags.last_token.text === ",");
    if (!this.start_of_statement(s)) {
      var R = !d;
      this.handle_whitespace_and_comments(s, R);
    }
    if (s.text === "*" && this._flags.last_token.type === u.DOT) {
      this.print_token(s);
      return;
    }
    if (s.text === "::") {
      this.print_token(s);
      return;
    }
    if (l(s.text, ["-", "+"]) && this.start_of_object_property()) {
      this.print_token(s);
      return;
    }
    if (this._flags.last_token.type === u.OPERATOR && l(this._options.operator_position, o) && this.allow_wrap_or_preserved_newline(s), s.text === ":" && this._flags.in_case) {
      this.print_token(s), this._flags.in_case = !1, this._flags.case_body = !0, this._tokens.peek().type !== u.START_BLOCK ? (this.indent(), this.print_newline(), this._flags.case_block = !1) : (this._flags.case_block = !0, this._output.space_before_token = !0);
      return;
    }
    var N = !0, W = !0, M = !1;
    if (s.text === ":" ? this._flags.ternary_depth === 0 ? N = !1 : (this._flags.ternary_depth -= 1, M = !0) : s.text === "?" && (this._flags.ternary_depth += 1), !m && !d && this._options.preserve_newlines && l(s.text, h)) {
      var $ = s.text === ":", ct = $ && M, Wt = $ && !M;
      switch (this._options.operator_position) {
        case x.before_newline:
          this._output.space_before_token = !Wt, this.print_token(s), (!$ || ct) && this.allow_wrap_or_preserved_newline(s), this._output.space_before_token = !0;
          return;
        case x.after_newline:
          this._output.space_before_token = !0, !$ || ct ? this._tokens.peek().newlines ? this.print_newline(!1, !0) : this.allow_wrap_or_preserved_newline(s) : this._output.space_before_token = !1, this.print_token(s), this._output.space_before_token = !0;
          return;
        case x.preserve_newline:
          Wt || this.allow_wrap_or_preserved_newline(s), N = !(this._output.just_added_newline() || Wt), this._output.space_before_token = N, this.print_token(s), this._output.space_before_token = !0;
          return;
      }
    }
    if (d) {
      this.allow_wrap_or_preserved_newline(s), N = !1;
      var xe = this._tokens.peek();
      W = xe && l(xe.type, [u.WORD, u.RESERVED]);
    } else if (s.text === "...")
      this.allow_wrap_or_preserved_newline(s), N = this._flags.last_token.type === u.START_BLOCK, W = !1;
    else if (l(s.text, ["--", "++", "!", "~"]) || m) {
      if ((this._flags.last_token.type === u.COMMA || this._flags.last_token.type === u.START_EXPR) && this.allow_wrap_or_preserved_newline(s), N = !1, W = !1, s.newlines && (s.text === "--" || s.text === "++" || s.text === "~")) {
        var Ee = g(this._flags.last_token, C) && s.newlines;
        Ee && (this._previous_flags.if_block || this._previous_flags.else_block) && this.restore_mode(), this.print_newline(Ee, !0);
      }
      this._flags.last_token.text === ";" && k(this._flags.mode) && (N = !0), this._flags.last_token.type === u.RESERVED ? N = !0 : this._flags.last_token.type === u.END_EXPR ? N = !(this._flags.last_token.text === "]" && (s.text === "--" || s.text === "++")) : this._flags.last_token.type === u.OPERATOR && (N = l(s.text, ["--", "-", "++", "+"]) && l(this._flags.last_token.text, ["--", "-", "++", "+"]), l(s.text, ["+", "-"]) && l(this._flags.last_token.text, ["--", "++"]) && (W = !0)), (this._flags.mode === n.BlockStatement && !this._flags.inline_frame || this._flags.mode === n.Statement) && (this._flags.last_token.text === "{" || this._flags.last_token.text === ";") && this.print_newline();
    }
    this._output.space_before_token = this._output.space_before_token || N, this.print_token(s), this._output.space_before_token = W;
  }, f.prototype.handle_block_comment = function(s, d) {
    if (this._output.raw) {
      this._output.add_raw_token(s), s.directives && s.directives.preserve === "end" && (this._output.raw = this._options.test_output_raw);
      return;
    }
    if (s.directives) {
      this.print_newline(!1, d), this.print_token(s), s.directives.preserve === "start" && (this._output.raw = !0), this.print_newline(!1, !0);
      return;
    }
    if (!i.newline.test(s.text) && !s.newlines) {
      this._output.space_before_token = !0, this.print_token(s), this._output.space_before_token = !0;
      return;
    } else
      this.print_block_commment(s, d);
  }, f.prototype.print_block_commment = function(s, d) {
    var m = T(s.text), R, N = !1, W = !1, M = s.whitespace_before, $ = M.length;
    if (this.print_newline(!1, d), this.print_token_line_indentation(s), this._output.add_token(m[0]), this.print_newline(!1, d), m.length > 1) {
      for (m = m.slice(1), N = c(m, "*"), W = b(m, M), N && (this._flags.alignment = 1), R = 0; R < m.length; R++)
        N ? (this.print_token_line_indentation(s), this._output.add_token(O(m[R]))) : W && m[R] ? (this.print_token_line_indentation(s), this._output.add_token(m[R].substring($))) : (this._output.current_line.set_indent(-1), this._output.add_token(m[R])), this.print_newline(!1, d);
      this._flags.alignment = 0;
    }
  }, f.prototype.handle_comment = function(s, d) {
    s.newlines ? this.print_newline(!1, d) : this._output.trim(!0), this._output.space_before_token = !0, this.print_token(s), this.print_newline(!1, d);
  }, f.prototype.handle_dot = function(s) {
    this.start_of_statement(s) || this.handle_whitespace_and_comments(s, !0), this._flags.last_token.text.match("^[0-9]+$") && (this._output.space_before_token = !0), g(this._flags.last_token, C) ? this._output.space_before_token = !1 : this.allow_wrap_or_preserved_newline(
      s,
      this._flags.last_token.text === ")" && this._options.break_chained_methods
    ), this._options.unindent_chained_methods && this._output.just_added_newline() && this.deindent(), this.print_token(s);
  }, f.prototype.handle_unknown = function(s, d) {
    this.print_token(s), s.text[s.text.length - 1] === `
` && this.print_newline(!1, d);
  }, f.prototype.handle_eof = function(s) {
    for (; this._flags.mode === n.Statement; )
      this.restore_mode();
    this.handle_whitespace_and_comments(s);
  }, Kt.Beautifier = f, Kt;
}
var Qe;
function Xi() {
  if (Qe) return gt.exports;
  Qe = 1;
  var a = Gi().Beautifier, t = vi().Options;
  function i(e, r) {
    var _ = new a(e, r);
    return _.beautify();
  }
  return gt.exports = i, gt.exports.defaultOptions = function() {
    return new t();
  }, gt.exports;
}
var vt = { exports: {} }, Jt = {}, Zt = {}, He;
function wi() {
  if (He) return Zt;
  He = 1;
  var a = ve().Options;
  function t(i) {
    a.call(this, i, "css"), this.selector_separator_newline = this._get_boolean("selector_separator_newline", !0), this.newline_between_rules = this._get_boolean("newline_between_rules", !0);
    var e = this._get_boolean("space_around_selector_separator");
    this.space_around_combinator = this._get_boolean("space_around_combinator") || e;
    var r = this._get_selection_list("brace_style", ["collapse", "expand", "end-expand", "none", "preserve-inline"]);
    this.brace_style = "collapse";
    for (var _ = 0; _ < r.length; _++)
      r[_] !== "expand" ? this.brace_style = "collapse" : this.brace_style = r[_];
  }
  return t.prototype = new a(), Zt.Options = t, Zt;
}
var Je;
function Qi() {
  if (Je) return Jt;
  Je = 1;
  var a = wi().Options, t = me().Output, i = ye().InputScanner, e = we().Directives, r = new e(/\/\*/, /\*\//), _ = /\r\n|[\r\n]/, h = /\r\n|[\r\n]/g, u = /\s/, l = /(?:\s|\n)+/g, O = /\/\*(?:[\s\S]*?)((?:\*\/)|$)/g, v = /\/\/(?:[^\n\r\u2028\u2029]*)/g;
  function y(g, C) {
    this._source_text = g || "", this._options = new a(C), this._ch = null, this._input = null, this.NESTED_AT_RULE = {
      page: !0,
      "font-face": !0,
      keyframes: !0,
      // also in CONDITIONAL_GROUP_RULE below
      media: !0,
      supports: !0,
      document: !0
    }, this.CONDITIONAL_GROUP_RULE = {
      media: !0,
      supports: !0,
      document: !0
    }, this.NON_SEMICOLON_NEWLINE_PROPERTY = [
      "grid-template-areas",
      "grid-template"
    ];
  }
  return y.prototype.eatString = function(g) {
    var C = "";
    for (this._ch = this._input.next(); this._ch; ) {
      if (C += this._ch, this._ch === "\\")
        C += this._input.next();
      else if (g.indexOf(this._ch) !== -1 || this._ch === `
`)
        break;
      this._ch = this._input.next();
    }
    return C;
  }, y.prototype.eatWhitespace = function(g) {
    for (var C = u.test(this._input.peek()), A = 0; u.test(this._input.peek()); )
      this._ch = this._input.next(), g && this._ch === `
` && (A === 0 || A < this._options.max_preserve_newlines) && (A++, this._output.add_new_line(!0));
    return C;
  }, y.prototype.foundNestedPseudoClass = function() {
    for (var g = 0, C = 1, A = this._input.peek(C); A; ) {
      if (A === "{")
        return !0;
      if (A === "(")
        g += 1;
      else if (A === ")") {
        if (g === 0)
          return !1;
        g -= 1;
      } else if (A === ";" || A === "}")
        return !1;
      C++, A = this._input.peek(C);
    }
    return !1;
  }, y.prototype.print_string = function(g) {
    this._output.set_indent(this._indentLevel), this._output.non_breaking_space = !0, this._output.add_token(g);
  }, y.prototype.preserveSingleSpace = function(g) {
    g && (this._output.space_before_token = !0);
  }, y.prototype.indent = function() {
    this._indentLevel++;
  }, y.prototype.outdent = function() {
    this._indentLevel > 0 && this._indentLevel--;
  }, y.prototype.beautify = function() {
    if (this._options.disabled)
      return this._source_text;
    var g = this._source_text, C = this._options.eol;
    C === "auto" && (C = `
`, g && _.test(g || "") && (C = g.match(_)[0])), g = g.replace(h, `
`);
    var A = g.match(/^[\t ]*/)[0];
    this._output = new t(this._options, A), this._input = new i(g), this._indentLevel = 0, this._nestedLevel = 0, this._ch = null;
    for (var x = 0, o = !1, n = !1, p = !1, T = !1, E = !1, k = this._ch, c = !1, b, f, S; b = this._input.read(l), f = b !== "", S = k, this._ch = this._input.next(), this._ch === "\\" && this._input.hasNext() && (this._ch += this._input.next()), k = this._ch, this._ch; )
      if (this._ch === "/" && this._input.peek() === "*") {
        this._output.add_new_line(), this._input.back();
        var s = this._input.read(O), d = r.get_directives(s);
        d && d.ignore === "start" && (s += r.readIgnored(this._input)), this.print_string(s), this.eatWhitespace(!0), this._output.add_new_line();
      } else if (this._ch === "/" && this._input.peek() === "/")
        this._output.space_before_token = !0, this._input.back(), this.print_string(this._input.read(v)), this.eatWhitespace(!0);
      else if (this._ch === "$") {
        this.preserveSingleSpace(f), this.print_string(this._ch);
        var m = this._input.peekUntilAfter(/[: ,;{}()[\]\/='"]/g);
        m.match(/[ :]$/) && (m = this.eatString(": ").replace(/\s+$/, ""), this.print_string(m), this._output.space_before_token = !0), x === 0 && m.indexOf(":") !== -1 && (n = !0, this.indent());
      } else if (this._ch === "@")
        if (this.preserveSingleSpace(f), this._input.peek() === "{")
          this.print_string(this._ch + this.eatString("}"));
        else {
          this.print_string(this._ch);
          var R = this._input.peekUntilAfter(/[: ,;{}()[\]\/='"]/g);
          R.match(/[ :]$/) && (R = this.eatString(": ").replace(/\s+$/, ""), this.print_string(R), this._output.space_before_token = !0), x === 0 && R.indexOf(":") !== -1 ? (n = !0, this.indent()) : R in this.NESTED_AT_RULE ? (this._nestedLevel += 1, R in this.CONDITIONAL_GROUP_RULE && (p = !0)) : x === 0 && !n && (T = !0);
        }
      else if (this._ch === "#" && this._input.peek() === "{")
        this.preserveSingleSpace(f), this.print_string(this._ch + this.eatString("}"));
      else if (this._ch === "{")
        n && (n = !1, this.outdent()), T = !1, p ? (p = !1, o = this._indentLevel >= this._nestedLevel) : o = this._indentLevel >= this._nestedLevel - 1, this._options.newline_between_rules && o && this._output.previous_line && this._output.previous_line.item(-1) !== "{" && this._output.ensure_empty_line_above("/", ","), this._output.space_before_token = !0, this._options.brace_style === "expand" ? (this._output.add_new_line(), this.print_string(this._ch), this.indent(), this._output.set_indent(this._indentLevel)) : (S === "(" ? this._output.space_before_token = !1 : S !== "," && this.indent(), this.print_string(this._ch)), this.eatWhitespace(!0), this._output.add_new_line();
      else if (this._ch === "}")
        this.outdent(), this._output.add_new_line(), S === "{" && this._output.trim(!0), n && (this.outdent(), n = !1), this.print_string(this._ch), o = !1, this._nestedLevel && this._nestedLevel--, this.eatWhitespace(!0), this._output.add_new_line(), this._options.newline_between_rules && !this._output.just_added_blankline() && this._input.peek() !== "}" && this._output.add_new_line(!0), this._input.peek() === ")" && (this._output.trim(!0), this._options.brace_style === "expand" && this._output.add_new_line(!0));
      else if (this._ch === ":") {
        for (var N = 0; N < this.NON_SEMICOLON_NEWLINE_PROPERTY.length; N++)
          if (this._input.lookBack(this.NON_SEMICOLON_NEWLINE_PROPERTY[N])) {
            c = !0;
            break;
          }
        (o || p) && !(this._input.lookBack("&") || this.foundNestedPseudoClass()) && !this._input.lookBack("(") && !T && x === 0 ? (this.print_string(":"), n || (n = !0, this._output.space_before_token = !0, this.eatWhitespace(!0), this.indent())) : (this._input.lookBack(" ") && (this._output.space_before_token = !0), this._input.peek() === ":" ? (this._ch = this._input.next(), this.print_string("::")) : this.print_string(":"));
      } else if (this._ch === '"' || this._ch === "'") {
        var W = S === '"' || S === "'";
        this.preserveSingleSpace(W || f), this.print_string(this._ch + this.eatString(this._ch)), this.eatWhitespace(!0);
      } else if (this._ch === ";")
        c = !1, x === 0 ? (n && (this.outdent(), n = !1), T = !1, this.print_string(this._ch), this.eatWhitespace(!0), this._input.peek() !== "/" && this._output.add_new_line()) : (this.print_string(this._ch), this.eatWhitespace(!0), this._output.space_before_token = !0);
      else if (this._ch === "(")
        if (this._input.lookBack("url"))
          this.print_string(this._ch), this.eatWhitespace(), x++, this.indent(), this._ch = this._input.next(), this._ch === ")" || this._ch === '"' || this._ch === "'" ? this._input.back() : this._ch && (this.print_string(this._ch + this.eatString(")")), x && (x--, this.outdent()));
        else {
          var M = !1;
          this._input.lookBack("with") && (M = !0), this.preserveSingleSpace(f || M), this.print_string(this._ch), n && S === "$" && this._options.selector_separator_newline ? (this._output.add_new_line(), E = !0) : (this.eatWhitespace(), x++, this.indent());
        }
      else if (this._ch === ")")
        x && (x--, this.outdent()), E && this._input.peek() === ";" && this._options.selector_separator_newline && (E = !1, this.outdent(), this._output.add_new_line()), this.print_string(this._ch);
      else if (this._ch === ",")
        this.print_string(this._ch), this.eatWhitespace(!0), this._options.selector_separator_newline && (!n || E) && x === 0 && !T ? this._output.add_new_line() : this._output.space_before_token = !0;
      else if ((this._ch === ">" || this._ch === "+" || this._ch === "~") && !n && x === 0)
        this._options.space_around_combinator ? (this._output.space_before_token = !0, this.print_string(this._ch), this._output.space_before_token = !0) : (this.print_string(this._ch), this.eatWhitespace(), this._ch && u.test(this._ch) && (this._ch = ""));
      else if (this._ch === "]")
        this.print_string(this._ch);
      else if (this._ch === "[")
        this.preserveSingleSpace(f), this.print_string(this._ch);
      else if (this._ch === "=")
        this.eatWhitespace(), this.print_string("="), u.test(this._ch) && (this._ch = "");
      else if (this._ch === "!" && !this._input.lookBack("\\"))
        this._output.space_before_token = !0, this.print_string(this._ch);
      else {
        var $ = S === '"' || S === "'";
        this.preserveSingleSpace($ || f), this.print_string(this._ch), !this._output.just_added_newline() && this._input.peek() === `
` && c && this._output.add_new_line();
      }
    var ct = this._output.get_code(C);
    return ct;
  }, Jt.Beautifier = y, Jt;
}
var Ze;
function Hi() {
  if (Ze) return vt.exports;
  Ze = 1;
  var a = Qi().Beautifier, t = wi().Options;
  function i(e, r) {
    var _ = new a(e, r);
    return _.beautify();
  }
  return vt.exports = i, vt.exports.defaultOptions = function() {
    return new t();
  }, vt.exports;
}
var yt = { exports: {} }, Yt = {}, te = {}, Ye;
function xi() {
  if (Ye) return te;
  Ye = 1;
  var a = ve().Options;
  function t(i) {
    a.call(this, i, "html"), this.templating.length === 1 && this.templating[0] === "auto" && (this.templating = ["django", "erb", "handlebars", "php"]), this.indent_inner_html = this._get_boolean("indent_inner_html"), this.indent_body_inner_html = this._get_boolean("indent_body_inner_html", !0), this.indent_head_inner_html = this._get_boolean("indent_head_inner_html", !0), this.indent_handlebars = this._get_boolean("indent_handlebars", !0), this.wrap_attributes = this._get_selection(
      "wrap_attributes",
      ["auto", "force", "force-aligned", "force-expand-multiline", "aligned-multiple", "preserve", "preserve-aligned"]
    ), this.wrap_attributes_min_attrs = this._get_number("wrap_attributes_min_attrs", 2), this.wrap_attributes_indent_size = this._get_number("wrap_attributes_indent_size", this.indent_size), this.extra_liners = this._get_array("extra_liners", ["head", "body", "/html"]), this.inline = this._get_array("inline", [
      "a",
      "abbr",
      "area",
      "audio",
      "b",
      "bdi",
      "bdo",
      "br",
      "button",
      "canvas",
      "cite",
      "code",
      "data",
      "datalist",
      "del",
      "dfn",
      "em",
      "embed",
      "i",
      "iframe",
      "img",
      "input",
      "ins",
      "kbd",
      "keygen",
      "label",
      "map",
      "mark",
      "math",
      "meter",
      "noscript",
      "object",
      "output",
      "progress",
      "q",
      "ruby",
      "s",
      "samp",
      /* 'script', */
      "select",
      "small",
      "span",
      "strong",
      "sub",
      "sup",
      "svg",
      "template",
      "textarea",
      "time",
      "u",
      "var",
      "video",
      "wbr",
      "text",
      // obsolete inline tags
      "acronym",
      "big",
      "strike",
      "tt"
    ]), this.inline_custom_elements = this._get_boolean("inline_custom_elements", !0), this.void_elements = this._get_array("void_elements", [
      // HTLM void elements - aka self-closing tags - aka singletons
      // https://www.w3.org/html/wg/drafts/html/master/syntax.html#void-elements
      "area",
      "base",
      "br",
      "col",
      "embed",
      "hr",
      "img",
      "input",
      "keygen",
      "link",
      "menuitem",
      "meta",
      "param",
      "source",
      "track",
      "wbr",
      // NOTE: Optional tags are too complex for a simple list
      // they are hard coded in _do_optional_end_element
      // Doctype and xml elements
      "!doctype",
      "?xml",
      // obsolete tags
      // basefont: https://www.computerhope.com/jargon/h/html-basefont-tag.htm
      // isndex: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/isindex
      "basefont",
      "isindex"
    ]), this.unformatted = this._get_array("unformatted", []), this.content_unformatted = this._get_array("content_unformatted", [
      "pre",
      "textarea"
    ]), this.unformatted_content_delimiter = this._get_characters("unformatted_content_delimiter"), this.indent_scripts = this._get_selection("indent_scripts", ["normal", "keep", "separate"]);
  }
  return t.prototype = new a(), te.Options = t, te;
}
var wt = {}, ti;
function ei() {
  if (ti) return wt;
  ti = 1;
  var a = Lt().Tokenizer, t = Lt().TOKEN, i = we().Directives, e = yi().TemplatablePattern, r = Dt().Pattern, _ = {
    TAG_OPEN: "TK_TAG_OPEN",
    TAG_CLOSE: "TK_TAG_CLOSE",
    CONTROL_FLOW_OPEN: "TK_CONTROL_FLOW_OPEN",
    CONTROL_FLOW_CLOSE: "TK_CONTROL_FLOW_CLOSE",
    ATTRIBUTE: "TK_ATTRIBUTE",
    EQUALS: "TK_EQUALS",
    VALUE: "TK_VALUE",
    COMMENT: "TK_COMMENT",
    TEXT: "TK_TEXT",
    UNKNOWN: "TK_UNKNOWN",
    START: t.START,
    RAW: t.RAW,
    EOF: t.EOF
  }, h = new i(/<\!--/, /-->/), u = function(l, O) {
    a.call(this, l, O), this._current_tag_name = "";
    var v = new e(this._input).read_options(this._options), y = new r(this._input);
    if (this.__patterns = {
      word: v.until(/[\n\r\t <]/),
      word_control_flow_close_excluded: v.until(/[\n\r\t <}]/),
      single_quote: v.until_after(/'/),
      double_quote: v.until_after(/"/),
      attribute: v.until(/[\n\r\t =>]|\/>/),
      element_name: v.until(/[\n\r\t >\/]/),
      angular_control_flow_start: y.matching(/\@[a-zA-Z]+[^({]*[({]/),
      handlebars_comment: y.starting_with(/{{!--/).until_after(/--}}/),
      handlebars: y.starting_with(/{{/).until_after(/}}/),
      handlebars_open: y.until(/[\n\r\t }]/),
      handlebars_raw_close: y.until(/}}/),
      comment: y.starting_with(/<!--/).until_after(/-->/),
      cdata: y.starting_with(/<!\[CDATA\[/).until_after(/]]>/),
      // https://en.wikipedia.org/wiki/Conditional_comment
      conditional_comment: y.starting_with(/<!\[/).until_after(/]>/),
      processing: y.starting_with(/<\?/).until_after(/\?>/)
    }, this._options.indent_handlebars && (this.__patterns.word = this.__patterns.word.exclude("handlebars"), this.__patterns.word_control_flow_close_excluded = this.__patterns.word_control_flow_close_excluded.exclude("handlebars")), this._unformatted_content_delimiter = null, this._options.unformatted_content_delimiter) {
      var g = this._input.get_literal_regexp(this._options.unformatted_content_delimiter);
      this.__patterns.unformatted_content_delimiter = y.matching(g).until_after(g);
    }
  };
  return u.prototype = new a(), u.prototype._is_comment = function(l) {
    return !1;
  }, u.prototype._is_opening = function(l) {
    return l.type === _.TAG_OPEN || l.type === _.CONTROL_FLOW_OPEN;
  }, u.prototype._is_closing = function(l, O) {
    return l.type === _.TAG_CLOSE && O && ((l.text === ">" || l.text === "/>") && O.text[0] === "<" || l.text === "}}" && O.text[0] === "{" && O.text[1] === "{") || l.type === _.CONTROL_FLOW_CLOSE && l.text === "}" && O.text.endsWith("{");
  }, u.prototype._reset = function() {
    this._current_tag_name = "";
  }, u.prototype._get_next_token = function(l, O) {
    var v = null;
    this._readWhitespace();
    var y = this._input.peek();
    return y === null ? this._create_token(_.EOF, "") : (v = v || this._read_open_handlebars(y, O), v = v || this._read_attribute(y, l, O), v = v || this._read_close(y, O), v = v || this._read_control_flows(y, O), v = v || this._read_raw_content(y, l, O), v = v || this._read_content_word(y, O), v = v || this._read_comment_or_cdata(y), v = v || this._read_processing(y), v = v || this._read_open(y, O), v = v || this._create_token(_.UNKNOWN, this._input.next()), v);
  }, u.prototype._read_comment_or_cdata = function(l) {
    var O = null, v = null, y = null;
    if (l === "<") {
      var g = this._input.peek(1);
      g === "!" && (v = this.__patterns.comment.read(), v ? (y = h.get_directives(v), y && y.ignore === "start" && (v += h.readIgnored(this._input))) : v = this.__patterns.cdata.read()), v && (O = this._create_token(_.COMMENT, v), O.directives = y);
    }
    return O;
  }, u.prototype._read_processing = function(l) {
    var O = null, v = null, y = null;
    if (l === "<") {
      var g = this._input.peek(1);
      (g === "!" || g === "?") && (v = this.__patterns.conditional_comment.read(), v = v || this.__patterns.processing.read()), v && (O = this._create_token(_.COMMENT, v), O.directives = y);
    }
    return O;
  }, u.prototype._read_open = function(l, O) {
    var v = null, y = null;
    return (!O || O.type === _.CONTROL_FLOW_OPEN) && l === "<" && (v = this._input.next(), this._input.peek() === "/" && (v += this._input.next()), v += this.__patterns.element_name.read(), y = this._create_token(_.TAG_OPEN, v)), y;
  }, u.prototype._read_open_handlebars = function(l, O) {
    var v = null, y = null;
    return (!O || O.type === _.CONTROL_FLOW_OPEN) && this._options.indent_handlebars && l === "{" && this._input.peek(1) === "{" && (this._input.peek(2) === "!" ? (v = this.__patterns.handlebars_comment.read(), v = v || this.__patterns.handlebars.read(), y = this._create_token(_.COMMENT, v)) : (v = this.__patterns.handlebars_open.read(), y = this._create_token(_.TAG_OPEN, v))), y;
  }, u.prototype._read_control_flows = function(l, O) {
    var v = "", y = null;
    if (!this._options.templating.includes("angular") || !this._options.indent_handlebars)
      return y;
    if (l === "@") {
      if (v = this.__patterns.angular_control_flow_start.read(), v === "")
        return y;
      for (var g = v.endsWith("(") ? 1 : 0, C = 0; !(v.endsWith("{") && g === C); ) {
        var A = this._input.next();
        if (A === null)
          break;
        A === "(" ? g++ : A === ")" && C++, v += A;
      }
      y = this._create_token(_.CONTROL_FLOW_OPEN, v);
    } else l === "}" && O && O.type === _.CONTROL_FLOW_OPEN && (v = this._input.next(), y = this._create_token(_.CONTROL_FLOW_CLOSE, v));
    return y;
  }, u.prototype._read_close = function(l, O) {
    var v = null, y = null;
    return O && O.type === _.TAG_OPEN && (O.text[0] === "<" && (l === ">" || l === "/" && this._input.peek(1) === ">") ? (v = this._input.next(), l === "/" && (v += this._input.next()), y = this._create_token(_.TAG_CLOSE, v)) : O.text[0] === "{" && l === "}" && this._input.peek(1) === "}" && (this._input.next(), this._input.next(), y = this._create_token(_.TAG_CLOSE, "}}"))), y;
  }, u.prototype._read_attribute = function(l, O, v) {
    var y = null, g = "";
    if (v && v.text[0] === "<")
      if (l === "=")
        y = this._create_token(_.EQUALS, this._input.next());
      else if (l === '"' || l === "'") {
        var C = this._input.next();
        l === '"' ? C += this.__patterns.double_quote.read() : C += this.__patterns.single_quote.read(), y = this._create_token(_.VALUE, C);
      } else
        g = this.__patterns.attribute.read(), g && (O.type === _.EQUALS ? y = this._create_token(_.VALUE, g) : y = this._create_token(_.ATTRIBUTE, g));
    return y;
  }, u.prototype._is_content_unformatted = function(l) {
    return this._options.void_elements.indexOf(l) === -1 && (this._options.content_unformatted.indexOf(l) !== -1 || this._options.unformatted.indexOf(l) !== -1);
  }, u.prototype._read_raw_content = function(l, O, v) {
    var y = "";
    if (v && v.text[0] === "{")
      y = this.__patterns.handlebars_raw_close.read();
    else if (O.type === _.TAG_CLOSE && O.opened.text[0] === "<" && O.text[0] !== "/") {
      var g = O.opened.text.substr(1).toLowerCase();
      if (g === "script" || g === "style") {
        var C = this._read_comment_or_cdata(l);
        if (C)
          return C.type = _.TEXT, C;
        y = this._input.readUntil(new RegExp("</" + g + "[\\n\\r\\t ]*?>", "ig"));
      } else this._is_content_unformatted(g) && (y = this._input.readUntil(new RegExp("</" + g + "[\\n\\r\\t ]*?>", "ig")));
    }
    return y ? this._create_token(_.TEXT, y) : null;
  }, u.prototype._read_content_word = function(l, O) {
    var v = "";
    if (this._options.unformatted_content_delimiter && l === this._options.unformatted_content_delimiter[0] && (v = this.__patterns.unformatted_content_delimiter.read()), v || (v = O && O.type === _.CONTROL_FLOW_OPEN ? this.__patterns.word_control_flow_close_excluded.read() : this.__patterns.word.read()), v)
      return this._create_token(_.TEXT, v);
  }, wt.Tokenizer = u, wt.TOKEN = _, wt;
}
var ii;
function Ji() {
  if (ii) return Yt;
  ii = 1;
  var a = xi().Options, t = me().Output, i = ei().Tokenizer, e = ei().TOKEN, r = /\r\n|[\r\n]/, _ = /\r\n|[\r\n]/g, h = function(o, n) {
    this.indent_level = 0, this.alignment_size = 0, this.max_preserve_newlines = o.max_preserve_newlines, this.preserve_newlines = o.preserve_newlines, this._output = new t(o, n);
  };
  h.prototype.current_line_has_match = function(o) {
    return this._output.current_line.has_match(o);
  }, h.prototype.set_space_before_token = function(o, n) {
    this._output.space_before_token = o, this._output.non_breaking_space = n;
  }, h.prototype.set_wrap_point = function() {
    this._output.set_indent(this.indent_level, this.alignment_size), this._output.set_wrap_point();
  }, h.prototype.add_raw_token = function(o) {
    this._output.add_raw_token(o);
  }, h.prototype.print_preserved_newlines = function(o) {
    var n = 0;
    o.type !== e.TEXT && o.previous.type !== e.TEXT && (n = o.newlines ? 1 : 0), this.preserve_newlines && (n = o.newlines < this.max_preserve_newlines + 1 ? o.newlines : this.max_preserve_newlines + 1);
    for (var p = 0; p < n; p++)
      this.print_newline(p > 0);
    return n !== 0;
  }, h.prototype.traverse_whitespace = function(o) {
    return o.whitespace_before || o.newlines ? (this.print_preserved_newlines(o) || (this._output.space_before_token = !0), !0) : !1;
  }, h.prototype.previous_token_wrapped = function() {
    return this._output.previous_token_wrapped;
  }, h.prototype.print_newline = function(o) {
    this._output.add_new_line(o);
  }, h.prototype.print_token = function(o) {
    o.text && (this._output.set_indent(this.indent_level, this.alignment_size), this._output.add_token(o.text));
  }, h.prototype.indent = function() {
    this.indent_level++;
  }, h.prototype.deindent = function() {
    this.indent_level > 0 && (this.indent_level--, this._output.set_indent(this.indent_level, this.alignment_size));
  }, h.prototype.get_full_indent = function(o) {
    return o = this.indent_level + (o || 0), o < 1 ? "" : this._output.get_indent_string(o);
  };
  var u = function(o) {
    for (var n = null, p = o.next; p.type !== e.EOF && o.closed !== p; ) {
      if (p.type === e.ATTRIBUTE && p.text === "type") {
        p.next && p.next.type === e.EQUALS && p.next.next && p.next.next.type === e.VALUE && (n = p.next.next.text);
        break;
      }
      p = p.next;
    }
    return n;
  }, l = function(o, n) {
    var p = null, T = null;
    return n.closed ? (o === "script" ? p = "text/javascript" : o === "style" && (p = "text/css"), p = u(n) || p, p.search("text/css") > -1 ? T = "css" : p.search(/module|((text|application|dojo)\/(x-)?(javascript|ecmascript|jscript|livescript|(ld\+)?json|method|aspect))/) > -1 ? T = "javascript" : p.search(/(text|application|dojo)\/(x-)?(html)/) > -1 ? T = "html" : p.search(/test\/null/) > -1 && (T = "null"), T) : null;
  };
  function O(o, n) {
    return n.indexOf(o) !== -1;
  }
  function v(o, n, p) {
    this.parent = o || null, this.tag = n ? n.tag_name : "", this.indent_level = p || 0, this.parser_token = n || null;
  }
  function y(o) {
    this._printer = o, this._current_frame = null;
  }
  y.prototype.get_parser_token = function() {
    return this._current_frame ? this._current_frame.parser_token : null;
  }, y.prototype.record_tag = function(o) {
    var n = new v(this._current_frame, o, this._printer.indent_level);
    this._current_frame = n;
  }, y.prototype._try_pop_frame = function(o) {
    var n = null;
    return o && (n = o.parser_token, this._printer.indent_level = o.indent_level, this._current_frame = o.parent), n;
  }, y.prototype._get_frame = function(o, n) {
    for (var p = this._current_frame; p && o.indexOf(p.tag) === -1; ) {
      if (n && n.indexOf(p.tag) !== -1) {
        p = null;
        break;
      }
      p = p.parent;
    }
    return p;
  }, y.prototype.try_pop = function(o, n) {
    var p = this._get_frame([o], n);
    return this._try_pop_frame(p);
  }, y.prototype.indent_to_tag = function(o) {
    var n = this._get_frame(o);
    n && (this._printer.indent_level = n.indent_level);
  };
  function g(o, n, p, T) {
    this._source_text = o || "", n = n || {}, this._js_beautify = p, this._css_beautify = T, this._tag_stack = null;
    var E = new a(n, "html");
    this._options = E, this._is_wrap_attributes_force = this._options.wrap_attributes.substr(0, 5) === "force", this._is_wrap_attributes_force_expand_multiline = this._options.wrap_attributes === "force-expand-multiline", this._is_wrap_attributes_force_aligned = this._options.wrap_attributes === "force-aligned", this._is_wrap_attributes_aligned_multiple = this._options.wrap_attributes === "aligned-multiple", this._is_wrap_attributes_preserve = this._options.wrap_attributes.substr(0, 8) === "preserve", this._is_wrap_attributes_preserve_aligned = this._options.wrap_attributes === "preserve-aligned";
  }
  g.prototype.beautify = function() {
    if (this._options.disabled)
      return this._source_text;
    var o = this._source_text, n = this._options.eol;
    this._options.eol === "auto" && (n = `
`, o && r.test(o) && (n = o.match(r)[0])), o = o.replace(_, `
`);
    var p = o.match(/^[\t ]*/)[0], T = {
      text: "",
      type: ""
    }, E = new C(), k = new h(this._options, p), c = new i(o, this._options).tokenize();
    this._tag_stack = new y(k);
    for (var b = null, f = c.next(); f.type !== e.EOF; )
      f.type === e.TAG_OPEN || f.type === e.COMMENT ? (b = this._handle_tag_open(k, f, E, T, c), E = b) : f.type === e.ATTRIBUTE || f.type === e.EQUALS || f.type === e.VALUE || f.type === e.TEXT && !E.tag_complete ? b = this._handle_inside_tag(k, f, E, T) : f.type === e.TAG_CLOSE ? b = this._handle_tag_close(k, f, E) : f.type === e.TEXT ? b = this._handle_text(k, f, E) : f.type === e.CONTROL_FLOW_OPEN ? b = this._handle_control_flow_open(k, f) : f.type === e.CONTROL_FLOW_CLOSE ? b = this._handle_control_flow_close(k, f) : k.add_raw_token(f), T = b, f = c.next();
    var S = k._output.get_code(n);
    return S;
  }, g.prototype._handle_control_flow_open = function(o, n) {
    var p = {
      text: n.text,
      type: n.type
    };
    return o.set_space_before_token(n.newlines || n.whitespace_before !== "", !0), n.newlines ? o.print_preserved_newlines(n) : o.set_space_before_token(n.newlines || n.whitespace_before !== "", !0), o.print_token(n), o.indent(), p;
  }, g.prototype._handle_control_flow_close = function(o, n) {
    var p = {
      text: n.text,
      type: n.type
    };
    return o.deindent(), n.newlines ? o.print_preserved_newlines(n) : o.set_space_before_token(n.newlines || n.whitespace_before !== "", !0), o.print_token(n), p;
  }, g.prototype._handle_tag_close = function(o, n, p) {
    var T = {
      text: n.text,
      type: n.type
    };
    return o.alignment_size = 0, p.tag_complete = !0, o.set_space_before_token(n.newlines || n.whitespace_before !== "", !0), p.is_unformatted ? o.add_raw_token(n) : (p.tag_start_char === "<" && (o.set_space_before_token(n.text[0] === "/", !0), this._is_wrap_attributes_force_expand_multiline && p.has_wrapped_attrs && o.print_newline(!1)), o.print_token(n)), p.indent_content && !(p.is_unformatted || p.is_content_unformatted) && (o.indent(), p.indent_content = !1), !p.is_inline_element && !(p.is_unformatted || p.is_content_unformatted) && o.set_wrap_point(), T;
  }, g.prototype._handle_inside_tag = function(o, n, p, T) {
    var E = p.has_wrapped_attrs, k = {
      text: n.text,
      type: n.type
    };
    return o.set_space_before_token(n.newlines || n.whitespace_before !== "", !0), p.is_unformatted ? o.add_raw_token(n) : p.tag_start_char === "{" && n.type === e.TEXT ? o.print_preserved_newlines(n) ? (n.newlines = 0, o.add_raw_token(n)) : o.print_token(n) : (n.type === e.ATTRIBUTE ? o.set_space_before_token(!0) : (n.type === e.EQUALS || n.type === e.VALUE && n.previous.type === e.EQUALS) && o.set_space_before_token(!1), n.type === e.ATTRIBUTE && p.tag_start_char === "<" && ((this._is_wrap_attributes_preserve || this._is_wrap_attributes_preserve_aligned) && (o.traverse_whitespace(n), E = E || n.newlines !== 0), this._is_wrap_attributes_force && p.attr_count >= this._options.wrap_attributes_min_attrs && (T.type !== e.TAG_OPEN || // ie. second attribute and beyond
    this._is_wrap_attributes_force_expand_multiline) && (o.print_newline(!1), E = !0)), o.print_token(n), E = E || o.previous_token_wrapped(), p.has_wrapped_attrs = E), k;
  }, g.prototype._handle_text = function(o, n, p) {
    var T = {
      text: n.text,
      type: "TK_CONTENT"
    };
    return p.custom_beautifier_name ? this._print_custom_beatifier_text(o, n, p) : p.is_unformatted || p.is_content_unformatted ? o.add_raw_token(n) : (o.traverse_whitespace(n), o.print_token(n)), T;
  }, g.prototype._print_custom_beatifier_text = function(o, n, p) {
    var T = this;
    if (n.text !== "") {
      var E = n.text, k, c = 1, b = "", f = "";
      p.custom_beautifier_name === "javascript" && typeof this._js_beautify == "function" ? k = this._js_beautify : p.custom_beautifier_name === "css" && typeof this._css_beautify == "function" ? k = this._css_beautify : p.custom_beautifier_name === "html" && (k = function(N, W) {
        var M = new g(N, W, T._js_beautify, T._css_beautify);
        return M.beautify();
      }), this._options.indent_scripts === "keep" ? c = 0 : this._options.indent_scripts === "separate" && (c = -o.indent_level);
      var S = o.get_full_indent(c);
      if (E = E.replace(/\n[ \t]*$/, ""), p.custom_beautifier_name !== "html" && E[0] === "<" && E.match(/^(<!--|<!\[CDATA\[)/)) {
        var s = /^(<!--[^\n]*|<!\[CDATA\[)(\n?)([ \t\n]*)([\s\S]*)(-->|]]>)$/.exec(E);
        if (!s) {
          o.add_raw_token(n);
          return;
        }
        b = S + s[1] + `
`, E = s[4], s[5] && (f = S + s[5]), E = E.replace(/\n[ \t]*$/, ""), (s[2] || s[3].indexOf(`
`) !== -1) && (s = s[3].match(/[ \t]+$/), s && (n.whitespace_before = s[0]));
      }
      if (E)
        if (k) {
          var d = function() {
            this.eol = `
`;
          };
          d.prototype = this._options.raw_options;
          var m = new d();
          E = k(S + E, m);
        } else {
          var R = n.whitespace_before;
          R && (E = E.replace(new RegExp(`
(` + R + ")?", "g"), `
`)), E = S + E.replace(/\n/g, `
` + S);
        }
      b && (E ? E = b + E + `
` + f : E = b + f), o.print_newline(!1), E && (n.text = E, n.whitespace_before = "", n.newlines = 0, o.add_raw_token(n), o.print_newline(!0));
    }
  }, g.prototype._handle_tag_open = function(o, n, p, T, E) {
    var k = this._get_tag_open_token(n);
    if ((p.is_unformatted || p.is_content_unformatted) && !p.is_empty_element && n.type === e.TAG_OPEN && !k.is_start_tag ? (o.add_raw_token(n), k.start_tag_token = this._tag_stack.try_pop(k.tag_name)) : (o.traverse_whitespace(n), this._set_tag_position(o, n, k, p, T), k.is_inline_element || o.set_wrap_point(), o.print_token(n)), k.is_start_tag && this._is_wrap_attributes_force) {
      var c = 0, b;
      do
        b = E.peek(c), b.type === e.ATTRIBUTE && (k.attr_count += 1), c += 1;
      while (b.type !== e.EOF && b.type !== e.TAG_CLOSE);
    }
    return (this._is_wrap_attributes_force_aligned || this._is_wrap_attributes_aligned_multiple || this._is_wrap_attributes_preserve_aligned) && (k.alignment_size = n.text.length + 1), !k.tag_complete && !k.is_unformatted && (o.alignment_size = k.alignment_size), k;
  };
  var C = function(o, n) {
    if (this.parent = o || null, this.text = "", this.type = "TK_TAG_OPEN", this.tag_name = "", this.is_inline_element = !1, this.is_unformatted = !1, this.is_content_unformatted = !1, this.is_empty_element = !1, this.is_start_tag = !1, this.is_end_tag = !1, this.indent_content = !1, this.multiline_content = !1, this.custom_beautifier_name = null, this.start_tag_token = null, this.attr_count = 0, this.has_wrapped_attrs = !1, this.alignment_size = 0, this.tag_complete = !1, this.tag_start_char = "", this.tag_check = "", !n)
      this.tag_complete = !0;
    else {
      var p;
      this.tag_start_char = n.text[0], this.text = n.text, this.tag_start_char === "<" ? (p = n.text.match(/^<([^\s>]*)/), this.tag_check = p ? p[1] : "") : (p = n.text.match(/^{{~?(?:[\^]|#\*?)?([^\s}]+)/), this.tag_check = p ? p[1] : "", (n.text.startsWith("{{#>") || n.text.startsWith("{{~#>")) && this.tag_check[0] === ">" && (this.tag_check === ">" && n.next !== null ? this.tag_check = n.next.text.split(" ")[0] : this.tag_check = n.text.split(">")[1])), this.tag_check = this.tag_check.toLowerCase(), n.type === e.COMMENT && (this.tag_complete = !0), this.is_start_tag = this.tag_check.charAt(0) !== "/", this.tag_name = this.is_start_tag ? this.tag_check : this.tag_check.substr(1), this.is_end_tag = !this.is_start_tag || n.closed && n.closed.text === "/>";
      var T = 2;
      this.tag_start_char === "{" && this.text.length >= 3 && this.text.charAt(2) === "~" && (T = 3), this.is_end_tag = this.is_end_tag || this.tag_start_char === "{" && (this.text.length < 3 || /[^#\^]/.test(this.text.charAt(T)));
    }
  };
  g.prototype._get_tag_open_token = function(o) {
    var n = new C(this._tag_stack.get_parser_token(), o);
    return n.alignment_size = this._options.wrap_attributes_indent_size, n.is_end_tag = n.is_end_tag || O(n.tag_check, this._options.void_elements), n.is_empty_element = n.tag_complete || n.is_start_tag && n.is_end_tag, n.is_unformatted = !n.tag_complete && O(n.tag_check, this._options.unformatted), n.is_content_unformatted = !n.is_empty_element && O(n.tag_check, this._options.content_unformatted), n.is_inline_element = O(n.tag_name, this._options.inline) || this._options.inline_custom_elements && n.tag_name.includes("-") || n.tag_start_char === "{", n;
  }, g.prototype._set_tag_position = function(o, n, p, T, E) {
    if (p.is_empty_element || (p.is_end_tag ? p.start_tag_token = this._tag_stack.try_pop(p.tag_name) : (this._do_optional_end_element(p) && (p.is_inline_element || o.print_newline(!1)), this._tag_stack.record_tag(p), (p.tag_name === "script" || p.tag_name === "style") && !(p.is_unformatted || p.is_content_unformatted) && (p.custom_beautifier_name = l(p.tag_check, n)))), O(p.tag_check, this._options.extra_liners) && (o.print_newline(!1), o._output.just_added_blankline() || o.print_newline(!0)), p.is_empty_element) {
      if (p.tag_start_char === "{" && p.tag_check === "else") {
        this._tag_stack.indent_to_tag(["if", "unless", "each"]), p.indent_content = !0;
        var k = o.current_line_has_match(/{{#if/);
        k || o.print_newline(!1);
      }
      p.tag_name === "!--" && E.type === e.TAG_CLOSE && T.is_end_tag && p.text.indexOf(`
`) === -1 || (p.is_inline_element || p.is_unformatted || o.print_newline(!1), this._calcluate_parent_multiline(o, p));
    } else if (p.is_end_tag) {
      var c = !1;
      c = p.start_tag_token && p.start_tag_token.multiline_content, c = c || !p.is_inline_element && !(T.is_inline_element || T.is_unformatted) && !(E.type === e.TAG_CLOSE && p.start_tag_token === T) && E.type !== "TK_CONTENT", (p.is_content_unformatted || p.is_unformatted) && (c = !1), c && o.print_newline(!1);
    } else
      p.indent_content = !p.custom_beautifier_name, p.tag_start_char === "<" && (p.tag_name === "html" ? p.indent_content = this._options.indent_inner_html : p.tag_name === "head" ? p.indent_content = this._options.indent_head_inner_html : p.tag_name === "body" && (p.indent_content = this._options.indent_body_inner_html)), !(p.is_inline_element || p.is_unformatted) && (E.type !== "TK_CONTENT" || p.is_content_unformatted) && o.print_newline(!1), this._calcluate_parent_multiline(o, p);
  }, g.prototype._calcluate_parent_multiline = function(o, n) {
    n.parent && o._output.just_added_newline() && !((n.is_inline_element || n.is_unformatted) && n.parent.is_inline_element) && (n.parent.multiline_content = !0);
  };
  var A = ["address", "article", "aside", "blockquote", "details", "div", "dl", "fieldset", "figcaption", "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6", "header", "hr", "main", "menu", "nav", "ol", "p", "pre", "section", "table", "ul"], x = ["a", "audio", "del", "ins", "map", "noscript", "video"];
  return g.prototype._do_optional_end_element = function(o) {
    var n = null;
    if (!(o.is_empty_element || !o.is_start_tag || !o.parent)) {
      if (o.tag_name === "body")
        n = n || this._tag_stack.try_pop("head");
      else if (o.tag_name === "li")
        n = n || this._tag_stack.try_pop("li", ["ol", "ul", "menu"]);
      else if (o.tag_name === "dd" || o.tag_name === "dt")
        n = n || this._tag_stack.try_pop("dt", ["dl"]), n = n || this._tag_stack.try_pop("dd", ["dl"]);
      else if (o.parent.tag_name === "p" && A.indexOf(o.tag_name) !== -1) {
        var p = o.parent.parent;
        (!p || x.indexOf(p.tag_name) === -1) && (n = n || this._tag_stack.try_pop("p"));
      } else o.tag_name === "rp" || o.tag_name === "rt" ? (n = n || this._tag_stack.try_pop("rt", ["ruby", "rtc"]), n = n || this._tag_stack.try_pop("rp", ["ruby", "rtc"])) : o.tag_name === "optgroup" ? n = n || this._tag_stack.try_pop("optgroup", ["select"]) : o.tag_name === "option" ? n = n || this._tag_stack.try_pop("option", ["select", "datalist", "optgroup"]) : o.tag_name === "colgroup" ? n = n || this._tag_stack.try_pop("caption", ["table"]) : o.tag_name === "thead" ? (n = n || this._tag_stack.try_pop("caption", ["table"]), n = n || this._tag_stack.try_pop("colgroup", ["table"])) : o.tag_name === "tbody" || o.tag_name === "tfoot" ? (n = n || this._tag_stack.try_pop("caption", ["table"]), n = n || this._tag_stack.try_pop("colgroup", ["table"]), n = n || this._tag_stack.try_pop("thead", ["table"]), n = n || this._tag_stack.try_pop("tbody", ["table"])) : o.tag_name === "tr" ? (n = n || this._tag_stack.try_pop("caption", ["table"]), n = n || this._tag_stack.try_pop("colgroup", ["table"]), n = n || this._tag_stack.try_pop("tr", ["table", "thead", "tbody", "tfoot"])) : (o.tag_name === "th" || o.tag_name === "td") && (n = n || this._tag_stack.try_pop("td", ["table", "thead", "tbody", "tfoot", "tr"]), n = n || this._tag_stack.try_pop("th", ["table", "thead", "tbody", "tfoot", "tr"]));
      return o.parent = this._tag_stack.get_parser_token(), n;
    }
  }, Yt.Beautifier = g, Yt;
}
var si;
function Zi() {
  if (si) return yt.exports;
  si = 1;
  var a = Ji().Beautifier, t = xi().Options;
  function i(e, r, _, h) {
    var u = new a(e, r, _, h);
    return u.beautify();
  }
  return yt.exports = i, yt.exports.defaultOptions = function() {
    return new t();
  }, yt.exports;
}
var ni;
function Yi() {
  if (ni) return it;
  ni = 1;
  var a = Xi(), t = Hi(), i = Zi();
  function e(r, _, h, u) {
    return h = h || a, u = u || t, i(r, _, h, u);
  }
  return e.defaultOptions = i.defaultOptions, it.js = a, it.css = t, it.html = e, it;
}
var ri;
function ts() {
  return ri || (ri = 1, function(a) {
    function t(i, e, r) {
      var _ = function(h, u) {
        return i.js_beautify(h, u);
      };
      return _.js = i.js_beautify, _.css = e.css_beautify, _.html = r.html_beautify, _.js_beautify = i.js_beautify, _.css_beautify = e.css_beautify, _.html_beautify = r.html_beautify, _;
    }
    (function(i) {
      var e = Yi();
      e.js_beautify = e.js, e.css_beautify = e.css, e.html_beautify = e.html, i.exports = t(e, e, e);
    })(a);
  }(Bt)), Bt.exports;
}
var es = ts();
class is {
  constructor(t) {
    this.partsLength = t, this.keys = {};
  }
  getKey(...t) {
    if (t.length !== this.partsLength)
      return null;
    let i = this.keys;
    for (let e = 0; e < t.length - 1; e += 1)
      i[t[e]] || (i[t[e]] = {}), i = i[t[e]];
    return i[t[t.length - 1]] || (i[t[t.length - 1]] = Symbol("")), i[t[t.length - 1]];
  }
}
function ss(a) {
  return es.html(a, {
    wrap_attributes: "force-expand-multiline"
  });
}
class ns {
  constructor(t) {
    this.store = new Oe(), this.components = /* @__PURE__ */ new Map(), this.mountedElements = /* @__PURE__ */ new Map(), this.subscribers = /* @__PURE__ */ new Map(), this.componentKeys = new is(2), this.dashboardState = new Oe(), this.provider = new Ei(), this.connector = new Fi(
      t ?? document.createElement("div"),
      this.store
    ), this.dashboardState.addSourceProvider("dashboardState", this.provider), this.dashboardState.subscribeAll(
      "dashboardState",
      (i, e) => {
        this.publish("storeValueChange", { value: i, key: e });
      },
      !1
    );
  }
  addSourceProvider(t, i) {
    this.store.addSourceProvider(t, i);
  }
  setDefaultSourceProvider(t) {
    this.store.setDefaultSourceProvider(t);
  }
  setStoreValue(t, i) {
    this.provider.updateSource(t, i);
  }
  getStoreValue(t, i = void 0) {
    return this.dashboardState.getSourceValue("dashboardState", t) ?? i;
  }
  getConnector() {
    return this.connector;
  }
  getRootElement() {
    return this.connector.getRootElement();
  }
  getStore() {
    return this.connector.getStore();
  }
  addComponent(t) {
    const i = this.componentKeys.getKey(t.type, t.id);
    i && this.components.set(i, t);
  }
  hasComponent(t, i) {
    const e = this.componentKeys.getKey(t, i);
    return e ? this.components.has(e) : !1;
  }
  getComponentIdsOfType(t) {
    return [...this.components.values()].filter(({ type: i }) => i === t).map(({ id: i }) => i);
  }
  create(t, i, e = {}) {
    const r = document.createElement("div"), _ = this.componentKeys.getKey(t, i);
    if (!_)
      return;
    const h = this.components.get(_);
    if (h) {
      const u = h.mount({
        dashboard: this,
        element: r,
        config: e
      });
      return this.mountedElements.set(r, {
        componentId: i,
        componentType: t,
        config: e,
        unmount: u
      }), r;
    }
  }
  /**
   * Unmounts the mounted component from a particular element
   * @param {HTMLElement} element
   */
  unmount(t) {
    const i = this.mountedElements.get(t);
    if (i) {
      const { config: e, unmount: r } = i;
      r && r({
        dashboard: this,
        element: t,
        config: e
      }), this.mountedElements.delete(t);
    }
  }
  /**
   *
   * @returns {() => void} A function to unsubscribe
   */
  subscribe(t, i) {
    this.subscribers.has(t) || this.subscribers.set(t, /* @__PURE__ */ new Map());
    const e = Symbol(""), r = this.subscribers.get(t);
    return r == null || r.set(e, i), () => {
      r == null || r.delete(e);
    };
  }
  /**
   *
   * @param {string} subscriberId
   * @param {unknown} context
   */
  publish(t, i = {}) {
    var e;
    this.subscribers.has(t) && ((e = this.subscribers.get(t)) == null || e.forEach((r) => {
      r(i);
    }));
  }
  addElements(t, i = "default") {
    this.connector.addElementConfigs(t, i);
  }
  getHtml() {
    return this.getElementHtml(this.connector.getRootElement(), !0);
  }
  setHtml(t) {
    this.connector.getRootElement().querySelectorAll("dashboard-tab").forEach((e) => {
      e.remove();
    });
    const i = document.createElement("div");
    i.innerHTML = t, i.querySelectorAll("dashboard-tab").forEach((e) => {
      e.setAttribute("slot", "tab"), this.connector.getRootElement().append(e);
    });
  }
  getElementDisplayName(t) {
    const i = this.connector.getMatchingElementConfig(t), e = t.tagName.toLowerCase();
    return i != null && i.dashboard ? typeof i.dashboard.displayName == "string" ? i.dashboard.displayName : typeof i.dashboard.displayName == "function" ? i.dashboard.displayName(t) : e : e;
  }
  getSelectorDisplayName(t) {
    const i = this.connector.getElementConfig(t);
    return i != null && i.dashboard ? typeof i.dashboard.displayName == "string" ? i.dashboard.displayName : typeof i.dashboard.displayName == "function" ? i.dashboard.displayName() : t : t;
  }
  getElementHtml(t, i = !1) {
    const e = t.cloneNode(!0);
    return this.setClonedElementPropAttributes(t, e), ss(
      i ? e.innerHTML : e.outerHTML
    );
  }
  setClonedElementPropAttributes(t, i) {
    const e = this.connector.getElementWebbit(t);
    if (e) {
      const { properties: r } = e.getConfig();
      Object.entries(r).forEach(
        ([_, { attribute: h, defaultValue: u }]) => {
          const l = e.getPropertyHandler(_);
          if (l != null && l.isAttribute() && h) {
            const O = l.getStoredAttribute();
            O === null || l.getStoredValue() === u ? i.removeAttribute(h) : i.setAttribute(h, O);
          }
        }
      );
    }
    for (let r = 0; r < t.children.length; r += 1)
      this.setClonedElementPropAttributes(
        t.children[r],
        i.children[r]
      );
  }
}
function rs(a, t) {
  return a.slots.map(({ name: i, allowedChildren: e }) => {
    const r = t.getElementConfigSelectors().filter(
      (_) => {
        var h, u;
        return e instanceof Array ? e.includes(_) : (u = (h = t.getElementConfig(_)) == null ? void 0 : h.dashboard) == null ? void 0 : u.topLevel;
      }
    );
    return {
      slot: i,
      allowedChildren: r
    };
  });
}
function ai(a, t) {
  if (!a)
    return [];
  const i = t.getMatchingElementConfig(a);
  return i ? rs(i, t) : [];
}
function as(a) {
  const t = document.createElement("div");
  return t.setAttribute("slot", "layer"), t.setAttribute("layer-id", a), t.style.width = "100%", t.style.height = "100%", t.style.position = "absolute", t.style.top = "0", t.style.boxSizing = "border-box", t.style.pointerEvents = "none", t;
}
function _s() {
  const a = document.createElement("style");
  return a.appendChild(document.createTextNode("")), document.head.appendChild(a), a.sheet;
}
function os(a, t) {
  a.insertRule(t);
}
class ls extends ns {
  constructor(t) {
    super(t), this.layers = {}, this.themeSheets = {}, this.subscribe("storeValueChange", (i) => {
      const { key: e, value: r } = i;
      e === "drawerOpened" ? this.publish("drawerToggle", { opened: r }) : e === "selectedElement" ? this.publish("elementSelect", { element: this.getSelectedElement() }) : e === "previewedElement" && this.publish("elementPreview", { element: this.getPreviewedElement() });
    }), this.getRootElement().setAttribute("data-theme", this.getTheme());
  }
  openDrawer() {
    this.setStoreValue("drawerOpened", !0);
  }
  closeDrawer() {
    this.setStoreValue("drawerOpened", !1);
  }
  isDrawerOpened() {
    return this.getStoreValue("drawerOpened", !1);
  }
  setSelectedElement(t) {
    this.getSelectedElement() !== t && (this.setStoreValue("selectedElement", t), this.setStoreValue(
      "allowedChildren",
      ai(t, this.getConnector())
    ));
  }
  setPreviewedElement(t) {
    this.getPreviewedElement() !== t && this.setStoreValue("previewedElement", t);
  }
  addElements(t, i = "default") {
    super.addElements(t, i);
    const e = this.getSelectedElement();
    e && this.setStoreValue(
      "allowedChildren",
      ai(e, this.getConnector())
    ), this.publish("elementsAdd");
  }
  setHtml(t) {
    super.setHtml(t);
    const i = this.getConnector().getRootElement().querySelector("dashboard-tab");
    i && this.setSelectedElement(i), this.publish("htmlSet");
  }
  resetHtml() {
    this.setHtml(`
      <dashboard-tab tab-name="TeleOperated" slot="tab" selected></dashboard-tab>
      <dashboard-tab slot="tab" tab-name="Autonomous"></dashboard-tab>
    `);
  }
  reloadHtml() {
    this.setHtml(this.getHtml());
  }
  getSelectedElement() {
    return this.getStoreValue("selectedElement", null);
  }
  getPreviewedElement() {
    return this.getStoreValue("previewedElement", null);
  }
  getAllowedChildren() {
    return this.getStoreValue("allowedChildren", []);
  }
  /**
   *
   * @param {string} id
   * @param {string} elementName
   */
  addPropertyInput(t, i) {
    this.addComponent({
      type: "propertyInput",
      id: t,
      mount({ dashboard: e, element: r, config: _ }) {
        const h = document.createElement(i);
        h.element = _.element, h.propertyHandler = _.propertyHandler, h.propertyName = _.propertyName, h.dashboard = e, r.append(h);
        const u = (O) => {
          e.publish("propertyInputChange", O.detail);
        };
        h.addEventListener("change", u);
        const l = e.subscribe(
          "propertyInputChange",
          () => {
            h == null || h.requestUpdate();
          }
        );
        return () => {
          h.removeEventListener("change", u), l(), h.remove();
        };
      }
    });
  }
  addElementEditor(t, i) {
    this.addComponent({
      type: "elementEditor",
      id: t,
      mount({ dashboard: e, element: r }) {
        const _ = document.createElement(i);
        return _.dashboard = e, r.append(_), () => {
          _.remove();
        };
      }
    });
  }
  addTab(t, i) {
    const e = document.createElement("dashboard-tab");
    return i && (e.innerHTML = i), e.setAttribute("tab-name", t), e.setAttribute("slot", "tab"), this.getRootElement().append(e), e;
  }
  isElementEditable(t = this.getSelectedElement()) {
    return t ? t.tagName.toLowerCase() !== "dashboard-tab" || !t.hasAttribute("tutorial") : !0;
  }
  addLayer(t) {
    return this.layers[t] ? this.layers[t] : (this.layers[t] = as(t), this.publish("layerAdd", { layer: this.layers[t] }), this.layers[t]);
  }
  hasLayer(t) {
    return !!this.layers[t];
  }
  getLayer(t) {
    return this.layers[t];
  }
  getLayers() {
    return this.layers;
  }
  addThemeRules(t, i) {
    typeof this.themeSheets[t] > "u" && (this.themeSheets[t] = _s());
    const e = Object.entries(i).map(
      ([r, _]) => `${r}: ${_};`
    );
    os(
      this.themeSheets[t],
      `
      [data-theme="${t}"] {
        ${e.join(`
`)}
      }
    `
    ), this.publish("themeRulesAdd");
  }
  setTheme(t) {
    this.getRootElement().setAttribute("data-theme", t), localStorage.setItem("theme", t), this.setStoreValue("theme", t), this.publish("themeSet");
  }
  getTheme() {
    const t = localStorage.getItem("theme") ?? "dark";
    return this.getStoreValue("theme", t);
  }
  getThemes() {
    const t = ["light", ...Object.keys(this.themeSheets)];
    return [...new Set(t).values()];
  }
}
export {
  ln as Accelerometer,
  vs as Axis,
  xs as Bar,
  fn as BasicFmsInfo,
  gn as BooleanBox,
  Ir as CameraWrapper,
  Br as Canvas,
  Kr as CanvasCircle,
  Vr as CanvasGroup,
  Ur as CanvasLine,
  zr as CanvasMjpgStream,
  Fr as CanvasNGon,
  qr as CanvasRect,
  $r as CanvasText,
  Ln as ChartLegendPositions,
  aa as DashboardThemes,
  Ss as DifferentialDrivebase,
  br as Encoder,
  Ps as Field,
  $s as Field3dObject,
  Gs as Field3dPoseVisualizer,
  Xs as Field3dUrdf,
  Qs as FieldObject,
  Ls as FieldPath,
  Ds as FieldRobot,
  Ws as FieldWrapper,
  ls as FrcDashboard,
  fs as GamepadProvider,
  vn as Gauge,
  xn as Gyro,
  sn as Icon,
  Qr as Label,
  Dn as LineChart,
  Wn as LineChartAxis,
  Mn as LineChartData,
  jn as LineChartLegend,
  Ts as MecanumDrivebase,
  hr as Mechanism2d,
  pr as Mechanism2dWrapper,
  yr as NetworkAlerts,
  ds as Nt4Client,
  gs as Nt4Provider,
  bs as Nt4Topic,
  Sn as NumberBar,
  Hr as NumberLabel,
  Hn as NumberSlider,
  Fn as Pdp,
  Er as PidCommand,
  Or as PidController,
  ta as Preferences,
  ea as PreferencesWrapper,
  Sr as ProfiledPidController,
  or as Relay,
  Lr as RobotCommand,
  Dr as RobotSubsystem,
  Ar as ScoringGrid,
  Cn as SendableChooser,
  Rn as SendableChooserWrapper,
  ks as SwerveDrivebase,
  _n as ThreeAxisAccelerometer,
  ir as ToggleButton,
  rr as ToggleGroup,
  Yn as ToggleSwitch,
  Hs as Urdf,
  Gn as VoltageView,
  hn as accelerometerDashboardConfig,
  ys as axisDashboardConfig,
  Es as barDashboardConfig,
  Vs as baseUnit,
  cn as basicFmsInfoDashboardConfig,
  bn as booleanBoxDashboardConfig,
  Gr as cameraDashboardConfig,
  Us as convertUnit,
  _a as darkTheme,
  na as dashboardElementConfigs,
  Cs as differentialDrivebaseDashboardConfig,
  qn as elementName,
  mr as encoderDashboardConfig,
  Js as field3dConfigs,
  Ms as fieldConfigs,
  js as fieldDashboardConfig,
  Is as fieldPathDashboardConfig,
  Bs as fieldRobotDashboardConfig,
  yn as gaugeDashboardConfig,
  Zs as getField3dDashboardConfigs,
  fr as getMechanism2dLines,
  cr as getMechanism2dRoots,
  En as gyroDashboardConfig,
  nn as iconDashboardConfig,
  rn as iconset,
  Jr as labelDashboardConfig,
  In as lineChartAxisDashboardConfig,
  Bn as lineChartDashboardConfig,
  Kn as lineChartDashboardConfigs,
  Vn as lineChartDataDashboardConfig,
  Un as lineChartLegendDashboardConfig,
  Rs as mecanumDrivebaseDashboardConfig,
  dr as mechanism2dDashboardConfig,
  wr as networkAlertsDashboardConfig,
  Tn as numberBarDashboardConfig,
  Zr as numberLabelDashboardConfig,
  Jn as numberSliderDashboardConfig,
  Ys as objectConfigs,
  Tr as pidCommandDashboardConfig,
  kr as pidControllerDashboardConfig,
  ia as preferencesDashboardConfig,
  Cr as profiledPidControllerDashboardConfig,
  ur as relayDashboardConfig,
  Wr as robotCommandDashboardConfig,
  Mr as robotSubsystemDashboardConfig,
  Nr as scoringGridDashboardConfig,
  An as sendableChooserDashboardConfig,
  Nn as sendableChooserOldDashboardConfig,
  As as swerveDrivebaseDashboardConfig,
  on as threeAxisAccelerometerDashboardConfig,
  zs as toBaseUnitConversions,
  sr as toggleButtonDashboardConfig,
  ar as toggleGroupDashboardConfig,
  tr as toggleSwitchDashboardConfig,
  Fs as unitAliases,
  tn as urdfConfigs,
  Xn as voltageViewDashboardConfig
};
