// @frc-web-components/fwc/dist/index-BXxACewW.js
import { c as j, g as Z } from "./_commonjsHelpers-DaMA6jEr.js";
function N(i, o) {
  const u = [];
  for (let s = 0; s < i.length / o; s += 1)
    u.push(i.slice(o * s, o * (s + 1)));
  return u;
}
function z(i) {
  const o = new ArrayBuffer(8), u = new DataView(o);
  return i.forEach((s, T) => {
    u.setUint8(T, s);
  }), u.getFloat64(0);
}
function tt(i) {
  return N(i, 8).map(z);
}
function it(i, o = 3) {
  const u = i instanceof Uint8Array ? tt(i) : i;
  return N(u, o);
}
var k, C;
function et() {
  if (C) return k;
  C = 1;
  var i = "Expected a function", o = NaN, u = "[object Symbol]", s = /^\s+|\s+$/g, T = /^[-+]0x[0-9a-f]+$/i, W = /^0b[01]+$/i, B = /^0o[0-7]+$/i, M = parseInt, P = typeof j == "object" && j && j.Object === Object && j, R = typeof self == "object" && self && self.Object === Object && self, $ = P || R || Function("return this")(), D = Object.prototype, U = D.toString, q = Math.max, w = Math.min, x = function() {
    return $.Date.now();
  };
  function G(t, n, r) {
    var f, c, p, g, a, d, m = 0, F = !1, b = !1, E = !0;
    if (typeof t != "function")
      throw new TypeError(i);
    n = A(n) || 0, y(r) && (F = !!r.leading, b = "maxWait" in r, p = b ? q(A(r.maxWait) || 0, n) : p, E = "trailing" in r ? !!r.trailing : E);
    function I(e) {
      var l = f, h = c;
      return f = c = void 0, m = e, g = t.apply(h, l), g;
    }
    function J(e) {
      return m = e, a = setTimeout(v, n), F ? I(e) : g;
    }
    function K(e) {
      var l = e - d, h = e - m, _ = n - l;
      return b ? w(_, p - h) : _;
    }
    function L(e) {
      var l = e - d, h = e - m;
      return d === void 0 || l >= n || l < 0 || b && h >= p;
    }
    function v() {
      var e = x();
      if (L(e))
        return S(e);
      a = setTimeout(v, K(e));
    }
    function S(e) {
      return a = void 0, E && f ? I(e) : (f = c = void 0, g);
    }
    function Q() {
      a !== void 0 && clearTimeout(a), m = 0, f = d = c = a = void 0;
    }
    function Y() {
      return a === void 0 ? g : S(x());
    }
    function O() {
      var e = x(), l = L(e);
      if (f = arguments, c = this, d = e, l) {
        if (a === void 0)
          return J(d);
        if (b)
          return a = setTimeout(v, n), I(d);
      }
      return a === void 0 && (a = setTimeout(v, n)), g;
    }
    return O.cancel = Q, O.flush = Y, O;
  }
  function H(t, n, r) {
    var f = !0, c = !0;
    if (typeof t != "function")
      throw new TypeError(i);
    return y(r) && (f = "leading" in r ? !!r.leading : f, c = "trailing" in r ? !!r.trailing : c), G(t, n, {
      leading: f,
      maxWait: n,
      trailing: c
    });
  }
  function y(t) {
    var n = typeof t;
    return !!t && (n == "object" || n == "function");
  }
  function V(t) {
    return !!t && typeof t == "object";
  }
  function X(t) {
    return typeof t == "symbol" || V(t) && U.call(t) == u;
  }
  function A(t) {
    if (typeof t == "number")
      return t;
    if (X(t))
      return o;
    if (y(t)) {
      var n = typeof t.valueOf == "function" ? t.valueOf() : t;
      t = y(n) ? n + "" : n;
    }
    if (typeof t != "string")
      return t === 0 ? t : +t;
    t = t.replace(s, "");
    var r = W.test(t);
    return r || B.test(t) ? M(t.slice(2), r ? 2 : 8) : T.test(t) ? o : +t;
  }
  return k = H, k;
}
var nt = et();
const ot = /* @__PURE__ */ Z(nt);
export {
  it as g,
  ot as t
};
