// @frc-web-components/react/dist/accelerometer.es-1e0be328.js
import { M as e } from "./property-C4BPH-wS-7fb0a59c.js";
import { t as c } from "./number-bar.es-14bfc169.js";
var u = Object.defineProperty, r = (p, m, s, y) => {
  for (var o = void 0, i = p.length - 1, n; i >= 0; i--)
    (n = p[i]) && (o = n(m, s, o) || o);
  return o && u(m, s, o), o;
};
class t extends c {
  constructor() {
    super(...arguments), this.value = 0, this.min = -1, this.max = 1, this.center = 0, this.precision = 2, this.hideText = !1, this.numTickMarks = 3, this.unit = "g";
  }
}
r([
  e({ type: Number })
], t.prototype, "value");
r([
  e({ type: Number })
], t.prototype, "min");
r([
  e({ type: Number })
], t.prototype, "max");
r([
  e({ type: Number })
], t.prototype, "center");
r([
  e({ type: Number })
], t.prototype, "precision");
r([
  e({ type: Boolean, attribute: "hide-text" })
], t.prototype, "hideText");
r([
  e({ type: Number, attribute: "num-tick-marks" })
], t.prototype, "numTickMarks");
r([
  e({ type: String })
], t.prototype, "unit");
customElements.get("frc-accelerometer") || customElements.define("frc-accelerometer", t);
export {
  t as e
};
