// @frc-web-components/react/dist/index-BXxACewW-cecd5a05.js
import { o as j, l as _ } from "./_commonjsHelpers-DaMA6jEr-bedfce41.js";
function S(i, u) {
  const a = [];
  for (let l = 0; l < i.length / u; l += 1)
    a.push(i.slice(u * l, u * (l + 1)));
  return a;
}
function Z(i) {
  const u = new ArrayBuffer(8), a = new DataView(u);
  return i.forEach((l, w) => {
    a.setUint8(w, l);
  }), a.getFloat64(0);
}
function tt(i) {
  return S(i, 8).map(Z);
}
function it(i, u = 3) {
  const a = i instanceof Uint8Array ? tt(i) : i;
  return S(a, u);
}
var E, N;
function nt() {
  if (N)
    return E;
  N = 1;
  var i = "Expected a function", u = NaN, a = "[object Symbol]", l = /^\s+|\s+$/g, w = /^[-+]0x[0-9a-f]+$/i, U = /^0b[01]+$/i, B = /^0o[0-7]+$/i, I = parseInt, V = typeof j == "object" && j && j.Object === Object && j, k = typeof self == "object" && self && self.Object === Object && self, q = V || k || Function("return this")(), z = Object.prototype, C = z.toString, G = Math.max, H = Math.min, x = function() {
    return q.Date.now();
  };
  function J(t, e, r) {
    var f, c, d, p, o, v, y = 0, A = !1, m = !1, O = !0;
    if (typeof t != "function")
      throw new TypeError(i);
    e = W(e) || 0, b(r) && (A = !!r.leading, m = "maxWait" in r, d = m ? G(W(r.maxWait) || 0, e) : d, O = "trailing" in r ? !!r.trailing : O);
    function T(n) {
      var s = f, g = c;
      return f = c = void 0, y = n, p = t.apply(g, s), p;
    }
    function Q(n) {
      return y = n, o = setTimeout(h, e), A ? T(n) : p;
    }
    function R(n) {
      var s = n - v, g = n - y, M = e - s;
      return m ? H(M, d - g) : M;
    }
    function D(n) {
      var s = n - v, g = n - y;
      return v === void 0 || s >= e || s < 0 || m && g >= d;
    }
    function h() {
      var n = x();
      if (D(n))
        return F(n);
      o = setTimeout(h, R(n));
    }
    function F(n) {
      return o = void 0, O && f ? T(n) : (f = c = void 0, p);
    }
    function X() {
      o !== void 0 && clearTimeout(o), y = 0, f = v = c = o = void 0;
    }
    function Y() {
      return o === void 0 ? p : F(x());
    }
    function $() {
      var n = x(), s = D(n);
      if (f = arguments, c = this, v = n, s) {
        if (o === void 0)
          return Q(v);
        if (m)
          return o = setTimeout(h, e), T(v);
      }
      return o === void 0 && (o = setTimeout(h, e)), p;
    }
    return $.cancel = X, $.flush = Y, $;
  }
  function K(t, e, r) {
    var f = !0, c = !0;
    if (typeof t != "function")
      throw new TypeError(i);
    return b(r) && (f = "leading" in r ? !!r.leading : f, c = "trailing" in r ? !!r.trailing : c), J(t, e, {
      leading: f,
      maxWait: e,
      trailing: c
    });
  }
  function b(t) {
    var e = typeof t;
    return !!t && (e == "object" || e == "function");
  }
  function L(t) {
    return !!t && typeof t == "object";
  }
  function P(t) {
    return typeof t == "symbol" || L(t) && C.call(t) == a;
  }
  function W(t) {
    if (typeof t == "number")
      return t;
    if (P(t))
      return u;
    if (b(t)) {
      var e = typeof t.valueOf == "function" ? t.valueOf() : t;
      t = b(e) ? e + "" : e;
    }
    if (typeof t != "string")
      return t === 0 ? t : +t;
    t = t.replace(l, "");
    var r = U.test(t);
    return r || B.test(t) ? I(t.slice(2), r ? 2 : 8) : w.test(t) ? u : +t;
  }
  return E = K, E;
}
var et = nt();
const ot = /* @__PURE__ */ _(et);
export {
  it as i,
  ot as o
};
