// @webbitjs/webbit/dist/index.es.js
var Ce = (s, e, t) => {
  if (!e.has(s))
    throw TypeError("Cannot " + t);
};
var i = (s, e, t) => (Ce(s, e, "read from private field"), t ? t.call(s) : e.get(s)), h = (s, e, t) => {
  if (e.has(s))
    throw TypeError("Cannot add the same private member more than once");
  e instanceof WeakSet ? e.add(s) : e.set(s, t);
}, d = (s, e, t, r) => (Ce(s, e, "write to private field"), r ? r.call(s, t) : e.set(s, t), t);
var y = (s, e, t) => (Ce(s, e, "access private method"), t);
var tt = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, pe = { exports: {} };
pe.exports;
(function(s, e) {
  (function(t, r) {
    var n = {};
    t.PubSub ? (n = t.PubSub, console.warn("PubSub already loaded, using existing version")) : (t.PubSub = n, r(n)), s !== void 0 && s.exports && (e = s.exports = n), e.PubSub = n, s.exports = e = n;
  })(typeof window == "object" && window || tt, function(t) {
    var r = {}, n = -1, u = "*";
    function f(o) {
      var c;
      for (c in o)
        if (Object.prototype.hasOwnProperty.call(o, c))
          return !0;
      return !1;
    }
    function l(o) {
      return function() {
        throw o;
      };
    }
    function p(o, c, a) {
      try {
        o(c, a);
      } catch (b) {
        setTimeout(l(b), 0);
      }
    }
    function P(o, c, a) {
      o(c, a);
    }
    function S(o, c, a, b) {
      var g = r[c], v = b ? P : p, F;
      if (Object.prototype.hasOwnProperty.call(r, c))
        for (F in g)
          Object.prototype.hasOwnProperty.call(g, F) && v(g[F], o, a);
    }
    function w(o, c, a) {
      return function() {
        var g = String(o), v = g.lastIndexOf(".");
        for (S(o, o, c, a); v !== -1; )
          g = g.substr(0, v), v = g.lastIndexOf("."), S(o, g, c, a);
        S(o, u, c, a);
      };
    }
    function E(o) {
      var c = String(o), a = !!(Object.prototype.hasOwnProperty.call(r, c) && f(r[c]));
      return a;
    }
    function A(o) {
      for (var c = String(o), a = E(c) || E(u), b = c.lastIndexOf("."); !a && b !== -1; )
        c = c.substr(0, b), b = c.lastIndexOf("."), a = E(c);
      return a;
    }
    function O(o, c, a, b) {
      o = typeof o == "symbol" ? o.toString() : o;
      var g = w(o, c, b), v = A(o);
      return v ? (a === !0 ? g() : setTimeout(g, 0), !0) : !1;
    }
    t.publish = function(o, c) {
      return O(o, c, !1, t.immediateExceptions);
    }, t.publishSync = function(o, c) {
      return O(o, c, !0, t.immediateExceptions);
    }, t.subscribe = function(o, c) {
      if (typeof c != "function")
        return !1;
      o = typeof o == "symbol" ? o.toString() : o, Object.prototype.hasOwnProperty.call(r, o) || (r[o] = {});
      var a = "uid_" + String(++n);
      return r[o][a] = c, a;
    }, t.subscribeAll = function(o) {
      return t.subscribe(u, o);
    }, t.subscribeOnce = function(o, c) {
      var a = t.subscribe(o, function() {
        t.unsubscribe(a), c.apply(this, arguments);
      });
      return t;
    }, t.clearAllSubscriptions = function() {
      r = {};
    }, t.clearSubscriptions = function(c) {
      var a;
      for (a in r)
        Object.prototype.hasOwnProperty.call(r, a) && a.indexOf(c) === 0 && delete r[a];
    }, t.countSubscriptions = function(c) {
      var a, b, g = 0;
      for (a in r)
        if (Object.prototype.hasOwnProperty.call(r, a) && a.indexOf(c) === 0) {
          for (b in r[a])
            g++;
          break;
        }
      return g;
    }, t.getSubscriptions = function(c) {
      var a, b = [];
      for (a in r)
        Object.prototype.hasOwnProperty.call(r, a) && a.indexOf(c) === 0 && b.push(a);
      return b;
    }, t.unsubscribe = function(o) {
      var c = function(et) {
        var Oe;
        for (Oe in r)
          if (Object.prototype.hasOwnProperty.call(r, Oe) && Oe.indexOf(et) === 0)
            return !0;
        return !1;
      }, a = typeof o == "string" && (Object.prototype.hasOwnProperty.call(r, o) || c(o)), b = !a && typeof o == "string", g = typeof o == "function", v = !1, F, k, ue;
      if (a) {
        t.clearSubscriptions(o);
        return;
      }
      for (F in r)
        if (Object.prototype.hasOwnProperty.call(r, F)) {
          if (k = r[F], b && k[o]) {
            delete k[o], v = o;
            break;
          }
          if (g)
            for (ue in k)
              Object.prototype.hasOwnProperty.call(k, ue) && k[ue] === o && (delete k[ue], v = !0);
        }
      return v;
    };
  });
})(pe, pe.exports);
var U = pe.exports, de = function() {
  return de = Object.assign || function(e) {
    for (var t, r = 1, n = arguments.length; r < n; r++) {
      t = arguments[r];
      for (var u in t)
        Object.prototype.hasOwnProperty.call(t, u) && (e[u] = t[u]);
    }
    return e;
  }, de.apply(this, arguments);
};
function rt(s) {
  return s.toLowerCase();
}
var it = [/([a-z0-9])([A-Z])/g, /([A-Z])([A-Z][a-z])/g], st = /[^A-Z0-9]+/gi;
function nt(s, e) {
  e === void 0 && (e = {});
  for (var t = e.splitRegexp, r = t === void 0 ? it : t, n = e.stripRegexp, u = n === void 0 ? st : n, f = e.transform, l = f === void 0 ? rt : f, p = e.delimiter, P = p === void 0 ? " " : p, S = Ue(Ue(s, r, "$1\0$2"), u, "\0"), w = 0, E = S.length; S.charAt(w) === "\0"; )
    w++;
  for (; S.charAt(E - 1) === "\0"; )
    E--;
  return S.slice(w, E).split("\0").map(l).join(P);
}
function Ue(s, e, t) {
  return e instanceof RegExp ? s.replace(e, t) : e.reduce(function(r, n) {
    return r.replace(n, t);
  }, s);
}
function Ke(s, e) {
  var t = s.charAt(0), r = s.substr(1).toLowerCase();
  return e > 0 && t >= "0" && t <= "9" ? "_" + t + r : "" + t.toUpperCase() + r;
}
function ot(s) {
  return s.charAt(0).toUpperCase() + s.slice(1).toLowerCase();
}
function ct(s, e) {
  return e === void 0 && (e = {}), nt(s, de({ delimiter: "", transform: Ke }, e));
}
function ut(s, e) {
  return e === 0 ? s.toLowerCase() : Ke(s, e);
}
function at(s, e) {
  return e === 0 ? s.toLowerCase() : ot(s);
}
function lt(s, e) {
  return e === void 0 && (e = {}), ct(s, de({ transform: ut }, e));
}
function ft(s) {
  switch (s) {
    case "String":
      return "";
    case "Boolean":
      return !1;
    case "Number":
      return 0;
    case "Array":
      return [];
    case "SourceProvider":
      return;
    case "Store":
      return;
    case "Object":
    default:
      return {};
  }
}
const ht = (s, {
  property: e = s,
  description: t = "",
  type: r = "String",
  defaultValue: n,
  attribute: u = s.toLowerCase(),
  reflect: f = !1,
  primary: l = !1,
  changeEvent: p,
  ...P
} = {}) => {
  const S = typeof r == "function" ? r.name : r, w = n ?? ft(S);
  return {
    property: e,
    description: t,
    type: S,
    defaultValue: w,
    attribute: u,
    reflect: !!f,
    primary: !!l,
    changeEvent: p,
    ...P
  };
}, He = ({
  description: s = "",
  group: e = "default",
  defaultSourceKey: t,
  defaultSourceProvider: r,
  dashboard: n,
  properties: u = {},
  events: f = [],
  slots: l = [],
  cssProperties: p = [],
  cssParts: P = [],
  ...S
} = {}) => ({
  description: s,
  group: e,
  defaultSourceKey: t,
  defaultSourceProvider: r,
  dashboard: {
    topLevel: !0,
    ...n
  },
  properties: Object.fromEntries(
    Object.entries(u).map(([w, E]) => [w, ht(w, E)])
  ),
  events: f,
  slots: l,
  cssProperties: p,
  cssParts: P,
  ...S
}), ae = (s) => typeof s == "string" ? "String" : typeof s == "number" ? "Number" : typeof s == "boolean" ? "Boolean" : s instanceof Array ? "Array" : s instanceof Object ? "Object" : null, M = (s, e) => {
  if (s === e)
    return !0;
  if (typeof s != typeof e)
    return !1;
  if (s instanceof Array && e instanceof Array) {
    if (s.length !== e.length)
      return !1;
    for (let t = 0; t < s.length; t++)
      if (s[t] !== e[t])
        return !1;
    return !0;
  }
  return !1;
};
function Ve() {
}
class De {
  static prop2PropValue(e) {
    if (e === null)
      return null;
    if (e instanceof Array)
      return e;
    if (typeof e == "string")
      try {
        const t = JSON.parse(e);
        return t instanceof Array ? t : [];
      } catch {
        return [];
      }
    return [];
  }
  static prop2AttrValue(e) {
    const t = De.prop2PropValue(e);
    return t === null ? null : JSON.stringify(t);
  }
  static attr2PropValue(e) {
    if (e === null)
      return null;
    try {
      const t = JSON.parse(e);
      return t instanceof Array ? t : [];
    } catch {
      return [];
    }
  }
}
class pt {
  static prop2PropValue(e) {
    return e === null ? null : !!e;
  }
  static prop2AttrValue(e) {
    return e ? "" : null;
  }
  static attr2PropValue(e) {
    return e !== null;
  }
}
class dt {
  static prop2PropValue(e) {
    return e === null ? null : parseFloat(e);
  }
  static prop2AttrValue(e) {
    return e === null ? null : parseFloat(e).toString();
  }
  static attr2PropValue(e, t) {
    return e === null ? null : parseFloat(e);
  }
}
class Re {
  static prop2PropValue(e) {
    if (e === null)
      return null;
    if (e instanceof Object)
      return e;
    if (typeof e == "string")
      try {
        const t = JSON.parse(e);
        return t instanceof Object ? t : {};
      } catch {
        return {};
      }
    return {};
  }
  static prop2AttrValue(e) {
    const t = Re.prop2PropValue(e);
    return t === null ? null : JSON.stringify(t);
  }
  static attr2PropValue(e) {
    if (e === null)
      return null;
    try {
      const t = JSON.parse(e);
      return t instanceof Object ? t : {};
    } catch {
      return {};
    }
  }
}
class bt {
  static prop2PropValue(e) {
    return e === null ? null : (e == null ? void 0 : e.toString()) || "";
  }
  static prop2AttrValue(e) {
    return e === null ? null : (e == null ? void 0 : e.toString()) || "";
  }
  static attr2PropValue(e, t) {
    return e === null ? null : e;
  }
}
const Le = {
  Array: De,
  Boolean: pt,
  Number: dt,
  Object: Re,
  String: bt
}, B = (s, e) => {
  var t;
  return ((t = Le[e]) == null ? void 0 : t.prop2PropValue(s)) ?? null;
}, Fe = (s, e) => {
  var t;
  return ((t = Le[e]) == null ? void 0 : t.prop2AttrValue(s)) ?? null;
}, we = (s, e) => {
  var t;
  return ((t = Le[e]) == null ? void 0 : t.attr2PropValue(s)) ?? null;
};
var C, T, I, x, Q, be, Qe, ie, Te, se, Ne;
const te = class te {
  constructor(e, t) {
    h(this, be);
    h(this, ie);
    h(this, se);
    h(this, C, void 0);
    h(this, T, void 0);
    h(this, I, !1);
    h(this, x, void 0);
    h(this, Q, void 0);
    te.UNIQUE_ID += 1, d(this, Q, Symbol(`WEBBIT_PROPERTY_CHANGE_${te.UNIQUE_ID}`)), d(this, C, e), d(this, T, t), d(this, I, !1), d(this, x, this.value), y(this, be, Qe).call(this);
  }
  get value() {
    const {
      reflect: e,
      attribute: t,
      property: r,
      type: n
    } = i(this, T);
    return typeof r == "string" && r in i(this, C) ? i(this, C)[r] : we(t && e ? i(this, C).getAttribute(t) : t ? i(this, C).getAttribute(t) : null, n);
  }
  set value(e) {
    const { attribute: t, property: r, type: n } = i(this, T), u = ae(e), f = this.value;
    if (e === null || typeof e > "u") {
      if (e === f)
        return;
    } else if (M(B(f, u), e))
      return;
    if (r) {
      const l = B(e, n), p = B(l, u);
      M(e, p) && (i(this, C)[r] = l);
    } else if (t) {
      const l = Fe(e, n), p = we(l, u);
      M(e, p) && (l === null ? i(this, C).removeAttribute(t) : i(this, C).setAttribute(t, l));
    }
  }
  isAttribute() {
    return typeof i(this, T).attribute == "string";
  }
  getStoredAttribute() {
    const { type: e } = i(this, T);
    return Fe(i(this, x), e);
  }
  connect() {
    if (!i(this, I)) {
      const e = this.value;
      d(this, x, typeof e < "u" ? e : i(this, T).defaultValue), d(this, I, !0);
    }
  }
  disconnect() {
    i(this, I) && (d(this, I, !1), y(this, se, Ne).call(this));
  }
  isConnected() {
    return i(this, I);
  }
  update(e) {
    this.connect(), this.value = e;
  }
  subscribe(e, t = !1) {
    U.subscribe(i(this, Q), (r, n) => {
      (i(this, I) || t) && e(n);
    });
  }
  getProperty() {
    return i(this, T);
  }
  setStoredValue(e) {
    d(this, x, e), this.isConnected() || y(this, se, Ne).call(this);
  }
  getStoredValue() {
    return i(this, x);
  }
};
C = new WeakMap(), T = new WeakMap(), I = new WeakMap(), x = new WeakMap(), Q = new WeakMap(), be = new WeakSet(), Qe = function() {
  const { changeEvent: e, attribute: t } = i(this, T);
  if (e) {
    const r = () => {
      y(this, ie, Te).call(this);
    };
    i(this, C).addEventListener(e, r, !1);
  } else
    t && new MutationObserver(() => {
      y(this, ie, Te).call(this);
    }).observe(i(this, C), {
      attributes: !0,
      attributeFilter: [t]
    });
}, ie = new WeakSet(), Te = function() {
  const { value: e } = this;
  this.isConnected() || d(this, x, e), U.publish(i(this, Q), e);
}, se = new WeakSet(), Ne = function() {
  this.value = i(this, x);
}, te.UNIQUE_ID = 0;
let Ae = te;
function le(s) {
  return lt(s, { transform: at });
}
function ke(s, e) {
  const r = Object.entries(s).find(([f]) => le(f) === e);
  if (!r)
    return;
  const [n, u] = r;
  return { name: n, source: u };
}
var m, _, j, V, G, $, W, J, z, D, Y, ge, Ge, ye, $e, Z, fe, Se, Je, Ee, ze, Pe, Ye;
const re = class re {
  constructor(e, t, r) {
    h(this, ge);
    h(this, ye);
    h(this, Z);
    h(this, Se);
    h(this, Ee);
    h(this, Pe);
    h(this, m, void 0);
    h(this, _, void 0);
    h(this, j, void 0);
    h(this, V, void 0);
    h(this, G, void 0);
    h(this, $, void 0);
    h(this, W, !1);
    h(this, J, void 0);
    h(this, z, {});
    h(this, D, Ve);
    h(this, Y, void 0);
    re.UNIQUE_ID += 1, d(this, Y, Symbol(`WEBBIT_ANY_PROPERTY_CHANGE_${re.UNIQUE_ID}`)), d(this, m, e), d(this, _, t), d(this, j, He(r || {
      name: e.tagName.toLowerCase()
    }));
    const n = Object.entries(i(this, j).properties).filter(([, u]) => !["SourceProvider", "Store", "Source"].includes(u.type)).map(([u, f]) => ({ ...f, name: u }));
    d(this, V, new Map(
      n.map((u) => {
        const f = new Ae(i(this, m), u);
        return f.subscribe((l) => {
          f.isConnected() && y(this, Pe, Ye).call(this, u, l), U.publish(i(this, Y), {
            property: u,
            value: l,
            connected: i(this, W)
          });
        }, !0), [u.name, f];
      })
    )), d(this, G, n.find(({ primary: u }) => u)), d(this, $, i(this, G) ? i(this, V).get(i(this, G).name) : void 0), d(this, J, y(this, ge, Ge).call(this)), i(this, _).defaultSourceProviderSet((u) => {
      this.sourceProvider === null && (this.sourceProvider = u);
    }), this.connect();
  }
  get sourceProvider() {
    return i(this, m).getAttribute("source-provider") ?? void 0;
  }
  set sourceProvider(e) {
    typeof e < "u" ? i(this, m).setAttribute("source-provider", e) : i(this, m).removeAttribute("source-provider");
  }
  get sourceKey() {
    return i(this, m).getAttribute("source-key") ?? void 0;
  }
  set sourceKey(e) {
    typeof e < "u" ? i(this, m).setAttribute("source-key", e) : i(this, m).removeAttribute("source-key");
  }
  get source() {
    if (typeof this.sourceProvider < "u" && typeof this.sourceKey < "u")
      return i(this, _).getSource(this.sourceProvider, this.sourceKey);
  }
  get hasSource() {
    return typeof this.source < "u";
  }
  subscribe(e) {
    const t = U.subscribe(i(this, Y), (r, n) => {
      e(n);
    });
    return () => U.unsubscribe(t);
  }
  getPropertyHandler(e) {
    return i(this, V).get(e);
  }
  connect() {
    d(this, W, !0), i(this, J).connect(), y(this, Z, fe).call(this);
  }
  disconnect() {
    d(this, W, !1), i(this, J).disconnect(), y(this, Z, fe).call(this);
  }
  setDefaultPropertyValue(e, t) {
    i(this, z)[e] = t;
  }
  getDefaultPropertyValue(e) {
    return i(this, z)[e];
  }
  getDefaultPropertyValues() {
    return i(this, z);
  }
  getConfig() {
    return i(this, j);
  }
};
m = new WeakMap(), _ = new WeakMap(), j = new WeakMap(), V = new WeakMap(), G = new WeakMap(), $ = new WeakMap(), W = new WeakMap(), J = new WeakMap(), z = new WeakMap(), D = new WeakMap(), Y = new WeakMap(), ge = new WeakSet(), Ge = function() {
  const e = new MutationObserver(() => {
    y(this, Z, fe).call(this);
  });
  return {
    connect: () => {
      e.observe(i(this, m), {
        attributes: !0,
        attributeFilter: ["source-provider", "source-key"]
      });
    },
    disconnect: () => {
      e.disconnect();
    }
  };
}, ye = new WeakSet(), $e = function() {
  const e = i(this, m).closest("[default-source-provider]");
  return (e == null ? void 0 : e.getAttribute("default-source-provider")) ?? void 0;
}, Z = new WeakSet(), fe = function() {
  if (y(this, Se, Je).call(this), !i(this, W)) {
    i(this, D).call(this), d(this, D, Ve), i(this, V).forEach((e) => e.disconnect());
    return;
  }
  this.sourceKey || (this.sourceKey = i(this, j).defaultSourceKey), this.sourceProvider || (this.sourceProvider = i(this, j).defaultSourceProvider || y(this, ye, $e).call(this) || i(this, _).getDefaultSourceProvider()), !this.sourceKey || !this.sourceProvider ? (i(this, D).call(this), d(this, D, Ve), i(this, V).forEach((e) => e.disconnect())) : (i(this, D).call(this), i(this, V).forEach((e) => {
    e.disconnect();
  }), d(this, D, i(this, _).subscribe(
    this.sourceProvider,
    this.sourceKey,
    (e, t, r) => {
      y(this, Ee, ze).call(this, t, r);
    },
    !0
  )));
}, Se = new WeakSet(), Je = function() {
  const e = this.sourceProvider ? i(this, _).getSourceProvider(this.sourceProvider) : void 0;
  Object.entries(i(this, j).properties).forEach(([, { property: t, type: r }]) => {
    t && (r === "SourceProvider" ? i(this, m)[t] = e : r === "Store" ? i(this, m)[t] = i(this, _) : r === "Source" && (i(this, m)[t] = this.source));
  });
}, Ee = new WeakSet(), ze = function(e, t) {
  var r, n;
  if (typeof this.source > "u")
    i(this, V).forEach((u) => {
      u.disconnect();
    });
  else if (this.source.hasChildren()) {
    const u = this.source.getChildren();
    if (e === t) {
      const f = Object.entries(u);
      [...i(this, V).entries()].forEach(([l, p]) => {
        const P = f.find(([S]) => le(S) === l);
        if (P) {
          const [, S] = P;
          p.update(S.getSourceValue());
        }
      });
    } else {
      const f = le(t.replace(`${e}/`, ""));
      if (i(this, V).has(f)) {
        const l = i(this, V).get(f), p = (n = (r = ke(u, f)) == null ? void 0 : r.source) == null ? void 0 : n.getSourceValue();
        typeof p > "u" ? l == null || l.disconnect() : l == null || l.update(p);
      }
    }
  } else
    i(this, $) && i(this, $).update(this.source.getSourceValue());
}, Pe = new WeakSet(), Ye = function({ name: e, primary: t }, r) {
  if (typeof this.source > "u")
    return;
  const n = ae(r), u = this.source.getChildren(), f = Object.entries(u);
  if (this.source.hasChildren())
    if (r === null) {
      const l = f.find(([p]) => le(p) === e);
      if (l) {
        const [, p] = l;
        p.getSourceValue() !== null && p.setSourceValue(r);
      }
    } else {
      const l = ke(u, e);
      if (!l)
        return;
      const p = ae(l.source.getSourceValue()) || n, P = B(r, p), S = B(P, n);
      M(r, S) && !M(l.source.getSourceValue(), P) && l.source.setSourceValue(r);
    }
  else {
    if (!t)
      return;
    if (r === null)
      this.source.getSourceValue() !== null && this.source.setSourceValue(r);
    else {
      const l = ae(this.source.getSourceValue()) || n, p = B(r, l), P = B(p, n);
      M(r, P) && !M(this.source.getSourceValue(), p) && this.source.setSourceValue(p);
    }
  }
}, re.UNIQUE_ID = 0;
let Ie = re;
var Me = function(s) {
  var e = s, t, r = {
    a: 0,
    b: 0,
    c: 0
  }, n = [], u = /(\[[^\]]+\])/g, f = /(#[^\#\s\+>~\.\[:\)]+)/g, l = /(\.[^\s\+>~\.\[:\)]+)/g, p = /(::[^\s\+>~\.\[:]+|:first-line|:first-letter|:before|:after)/gi, P = /(:(?!not|global|local)[\w-]+\([^\)]*\))/gi, S = /(:(?!not|global|local)[^\s\+>~\.\[:]+)/g, w = /([^\s\+>~\.\[:]+)/g;
  return t = function(E, A) {
    var O, o, c, a, b, g;
    if (E.test(e))
      for (O = e.match(E), o = 0, c = O.length; o < c; o += 1)
        r[A] += 1, a = O[o], b = e.indexOf(a), g = a.length, n.push({
          selector: s.substr(b, g),
          type: A,
          index: b,
          length: g
        }), e = e.replace(a, Array(g + 1).join(" "));
  }, function() {
    var E = function(c) {
      var a, b, g, v;
      if (c.test(e))
        for (a = e.match(c), b = 0, g = a.length; b < g; b += 1)
          v = a[b], e = e.replace(v, Array(v.length + 1).join("A"));
    }, A = /\\[0-9A-Fa-f]{6}\s?/g, O = /\\[0-9A-Fa-f]{1,5}\s/g, o = /\\./g;
    E(A), E(O), E(o);
  }(), function() {
    var E = /{[^]*/gm, A, O, o, c;
    if (E.test(e))
      for (A = e.match(E), O = 0, o = A.length; O < o; O += 1)
        c = A[O], e = e.replace(c, Array(c.length + 1).join(" "));
  }(), t(u, "b"), t(f, "a"), t(l, "b"), t(p, "c"), t(P, "b"), t(S, "b"), e = e.replace(/[\*\s\+>~]/g, " "), e = e.replace(/[#\.]/g, " "), e = e.replace(/:not/g, "    "), e = e.replace(/:local/g, "      "), e = e.replace(/:global/g, "       "), e = e.replace(/[\(\)]/g, " "), t(w, "c"), n.sort(function(E, A) {
    return E.index - A.index;
  }), {
    selector: s,
    specificity: "0," + r.a.toString() + "," + r.b.toString() + "," + r.c.toString(),
    specificityArray: [0, r.a, r.b, r.c],
    parts: n
  };
}, Be = function(s, e) {
  var t, r, n;
  if (typeof s == "string") {
    if (s.indexOf(",") !== -1)
      throw "Invalid CSS selector";
    t = Me(s).specificityArray;
  } else if (Array.isArray(s)) {
    if (s.filter(function(u) {
      return typeof u == "number";
    }).length !== 4)
      throw "Invalid specificity array";
    t = s;
  } else
    throw "Invalid CSS selector or specificity array";
  if (typeof e == "string") {
    if (e.indexOf(",") !== -1)
      throw "Invalid CSS selector";
    r = Me(e).specificityArray;
  } else if (Array.isArray(e)) {
    if (e.filter(function(u) {
      return typeof u == "number";
    }).length !== 4)
      throw "Invalid specificity array";
    r = e;
  } else
    throw "Invalid CSS selector or specificity array";
  for (n = 0; n < 4; n += 1) {
    if (t[n] < r[n])
      return -1;
    if (t[n] > r[n])
      return 1;
  }
  return 0;
};
function gt(s, e) {
  const t = [...s.entries()].find(([r]) => e.matches(r));
  return t == null ? void 0 : t[1];
}
function Ze(s, e, t) {
  const r = s, n = gt(t, r);
  return typeof n > "u" || !r.hasAttribute("source-provider") && typeof n.defaultSourceProvider > "u" && typeof e > "u" ? NodeFilter.FILTER_SKIP : NodeFilter.FILTER_ACCEPT;
}
function yt(s, e) {
  return (t) => Ze(t, s, e);
}
function St(s, e, t) {
  return document.createNodeIterator(
    s,
    NodeFilter.SHOW_ELEMENT,
    yt(e, t)
  );
}
var R, N, L, K, X, q, ee, he, ne, xe, oe, _e, me, Xe, ve, qe, ce, je;
const H = class H {
  constructor(e, t, r = {}, n = "default") {
    h(this, ee);
    h(this, ne);
    h(this, oe);
    h(this, me);
    h(this, ve);
    h(this, ce);
    h(this, R, void 0);
    h(this, N, /* @__PURE__ */ new Map());
    h(this, L, /* @__PURE__ */ new Map());
    h(this, K, void 0);
    h(this, X, void 0);
    h(this, q, void 0);
    H.UNIQUE_ID += 1, d(this, X, Symbol(`WEBBIT_CONNECTOR_ELEMENT_CONNECTED_${H.UNIQUE_ID}`)), d(this, q, Symbol(`WEBBIT_CONNECTOR_ELEMENT_DISCONNECTED_${H.UNIQUE_ID}`)), d(this, R, t), i(this, R).defaultSourceProviderSet((u) => {
      i(this, L).forEach((f, l) => {
        f.sourceProvider === null && this.setSourceProvider(l, u);
      });
    }), d(this, K, e), y(this, me, Xe).call(this), this.addElementConfigs(r, n);
  }
  addElementConfigs(e = {}, t = "default") {
    Object.entries(e).forEach(([r, n]) => {
      i(this, N).set(r, He({
        group: t,
        ...n
      }));
    }), y(this, oe, _e).call(this, i(this, K));
  }
  subscribeElementConnected(e) {
    U.subscribe(i(this, X), (t, r) => {
      e(r);
    });
  }
  subscribeElementDisconnected(e) {
    U.subscribe(i(this, q), (t, r) => {
      e(r);
    });
  }
  getRootElement() {
    return i(this, K);
  }
  /**
   * Gets the element config with the most specific selector that matches the element
   * @param element
   * @returns The element config or undefined if none is found
   */
  getMatchingElementConfig(e) {
    if (!e)
      return;
    const r = [
      ...i(this, N).keys()
    ].filter((n) => e.matches(n)).sort(Be).at(-1);
    return r ? this.getElementConfig(r) : void 0;
  }
  /**
   * Gets the most specific selector that matches the element
   * @param element
   * @returns The selector or undefined if none is found
   */
  getMatchingElementSelector(e) {
    return e ? [
      ...i(this, N).keys()
    ].filter((r) => e.matches(r)).sort(Be).at(-1) : void 0;
  }
  getElementConfigSelectors() {
    return [...i(this, N).keys()];
  }
  getElementConfig(e) {
    return i(this, N).get(e);
  }
  hasElementConfig(e) {
    return i(this, N).has(e);
  }
  setDefaultPropertyValue(e, t, r) {
    const n = this.getElementWebbit(e);
    n && n.setDefaultPropertyValue(t, r);
  }
  setSourceProvider(e, t) {
    const r = this.getElementWebbit(e);
    r && (r.sourceProvider = t);
  }
  setSourceKey(e, t) {
    const r = this.getElementWebbit(e);
    r && (r.sourceKey = t);
  }
  getDefaultPropertyValue(e, t) {
    const r = this.getElementWebbit(e);
    return r ? r.getDefaultPropertyValue(t) : void 0;
  }
  getDefaultPropertyValues(e) {
    const t = this.getElementWebbit(e);
    return t ? t.getDefaultPropertyValues() : {};
  }
  getSourceProvider(e) {
    const t = this.getElementWebbit(e);
    return t ? t.sourceProvider : i(this, R).getDefaultSourceProvider();
  }
  getSourceKey(e) {
    const t = this.getElementWebbit(e);
    return t ? t.sourceKey : void 0;
  }
  getElementWebbit(e) {
    return i(this, L).get(e);
  }
  hasElement(e) {
    return i(this, L).has(e);
  }
  getStore() {
    return i(this, R);
  }
};
R = new WeakMap(), N = new WeakMap(), L = new WeakMap(), K = new WeakMap(), X = new WeakMap(), q = new WeakMap(), ee = new WeakSet(), he = function(e) {
  const t = this.getMatchingElementConfig(e);
  if (this.hasElement(e) || !t)
    return;
  const r = new Ie(e, i(this, R), t);
  i(this, L).set(e, r), U.publish(i(this, X), { element: e });
}, ne = new WeakSet(), xe = function(e) {
  const t = this.getElementWebbit(e);
  t && (t.disconnect(), i(this, L).delete(e), U.publish(i(this, q), { element: e }));
}, oe = new WeakSet(), _e = function(e) {
  const t = y(this, ve, qe).call(this, e);
  for (; t.nextNode(); )
    y(this, ee, he).call(this, t.referenceNode);
}, me = new WeakSet(), Xe = function() {
  new MutationObserver((t) => {
    t.forEach((r) => {
      if (r.type === "childList") {
        const n = r.addedNodes || [], u = r.removedNodes || [];
        n.forEach((f) => {
          f instanceof HTMLElement && (y(this, ce, je).call(this, f) && y(this, ee, he).call(this, f), y(this, oe, _e).call(this, f));
        }), u.forEach((f) => {
          y(this, ne, xe).call(this, f), [...i(this, L).keys()].forEach((l) => {
            f.contains(l) && y(this, ne, xe).call(this, l);
          });
        });
      } else
        r.type === "attributes" && r.target instanceof HTMLElement && y(this, ce, je).call(this, r.target) && y(this, ee, he).call(this, r.target);
    });
  }).observe(i(this, K), {
    childList: !0,
    subtree: !0,
    attributes: !0,
    attributeFilter: ["source-key", "source-provider"]
  });
}, ve = new WeakSet(), qe = function(e) {
  return St(
    e,
    this.getStore().getDefaultSourceProvider(),
    i(this, N)
  );
}, ce = new WeakSet(), je = function(e) {
  return Ze(
    e,
    i(this, R).getDefaultSourceProvider(),
    i(this, N)
  ) === NodeFilter.FILTER_ACCEPT;
}, H.UNIQUE_ID = 0;
let We = H;
export {
  We as WebbitConnector,
  Ie as default,
  He as normalizeConfig
};
